

STAR build index
```sh
#!/bin/bash
export PATH=$PATH:~/software/STAR-master/bin/Linux_x86_64_static
fasta=/Share/home/tiangeng/reference_genome/human_grch38/Homo_sapiens.GRCh38.dna.toplevel.fa
gtf=/Share/home/tiangeng/reference_genome/human_grch38/Homo_sapiens.GRCh38.97.gtf
genomeout=/Share/home/tiangeng/reference_genome/human_hg38_star

STAR --runThreadN 14 --runMode genomeGenerate --genomeDir ${genomeout} --genomeFastaFiles ${fasta} --sjdbGTFfile ${gtf}

```
Bowtie build index
```sh
fasta=/Share/home/tiangeng/reference_genome/human_grch38/Homo_sapiens.GRCh38.dna.toplevel.fa
out=/Share/home/tiangeng/reference_genome/human_hg38_bowtie
bowtie-build ${fasta} ${out}/Homo_sapiens.GRCh38

```
## generate rRNA

/Share/home/tiangeng/reference_genome/human_grch38/a2.rRNA.sh
```sh

grep 'rRNA' Homo_sapiens.GRCh38.97.gtf > homo_rRNA.gtf
grep $'\t'gene$'\t' homo_rRNA.gtf > homo_rRNA2.gtf
mv homo_rRNA2.gtf homo_rRNA.gtf
fa_in=Homo_sapiens.GRCh38.dna.toplevel.fa
gtf=homo_rRNA.gtf
fa_rRNA=Homo_sapiens.rRNA.fa
bedtools getfasta -fi ${fa_in} -bed ${gtf} -fo ${fa_rRNA}
```
```sh
cd /Share/home/tiangeng/reference_genome/human_hg38_rRNA_bowtie
fasta=/Share/home/tiangeng/reference_genome/human_grch38/Homo_sapiens.rRNA.fa
out=/Share/home/tiangeng/reference_genome/human_hg38_rRNA_bowtie
bowtie-build ${fasta} ${out}/Homo_sapiens.rRNA

```

## generate human STAR index
```sh
#!/bin/bash
export PATH=$PATH:~/software/STAR-master/bin/Linux_x86_64_static
fasta=/Share/home/tiangeng/reference_genome/human_grch38/Homo_sapiens.GRCh38.dna.toplevel.fa
gtf=/Share/home/tiangeng/reference_genome/human_grch38/Homo_sapiens.GRCh38.97.gtf
genomeout=/Share/home/tiangeng/reference_genome/human_hg38_star

nohup STAR --runThreadN 14 --runMode genomeGenerate --limitGenomeGenerateRAM 162003700778 --genomeDir ${genomeout} --genomeFastaFiles ${fasta} --sjdbGTFfile ${gtf} >log.txt 2>&1 &
```

## remove rRNA
```sh
#!/bin/bash
data=/Share/home/tiangeng/ZY/RNA-seq/data/sequence_data/xhsample_zxy/trimmomatic
wkdir=/Share/home/tiangeng/project_result/RNAseq_resullt/project100_20190918_xiehe_zhang/a1-rRNA
name=(ZXY-A-A10 ZXY-A-A1 ZXY-A-A2 ZXY-A-A3 ZXY-A-A4 ZXY-A-A5 ZXY-A-A6 ZXY-A-A7 ZXY-A-A8 ZXY-A-A9 ZXY-B-B10 ZXY-B-B1 ZXY-B-B2 ZXY-B-B3 ZXY-B-B4 ZXY-B-B5 ZXY-B-B6 ZXY-B-B7 ZXY-B-B8 ZXY-B-B9)
bowtieindex=/Share/home/tiangeng/reference_genome/human_hg38_rRNA_bowtie/Homo_sapiens.rRNA
cd ${wkdir}
for i in ${name[@]}
do

echo " nohup bowtie -n 0 -norc --best -l 15 -p 15 --un=nocontam_${i} ${bowtieindex} -1 ${data}/${i}_1.paired.fq -2 ${data}/${i}_2.paired.fq ${i}.alin > ${i}.err 2>&1 &"
bowtie -n 0 -norc --best -l 15 -p 15 --un=nocontam_${i} ${bowtieindex} -1 ${data}/${i}_1.paired.fq -2 ${data}/${i}_2.paired.fq ${i}.alin > ${i}.err 2>&1
done
```
a2.merge.sh
```sh
name=(ZXY-A-A10 ZXY-A-A1 ZXY-A-A2 ZXY-A-A3 ZXY-A-A4 ZXY-A-A5 ZXY-A-A6 ZXY-A-A7 ZXY-A-A8 ZXY-A-A9 ZXY-B-B10 ZXY-B-B1 ZXY-B-B2 ZXY-B-B3 ZXY-B-B4 ZXY-B-B5 ZXY-B-B6 ZXY-B-B7 ZXY-B-B8 ZXY-B-B9)
head='Iterm'
for i in ${name[@]};
do
 head+="\t${i}";
done
echo -e $head >merge.counter;

for i in ${name[@]};
do
echo ${i}.err;
head -n 3 ${i}.err > ${i}.err.tmp;
sed -i 's/#//g' ${i}.err.tmp;
sed -i 's/:/\t/g' ${i}.err.tmp;
sed -i 's/^ //g' ${i}.err.tmp;
done
#
begin1=${name[0]};
begin2=${name[1]};
name2=("${name[@]:2}");
join -t $'\t' ${begin1}.err.tmp ${begin2}.err.tmp >merge.tmp

for i in ${name2[@]};
do 
echo ${i}.err.tmp;
join -t $'\t' merge.tmp ${i}.err.tmp >>merge.tmp2;
mv merge.tmp2 merge.tmp
done
#
cat merge.counter merge.tmp > merge2.tmp;
cut -f 2- merge2.tmp > summary.txt
rm merge.counter merge.tmp *.err.tmp
echo -e "Iterm\nTotal\nrRNA\nnon rRNA">name.txt;
paste -d $'\t' name.txt summary.txt > summary2.txt
mv summary2.txt summary.txt
rm name.txt merge2.tmp
```

## STAR mapping
/Share/home/tiangeng/project_result/RNAseq_resullt/project100_20190918_xiehe_zhang/a2-map

a1.map.sh
```sh
#!/bin/bash
export PATH=$PATH:~/software/STAR-master/bin/Linux_x86_64_static
genomeFile=/Share/home/tiangeng/reference_genome/human_hg38_star
fileName=(ZXY-A-A10 ZXY-A-A1 ZXY-A-A2 ZXY-A-A3 ZXY-A-A4 ZXY-A-A5 ZXY-A-A6 ZXY-A-A7 ZXY-A-A8 ZXY-A-A9)
for i in ${fileName[@]}
do
        mkdir -p ${i}_STAR
        cd ${i}_STAR
        echo "STAR --runThreadN 14 --alignEndsType EndToEnd --outFilterMismatchNmax 1 --outFilterMultimapNmax 1 --genomeDir $genomeFile --readFilesIn ../../a1-rRNA/nocontam_${i}_1 ../../a1-rRNA/nocontam_${i}_2 --outFileNamePrefix $i --outSAMtype BAM SortedByCoordinate --quantMode TranscriptomeSAM GeneCounts >${i}.log 2>&1"
        STAR --runThreadN 14 --alignEndsType EndToEnd --outFilterMismatchNmax 1 --outFilterMultimapNmax 1 --genomeDir $genomeFile --readFilesIn ../../a1-rRNA/nocontam_${i}_1 ../../a1-rRNA/nocontam_${i}_2 --outFileNamePrefix $i --outSAMtype BAM SortedByCoordinate --quantMode TranscriptomeSAM GeneCounts >${i}.log 2>&1
        cd ../
done

```




a2.map.sh
```sh
#!/bin/bash
export PATH=$PATH:~/software/STAR-master/bin/Linux_x86_64_static
genomeFile=/Share/home/tiangeng/reference_genome/human_hg38_star
fileName=(ZXY-B-B10 ZXY-B-B1 ZXY-B-B2 ZXY-B-B3 ZXY-B-B4 ZXY-B-B5 ZXY-B-B6 ZXY-B-B7 ZXY-B-B8 ZXY-B-B9)
for i in ${fileName[@]}
do
        mkdir -p ${i}_STAR
        cd ${i}_STAR
        echo "STAR --runThreadN 14 --alignEndsType EndToEnd --outFilterMismatchNmax 1 --outFilterMultimapNmax 1 --genomeDir $genomeFile --readFilesIn ../../a1-rRNA/nocontam_${i}_1 ../../a1-rRNA/nocontam_${i}_2 --outFileNamePrefix $i --outSAMtype BAM SortedByCoordinate --quantMode TranscriptomeSAM GeneCounts >${i}.log 2>&1"
        STAR --runThreadN 14 --alignEndsType EndToEnd --outFilterMismatchNmax 1 --outFilterMultimapNmax 1 --genomeDir $genomeFile --readFilesIn ../../a1-rRNA/nocontam_${i}_1 ../../a1-rRNA/nocontam_${i}_2 --outFileNamePrefix $i --outSAMtype BAM SortedByCoordinate --quantMode TranscriptomeSAM GeneCounts >${i}.log 2>&1
        cd ../
done

```
merge summary
```sh
name=(ZXY-A-A10 ZXY-A-A1 ZXY-A-A2 ZXY-A-A3 ZXY-A-A4 ZXY-A-A5 ZXY-A-A6 ZXY-A-A7 ZXY-A-A8 ZXY-A-A9 ZXY-B-B10 ZXY-B-B1 ZXY-B-B2 ZXY-B-B3 ZXY-B-B4 ZXY-B-B5 ZXY-B-B6 ZXY-B-B7 ZXY-B-B8 ZXY-B-B9)
head='Iterm'
for i in ${name[@]};
do
 head+="\t${i}";
 done
 echo -e $head >merge.counter;

 for i in ${name[@]};
 do
     echo "${i}_STAR/${i}Log.final.out";

 cat ${i}_STAR/${i}Log.final.out > ${i}.err.tmp;
 done
 #
 begin1=${name[0]};
 begin2=${name[1]};
 name2=("${name[@]:2}");
 join -t $'\t' ${begin1}.err.tmp ${begin2}.err.tmp >merge.tmp

 for i in ${name2[@]};
 do 
 echo ${i}.err.tmp;
 join -t $'\t' merge.tmp ${i}.err.tmp >>merge.tmp2;
 mv merge.tmp2 merge.tmp
 done

 cat merge.counter merge.tmp > merge2.tmp;
mv merge2.tmp summary.txt
rm *.err.tmp
rm merge.counter merge.tmp

sed -i 's/^[ ]*//g' summary.txt;
sed -i 's/|//g' summary.txt;
```

## remove duplication
/Share/home/tiangeng/project_result/RNAseq_resullt/project100_20190918_xiehe_zhang/a3-rmdup

a1.rmdup.sh
```sh
#!/bin/bash
data=/Share/home/tiangeng/project_result/RNAseq_resullt/project100_20190918_xiehe_zhang/a2-map
wkdir=/Share/home/tiangeng/project_result/RNAseq_resullt/project100_20190918_xiehe_zhang/a3-rmdup
fileName=(ZXY-B-B10 ZXY-B-B1 ZXY-B-B2 ZXY-B-B3 ZXY-B-B4 ZXY-B-B5 ZXY-B-B6 ZXY-B-B7 ZXY-B-B8 ZXY-B-B9)
for i in ${fileName[@]}
do
        mkdir -p ${wkdir}/${i}_rmdup
        echo "nohup samtools rmdup ${data}/${i}_STAR/${i}Aligned.sortedByCoord.out.bam ${wkdir}/${i}_rmdup/${i}.rmdup.bam > ${wkdir}/${i}_rmdup/${i}.log 2>&1 &"
        nohup samtools rmdup ${data}/${i}_STAR/${i}Aligned.sortedByCoord.out.bam ${wkdir}/${i}_rmdup/${i}.rmdup.bam > ${wkdir}/${i}_rmdup/${i}.log 2>&1 &
done
```
a2.rmdup.sh
```sh
#!/bin/bash
data=/Share/home/tiangeng/project_result/RNAseq_resullt/project100_20190918_xiehe_zhang/a2-map
wkdir=/Share/home/tiangeng/project_result/RNAseq_resullt/project100_20190918_xiehe_zhang/a3-rmdup
fileName=(ZXY-A-A10 ZXY-A-A1 ZXY-A-A2 ZXY-A-A3 ZXY-A-A4 ZXY-A-A5 ZXY-A-A6 ZXY-A-A7 ZXY-A-A8 ZXY-A-A9)
for i in ${fileName[@]}
do
        mkdir -p ${wkdir}/${i}_rmdup
        echo "nohup samtools rmdup ${data}/${i}_STAR/${i}Aligned.sortedByCoord.out.bam ${wkdir}/${i}_rmdup/${i}.rmdup.bam > ${wkdir}/${i}_rmdup/${i}.log 2>&1 &"
        nohup samtools rmdup ${data}/${i}_STAR/${i}Aligned.sortedByCoord.out.bam ${wkdir}/${i}_rmdup/${i}.rmdup.bam > ${wkdir}/${i}_rmdup/${i}.log 2>&1 &
done
```
```sh
name=(ZXY-A-A10 ZXY-A-A1 ZXY-A-A2 ZXY-A-A3 ZXY-A-A4 ZXY-A-A5 ZXY-A-A6 ZXY-A-A7 ZXY-A-A8 ZXY-A-A9 ZXY-B-B10 ZXY-B-B1 ZXY-B-B2 ZXY-B-B3 ZXY-B-B4 ZXY-B-B5 ZXY-B-B6 ZXY-B-B7 ZXY-B-B8 ZXY-B-B9)
head=''
for i in ${name[@]};
do
 head+="${i}\t";
done

echo -e $head >merge.counter;

for i in ${name[@]};
  do
    echo "${i}_rmdup/${i}.log";
    tail -n 1 ${i}_rmdup/${i}.log > ${i}.err.tmp;
    sed -i "s/\[bam_rmdup_core\] //g" ${i}.err.tmp;
    sed -i 's/\t//g' ${i}.err.tmp;
    sed -i "s/'//g" ${i}.err.tmp;
    sed -i "s/ in library//g" ${i}.err.tmp;
    sed -i "s/ //g" ${i}.err.tmp;
done

#
begin1=${name[0]};
begin2=${name[1]};
name2=("${name[@]:2}");
paste -d $'\t' ${begin1}.err.tmp ${begin2}.err.tmp >merge.tmp

for i in ${name2[@]};
   do
     echo ${i}.err.tmp;
     paste -d $'\t' merge.tmp ${i}.err.tmp >>merge.tmp2;
     mv merge.tmp2 merge.tmp
done

 cat merge.counter merge.tmp > merge2.tmp;
 mv merge2.tmp summary.txt
 rm *.err.tmp
 rm merge.counter merge.tmp
 ```

## Read count
/Share/home/tiangeng/project_result/RNAseq_resullt/project100_20190918_xiehe_zhang/a4-count
a1.count.sh
```sh
#!/bin/bash
data=/Share/home/tiangeng/project_result/RNAseq_resullt/project100_20190918_xiehe_zhang/a3-rmdup
wkdir=/Share/home/tiangeng/project_result/RNAseq_resullt/project100_20190918_xiehe_zhang/a4-count
gtf=/Share/home/tiangeng/reference_genome/human_grch38/Homo_sapiens.GRCh38.97.gtf
fileName=(ZXY-A-A10 ZXY-A-A1 ZXY-A-A2 ZXY-A-A3 ZXY-A-A4 ZXY-A-A5 ZXY-A-A6 ZXY-A-A7 ZXY-A-A8 ZXY-A-A9)
for i in ${fileName[@]}
do
echo "nohup htseq-count -r pos -q -f bam -s no ${data}/${i}_rmdup/${i}.rmdup.bam ${gtf} > ${wkdir}/${i}.count 2>&1 &"
nohup htseq-count -r pos -q -f bam -s no ${data}/${i}_rmdup/${i}.rmdup.bam ${gtf} > ${wkdir}/${i}.count 2>&1 &
done
```
a2.count.sh
```sh
#!/bin/bash
data=/Share/home/tiangeng/project_result/RNAseq_resullt/project100_20190918_xiehe_zhang/a3-rmdup
wkdir=/Share/home/tiangeng/project_result/RNAseq_resullt/project100_20190918_xiehe_zhang/a4-count
gtf=/Share/home/tiangeng/reference_genome/human_grch38/Homo_sapiens.GRCh38.97.gtf
fileName=(ZXY-B-B10 ZXY-B-B1 ZXY-B-B2 ZXY-B-B3 ZXY-B-B4 ZXY-B-B5 ZXY-B-B6 ZXY-B-B7 ZXY-B-B8 ZXY-B-B9)
for i in ${fileName[@]}
do
echo "nohup htseq-count -r pos -q -f bam -s no ${data}/${i}_rmdup/${i}.rmdup.bam ${gtf} > ${wkdir}/${i}.count 2>&1 &"
nohup htseq-count -r pos -q -f bam -s no ${data}/${i}_rmdup/${i}.rmdup.bam ${gtf} > ${wkdir}/${i}.count 2>&1 &
done
```

a3.merge.sh
```sh
name=(ZXY-A-A10 ZXY-A-A1 ZXY-A-A2 ZXY-A-A3 ZXY-A-A4 ZXY-A-A5 ZXY-A-A6 ZXY-A-A7 ZXY-A-A8 ZXY-A-A9 ZXY-B-B10 ZXY-B-B1 ZXY-B-B2 ZXY-B-B3 ZXY-B-B4 ZXY-B-B5 ZXY-B-B6 ZXY-B-B7 ZXY-B-B8 ZXY-B-B9)
head='gene'
for i in ${name[@]};
do
 head+=" ${i}";
done
echo -e $head >merge.counter;

begin1=${name[0]};
begin2=${name[1]};
name2=("${name[@]:2}");
join ${begin1}.count ${begin2}.count >merge.tmp
commander='join';
for i in ${name2[@]};
do 
echo ${i}.count;
join merge.tmp ${i}.count >>merge.tmp2;
mv merge.tmp2 merge.tmp
done

cat merge.counter merge.tmp > merge2.tmp;
rm merge.tmp
mv merge2.tmp merge.counter
sed -i 's/ \+/\t/g' merge.counter

grep -v '^__' merge.counter > merge.counter2
mv merge.counter2 merge.counter 
```

## summary the result
> /Share/home/tiangeng/project_result/RNAseq_resullt/project100_20190918_xiehe_zhang/a5-summary
a1.count.sh
```sh
#!/bin/bash
export PATH=/Share/home/tiangeng/anaconda2/bin:$PATH
name=(ZXY-A-A10 ZXY-A-A1 ZXY-A-A2 ZXY-A-A3 ZXY-A-A4 ZXY-A-A5 ZXY-A-A6 ZXY-A-A7 ZXY-A-A8 ZXY-A-A9)
gtf=/Share/home/tiangeng/reference_genome/human_grch38/Homo_sapiens.GRCh38.97.gtf
data_dir=/Share/home/tiangeng/project_result/RNAseq_resullt/project100_20190918_xiehe_zhang/a3-rmdup
work_dir=/Share/home/tiangeng/project_result/RNAseq_resullt/project100_20190918_xiehe_zhang/a5-summary
for i in ${name[@]}
do
       echo " nohup python /Share/home/tiangeng/bin/readsNumCal_intron_v3.py ${data_dir}/${i}_rmdup/${i}.rmdup.bam $gtf ${work_dir}/nocontam_${i}_mappedNum_intron.txt ${work_dir}/nocontam_${i} > ${work_dir}/read.${i}.log 2>&1 &"
       nohup python /Share/home/tiangeng/bin/readsNumCal_intron_v3.py ${data_dir}/${i}_rmdup/${i}.rmdup.bam $gtf ${work_dir}/nocontam_${i}_mappedNum_intron.txt ${work_dir}/nocontam_${i} > ${work_dir}/read.${i}.log 2>&1 &

done
```

a2.count.sh
```sh
#!/bin/bash
export PATH=/Share/home/tiangeng/anaconda2/bin:$PATH
name=(ZXY-B-B10 ZXY-B-B1 ZXY-B-B2 ZXY-B-B3 ZXY-B-B4 ZXY-B-B5 ZXY-B-B6 ZXY-B-B7 ZXY-B-B8 ZXY-B-B9)
gtf=/Share/home/tiangeng/reference_genome/human_grch38/Homo_sapiens.GRCh38.97.gtf
data_dir=/Share/home/tiangeng/project_result/RNAseq_resullt/project100_20190918_xiehe_zhang/a3-rmdup
work_dir=/Share/home/tiangeng/project_result/RNAseq_resullt/project100_20190918_xiehe_zhang/a5-summary
for i in ${name[@]}
do
       echo " nohup python /Share/home/tiangeng/bin/readsNumCal_intron_v3.py ${data_dir}/${i}_rmdup/${i}.rmdup.bam $gtf ${work_dir}/nocontam_${i}_mappedNum_intron.txt ${work_dir}/nocontam_${i} > ${work_dir}/read.${i}.log 2>&1 &"
       nohup python /Share/home/tiangeng/bin/readsNumCal_intron_v3.py ${data_dir}/${i}_rmdup/${i}.rmdup.bam $gtf ${work_dir}/nocontam_${i}_mappedNum_intron.txt ${work_dir}/nocontam_${i} > ${work_dir}/read.${i}.log 2>&1 &
done
```


## edge R
[Tutorial](https://bioconductor.org/packages/release/bioc/vignettes/edgeR/inst/doc/edgeRUsersGuide.pdf)
```R
library(edgeR)
# read data

x=read.table('/Share/home/tiangeng/project_result/Riboseq/project_190814_luboxun/a12-counter/lbx.merge.counter',header=T,row.names=1)
# set group
group <- factor(c(1,1,2,2))

y <- DGEList(counts=x,group=group)
y <- calcNormFactors(y)
design <- model.matrix(~group)
y <- estimateDisp(y,design)

# To perform quasi-likelihood F-tests:
fit <- glmQLFit(y,design)
qlf <- glmQLFTest(fit,coef=2)
topTags(qlf)

## filter reads
keep <- filterByExpr(y)
> y <- y[keep, , keep.lib.sizes=FALSE]


keep <- filterByExpr(counts, design)
v <- voom(counts[keep,], design, plot=TRUE)
fit <- lmFit(v, design)
fit <- eBayes(fit, robust=TRUE)
```
[voom](https://ucdavis-bioinformatics-training.github.io/2018-June-RNA-Seq-Workshop/thursday/DE.html)


```{r}
counts=read.table('/Users/xugang/Documents/r-rscript/merge.counter',header=T,row.names=1)
colnames(counts)=c("A10" ,"A1", "A2", "A3", "A4", "A5", "A6", "A7", "A8", "A9", "B10" ,"B1", "B2", "B3" ,"B4", "B5", "B6", "B7", "B8", "B9")
head(counts)
```

Create DGEList object
```{r}
d0 <- DGEList(counts)
d0
```

## 2. Preprocessing
Calculate normalization factors
```{r}
d0<- calcNormFactors(d0)
d0
```
Note: calcNormFactors doesn’t normalize the data, it just calculates normalization factors for use downstream.

Filter low-expressed genes
```{r}
cutoff <- 1
drop <- which(apply(cpm(d0), 1, max) < cutoff)
d <- d0[-drop,] 
dim(d)
```

```{r}
d$E
counts_cpm=cpm(counts)
write.csv(counts_cpm,'/Users/xugang/Documents/c-pycharm/git/sequencing_center/result/result_191008_xiehe/counts_cpm.csv')
```


The sample names are the cultivar, followed by the time, followed by the replicate
```{r}
snames <- colnames(counts) # Sample names
snames
```
```{r}
treat <- substr(snames, 1, 1) 
treat 

```
```{r}
people <- substr(snames, 2, nchar(snames))
people 

```

Create a new variable “group” that combines cultivar and time
```{r}
group <- interaction(people, treat)
group=interaction(treat)
group
```
Multidimensional scaling (MDS) plot
Multidimensional scaling plot of distances between gene expression profiles

Description

Plot samples on a two-dimensional scatterplot so that distances on the plot approximate the typical log2 fold changes between the samples.

Details

This function is a variation on the usual multdimensional scaling (or principle coordinate) plot, in that a distance measure particularly appropriate for the microarray context is used. The distance between each pair of samples (columns) is the root-mean-square deviation (Euclidean distance) for the top top genes. Distances on the plot can be interpreted as leading log2-fold-change, meaning the typical (root-mean-square) log2-fold-change between the samples for the genes that distinguish those samples.

If gene.selection is "common", then the top genes are those with the largest standard deviations between samples. If gene.selection is "pairwise", then a different set of top genes is selected for each pair of samples. The pairwise feature selection may be appropriate for microarray data when different molecular pathways are relevant for distinguishing different pairs of samples.

If pch=NULL, then each sample is represented by a text label, defaulting to the column names of x. If pch is not NULL, then plotting symbols are used.


```{r}
plotMDS(d, col = as.numeric(group),gene.selection = "pairwise")
pdf('/Users/xugang/Downloads/mds.pdf')
plotMDS(d, col = as.numeric(group),gene.selection = "pairwise")
dev.off()
?plotMDS
```



3. Voom transformation and calculation of variance weights
Specify the model to be fitted. We do this before using voom since voom uses variances of the model residuals (observed - fitted)


```{r}
mm <- model.matrix(~0 + treat)
mm
```

The above specifies a model where each coefficient corresponds to a group mean

Voom


```{r}
y = voom(d, mm, plot = T)
pdf('/Users/xugang/Downloads/mean-variance-remove-lowexpres.pdf')
y = voom(d, mm, plot = T)
dev.off()
```
```{r}
y$E
write.csv(y$E,'/Users/xugang/Documents/c-pycharm/git/sequencing_center/result/result_191008_xiehe/normal_gene_expression.csv')
```

What is voom doing?

1. Counts are transformed to log2 counts per million reads (CPM), where “per million reads” is defined based on the normalization factors we calculated earlier
2.A linear model is fitted to the log2 CPM for each gene, and the residuals are calculated
3.A smoothed curve is fitted to the sqrt(residual standard deviation) by average expression (see red line in plot above)
4. The smoothed curve is used to obtain weights for each gene and sample that are passed into limma along with the log2 CPMs.

More details at https://genomebiology.biomedcentral.com/articles/10.1186/gb-2014-15-2-r29

The above is a “good” voom plot. If your voom plot looks like the below, you might want to filter more:
```{r}
tmp <- voom(d0, mm, plot = T)
pdf('/Users/xugang/Downloads/mean-variance-raw.pdf')
tmp <- voom(d0, mm, plot = T)
dev.off()
```

## 4. Fitting linear models in limma

lmFit fits a linear model using weighted least squares for each gene:
```{r}
fit <- lmFit(y, mm)
head(coef(fit))
```
Comparisons between groups (log fold-changes) are obtained as contrasts of these fitted linear models:

Specify which groups to compare:

Comparison between times 6 and 9 for cultivar I5
```{r}
contr <- makeContrasts(treatA - treatB, levels = colnames(coef(fit)))
contr
```
Estimate contrast for each gene
```{r}
tmp <- contrasts.fit(fit, contr)
head(tmp$coefficients)
```

Empirical Bayes smoothing of standard errors (shrinks standard errors that are much larger or smaller than those from other genes towards the average standard error) (see https://www.degruyter.com/doi/10.2202/1544-6115.1027)
```{r}
tmp <- eBayes(tmp)
```

What genes are most differentially expressed?
```{r}
top.table <- topTable(tmp, sort.by = "P", n = Inf)
head(top.table, 40)
```
logFC: log2 fold change of TimeA/TimeB
AveExpr: Average expression across all samples, in log2 CPM
t: logFC divided by its standard error
P.Value: Raw p-value (based on t) from test that logFC differs from 0
adj.P.Val: Benjamini-Hochberg false discovery rate adjusted p-value
B: log-odds that gene is DE (arguably less useful than the other columns)

ENSG00000222675 has higher expression at time A than at time B (logFC is positive). 
ENSG00000229436 has lower expression at time A than at time B (logFC is negative).


How many DE genes are there?


```{r}
length(which(top.table$P.Value < 0.05))
write.csv(top.table,'/Users/xugang/Downloads/remove_low_expression.csv')
```

What if we refit our model as a two-factor model (rather than using the group variable)?

Create new model matrix:
Use concentration
```{r}
concen=c(3242, 679, 368, 1022.1, 1555.1, 2291.8, 971.5, 1008.4, 607, 1585.4, 695, 1345.2, 514.9, 785.8, 490.1, 1054.1, 977.1, 633.5, 580.8, 395)
concen
```

Persion
```{r}
people=c('10','1','2','3','4','5','6','7','8','9','6','11','12','13','14','15','16','17','18','2')
p1=c('10','1','2','3','4','5','6','7','8','9','6','11','12','13','14','15','16','17','18','2')
p2=c()
people
```

Treatment
```{r}
treat

```
library_time
```{r}
library_time=c('2018.05.09', '2018.04.02','2018.04.02','2018.04.04','2018.04.04','2018.04.04','2018.04.04','2018.04.23','2018.04.27','2018.04.27','2018.05.09','2018.04.19','2018.04.23','2018.04.23','2018.04.23','2018.04.23','2018.05.07','2018.04.27','2018.05.07','2018.05.07')

```


```{r}
mm <- model.matrix(~0+treat+concen)
mm
#time
```



We are specifying that model includes effects for cultivar, time, and the cultivar-time interaction (which allows the differences between cultivars to differ across time)
```{r}

y <- voom(d, mm, plot = T)
pdf('/Users/xugang/Downloads/mean-variance-remove-batch-effect.pdf')
y <- voom(d, mm, plot = T)
write.csv(y$E,'/Users/xugang/Documents/c-pycharm/git/sequencing_center/result/result_191008_xiehe/normal_gene_expression_concen.csv')

dev.off()
```

```{r}
fit <- lmFit(y, mm)
fit
```

```{r}
contr <- makeContrasts(treatA - treatB, levels = colnames(coef(fit)))
tmp <- contrasts.fit(fit, contr)
tmp <- eBayes(tmp)
top.table <- topTable(tmp, sort.by = "P",n = Inf)
head(top.table, 100)
write.table(top.table, file = "/Users/xugang/Documents/r-rscript/xiehe.txt", row.names = T, sep = "\t", quote = F)
write.csv(top.table, file = "/Users/xugang/Downloads/remove_centren.csv")
```
```{r}
mm <- model.matrix(~0+treat+library_time)
mm
y <- voom(d, mm, plot = T)
pdf('/Users/xugang/Downloads/mean-variance-remove-library_time.pdf')
y <- voom(d, mm, plot = T)

dev.off()

y$E
write.csv(y$E,'/Users/xugang/Documents/c-pycharm/git/sequencing_center/result/result_191008_xiehe/normal_gene_expression_library_time.csv')
fit <- lmFit(y, mm)
fit
contr <- makeContrasts(treatA - treatB, levels = colnames(coef(fit)))
tmp <- contrasts.fit(fit, contr)
tmp <- eBayes(tmp)
top.table <- topTable(tmp, sort.by = "P",n = Inf)
head(top.table, 100)

write.csv(top.table, file = "/Users/xugang/Downloads/remove_library_time.csv")
```
```{r}
mm <- model.matrix(~0+treat+concen)
mm
y <- voom(d, mm, plot = T)
pdf('/Users/xugang/Downloads/mean-variance-remove-concen.pdf')
y <- voom(d, mm, plot = T)
dev.off()
fit <- lmFit(y, mm)
fit
contr <- makeContrasts(treatA - treatB, levels = colnames(coef(fit)))
tmp <- contrasts.fit(fit, contr)
tmp <- eBayes(tmp)
top.table <- topTable(tmp, sort.by = "P",n = Inf)
head(top.table, 100)

write.csv(top.table, file = "/Users/xugang/Downloads/remove_concen.csv")
```




```{r}
length(which(top.table$P.Value < 0.01))
which(top.table$P.Value < 0.01)
```

In this case, limma is fitting a linear regression model, which here is a straight line fit, with the slope and intercept defined by the model coefficients:


```{r}
ENSG00000071282 <- y$E["ENSG00000071282",]
plot( concen ~ ENSG00000071282)
coef(fit)
intercept <- 0
slope <- coef(fit)["ENSG00000071282","concen"]
plot(concen ~ ENSG00000071282)
tmp
```
Install from CRAN as follow:
```{r}
install.packages("ggpubr")
```
```{r}
library(ggpubr)
```
Demo data sets
Data: ToothGrowth data sets.


```{r}
data("ToothGrowth")
head(ToothGrowth)
```
```{r}
#ex=read.csv('/Users/xugang/Documents/c-pycharm/git/sequencing_center/result/result_191008_xiehe/normal_gene_expression.csv')
ex=read.csv('/Users/xugang/Documents/c-pycharm/git/sequencing_center/result/result_191008_xiehe/counts_cpm.csv')
# 
dim(ex)

```
```{r}
data.use = data.frame(group=substr(colnames(ex)[-1],1,1),t(ex[ex[,1]=="ENSG00000268658",-1]))

colnames(data.use)=c("group","X1")

compare_means(X1 ~ group,  data = data.use,
              method = "t.test")

png('/Users/xugang/Documents/c-pycharm/git/sequencing_center/result/result_191008_xiehe/image/ENSG00000268658.png')
# Visualize
ggboxplot(data.use, x =  "group", y =  "X1",title="ENSG00000268658",
          color = "group", palette = "jco",add = "jitter")+
  stat_compare_means(method = "t.test")+      # Add global p-value
  stat_compare_means(label = "p.signif", method = "t.test",aes(label = ..p.signif..),label.x = 1.5)
dev.off()

#plot_exp('ENSG00000211667')

```
```{r}
plot_exp <- function(gene_name) {
data.use = data.frame(group=substr(colnames(ex)[-1],1,1),t(ex[ex[,1]==gene_name,-1]))
colnames(data.use)=c("group","X1")
compare_means(X1 ~ group,  data = data.use,
              method = "t.test")
png_file=paste("/Users/xugang/Documents/c-pycharm/git/sequencing_center/result/result_191008_xiehe/autophagy_image/",gene_name, ".png",sep="")
png(png_file)
# Visualize
ggboxplot(data.use, x =  "group", y =  "X1",title=gene_name,
          color = "group", palette = "jco",add = "jitter")+
  stat_compare_means(method = "t.test")+      # Add global p-value
  stat_compare_means(label = "p.signif", method = "t.test",aes(label = ..p.signif..),label.x = 1.5)

}

```


```{r}
plot_exp('ENSG00000183625')
dev.off()
plot_exp('ENSG00000234832')
dev.off()
plot_exp('ENSG00000211667')
dev.off()
plot_exp('ENSG00000262074')
dev.off()
plot_exp('ENSG00000188266')
dev.off()
plot_exp('ENSG00000260078')
dev.off()
plot_exp('ENSG00000268240')
dev.off()
plot_exp('ENSG00000272831')
dev.off()
plot_exp('ENSG00000197013')
dev.off()
plot_exp('ENSG00000271532')
dev.off()
plot_exp('ENSG00000169429')
dev.off()
plot_exp('ENSG00000260898')
dev.off()



```
```{r}
plot_exp('ENSG00000244734')
dev.off()
plot_exp('ENSG00000188536')
dev.off()
plot_exp('ENSG00000206172')
dev.off()
plot_exp('ENSG00000131203')
dev.off()
```

```{r}
#malaria_image
plot_exp('ENSG00000244734')
dev.off()

plot_exp('ENSG00000188536')
dev.off()

plot_exp('ENSG00000206172')
dev.off()

plot_exp('ENSG00000169429')
dev.off()
```
```{r}
#autophagy_image
plot_exp('ENSG00000170296')

dev.off()
plot_exp('ENSG00000139112')

dev.off()
plot_exp('ENSG00000034713')

dev.off()
```




# Methods for comparing means
The standard methods to compare the means of two or more groups in R, have been largely described at: comparing means in R.

The most common methods for comparing means include:

METHODS	R FUNCTION	DESCRIPTION
T-test	t.test()	Compare two groups (parametric) # each group 30 mean
Wilcoxon test	wilcox.test()	Compare two groups (non-parametric) # 不用假设分布,中位数
ANOVA	aov() or anova()	Compare multiple groups (parametric) # 方差分析
Kruskal-Wallis	kruskal.test()	Compare multiple groups (non-parametric)

# Compare two independent groups
```{r}



```


```{r}
compare_means(len ~ supp, data = ToothGrowth)
p <- ggboxplot(ToothGrowth, x = "supp", y = "len",
          color = "supp", palette = "jco",
          add = "jitter")
#  Add p-value
p + stat_compare_means()

# Change method
p + stat_compare_means(method = "t.test")
p + stat_compare_means( aes(label = ..p.signif..), 
                        label.x = 1.5, label.y = 40)
```




# Compare two paired samples
```{r}
compare_means(len ~ supp, data = ToothGrowth, paired = TRUE)
ggpaired(ToothGrowth, x = "supp", y = "len",
         color = "supp", line.color = "gray", line.size = 0.4,
         palette = "jco")+
  stat_compare_means(paired = TRUE)
```

# Compare more than two groups
```{r}
compare_means(len ~ dose,  data = ToothGrowth, method = "anova")
#Plot with global p-value:

# Default method = "kruskal.test" for multiple groups
ggboxplot(ToothGrowth, x = "dose", y = "len",
          color = "dose", palette = "jco")+
  stat_compare_means()

# Change method to anova
ggboxplot(ToothGrowth, x = "dose", y = "len",
          color = "dose", palette = "jco")+
  stat_compare_means(method = "anova")
```

# Perorm pairwise comparisons
```{r}

compare_means(len ~ dose,  data = ToothGrowth)

# Visualize: Specify the comparisons you want
my_comparisons <- list( c("0.5", "1"), c("1", "2"), c("0.5", "2") )
ggboxplot(ToothGrowth, x = "dose", y = "len",
          color = "dose", palette = "jco")+ 
  stat_compare_means(comparisons = my_comparisons)+ # Add pairwise comparisons p-value
  stat_compare_means(label.y = 50)     # Add global p-value

ggboxplot(ToothGrowth, x = "dose", y = "len",
          color = "dose", palette = "jco")+ 
  stat_compare_means(comparisons = my_comparisons, label.y = c(29, 35, 40))+
  stat_compare_means(label.y = 45)
```

# Multiple pairwise tests against a reference group:
```{r}
compare_means(len ~ dose,  data = ToothGrowth, ref.group = "0.5",
              method = "t.test")
# Visualize
ggboxplot(ToothGrowth, x = "dose", y = "len",
          color = "dose", palette = "jco")+
  stat_compare_means(method = "anova", label.y = 40)+      # Add global p-value
  stat_compare_means(label = "p.signif", method = "t.test",
                     ref.group = "0.5")
```

# Multiple pairwise tests against all (base-mean):
```{r}
compare_means(len ~ dose,  data = ToothGrowth, ref.group = ".all.",
              method = "t.test")
# Visualize
ggboxplot(ToothGrowth, x = "dose", y = "len",
          color = "dose", palette = "jco")+
  stat_compare_means(method = "anova", label.y = 40)+      # Add global p-value
  stat_compare_means(label = "p.signif", method = "t.test",
                     ref.group = ".all.")                  # Pairwise comparison against all
```
# Load myeloma data from survminer package

```{r}
if(!require(survminer)) install.packages("survminer")
data("myeloma", package = "survminer")
```

#

```{r}
compare_means(DEPDC1 ~ molecular_group,  data = myeloma,
              ref.group = ".all.", method = "t.test")
```

#

```{r}
# Perform the test
compare_means(DEPDC1 ~ molecular_group,  data = myeloma,
              ref.group = ".all.", method = "t.test")
# Visualize the expression profile
ggboxplot(myeloma, x = "molecular_group", y = "DEPDC1", color = "molecular_group", 
          add = "jitter", legend = "none") +
  rotate_x_text(angle = 45)+
  geom_hline(yintercept = mean(myeloma$DEPDC1), linetype = 2)+ # Add horizontal line at base mean
  stat_compare_means(method = "anova", label.y = 1600)+        # Add global annova p-value
  stat_compare_means(label = "p.signif", method = "t.test",
                     ref.group = ".all.")
# Pairwise comparison against all
# Visualize the expression profile
ggboxplot(myeloma, x = "molecular_group", y = "DEPDC1", color = "molecular_group", 
          add = "jitter", legend = "none") +
  rotate_x_text(angle = 45)+
  geom_hline(yintercept = mean(myeloma$DEPDC1), linetype = 2)+ # Add horizontal line at base mean
  stat_compare_means(method = "anova", label.y = 1600)+        # Add global annova p-value
  stat_compare_means(label = "p.signif", method = "t.test",
                     ref.group = ".all.", hide.ns = TRUE)      # Pairwise comparison against all
```
