

STAR build index
```sh
#!/bin/bash
export PATH=$PATH:~/software/STAR-master/bin/Linux_x86_64_static
fasta=/Share/home/tiangeng/reference_genome/human_grch38/Homo_sapiens.GRCh38.dna.toplevel.fa
gtf=/Share/home/tiangeng/reference_genome/human_grch38/Homo_sapiens.GRCh38.97.gtf
genomeout=/Share/home/tiangeng/reference_genome/human_hg38_star

STAR --runThreadN 14 --runMode genomeGenerate --genomeDir ${genomeout} --genomeFastaFiles ${fasta} --sjdbGTFfile ${gtf}

```
Bowtie build index
```sh
fasta=/Share/home/tiangeng/reference_genome/human_grch38/Homo_sapiens.GRCh38.dna.toplevel.fa
out=/Share/home/tiangeng/reference_genome/human_hg38_bowtie
bowtie-build ${fasta} ${out}/Homo_sapiens.GRCh38

```
## generate rRNA

/Share/home/tiangeng/reference_genome/human_grch38/a2.rRNA.sh
```sh

grep 'rRNA' Homo_sapiens.GRCh38.97.gtf > homo_rRNA.gtf
grep $'\t'gene$'\t' homo_rRNA.gtf > homo_rRNA2.gtf
mv homo_rRNA2.gtf homo_rRNA.gtf
fa_in=Homo_sapiens.GRCh38.dna.toplevel.fa
gtf=homo_rRNA.gtf
fa_rRNA=Homo_sapiens.rRNA.fa
bedtools getfasta -fi ${fa_in} -bed ${gtf} -fo ${fa_rRNA}
```
```sh
cd /Share/home/tiangeng/reference_genome/human_hg38_rRNA_bowtie
fasta=/Share/home/tiangeng/reference_genome/human_grch38/Homo_sapiens.rRNA.fa
out=/Share/home/tiangeng/reference_genome/human_hg38_rRNA_bowtie
bowtie-build ${fasta} ${out}/Homo_sapiens.rRNA

```

## generate human STAR index
```sh
#!/bin/bash
export PATH=$PATH:~/software/STAR-master/bin/Linux_x86_64_static
fasta=/Share/home/tiangeng/reference_genome/human_grch38/Homo_sapiens.GRCh38.dna.toplevel.fa
gtf=/Share/home/tiangeng/reference_genome/human_grch38/Homo_sapiens.GRCh38.97.gtf
genomeout=/Share/home/tiangeng/reference_genome/human_hg38_star

nohup STAR --runThreadN 14 --runMode genomeGenerate --limitGenomeGenerateRAM 162003700778 --genomeDir ${genomeout} --genomeFastaFiles ${fasta} --sjdbGTFfile ${gtf} >log.txt 2>&1 &
```

## remove rRNA
```sh
#!/bin/bash
data=/Share/home/tiangeng/ZY/RNA-seq/data/sequence_data/xhsample_zxy/trimmomatic
wkdir=/Share/home/tiangeng/project_result/RNAseq_resullt/project100_20190918_xiehe_zhang/a1-rRNA
name=(ZXY-A-A10 ZXY-A-A1 ZXY-A-A2 ZXY-A-A3 ZXY-A-A4 ZXY-A-A5 ZXY-A-A6 ZXY-A-A7 ZXY-A-A8 ZXY-A-A9 ZXY-B-B10 ZXY-B-B1 ZXY-B-B2 ZXY-B-B3 ZXY-B-B4 ZXY-B-B5 ZXY-B-B6 ZXY-B-B7 ZXY-B-B8 ZXY-B-B9)
bowtieindex=/Share/home/tiangeng/reference_genome/human_hg38_rRNA_bowtie/Homo_sapiens.rRNA
cd ${wkdir}
for i in ${name[@]}
do

echo " nohup bowtie -n 0 -norc --best -l 15 -p 15 --un=nocontam_${i} ${bowtieindex} -1 ${data}/${i}_1.paired.fq -2 ${data}/${i}_2.paired.fq ${i}.alin > ${i}.err 2>&1 &"
bowtie -n 0 -norc --best -l 15 -p 15 --un=nocontam_${i} ${bowtieindex} -1 ${data}/${i}_1.paired.fq -2 ${data}/${i}_2.paired.fq ${i}.alin > ${i}.err 2>&1
done
```
a2.merge.sh
```sh
name=(ZXY-A-A10 ZXY-A-A1 ZXY-A-A2 ZXY-A-A3 ZXY-A-A4 ZXY-A-A5 ZXY-A-A6 ZXY-A-A7 ZXY-A-A8 ZXY-A-A9 ZXY-B-B10 ZXY-B-B1 ZXY-B-B2 ZXY-B-B3 ZXY-B-B4 ZXY-B-B5 ZXY-B-B6 ZXY-B-B7 ZXY-B-B8 ZXY-B-B9)
head='Iterm'
for i in ${name[@]};
do
 head+="\t${i}";
done
echo -e $head >merge.counter;

for i in ${name[@]};
do
echo ${i}.err;
head -n 3 ${i}.err > ${i}.err.tmp;
sed -i 's/#//g' ${i}.err.tmp;
sed -i 's/:/\t/g' ${i}.err.tmp;
sed -i 's/^ //g' ${i}.err.tmp;
done
#
begin1=${name[0]};
begin2=${name[1]};
name2=("${name[@]:2}");
join -t $'\t' ${begin1}.err.tmp ${begin2}.err.tmp >merge.tmp

for i in ${name2[@]};
do 
echo ${i}.err.tmp;
join -t $'\t' merge.tmp ${i}.err.tmp >>merge.tmp2;
mv merge.tmp2 merge.tmp
done
#
cat merge.counter merge.tmp > merge2.tmp;
cut -f 2- merge2.tmp > summary.txt
rm merge.counter merge.tmp *.err.tmp
echo -e "Iterm\nTotal\nrRNA\nnon rRNA">name.txt;
paste -d $'\t' name.txt summary.txt > summary2.txt
mv summary2.txt summary.txt
rm name.txt merge2.tmp
```

## STAR mapping
/Share/home/tiangeng/project_result/RNAseq_resullt/project100_20190918_xiehe_zhang/a2-map

a1.map.sh
```sh
#!/bin/bash
export PATH=$PATH:~/software/STAR-master/bin/Linux_x86_64_static
genomeFile=/Share/home/tiangeng/reference_genome/human_hg38_star
fileName=(ZXY-A-A10 ZXY-A-A1 ZXY-A-A2 ZXY-A-A3 ZXY-A-A4 ZXY-A-A5 ZXY-A-A6 ZXY-A-A7 ZXY-A-A8 ZXY-A-A9)
for i in ${fileName[@]}
do
        mkdir -p ${i}_STAR
        cd ${i}_STAR
        echo "STAR --runThreadN 14 --alignEndsType EndToEnd --outFilterMismatchNmax 1 --outFilterMultimapNmax 1 --genomeDir $genomeFile --readFilesIn ../../a1-rRNA/nocontam_${i}_1 ../../a1-rRNA/nocontam_${i}_2 --outFileNamePrefix $i --outSAMtype BAM SortedByCoordinate --quantMode TranscriptomeSAM GeneCounts >${i}.log 2>&1"
        STAR --runThreadN 14 --alignEndsType EndToEnd --outFilterMismatchNmax 1 --outFilterMultimapNmax 1 --genomeDir $genomeFile --readFilesIn ../../a1-rRNA/nocontam_${i}_1 ../../a1-rRNA/nocontam_${i}_2 --outFileNamePrefix $i --outSAMtype BAM SortedByCoordinate --quantMode TranscriptomeSAM GeneCounts >${i}.log 2>&1
        cd ../
done

```




a2.map.sh
```sh
#!/bin/bash
export PATH=$PATH:~/software/STAR-master/bin/Linux_x86_64_static
genomeFile=/Share/home/tiangeng/reference_genome/human_hg38_star
fileName=(ZXY-B-B10 ZXY-B-B1 ZXY-B-B2 ZXY-B-B3 ZXY-B-B4 ZXY-B-B5 ZXY-B-B6 ZXY-B-B7 ZXY-B-B8 ZXY-B-B9)
for i in ${fileName[@]}
do
        mkdir -p ${i}_STAR
        cd ${i}_STAR
        echo "STAR --runThreadN 14 --alignEndsType EndToEnd --outFilterMismatchNmax 1 --outFilterMultimapNmax 1 --genomeDir $genomeFile --readFilesIn ../../a1-rRNA/nocontam_${i}_1 ../../a1-rRNA/nocontam_${i}_2 --outFileNamePrefix $i --outSAMtype BAM SortedByCoordinate --quantMode TranscriptomeSAM GeneCounts >${i}.log 2>&1"
        STAR --runThreadN 14 --alignEndsType EndToEnd --outFilterMismatchNmax 1 --outFilterMultimapNmax 1 --genomeDir $genomeFile --readFilesIn ../../a1-rRNA/nocontam_${i}_1 ../../a1-rRNA/nocontam_${i}_2 --outFileNamePrefix $i --outSAMtype BAM SortedByCoordinate --quantMode TranscriptomeSAM GeneCounts >${i}.log 2>&1
        cd ../
done

```

## remove duplication
/Share/home/tiangeng/project_result/RNAseq_resullt/project100_20190918_xiehe_zhang/a3-rmdup

a1.rmdup.sh
```sh
#!/bin/bash
data=/Share/home/tiangeng/project_result/RNAseq_resullt/project100_20190918_xiehe_zhang/a2-map
wkdir=/Share/home/tiangeng/project_result/RNAseq_resullt/project100_20190918_xiehe_zhang/a3-rmdup
fileName=(ZXY-B-B10 ZXY-B-B1 ZXY-B-B2 ZXY-B-B3 ZXY-B-B4 ZXY-B-B5 ZXY-B-B6 ZXY-B-B7 ZXY-B-B8 ZXY-B-B9)
for i in ${fileName[@]}
do
        mkdir -p ${wkdir}/${i}_rmdup
        echo "nohup samtools rmdup ${data}/${i}_STAR/${i}Aligned.sortedByCoord.out.bam ${wkdir}/${i}_rmdup/${i}.rmdup.bam > ${wkdir}/${i}_rmdup/${i}.log 2>&1 &"
        nohup samtools rmdup ${data}/${i}_STAR/${i}Aligned.sortedByCoord.out.bam ${wkdir}/${i}_rmdup/${i}.rmdup.bam > ${wkdir}/${i}_rmdup/${i}.log 2>&1 &
done
```
a2.rmdup.sh
```sh
#!/bin/bash
data=/Share/home/tiangeng/project_result/RNAseq_resullt/project100_20190918_xiehe_zhang/a2-map
wkdir=/Share/home/tiangeng/project_result/RNAseq_resullt/project100_20190918_xiehe_zhang/a3-rmdup
fileName=(ZXY-A-A10 ZXY-A-A1 ZXY-A-A2 ZXY-A-A3 ZXY-A-A4 ZXY-A-A5 ZXY-A-A6 ZXY-A-A7 ZXY-A-A8 ZXY-A-A9)
for i in ${fileName[@]}
do
        mkdir -p ${wkdir}/${i}_rmdup
        echo "nohup samtools rmdup ${data}/${i}_STAR/${i}Aligned.sortedByCoord.out.bam ${wkdir}/${i}_rmdup/${i}.rmdup.bam > ${wkdir}/${i}_rmdup/${i}.log 2>&1 &"
        nohup samtools rmdup ${data}/${i}_STAR/${i}Aligned.sortedByCoord.out.bam ${wkdir}/${i}_rmdup/${i}.rmdup.bam > ${wkdir}/${i}_rmdup/${i}.log 2>&1 &
done
```

## Read count
/Share/home/tiangeng/project_result/RNAseq_resullt/project100_20190918_xiehe_zhang/a4-count
a1.count.sh
```sh
#!/bin/bash
data=/Share/home/tiangeng/project_result/RNAseq_resullt/project100_20190918_xiehe_zhang/a3-rmdup
wkdir=/Share/home/tiangeng/project_result/RNAseq_resullt/project100_20190918_xiehe_zhang/a4-count
gtf=/Share/home/tiangeng/reference_genome/human_grch38/Homo_sapiens.GRCh38.97.gtf
fileName=(ZXY-A-A10 ZXY-A-A1 ZXY-A-A2 ZXY-A-A3 ZXY-A-A4 ZXY-A-A5 ZXY-A-A6 ZXY-A-A7 ZXY-A-A8 ZXY-A-A9)
for i in ${fileName[@]}
do
echo "nohup htseq-count -r pos -q -f bam -s no ${data}/${i}_rmdup/${i}.rmdup.bam ${gtf} > ${wkdir}/${i}.count 2>&1 &"
nohup htseq-count -r pos -q -f bam -s no ${data}/${i}_rmdup/${i}.rmdup.bam ${gtf} > ${wkdir}/${i}.count 2>&1 &
done
```
a2.count.sh
```sh
#!/bin/bash
data=/Share/home/tiangeng/project_result/RNAseq_resullt/project100_20190918_xiehe_zhang/a3-rmdup
wkdir=/Share/home/tiangeng/project_result/RNAseq_resullt/project100_20190918_xiehe_zhang/a4-count
gtf=/Share/home/tiangeng/reference_genome/human_grch38/Homo_sapiens.GRCh38.97.gtf
fileName=(ZXY-B-B10 ZXY-B-B1 ZXY-B-B2 ZXY-B-B3 ZXY-B-B4 ZXY-B-B5 ZXY-B-B6 ZXY-B-B7 ZXY-B-B8 ZXY-B-B9)
for i in ${fileName[@]}
do
echo "nohup htseq-count -r pos -q -f bam -s no ${data}/${i}_rmdup/${i}.rmdup.bam ${gtf} > ${wkdir}/${i}.count 2>&1 &"
nohup htseq-count -r pos -q -f bam -s no ${data}/${i}_rmdup/${i}.rmdup.bam ${gtf} > ${wkdir}/${i}.count 2>&1 &
done
```

a3.merge.sh
```sh
name=(ZXY-A-A10 ZXY-A-A1 ZXY-A-A2 ZXY-A-A3 ZXY-A-A4 ZXY-A-A5 ZXY-A-A6 ZXY-A-A7 ZXY-A-A8 ZXY-A-A9 ZXY-B-B10 ZXY-B-B1 ZXY-B-B2 ZXY-B-B3 ZXY-B-B4 ZXY-B-B5 ZXY-B-B6 ZXY-B-B7 ZXY-B-B8 ZXY-B-B9)
head='gene'
for i in ${name[@]};
do
 head+=" ${i}";
done
echo -e $head >merge.counter;

begin1=${name[0]};
begin2=${name[1]};
name2=("${name[@]:2}");
join ${begin1}.count ${begin2}.count >merge.tmp
commander='join';
for i in ${name2[@]};
do 
echo ${i}.count;
join merge.tmp ${i}.count >>merge.tmp2;
mv merge.tmp2 merge.tmp
done

cat merge.counter merge.tmp > merge2.tmp;
rm merge.tmp
mv merge2.tmp merge.counter
sed -i 's/ \+/\t/g' merge.counter

grep -v '^__' merge.counter > merge.counter2
mv merge.counter2 merge.counter 
```

## summary the result
> /Share/home/tiangeng/project_result/RNAseq_resullt/project100_20190918_xiehe_zhang/a5-summary
a1.count.sh
```sh
#!/bin/bash
export PATH=/Share/home/tiangeng/anaconda2/bin:$PATH
name=(ZXY-A-A10 ZXY-A-A1 ZXY-A-A2 ZXY-A-A3 ZXY-A-A4 ZXY-A-A5 ZXY-A-A6 ZXY-A-A7 ZXY-A-A8 ZXY-A-A9)
gtf=/Share/home/tiangeng/reference_genome/human_grch38/Homo_sapiens.GRCh38.97.gtf
data_dir=/Share/home/tiangeng/project_result/RNAseq_resullt/project100_20190918_xiehe_zhang/a3-rmdup
work_dir=/Share/home/tiangeng/project_result/RNAseq_resullt/project100_20190918_xiehe_zhang/a5-summary
for i in ${name[@]}
do
       echo " nohup python /Share/home/tiangeng/bin/readsNumCal_intron_v3.py ${data_dir}/${i}_rmdup/${i}.rmdup.bam $gtf ${work_dir}/nocontam_${i}_mappedNum_intron.txt ${work_dir}/nocontam_${i} > ${work_dir}/read.${i}.log 2>&1 &"
       nohup python /Share/home/tiangeng/bin/readsNumCal_intron_v3.py ${data_dir}/${i}_rmdup/${i}.rmdup.bam $gtf ${work_dir}/nocontam_${i}_mappedNum_intron.txt ${work_dir}/nocontam_${i} > ${work_dir}/read.${i}.log 2>&1 &

done
```

a2.count.sh
```sh
#!/bin/bash
export PATH=/Share/home/tiangeng/anaconda2/bin:$PATH
name=(ZXY-B-B10 ZXY-B-B1 ZXY-B-B2 ZXY-B-B3 ZXY-B-B4 ZXY-B-B5 ZXY-B-B6 ZXY-B-B7 ZXY-B-B8 ZXY-B-B9)
gtf=/Share/home/tiangeng/reference_genome/human_grch38/Homo_sapiens.GRCh38.97.gtf
data_dir=/Share/home/tiangeng/project_result/RNAseq_resullt/project100_20190918_xiehe_zhang/a3-rmdup
work_dir=/Share/home/tiangeng/project_result/RNAseq_resullt/project100_20190918_xiehe_zhang/a5-summary
for i in ${name[@]}
do
       echo " nohup python /Share/home/tiangeng/bin/readsNumCal_intron_v3.py ${data_dir}/${i}_rmdup/${i}.rmdup.bam $gtf ${work_dir}/nocontam_${i}_mappedNum_intron.txt ${work_dir}/nocontam_${i} > ${work_dir}/read.${i}.log 2>&1 &"
       nohup python /Share/home/tiangeng/bin/readsNumCal_intron_v3.py ${data_dir}/${i}_rmdup/${i}.rmdup.bam $gtf ${work_dir}/nocontam_${i}_mappedNum_intron.txt ${work_dir}/nocontam_${i} > ${work_dir}/read.${i}.log 2>&1 &
done
```


## edge R
[Tutorial](https://bioconductor.org/packages/release/bioc/vignettes/edgeR/inst/doc/edgeRUsersGuide.pdf)
```R
library(edgeR)
# read data

x=read.table('/Share/home/tiangeng/project_result/Riboseq/project_190814_luboxun/a12-counter/lbx.merge.counter',header=T,row.names=1)
# set group
group <- factor(c(1,1,2,2))

y <- DGEList(counts=x,group=group)
y <- calcNormFactors(y)
design <- model.matrix(~group)
y <- estimateDisp(y,design)

# To perform quasi-likelihood F-tests:
fit <- glmQLFit(y,design)
qlf <- glmQLFTest(fit,coef=2)
topTags(qlf)

## filter reads
keep <- filterByExpr(y)
> y <- y[keep, , keep.lib.sizes=FALSE]


keep <- filterByExpr(counts, design)
v <- voom(counts[keep,], design, plot=TRUE)
fit <- lmFit(v, design)
fit <- eBayes(fit, robust=TRUE)
```
[voom](https://ucdavis-bioinformatics-training.github.io/2018-June-RNA-Seq-Workshop/thursday/DE.html)
