## 1. download genome annotation
```sh
wget ftp://ftp.ensemblgenomes.org/pub/plants/release-46/fasta/arabidopsis_thaliana/dna/Arabidopsis_thaliana.TAIR10.dna.toplevel.fa.gz
wet ftp://ftp.ensemblgenomes.org/pub/plants/release-46/gff3/arabidopsis_thaliana/Arabidopsis_thaliana.TAIR10.46.gff3.gz

scp -P 11811 Arabidopsis_thaliana.TAIR10.dna.toplevel.fa.clean.fa xugang@166.111.152.116:/WORK/teaching/project/xuzhiyu/reference
scp -P 11811 Arabidopsis_thaliana.TAIR10.46.gff3.gz xugang@166.111.152.116:/WORK/teaching/project/xuzhiyu/reference

cd /Share/home/xuzhiyu/ribo-seq/CLIP_db/tair10
for i in `find -iname *bam`;do echo $i; scp -P 11811 -r $i xugang@166.111.152.116:/WORK/teaching/project/xuzhiyu; done;

open DATA, "<$ARGV[0]";
open OUT, ">$ARGV[0].clean.fa";
while(<DATA>)
{
#chmop;
if($_=~/^>/)
{
@data=split / /,$_;
print OUT "$data[0]\n";
}else{
print OUT $_;
}
}

samtools fastq bam
```

## 2. Build the rRNA index.

```sh
grep $'\t'rRNA$'\t' Arabidopsis_thaliana.TAIR10.46.gff3 > tair10.rRNA.gtf
 bedtools getfasta -fi Arabidopsis_thaliana.TAIR10.dna.toplevel.fa.clean.fa -bed tair10.rRNA.gtf -fo tair.rRNA.fa
 bowtie-build tair.rRNA.fa tair.rRNA.fa
```

## 3.Build STAR index
```sh
#!/bin/bash
#SBATCH -J tophat_test
#SBATCH -p CN_BIOT
#SBATCH --nodes=1
#SBATCH --ntasks=16
#SBATCH --output=%j.out
#SBATH --error=%j.err

# Get the software
export PATH=/WORK/teaching/bin:$PATH


fasta=/WORK/teaching/project/xuzhiyu/reference/Arabidopsis_thaliana.TAIR10.dna.toplevel.fa.clean.fa
gtf=/WORK/teaching/project/xuzhiyu/reference/Arabidopsis_thaliana.TAIR10.43.gtf
genomeout=/WORK/teaching/project/xuzhiyu/reference/tair_star
[ -d $genomeout ] || mkdir -p $genomeout
STAR --runThreadN 8 --runMode genomeGenerate --genomeDir ${genomeout} --genomeFastaFiles ${fasta} --sjdbGTFfile ${gtf} >log.txt

```


## 4. contam (remove ribosome sequence)

mkdir a6-contam

> a6-contam
**a1.contam.sh**
```sh
echohead(){
   echo '#!/bin/bash
#SBATCH -J '$1'
#SBATCH -p CN_BIOT
#SBATCH --nodes=1
#SBATCH --ntasks=8
#SBATCH --output=%j.out
#SBATH --error=%j.err

# Get the software
export PATH=/WORK/teaching/bin:$PATH
export SINGULARITY_BINDPATH="/WORK"
' 
}

bowtieindex=/WORK/teaching/project/xuzhiyu/reference/tair_rRNA_bowtie_index/tair.rRNA.fa
out=/WORK/teaching/project/xuzhiyu/a2-contam
data_dir=/WORK/teaching/project/xuzhiyu/fastq


[ -d $out ] || mkdir -p $out
[ -d $out/clean ] || mkdir -p $out/clean
[ -d $out/log ] || mkdir -p $out/log
[ -d $out/script ] || mkdir -p $out/script

for i in `ls $data_dir|grep fq$`;
do echo $i;
echohead $i> $out/script/$i.sh;
echo "bowtie -n 0 -norc --best -l 15 -p 7 --un=${out}/clean/nocontam_${i}.fastq $bowtieindex -q $data_dir/$i $out/${i}.alin > $out/log/${i}.err " >> $out/script/$i.sh;
done


```

## a3-map
```sh
echohead(){
   echo '#!/bin/bash
#SBATCH -J '$1'
#SBATCH -p CN_BIOT
#SBATCH --nodes=1
#SBATCH --ntasks=16
#SBATCH --output=%'$1'.out
#SBATH --error=%'$1'.err

# Get the software
export PATH=/WORK/teaching/bin:$PATH
export SINGULARITY_BINDPATH="/WORK"
' 
}



genomeFile=/WORK/teaching/project/xuzhiyu/reference/tair_star
out=/WORK/teaching/project/xuzhiyu/a3-map
data_dir=/WORK/teaching/project/xuzhiyu/a2-contam/clean

[ -d $out ] || mkdir $out 
[ -d $out/script ] || mkdir $out/script
floop(){
for i in `ls $data_dir|grep fastq`;
do echo $i;
echohead $i > $out/script/$i.sh ;
echo "mkdir ${out}/${i}_STAR/" >> $out/script/$i.sh
echo "STAR --runThreadN 8 --alignEndsType EndToEnd --outFilterMismatchNmax 1 --outFilterMultimapNmax 1 --genomeDir $genomeFile --readFilesIn ${data_dir}/${i} --outFileNamePrefix ${out}/${i}_STAR/${i} --outSAMtype BAM SortedByCoordinate --quantMode TranscriptomeSAM GeneCounts" >> $out/script/$i.sh

done
}

upload(){
    for i in `ls $out/script|grep sh`;do
    cd $out;
    echo "$out/script/$i";
    sbatch $out/script/$i
    done;

}

#floop
#upload

```

## 4. RiboCode annotation



```sh

gtf=/WORK/teaching/project/xuzhiyu/reference/Arabidopsis_thaliana.TAIR10.43.gtf
fa=/WORK/teaching/project/xuzhiyu/reference/Arabidopsis_thaliana.TAIR10.dna.toplevel.fa.clean.fa
out=/WORK/teaching/project/xuzhiyu/reference/ribocode

[ -d $out ] || mkdir $out

/WORK/app/singularity-2.0-20191204/bin/singularity exec -B /WORK/teaching/project:/WORK/teaching/project /WORK/teaching/project/20200113luboxun/ribocodeminer /home/test/anaconda3/bin/prepare_transcripts  -g $gtf -f $fa -o $out

```

## 5. change name.

```sh
cd /WORK/teaching/project/xuzhiyu/a3-map

for i in `ls|grep STAR`;
do name=$(echo $i | cut -f 2 -d '_');
echo ${i};
cd $i;
mv Aligned.sortedByCoord.out.bam ${name}.Aligned.sortedByCoord.out.bam;
cd ..;
done;



for i in `ls|grep STAR`;
do name=$(echo $i | cut -f 2 -d '_');
echo ${i};
cd $i;
mv Aligned.toTranscriptome.out.bam ${name}.Aligned.toTranscriptome.out.bam;
cd ..;
done;

```
merge_transcript.sh
```sh
rm -rf /WORK/teaching/project/xuzhiyu/a3-map/b1.merge_transcriptome.txt;
for i in `ls /WORK/teaching/project/xuzhiyu/a3-map|grep STAR`;
do echo $i;
name=$(echo $i | cut -f 2 -d '_');
echo "/WORK/teaching/project/xuzhiyu/a3-map/$i/${name}.Aligned.toTranscriptome.out.bam" >> /WORK/teaching/project/xuzhiyu/a3-map/b1.merge_transcriptome.txt;
done;

```

## a5_metaplot.sh
```sh
#!/bin/bash
#SBATCH -J tophat_test
#SBATCH -p CN_BIOT
#SBATCH --nodes=1
#SBATCH --ntasks=4
#SBATCH --output=%j.out
#SBATH --error=%j.err

# Get the software
export PATH=/WORK/teaching/bin:$PATH

out=/WORK/teaching/project/xuzhiyu/a5-metaplot
rico_ann=/WORK/teaching/project/xuzhiyu/reference/ribocode
merge_transcript=/WORK/teaching/project/xuzhiyu/a3-map/b1.merge_transcriptome.txt

[ -d $out ] || mkdir $out

/WORK/app/singularity-2.0-20191204/bin/singularity exec -B /WORK/teaching/project:/WORK/teaching/project /WORK/teaching/project/20200113luboxun/ribocodeminer /home/test/anaconda3/bin/metaplots -a $rico_ann -i $merge_transcript -o $out/meta > $out/metaplot.log 

```

## 6. Ribocode 

```sh



#!/bin/bash
export PATH=~/anaconda3/bin:$PATH
mkdir b4-RiboCode

rico_ann=/WORK/teaching/project/xuzhiyu/reference/ribocode
out=/WORK/teaching/project/xuzhiyu/a4-ribocode

[ -d $out ] || mkdir $out

config=b3-metaplot/config.txt
nohup RiboCode -a ${rico_ann} -c ${config} -l no -g -o $out/RiboCode_ORFs_result  > b4-RiboCode/ribocode.log 2>&1 &
 
```
