## 1. download genome annotation
```sh
wget ftp://ftp.ensemblgenomes.org/pub/plants/release-46/fasta/arabidopsis_thaliana/dna/Arabidopsis_thaliana.TAIR10.dna.toplevel.fa.gz
wet ftp://ftp.ensemblgenomes.org/pub/plants/release-46/gff3/arabidopsis_thaliana/Arabidopsis_thaliana.TAIR10.46.gff3.gz
wget ftp://ftp.ensemblgenomes.org/pub/plants/release-43/gff3/arabidopsis_thaliana/Arabidopsis_thaliana.TAIR10.43.gff3.gz

scp -P 11811 Arabidopsis_thaliana.TAIR10.dna.toplevel.fa.clean.fa xugang@166.111.152.116:/WORK/teaching/project/xuzhiyu/reference
scp -P 11811 Arabidopsis_thaliana.TAIR10.46.gff3.gz xugang@166.111.152.116:/WORK/teaching/project/xuzhiyu/reference



cd /Share/home/xuzhiyu/ribo-seq/CLIP_db/tair10
for i in `find -iname *bam`;do echo $i; scp -P 11811 -r $i xugang@166.111.152.116:/WORK/teaching/project/xuzhiyu; done;

open DATA, "<$ARGV[0]";
open OUT, ">$ARGV[0].clean.fa";
while(<DATA>)
{
#chmop;
if($_=~/^>/)
{
@data=split / /,$_;
print OUT "$data[0]\n";
}else{
print OUT $_;
}
}

samtools fastq bam
```

## 2. Build the rRNA index.

```sh
grep $'\t'rRNA$'\t' Arabidopsis_thaliana.TAIR10.46.gff3 > tair10.rRNA.gtf
 bedtools getfasta -fi Arabidopsis_thaliana.TAIR10.dna.toplevel.fa.clean.fa -bed tair10.rRNA.gtf -fo tair.rRNA.fa
 bowtie-build tair.rRNA.fa tair.rRNA.fa
```

## 3.Build STAR index
```sh
#!/bin/bash
#SBATCH -J tophat_test
#SBATCH -p CN_BIOT
#SBATCH --nodes=1
#SBATCH --ntasks=16
#SBATCH --output=%j.out
#SBATH --error=%j.err

# Get the software
export PATH=/WORK/teaching/bin:$PATH


fasta=/WORK/teaching/project/xuzhiyu/reference/Arabidopsis_thaliana.TAIR10.dna.toplevel.fa.clean.fa
gtf=/WORK/teaching/project/xuzhiyu/reference/Arabidopsis_thaliana.TAIR10.43.gtf
genomeout=/WORK/teaching/project/xuzhiyu/reference/tair_star
[ -d $genomeout ] || mkdir -p $genomeout
STAR --runThreadN 8 --runMode genomeGenerate --genomeDir ${genomeout} --genomeFastaFiles ${fasta} --sjdbGTFfile ${gtf} >log.txt

```


## 4. contam (remove ribosome sequence)

mkdir a6-contam

> a6-contam
**a1.contam.sh**
```sh
echohead(){
   echo '#!/bin/bash
#SBATCH -J '$1'
#SBATCH -p CN_BIOT
#SBATCH --nodes=1
#SBATCH --ntasks=8
#SBATCH --output=%j.out
#SBATH --error=%j.err

# Get the software
export PATH=/WORK/teaching/bin:$PATH
export SINGULARITY_BINDPATH="/WORK"
' 
}

bowtieindex=/WORK/teaching/project/xuzhiyu/reference/tair_rRNA_bowtie_index/tair.rRNA.fa
out=/WORK/teaching/project/xuzhiyu/a2-contam
data_dir=/WORK/teaching/project/xuzhiyu/fastq


[ -d $out ] || mkdir -p $out
[ -d $out/clean ] || mkdir -p $out/clean
[ -d $out/log ] || mkdir -p $out/log
[ -d $out/script ] || mkdir -p $out/script

for i in `ls $data_dir|grep fq$`;
do echo $i;
echohead $i> $out/script/$i.sh;
echo "bowtie -n 0 -norc --best -l 15 -p 7 --un=${out}/clean/nocontam_${i}.fastq $bowtieindex -q $data_dir/$i $out/${i}.alin > $out/log/${i}.err " >> $out/script/$i.sh;
done


```

## a3-map
```sh
echohead(){
   echo '#!/bin/bash
#SBATCH -J '$1'
#SBATCH -p CN_BIOT
#SBATCH --nodes=1
#SBATCH --ntasks=16
#SBATCH --output=%'$1'.out
#SBATH --error=%'$1'.err

# Get the software
export PATH=/WORK/teaching/bin:$PATH
export SINGULARITY_BINDPATH="/WORK"
' 
}



genomeFile=/WORK/teaching/project/xuzhiyu/reference/tair_star
out=/WORK/teaching/project/xuzhiyu/a3-map
data_dir=/WORK/teaching/project/xuzhiyu/a2-contam/clean

[ -d $out ] || mkdir $out 
[ -d $out/script ] || mkdir $out/script
floop(){
for i in `ls $data_dir|grep fastq`;
do echo $i;
echohead $i > $out/script/$i.sh ;
echo "mkdir ${out}/${i}_STAR/" >> $out/script/$i.sh
echo "STAR --runThreadN 8 --alignEndsType EndToEnd --outFilterMismatchNmax 1 --outFilterMultimapNmax 1 --genomeDir $genomeFile --readFilesIn ${data_dir}/${i} --outFileNamePrefix ${out}/${i}_STAR/${i} --outSAMtype BAM SortedByCoordinate --quantMode TranscriptomeSAM GeneCounts" >> $out/script/$i.sh

done
}

upload(){
    for i in `ls $out/script|grep sh`;do
    cd $out;
    echo "$out/script/$i";
    sbatch $out/script/$i
    done;

}

#floop
#upload

```

## sort and index.

sort_and_index.sh

```sh
echohead(){
   echo '#!/bin/bash
#SBATCH -J '$1'
#SBATCH -p CN_BIOT
#SBATCH --nodes=1
#SBATCH --ntasks=16
#SBATCH --output=%'$1'.out
#SBATH --error=%'$1'.err

# Get the software
export PATH=/WORK/teaching/bin:$PATH
export SINGULARITY_BINDPATH="/WORK"
' 
}



genomeFile=/WORK/teaching/project/xuzhiyu/reference/tair_star
out=/WORK/teaching/project/xuzhiyu/a3-map
data_dir=/WORK/teaching/project/xuzhiyu/a3-map

[ -d $out ] || mkdir $out 
[ -d $out/script ] || mkdir $out/script
floop(){
for i in `ls $data_dir|grep _STAR `;
do 
echo $i;
name=$(echo $i| cut -d '_' -f 2);
echo $name;
echohead $i > $out/script/$i.sort_index.sh
echo "samtools index /WORK/teaching/project/xuzhiyu/a3-map/$i/${name}.Aligned.sortedByCoord.out.bam" >> $out/script/$i.sort_index.sh;
echo "samtools sort -T /WORK/teaching/project/xuzhiyu/a3-map/$i/${name}.Aligned.toTranscriptome.sort -o /WORK/teaching/project/xuzhiyu/a3-map/$i/${name}.Aligned.toTranscriptome.sort.bam /WORK/teaching/project/xuzhiyu/a3-map/$i/${name}.Aligned.toTranscriptome.out.bam" >> $out/script/$i.sort_index.sh;
echo "samtools index /WORK/teaching/project/xuzhiyu/a3-map/$i/${name}.Aligned.toTranscriptome.sort.bam" >> $out/script/$i.sort_index.sh;
done
}

upload(){
    for i in `ls $out/script|grep _STAR.sort_index.sh`;do
    cd $out;
    echo "$out/script/$i";
    sbatch $out/script/$i
    done;

}

#floop
upload

```


## 4. RiboCode annotation



```sh

gtf=/WORK/teaching/project/xuzhiyu/reference/Arabidopsis_thaliana.TAIR10.43.gtf
fa=/WORK/teaching/project/xuzhiyu/reference/Arabidopsis_thaliana.TAIR10.dna.toplevel.fa.clean.fa
out=/WORK/teaching/project/xuzhiyu/reference/ribocode

[ -d $out ] || mkdir $out

/WORK/app/singularity-2.0-20191204/bin/singularity exec -B /WORK/teaching/project:/WORK/teaching/project /WORK/teaching/project/20200113luboxun/ribocodeminer /home/test/anaconda3/bin/prepare_transcripts  -g $gtf -f $fa -o $out

```

## 5. change name.

```sh
cd /WORK/teaching/project/xuzhiyu/a3-map

for i in `ls|grep STAR`;
do name=$(echo $i | cut -f 2 -d '_');
echo ${i};
cd $i;
mv Aligned.sortedByCoord.out.bam ${name}.Aligned.sortedByCoord.out.bam;
cd ..;
done;



for i in `ls|grep STAR`;
do name=$(echo $i | cut -f 2 -d '_');
echo ${i};
cd $i;
mv Aligned.toTranscriptome.out.bam ${name}.Aligned.toTranscriptome.out.bam;
cd ..;
done;

```
merge_transcript.sh
```sh
rm -rf /WORK/teaching/project/xuzhiyu/a3-map/b1.merge_transcriptome.txt;
for i in `ls /WORK/teaching/project/xuzhiyu/a3-map|grep STAR`;
do echo $i;
name=$(echo $i | cut -f 2 -d '_');
echo "/WORK/teaching/project/xuzhiyu/a3-map/$i/${name}.Aligned.toTranscriptome.out.bam" >> /WORK/teaching/project/xuzhiyu/a3-map/b1.merge_transcriptome.txt;
done;

```

## a5_metaplot.sh
```sh
#!/bin/bash
#SBATCH -J tophat_test
#SBATCH -p CN_BIOT
#SBATCH --nodes=1
#SBATCH --ntasks=4
#SBATCH --output=%j.out
#SBATH --error=%j.err

# Get the software
export PATH=/WORK/teaching/bin:$PATH

out=/WORK/teaching/project/xuzhiyu/a5-metaplot
rico_ann=/WORK/teaching/project/xuzhiyu/reference/ribocode
merge_transcript=/WORK/teaching/project/xuzhiyu/a3-map/b1.merge_transcriptome.txt

[ -d $out ] || mkdir $out

/WORK/app/singularity-2.0-20191204/bin/singularity exec -B /WORK/teaching/project:/WORK/teaching/project /WORK/teaching/project/20200113luboxun/ribocodeminer /home/test/anaconda3/bin/metaplots -a $rico_ann -i $merge_transcript -o $out/meta > $out/metaplot.log 

```

## 6. Ribocode 
a6_ribocode.sh

```sh
#!/bin/bash
#SBATCH -J tophat_test
#SBATCH -p CN_BIOT
#SBATCH --nodes=1
#SBATCH --ntasks=4
#SBATCH --output=%j.out
#SBATH --error=%j.err

# Get the software
export PATH=/WORK/teaching/bin:$PATH

out=/WORK/teaching/project/xuzhiyu/a7-ribocode-test
rico_ann=/WORK/teaching/project/xuzhiyu/reference/ribocode
merge_transcript=/WORK/teaching/project/xuzhiyu/a3-map/b1.merge_transcriptome.txt

[ -d $out ] || mkdir $out

#config=/WORK/teaching/project/xuzhiyu/a5-metaplot/meta_pre_config_test.txt
config=/WORK/teaching/project/xuzhiyu/a5-metaplot/meta_pre_config_refine.txt

/WORK/app/singularity-2.0-20191204/bin/singularity exec -B /WORK/teaching/project:/WORK/teaching/project /WORK/teaching/project/20200113luboxun/ribocodeminer /home/test/anaconda3/bin/RiboCode -a ${rico_ann} -c ${config} -l no -g -o $out/RiboCode_ORFs_result
```

a8-ribominer.sh

```sh
#!/bin/bash
#SBATCH -J tophat_test
#SBATCH -p CN_BIOT
#SBATCH --nodes=1
#SBATCH --ntasks=4
#SBATCH --output=%j.out
#SBATH --error=%j.err

# Get the software
export PATH=/WORK/teaching/bin:$PATH

gtf=/WORK/teaching/project/xuzhiyu/reference/Arabidopsis_thaliana.TAIR10.43.gtf
fa=/WORK/teaching/project/xuzhiyu/reference/Arabidopsis_thaliana.TAIR10.dna.toplevel.fa.clean.fa
ribocode=/WORK/teaching/project/xuzhiyu/a8-Ribocode
ribominer=/WORK/teaching/project/xuzhiyu/a9-RiboMiner

# Prepare sequences and annotaiton files on transcriptome level.
echo "/WORK/app/singularity-2.0-20191204/bin/singularity exec -B /WORK/teaching/project:/WORK/teaching/project /WORK/teaching/project/singularity_images/ribocodeminer /home/test/anaconda3/bin/prepare_transcripts -g $gtf -f $fa -o ${ribocode}"
/WORK/app/singularity-2.0-20191204/bin/singularity exec -B /WORK/teaching/project:/WORK/teaching/project /WORK/teaching/project/singularity_images/ribocodeminer /home/test/anaconda3/bin/prepare_transcripts -g $gtf -f $fa -o ${ribocode}

# Prepare the longest transcript annotaion files.

cds=${ribocode}/transcripts_cds.txt
trans=${ribocode}/transcripts_sequence.fa

longest_info=${ribominer}/longest.transcripts.info.txt
all_info=${ribominer}/all.transcripts.info.txt
mkdir ${ribominer}
echo "OutputTranscriptInfo -c $cds -g $gff -f $trans -o $longest_info -O $all_info "
/WORK/app/singularity-2.0-20191204/bin/singularity exec -B /WORK/teaching/project:/WORK/teaching/project /WORK/teaching/project/singularity_images/ribocodeminer /home/test/anaconda3/bin/OutputTranscriptInfo -c $cds -g $gtf -f $trans -o $longest_info -O $all_info 

# Prepare the sequence file for the longest transcripts
transcripts_sequence=${ribocode}/transcripts_sequence.fa
longest_info=${ribominer}/longest.transcripts.info.txt
transcript=${ribominer}/transcript

echo "GetProteinCodingSequence -i $transcripts_sequence  -c $longest_info -o $transcript --mode whole --table 1"
/WORK/app/singularity-2.0-20191204/bin/singularity exec -B /WORK/teaching/project:/WORK/teaching/project /WORK/teaching/project/singularity_images/ribocodeminer /home/test/anaconda3/bin/GetProteinCodingSequence -i $transcripts_sequence  -c $longest_info -o $transcript --mode whole --table 1

# Sometines, UTR sequences are needed. In this case, GetUTRSequences maybe helpful:
transcripts_sequence=${ribocode}/transcripts_sequence.fa
utr=${ribominer}/utr
transcript_cds=${ribocode}/transcripts_cds.txt

echo "GetUTRSequences -i $transcripts_sequence -o $utr -c $transcript_cds"
/WORK/app/singularity-2.0-20191204/bin/singularity exec -B /WORK/teaching/project:/WORK/teaching/project /WORK/teaching/project/singularity_images/ribocodeminer /home/test/anaconda3/bin/GetUTRSequences -i $transcripts_sequence -o $utr -c $transcript_cds

```

## update the singularity images.

```sh
sudo singularity  shell -w -B /home/xugang/singularity_image/xuzhiyu/:/work /home/xugang/singularity_image/ribocodeminer
pip install --upgrade RiboMiner
exit
sudo tar -cvf ribocodeminer_new_v2.tar ribocodeminer
scp -P 11811 ribocodeminer_new_v2.tar xugang@166.111.152.116:/WORK/teaching/project/singularity_images

```

**b.qualitycontrol.sh** 
b1.qualitycontrol.sh

```sh


g1='upf2-air,upf2-ethylene'
r1='upf2-Air-RFP1,upf2-Air-RFP2__upf2-Ethylene-RFP1,upf2-Ethylene-RFP2'
o1='upf2-air_upf2-ethylene'

g2='col-air,col-ethylene'
r2='Col-Air-RFP1,Col-Air-RFP2__Col-Ethylene-RFP1,Col-Ethylene-RFP2'
o2='col-air_col-ethylene'

g3='col-ethylene,upf2-ethylene'
r3='Col-Ethylene-RFP1,Col-Ethylene-RFP2__upf2-Ethylene-RFP1,upf2-Ethylene-RFP2'
o3='col-ethylene_upf2-ethylene'

g4='root,shoot'
r4='riboseq_root_1,riboseq_root_2,riboseq_root_3__riboseq_shoot_1,riboseq_shoot_2,riboseq_shoot_3'
o4='root_shoot'

g5='dark,light'
r5='mRNARP-Dark,mRNARP-Dark_2__mRNARP-L4h,mRNARP-L4h_2'
o5='dark_light'

g6='no-stress,stress'
r6='Total_no_stress2__Total_stress2'
o6='no-stress_stress'

g7='col,d14'
r7='col__d14'
o7='col_d14'

echohead(){
   echo '#!/bin/bash
#SBATCH -J slum
#SBATCH -p CN_BIOT
#SBATCH --nodes=1
#SBATCH --ntasks=3
#SBATCH --output=%slum.out
#SBATH --error=%slum.err

# Get the software
export PATH=/WORK/teaching/bin:$PATH
export SINGULARITY_BINDPATH="/WORK"
' 
}

echo "bam must be sort and index."


#Periodicity checking
#Ribosome profiling data with a good quality tend to have a good 3-nt periodicity.
ribocode=/WORK/teaching/project/xuzhiyu/a8-Ribocode
ribominer=/WORK/teaching/project/xuzhiyu/a9-RiboMiner

out='/WORK/teaching/project/xuzhiyu/b1-qualitycontrol'
quality_out=$out
meta_out=$quality_out'/metaplot'
a111R='/home/share/riboseq/a9-STAR/7-111-R_STAR/7-111-R.toTranscriptome.sort.bam'
a111Rgenome='/home/share/riboseq/a9-STAR/7-111-R_STAR/7-111-RAligned.sortedByCoord.out.bam'
a111R_o=$meta_out'/7111r'
a7R='/home/share/riboseq/a9-STAR/7-7-R_STAR/7-7-R.toTranscriptome.sort.bam'
a7Rgenome='/home/share/riboseq/a9-STAR/7-7-R_STAR/7-7-RAligned.sortedByCoord.out.bam'
a7R_o=$meta_out'/77r'
long=${ribominer}'/longest.transcripts.info.txt'
periodicity_out=$quality_out'/a5-periodicity'
a111R_periodicity_o=$periodicity_out'/7111r'
a7R_periodicity_o=$periodicity_out'/77r'
config=/WORK/teaching/project/xuzhiyu/a5-metaplot/meta_pre_config_refine.txt

echo $meta_out

gtf=/WORK/teaching/project/xuzhiyu/reference/Arabidopsis_thaliana.TAIR10.43.gtf

[ -d $out ] || mkdir $out

generate_attribute()
{
awk '{print $2"\t"$4"\t"$5"\t"$1}' ${config}  > ${out}/attributes2.txt
cat ${out}/attributes2.txt | sed 's/Aligned.toTranscriptome.out.bam/Aligned.toTranscriptome.sort.bam/g' > ${out}/attributes.txt
cut -f 1  ${out}/attributes.txt > ${out}/transcript.list2
cut -f 1  ${out}/attributes.txt | sed 's/Aligned.toTranscriptome.sort.bam/Aligned.sortedByCoord.out.bam/g' > ${out}/genome.list2
sed '1d' ${out}/genome.list2 > ${out}/genome.list
sed '1d' ${out}/transcript.list2 > ${out}/transcript.list
rm ${out}/genome.list2 ${out}/transcript.list2 ${out}/attributes2.txt
}

periodicity()
{
echo "Periodicity start `date`"
[ -d ${periodicity_out} ] || mkdir ${periodicity_out}
[ -d ${periodicity_out}/script ] || mkdir ${periodicity_out}/script

for i in `cat ${out}/transcript.list`; 
do 
file_name=`echo $i|rev |cut -f 1 -d '/'|rev|sed 's/\.Aligned.toTranscriptome.sort.bam//g'`;
echohead > ${periodicity_out}/script/${file_name}.sh
echo "/WORK/app/singularity-2.0-20191204/bin/singularity exec -B /WORK/teaching/project:/WORK/teaching/project /WORK/teaching/project/20200113luboxun/ribocodeminer /home/test/anaconda3/bin/Periodicity -i ${i} -a $ribocode -o ${periodicity_out}/$file_name -c $long -L 25 -R 35" >> ${periodicity_out}/script/${file_name}.sh;

#sbatch ${periodicity_out}/script/${file_name}.sh;
done;

echo "Periodicity end `date`"
}



ribodensity()
{
echo "RiboDensityOfDiffFrames begin `date`"
#Reads distribution among different reading frames.
echohead > $out/a6-ribodensity.sh
echo -e "/WORK/app/singularity-2.0-20191204/bin/singularity exec -B /WORK/teaching/project:/WORK/teaching/project /WORK/teaching/project/20200113luboxun/ribocodeminer  /home/test/anaconda3/bin/RiboDensityOfDiffFrames -f ${out}/attributes.txt -c $long -o $out/a6-ribo-density-diff-fram" >> $out/a6-ribodensity.sh;
#sbatch $out/a6-ribodensity.sh;
echo "RiboDensityOfDiffFrames end `date`"
}

length_dis()
{
#Length distribution.
echo "LengthDistribution begin `date`"
echo -e "LengthDistribution -i $a111Rgenome -o $quality_out/r111.length -f bam"
[ -d ${out}/length ] || mkdir -p ${out}/length
[ -d ${out}/length/script ] || mkdir -p ${out}/length/script
for i in `cat ${out}/genome.list`; 
do 
file_name=`echo $i|rev |cut -f 1 -d '/'|rev|sed 's/\.Aligned.sortedByCoord.out.bam//g'`;
echohead > ${out}/length/script/${file_name}.sh
echo "/WORK/app/singularity-2.0-20191204/bin/singularity exec -B /WORK/teaching/project:/WORK/teaching/project /WORK/teaching/project/singularity_images/ribocodeminer /home/test/anaconda3/bin/LengthDistribution -i ${i}  -o ${out}/length/$file_name -f bam" >> ${out}/length/script/${file_name}.sh;
#sbatch ${out}/length/script/${file_name}.sh;
done;

echo "LengthDistribution end `date`"
}

DNA_contam()
{
#DNA contamination.
[ -d ${out}/dna_contam ] || mkdir -p ${out}/dna_contam
[ -d ${out}/dna_contam/script ] || mkdir -p ${out}/dna_contam/script
echo "StatisticReadsOnDNAsContam begin `date`"
echo "StatisticReadsOnDNAsContam -i $a111Rgenome  -g $gtf -o $out/a7-dna-contamination.111"
for i in `cat ${out}/genome.list`; 
do 
file_name=`echo $i|rev |cut -f 1 -d '/'|rev |sed 's/\.Aligned.sortedByCoord.out.bam//g'`;
echohead > ${out}/dna_contam/script/${file_name}.sh
echo "/WORK/app/singularity-2.0-20191204/bin/singularity exec -B /WORK/teaching/project:/WORK/teaching/project /WORK/teaching/project/singularity_images/ribocodeminer /home/test/anaconda3/bin/StatisticReadsOnDNAsContam -i ${i} -g $gtf -o ${out}/dna_contam/$file_name" >> ${out}/dna_contam/script/${file_name}.sh;
#sbatch ${out}/dna_contam/script/${file_name}.sh;
done;


echo "StatisticReadsOnDNAsContam end `date`"

}


#generate_attribute
#periodicity
#ribodensity
#length_dis
#DNA_contam

```

b2_metagene_analysis.sh


```sh

echohead(){
   echo '#!/bin/bash
#SBATCH -J slum
#SBATCH -p CN_BIOT
#SBATCH --nodes=1
#SBATCH --ntasks=3
#SBATCH --output=%slum.out
#SBATH --error=%slum.err

# Get the software
export PATH=/WORK/teaching/bin:$PATH
export SINGULARITY_BINDPATH="/WORK"
' 
}

g1='upf2-air,upf2-ethylene'
r1='upf2-Air-RFP1,upf2-Air-RFP2__upf2-Ethylene-RFP1,upf2-Ethylene-RFP2'
o1='upf2-air_upf2-ethylene'

g2='col-air,col-ethylene'
r2='Col-Air-RFP1,Col-Air-RFP2__Col-Ethylene-RFP1,Col-Ethylene-RFP2'
o2='col-air_col-ethylene'

g3='col-ethylene,upf2-ethylene'
r3='Col-Ethylene-RFP1,Col-Ethylene-RFP2__upf2-Ethylene-RFP1,upf2-Ethylene-RFP2'
o3='col-ethylene_upf2-ethylene'

g4='root,shoot'
r4='riboseq_root_1,riboseq_root_2,riboseq_root_3__riboseq_shoot_1,riboseq_shoot_2,riboseq_shoot_3'
o4='root_shoot'

g5='dark,light'
r5='mRNARP-Dark,mRNARP-Dark_2__mRNARP-L4h,mRNARP-L4h_2'
o5='dark_light'

g6='no-stress,stress'
r6='Total_no_stress2__Total_stress2'
o6='no-stress_stress'

g7='col,d14'
r7='col__d14'
o7='col_d14'

g8='upf2-air,upf2-ethylene,col-air,col-ethylene,d14'
r8='upf2-Air-RFP1,upf2-Air-RFP2__upf2-Ethylene-RFP1,upf2-Ethylene-RFP2__Col-Air-RFP1,Col-Air-RFP2__Col-Ethylene-RFP1,Col-Ethylene-RFP2__d14'
o8='mulitp1'

g9='upf2-air,upf2-ethylene,col-air,col-ethylene,root,shoot,dark,light,no-stress,stress,d14'
r9='upf2-Air-RFP1,upf2-Air-RFP2__upf2-Ethylene-RFP1,upf2-Ethylene-RFP2__Col-Air-RFP1,Col-Air-RFP2__Col-Ethylene-RFP1,Col-Ethylene-RFP2__riboseq_root_1,riboseq_root_2,riboseq_root_3__riboseq_shoot_1,riboseq_shoot_2,riboseq_shoot_3__mRNARP-Dark,mRNARP-Dark_2__mRNARP-L4h,mRNARP-L4h_2__Total_no_stress2__Total_stress2__d14'
o9='mulitp'

g10='root,shoot,dark,light,no-stress,stress'
r10='riboseq_root_1,riboseq_root_2,riboseq_root_3__riboseq_shoot_1,riboseq_shoot_2,riboseq_shoot_3__mRNARP-Dark,mRNARP-Dark_2__mRNARP-L4h,mRNARP-L4h_2__Total_no_stress2__Total_stress2'
o10='mulitp2'

meta_out='/WORK/teaching/project/xuzhiyu/b1-qualitycontrol/'
ribocode='/WORK/teaching/project/xuzhiyu/a8-Ribocode'
long='/WORK/teaching/project/xuzhiyu/a9-RiboMiner/longest.transcripts.info.txt'
out='/WORK/teaching/project/xuzhiyu/b2-metagene_analysis'
groupinfo='R111,R7'
replace='7-111-R.toTranscriptome.sort__7-7-R.toTranscriptome.sort'
attribute='/WORK/teaching/project/xuzhiyu/b1-qualitycontrol/attributes.txt'



[ -d $out ] || mkdir -p $out
[ -d $out/script ] || mkdir -p $out/script

wholeregion(){
echo "Name must from MetageneAnalysisForTheWholeRegions output"
echo "Start Metagene analysis along the whole transcript region. `date`"
echohead > $out/script/meta.sh
echo "/WORK/app/singularity-2.0-20191204/bin/singularity exec -B /WORK/teaching/project:/WORK/teaching/project /WORK/teaching/project/singularity_images/ribocodeminer /home/test/anaconda3/bin/MetageneAnalysisForTheWholeRegions -f ${attribute} -c $long -o $out/a8-metagene -b 15,90,60 -l 100 -n 10 -m 1 -e 5 --plot yes"  >> $out/script/meta.sh
#/WORK/app/singularity-2.0-20191204/bin/singularity exec -B /WORK/teaching/project:/WORK/teaching/project /WORK/teaching/project/singularity_images/ribocodeminer /home/test/anaconda3/bin/MetageneAnalysisForTheWholeRegions -f ${attribute} -c $long -o $out/a8-metagene_${o1} -b 15,90,60 -l 100 -n 10 -m 1 -e 5 --plot yes

echohead > $out/script/plot1.sh
echo "/WORK/app/singularity-2.0-20191204/bin/singularity exec -B /WORK/teaching/project:/WORK/teaching/project /WORK/teaching/project/singularity_images/ribocodeminer /home/test/anaconda3/bin/PlotMetageneAnalysisForTheWholeRegions -i $out/a8-metagene_scaled_density_dataframe.txt -o $out/a9-meta_gene_whole_regin_${o1} -g ${g1} -r ${r1} -b 15,90,60 --mode all" >> $out/script/plot1.sh

echohead > $out/script/plot2.sh
echo "/WORK/app/singularity-2.0-20191204/bin/singularity exec -B /WORK/teaching/project:/WORK/teaching/project /WORK/teaching/project/singularity_images/ribocodeminer /home/test/anaconda3/bin/PlotMetageneAnalysisForTheWholeRegions -i $out/a8-metagene_scaled_density_dataframe.txt -o $out/a9-meta_gene_whole_regin_${o2} -g ${g2} -r ${r2} -b 15,90,60 --mode all" >> $out/script/plot2.sh

echohead > $out/script/plot3.sh
echo "/WORK/app/singularity-2.0-20191204/bin/singularity exec -B /WORK/teaching/project:/WORK/teaching/project /WORK/teaching/project/singularity_images/ribocodeminer /home/test/anaconda3/bin/PlotMetageneAnalysisForTheWholeRegions -i $out/a8-metagene_scaled_density_dataframe.txt -o $out/a9-meta_gene_whole_regin_${o3} -g ${g3} -r ${r3} -b 15,90,60 --mode all" >> $out/script/plot3.sh

echohead > $out/script/plot4.sh
echo "/WORK/app/singularity-2.0-20191204/bin/singularity exec -B /WORK/teaching/project:/WORK/teaching/project /WORK/teaching/project/singularity_images/ribocodeminer /home/test/anaconda3/bin/PlotMetageneAnalysisForTheWholeRegions -i $out/a8-metagene_scaled_density_dataframe.txt -o $out/a9-meta_gene_whole_regin_${o4} -g ${g4} -r ${r4} -b 15,90,60 --mode all" >> $out/script/plot4.sh

echohead > $out/script/plot5.sh
echo "/WORK/app/singularity-2.0-20191204/bin/singularity exec -B /WORK/teaching/project:/WORK/teaching/project /WORK/teaching/project/singularity_images/ribocodeminer /home/test/anaconda3/bin/PlotMetageneAnalysisForTheWholeRegions -i $out/a8-metagene_scaled_density_dataframe.txt -o $out/a9-meta_gene_whole_regin_${o5} -g ${g5} -r ${r5} -b 15,90,60 --mode all" >> $out/script/plot5.sh

echohead > $out/script/plot6.sh
echo "/WORK/app/singularity-2.0-20191204/bin/singularity exec -B /WORK/teaching/project:/WORK/teaching/project /WORK/teaching/project/singularity_images/ribocodeminer /home/test/anaconda3/bin/PlotMetageneAnalysisForTheWholeRegions -i $out/a8-metagene_scaled_density_dataframe.txt -o $out/a9-meta_gene_whole_regin_${o6} -g ${g6} -r ${r6} -b 15,90,60 --mode all" >> $out/script/plot6.sh

echohead > $out/script/plot7.sh
echo "/WORK/app/singularity-2.0-20191204/bin/singularity exec -B /WORK/teaching/project:/WORK/teaching/project /WORK/teaching/project/singularity_images/ribocodeminer /home/test/anaconda3/bin/PlotMetageneAnalysisForTheWholeRegions -i $out/a8-metagene_scaled_density_dataframe.txt -o $out/a9-meta_gene_whole_regin_${o7} -g ${g7} -r ${r7} -b 15,90,60 --mode all" >> $out/script/plot7.sh

echo "End Metagene analysis along the whole transcript region. `date`"
}

cdsregion()
{
echo "Begin Metagene analysis on CDS regions. `date`";
#Metagene analysis on CDS regions.
echohead > $out/script/cdsregion.sh
echo "/WORK/app/singularity-2.0-20191204/bin/singularity exec -B /WORK/teaching/project:/WORK/teaching/project /WORK/teaching/project/singularity_images/ribocodeminer /home/test/anaconda3/bin/MetageneAnalysis -f ${attribute} -c $long -o $out/b1-meat-cds_noNorm -U codon -M RPKM -u 0 -d 500 -l 100 -n 10 -m 1 -e 5 --norm no -y 100 --CI 0.95 --type CDS" >> $out/script/cdsregion.sh


echo "End Metagene analysis on CDS regions. `date`";

}

utrregion(){

echo "Begin Metagene analysis on UTR regions. `date`"
#Metagene analysis on UTR regions.
echohead > $out/script/utrregion.sh
echo " /WORK/app/singularity-2.0-20191204/bin/singularity exec -B /WORK/teaching/project:/WORK/teaching/project /WORK/teaching/project/singularity_images/ribocodeminer /home/test/anaconda3/bin/MetageneAnalysis -f ${attribute} -c $long -o $out/b2-meat-utr_nonNorm -U nt -M RPKM -u 100 -d 100 -l 100 -n 10 -m 1 -e 5 --norm no -y 50 --CI 0.95 --type UTR" >> $out/script/utrregion.sh

echohead > $out/script/utrregion_plot.sh
echo " /WORK/app/singularity-2.0-20191204/bin/singularity exec -B /WORK/teaching/project:/WORK/teaching/project /WORK/teaching/project/singularity_images/ribocodeminer /home/test/anaconda3/bin/PlotMetageneAnalysis -i $out/b2-meat-utr_nonNorm_dataframe.txt  -o $out/b2-meat-urt-${o8} -u 50 -d 50 -g ${g8} -r ${r8} -U codon --CI 0.95" >> $out/script/utrregion_plot.sh
echo " /WORK/app/singularity-2.0-20191204/bin/singularity exec -B /WORK/teaching/project:/WORK/teaching/project /WORK/teaching/project/singularity_images/ribocodeminer /home/test/anaconda3/bin/PlotMetageneAnalysis -i $out/b2-meat-utr_nonNorm_dataframe.txt  -o $out/b2-meat-urt-${o9} -u 50 -d 50 -g ${g9} -r ${r9} -U codon --CI 0.95" >> $out/script/utrregion_plot.sh
echo " /WORK/app/singularity-2.0-20191204/bin/singularity exec -B /WORK/teaching/project:/WORK/teaching/project /WORK/teaching/project/singularity_images/ribocodeminer /home/test/anaconda3/bin/PlotMetageneAnalysis -i $out/b2-meat-utr_nonNorm_dataframe.txt  -o $out/b2-meat-urt-${o10} -u 50 -d 50 -g ${g10} -r ${r10} -U codon --CI 0.95" >> $out/script/utrregion_plot.sh


echo "End Metagene analysis on UTR regions. `date`"
}

polarity()
{
echo "Begin Polarity calculation. `date`"
#Polarity calculation.
echohead > $out/script/polarity.sh
echo "/WORK/app/singularity-2.0-20191204/bin/singularity exec -B /WORK/teaching/project:/WORK/teaching/project /WORK/teaching/project/singularity_images/ribocodeminer /home/test/anaconda3/bin/PolarityCalculation -f ${meta_out}/attributes.txt -c $long -o $out/b3-polarity -n 64" >> $out/script/polarity.sh
echohead > $out/script/polarity_plot.sh

echo "/WORK/app/singularity-2.0-20191204/bin/singularity exec -B /WORK/teaching/project:/WORK/teaching/project /WORK/teaching/project/singularity_images/ribocodeminer /home/test/anaconda3/bin/PlotPolarity -i $out/b3-polarity_polarity_dataframe.txt -g ${g8} -r ${r8} -o $out/b3-polarity-${o8}-4 -y 10" >> $out/script/polarity_plot.sh

echo "/WORK/app/singularity-2.0-20191204/bin/singularity exec -B /WORK/teaching/project:/WORK/teaching/project /WORK/teaching/project/singularity_images/ribocodeminer /home/test/anaconda3/bin/PlotPolarity -i $out/b3-polarity_polarity_dataframe.txt -g ${g9} -r ${r9} -o $out/b3-polarity-${o9}-4 -y 10" >> $out/script/polarity_plot.sh


echo "/WORK/app/singularity-2.0-20191204/bin/singularity exec -B /WORK/teaching/project:/WORK/teaching/project /WORK/teaching/project/singularity_images/ribocodeminer /home/test/anaconda3/bin/PlotPolarity -i $out/b3-polarity_polarity_dataframe.txt -g ${g10} -r ${r10} -o $out/b3-polarity-${o10}-4 -y 10" >> $out/script/polarity_plot.sh


echo "End Polarity calculation. `date`"
}

#wholeregion
#cdsregion
utrregion
#polarity



```

b3.generateList.sh

```sh
out=/WORK/teaching/project/xuzhiyu/b3-featureAnalysis
[ -d $out ] || mkdir $out
grep '>' /WORK/teaching/project/xuzhiyu/a9-RiboMiner/transcript_cds_sequences.fa | sed 's/>//g'| cut -f 1 -d ' '>$out/total_transcript.txt

echo "motifs
QQQ
PPP
PPD
DDP" > $out/tri_AA_motifs1.txt;
echo "motifs
QQQ
KKK
KKP
RRR" > $out/tri_AA_motifs2.txt;
```

b3_feature_analysis.sh

```sh


echohead(){
   echo '#!/bin/bash
#SBATCH -J slum
#SBATCH -p CN_BIOT
#SBATCH --nodes=1
#SBATCH --ntasks=3
#SBATCH --output=%slum.out
#SBATH --error=%slum.err

# Get the software
export PATH=/WORK/teaching/bin:$PATH
export SINGULARITY_BINDPATH="/WORK"
' 
}

g1='upf2-air,upf2-ethylene'
r1='upf2-Air-RFP1,upf2-Air-RFP2__upf2-Ethylene-RFP1,upf2-Ethylene-RFP2'
o1='upf2-air_upf2-ethylene'

g2='col-air,col-ethylene'
r2='Col-Air-RFP1,Col-Air-RFP2__Col-Ethylene-RFP1,Col-Ethylene-RFP2'
o2='col-air_col-ethylene'

g3='col-ethylene,upf2-ethylene'
r3='Col-Ethylene-RFP1,Col-Ethylene-RFP2__upf2-Ethylene-RFP1,upf2-Ethylene-RFP2'
o3='col-ethylene_upf2-ethylene'

g4='root,shoot'
r4='riboseq_root_1,riboseq_root_2,riboseq_root_3__riboseq_shoot_1,riboseq_shoot_2,riboseq_shoot_3'
o4='root_shoot'

g5='dark,light'
r5='mRNARP-Dark,mRNARP-Dark_2__mRNARP-L4h,mRNARP-L4h_2'
o5='dark_light'

g6='no-stress,stress'
r6='Total_no_stress2__Total_stress2'
o6='no-stress_stress'

g7='col,d14'
r7='col__d14'
o7='col_d14'

attribute='/WORK/teaching/project/xuzhiyu/b1-qualitycontrol/attributes.txt'

out='/WORK/teaching/project/xuzhiyu/b3-featureAnalysis'
meta_out='/WORK/teaching/project/xuzhiyu/b1-qualitycontrol/'

RiboMiner='/WORK/teaching/project/xuzhiyu/a9-RiboMiner'
long=$RiboMiner'/longest.transcripts.info.txt'
select_gene='/WORK/teaching/project/xuzhiyu/b3-featureAnalysis/total_transcript.txt'
home_dir='/WORK/teaching/project/xuzhiyu/'

total='/WORK/teaching/project/xuzhiyu/b3-featureAnalysis/total_transcript.txt'

[  -d $out ] ||  mkdir $out
[  -d $out/script ] ||  mkdir $out/script

densityspecific()
{
echo "Begin:Pick out transcripts enriched ribosomes on specific region `date`"
# Pick out transcripts enriched ribosomes on specific region.
echohead > ${out}/script/ribodensity.sh
echo "/WORK/app/singularity-2.0-20191204/bin/singularity exec -B /WORK/teaching/project:/WORK/teaching/project /WORK/teaching/project/singularity_images/ribocodeminer /home/test/anaconda3/bin/RiboDensityForSpecificRegion -f ${attribute} -c $long -o $out/b5-transcript-enrich -U codon -M RPKM -L 25 -R 75" >> ${out}/script/ribodensity.sh

echo "End: Pick out transcripts enriched ribosomes on specific region `date`"
}

densitycodon()
{
echo "Begin: Ribosome density at each kind of AA or codon. `date`"
# Ribosome density at each kind of AA or codon.
echohead > ${out}/script/densitycodon.sh
echo " /WORK/app/singularity-2.0-20191204/bin/singularity exec -B /WORK/teaching/project:/WORK/teaching/project /WORK/teaching/project/singularity_images/ribocodeminer /home/test/anaconda3/bin/RiboDensityAtEachKindAAOrCodon -f ${attribute} -c $long -o $out/b6_ -M counts -S ${select_gene} -l 100 -n 10 --table 1 -F $RiboMiner/transcript_cds_sequences.fa " >> ${out}/script/densitycodon.sh
#RiboDensityAtEachKindAAOrCodon -f $meta_out/attributes.txt -c $long -o $out/b6_${1} -M counts -S ${home_dir}/${1} -l 100 -n 10 --table 1 -F $RiboMiner/transcript_cds_sequences.fa
echohead > ${out}/script/densitycodonp1.sh
echo "/WORK/app/singularity-2.0-20191204/bin/singularity exec -B /WORK/teaching/project:/WORK/teaching/project /WORK/teaching/project/singularity_images/ribocodeminer /home/test/anaconda3/bin/PlotRiboDensityAtEachKindAAOrCodon -i $out/b6_all_codon_density.txt -o $out/b7-PlotRiboDensityAtEachKindAAOrCodon_${o1} -g $g1 -r $r1 --level AA" >> ${out}/script/densitycodonp1.sh

echohead > ${out}/script/densitycodonp2.sh
echo "/WORK/app/singularity-2.0-20191204/bin/singularity exec -B /WORK/teaching/project:/WORK/teaching/project /WORK/teaching/project/singularity_images/ribocodeminer /home/test/anaconda3/bin/PlotRiboDensityAtEachKindAAOrCodon -i $out/b6_all_codon_density.txt -o $out/b7-PlotRiboDensityAtEachKindAAOrCodon_${o2} -g $g2 -r $r2 --level AA" >> ${out}/script/densitycodonp2.sh

echohead > ${out}/script/densitycodonp3.sh
echo "/WORK/app/singularity-2.0-20191204/bin/singularity exec -B /WORK/teaching/project:/WORK/teaching/project /WORK/teaching/project/singularity_images/ribocodeminer /home/test/anaconda3/bin/PlotRiboDensityAtEachKindAAOrCodon -i $out/b6_all_codon_density.txt -o $out/b7-PlotRiboDensityAtEachKindAAOrCodon_${o3} -g $g3 -r $r3 --level AA" >> ${out}/script/densitycodonp3.sh

echohead > ${out}/script/densitycodonp4.sh
echo "/WORK/app/singularity-2.0-20191204/bin/singularity exec -B /WORK/teaching/project:/WORK/teaching/project /WORK/teaching/project/singularity_images/ribocodeminer /home/test/anaconda3/bin/PlotRiboDensityAtEachKindAAOrCodon -i $out/b6_all_codon_density.txt -o $out/b7-PlotRiboDensityAtEachKindAAOrCodon_${o4} -g $g4 -r $r4 --level AA" >> ${out}/script/densitycodonp4.sh

echohead > ${out}/script/densitycodonp5.sh
echo "/WORK/app/singularity-2.0-20191204/bin/singularity exec -B /WORK/teaching/project:/WORK/teaching/project /WORK/teaching/project/singularity_images/ribocodeminer /home/test/anaconda3/bin/PlotRiboDensityAtEachKindAAOrCodon -i $out/b6_all_codon_density.txt -o $out/b7-PlotRiboDensityAtEachKindAAOrCodon_${o5} -g $g5 -r $r5 --level AA" >> ${out}/script/densitycodonp5.sh

echohead > ${out}/script/densitycodonp6.sh
echo "/WORK/app/singularity-2.0-20191204/bin/singularity exec -B /WORK/teaching/project:/WORK/teaching/project /WORK/teaching/project/singularity_images/ribocodeminer /home/test/anaconda3/bin/PlotRiboDensityAtEachKindAAOrCodon -i $out/b6_all_codon_density.txt -o $out/b7-PlotRiboDensityAtEachKindAAOrCodon_${o6} -g $g6 -r $r6 --level AA" >> ${out}/script/densitycodonp6.sh

echohead > ${out}/script/densitycodonp7.sh
echo "/WORK/app/singularity-2.0-20191204/bin/singularity exec -B /WORK/teaching/project:/WORK/teaching/project /WORK/teaching/project/singularity_images/ribocodeminer /home/test/anaconda3/bin/PlotRiboDensityAtEachKindAAOrCodon -i $out/b6_all_codon_density.txt -o $out/b7-PlotRiboDensityAtEachKindAAOrCodon_${o7} -g $g7 -r $r7 --level AA" >> ${out}/script/densitycodonp7.sh

echo "End: Ribosome density at each kind of AA or codon. `date`"
}

triplete()
{
echo "Begin: Ribosome density around the triplete amino acid (tri-AA) motifs `date`"
## ribosome density at each tri-AA motif
echohead > ${out}/script/triple.sh
echo "/WORK/app/singularity-2.0-20191204/bin/singularity exec -B /WORK/teaching/project:/WORK/teaching/project /WORK/teaching/project/singularity_images/ribocodeminer /home/test/anaconda3/bin/RiboDensityAroundTripleteAAMotifs -f ${attribute} -c $RiboMiner/longest.transcripts.info.txt -o $out/b8_PPP -M RPKM -S ${select_gene} -l 100 -n 10 --table 1 -F $RiboMiner/transcript_cds_sequences.fa --type2 PPP --type1 PP" >> ${out}/script/triple.sh
#RiboDensityAroundTripleteAAMotifs -f $meta_out/attributes.txt -c $RiboMiner/longest.transcripts.info.txt -o $out/b8_PPP_${1} -M RPKM -S ${home_dir}/${1} -l 100 -n 10 --table 1 -F $RiboMiner/transcript_cds_sequences.fa --type2 PPP --type1 PP
## plot
echohead > ${out}/script/triple_p1.sh
echo "/WORK/app/singularity-2.0-20191204/bin/singularity exec -B /WORK/teaching/project:/WORK/teaching/project /WORK/teaching/project/singularity_images/ribocodeminer /home/test/anaconda3/bin/PlotRiboDensityAroundTriAAMotifs -i $out/b8_PPP_motifDensity_dataframe.txt -o $out/b9-PPP_plot_${o1} -g ${g1} -r ${r1} --mode mean --ymax 0.2" >> ${out}/script/triple_p1.sh

echohead > ${out}/script/triple_p2.sh
echo "/WORK/app/singularity-2.0-20191204/bin/singularity exec -B /WORK/teaching/project:/WORK/teaching/project /WORK/teaching/project/singularity_images/ribocodeminer /home/test/anaconda3/bin/PlotRiboDensityAroundTriAAMotifs -i $out/b8_PPP_motifDensity_dataframe.txt -o $out/b9-PPP_plot_${o2} -g ${g2} -r ${r2} --mode mean --ymax 0.2" >> ${out}/script/triple_p2.sh

echohead > ${out}/script/triple_p3.sh
echo "/WORK/app/singularity-2.0-20191204/bin/singularity exec -B /WORK/teaching/project:/WORK/teaching/project /WORK/teaching/project/singularity_images/ribocodeminer /home/test/anaconda3/bin/PlotRiboDensityAroundTriAAMotifs -i $out/b8_PPP_motifDensity_dataframe.txt -o $out/b9-PPP_plot_${o3} -g ${g3} -r ${r3} --mode mean --ymax 0.2" >> ${out}/script/triple_p3.sh

echohead > ${out}/script/triple_p4.sh
echo "/WORK/app/singularity-2.0-20191204/bin/singularity exec -B /WORK/teaching/project:/WORK/teaching/project /WORK/teaching/project/singularity_images/ribocodeminer /home/test/anaconda3/bin/PlotRiboDensityAroundTriAAMotifs -i $out/b8_PPP_motifDensity_dataframe.txt -o $out/b9-PPP_plot_${o4} -g ${g4} -r ${r4} --mode mean --ymax 0.2" >> ${out}/script/triple_p4.sh

echohead > ${out}/script/triple_p5.sh
echo "/WORK/app/singularity-2.0-20191204/bin/singularity exec -B /WORK/teaching/project:/WORK/teaching/project /WORK/teaching/project/singularity_images/ribocodeminer /home/test/anaconda3/bin/PlotRiboDensityAroundTriAAMotifs -i $out/b8_PPP_motifDensity_dataframe.txt -o $out/b9-PPP_plot_${o5} -g ${g5} -r ${r5} --mode mean --ymax 0.2" >> ${out}/script/triple_p5.sh

echohead > ${out}/script/triple_p6.sh
echo "/WORK/app/singularity-2.0-20191204/bin/singularity exec -B /WORK/teaching/project:/WORK/teaching/project /WORK/teaching/project/singularity_images/ribocodeminer /home/test/anaconda3/bin/PlotRiboDensityAroundTriAAMotifs -i $out/b8_PPP_motifDensity_dataframe.txt -o $out/b9-PPP_plot_${o6} -g ${g6} -r ${r6} --mode mean --ymax 0.2" >> ${out}/script/triple_p6.sh

echohead > ${out}/script/triple_p7.sh
echo "/WORK/app/singularity-2.0-20191204/bin/singularity exec -B /WORK/teaching/project:/WORK/teaching/project /WORK/teaching/project/singularity_images/ribocodeminer /home/test/anaconda3/bin/PlotRiboDensityAroundTriAAMotifs -i $out/b8_PPP_motifDensity_dataframe.txt -o $out/b9-PPP_plot_${o7} -g ${g7} -r ${r7} --mode mean --ymax 0.2" >> ${out}/script/triple_p7.sh



echohead > ${out}/script/triple_other.sh
echo "/WORK/app/singularity-2.0-20191204/bin/singularity exec -B /WORK/teaching/project:/WORK/teaching/project /WORK/teaching/project/singularity_images/ribocodeminer /home/test/anaconda3/bin/RiboDensityAroundTripleteAAMotifs -f ${attribute} -c $RiboMiner/longest.transcripts.info.txt -o $out/b9_triple_motif -M RPKM -S ${select_gene} -l 100 -n 10 --table 1 -F $RiboMiner/transcript_cds_sequences.fa --motifList1 $out/tri_AA_motifs1.txt --motifList2 $out/tri_AA_motifs2.txt" >> ${out}/script/triple_other.sh
#RiboDensityAroundTripleteAAMotifs -f $meta_out/attributes.txt -c $RiboMiner/longest.transcripts.info.txt -o $out/b9_triple_motif_${1} -M RPKM -S ${home_dir}/${1} -l 100 -n 10 --table 1 -F $RiboMiner/transcript_cds_sequences.fa --motifList1 $out/tri_AA_motifs1.txt --motifList2 $out/tri_AA_motifs2.txt
## plot
echohead > ${out}/script/triple_other_p1.sh
echo "/WORK/app/singularity-2.0-20191204/bin/singularity exec -B /WORK/teaching/project:/WORK/teaching/project /WORK/teaching/project/singularity_images/ribocodeminer /home/test/anaconda3/bin/PlotRiboDensityAroundTriAAMotifs -i $out/b9_triple_motif_motifDensity_dataframe.txt -o $out/c1-triple_motif_plot_${o1} -g ${g1} -r ${r1} --mode mean --ymax 0.2">> ${out}/script/triple_other_p1.sh

echohead > ${out}/script/triple_other_p2.sh
echo "/WORK/app/singularity-2.0-20191204/bin/singularity exec -B /WORK/teaching/project:/WORK/teaching/project /WORK/teaching/project/singularity_images/ribocodeminer /home/test/anaconda3/bin/PlotRiboDensityAroundTriAAMotifs -i $out/b9_triple_motif_motifDensity_dataframe.txt -o $out/c1-triple_motif_plot_${o2} -g ${g2} -r ${r2} --mode mean --ymax 0.2">> ${out}/script/triple_other_p2.sh

echohead > ${out}/script/triple_other_p3.sh
echo "/WORK/app/singularity-2.0-20191204/bin/singularity exec -B /WORK/teaching/project:/WORK/teaching/project /WORK/teaching/project/singularity_images/ribocodeminer /home/test/anaconda3/bin/PlotRiboDensityAroundTriAAMotifs -i $out/b9_triple_motif_motifDensity_dataframe.txt -o $out/c1-triple_motif_plot_${o3} -g ${g3} -r ${r3} --mode mean --ymax 0.2">> ${out}/script/triple_other_p3.sh

echohead > ${out}/script/triple_other_p4.sh
echo "/WORK/app/singularity-2.0-20191204/bin/singularity exec -B /WORK/teaching/project:/WORK/teaching/project /WORK/teaching/project/singularity_images/ribocodeminer /home/test/anaconda3/bin/PlotRiboDensityAroundTriAAMotifs -i $out/b9_triple_motif_motifDensity_dataframe.txt -o $out/c1-triple_motif_plot_${o4} -g ${g4} -r ${r4} --mode mean --ymax 0.2">> ${out}/script/triple_other_p4.sh

echohead > ${out}/script/triple_other_p5.sh
echo "/WORK/app/singularity-2.0-20191204/bin/singularity exec -B /WORK/teaching/project:/WORK/teaching/project /WORK/teaching/project/singularity_images/ribocodeminer /home/test/anaconda3/bin/PlotRiboDensityAroundTriAAMotifs -i $out/b9_triple_motif_motifDensity_dataframe.txt -o $out/c1-triple_motif_plot_${o5} -g ${g5} -r ${r5} --mode mean --ymax 0.2">> ${out}/script/triple_other_p5.sh

echohead > ${out}/script/triple_other_p6.sh
echo "/WORK/app/singularity-2.0-20191204/bin/singularity exec -B /WORK/teaching/project:/WORK/teaching/project /WORK/teaching/project/singularity_images/ribocodeminer /home/test/anaconda3/bin/PlotRiboDensityAroundTriAAMotifs -i $out/b9_triple_motif_motifDensity_dataframe.txt -o $out/c1-triple_motif_plot_${o6} -g ${g6} -r ${r6} --mode mean --ymax 0.2">> ${out}/script/triple_other_p6.sh

echohead > ${out}/script/triple_other_p7.sh
echo "/WORK/app/singularity-2.0-20191204/bin/singularity exec -B /WORK/teaching/project:/WORK/teaching/project /WORK/teaching/project/singularity_images/ribocodeminer /home/test/anaconda3/bin/PlotRiboDensityAroundTriAAMotifs -i $out/b9_triple_motif_motifDensity_dataframe.txt -o $out/c1-triple_motif_plot_${o7} -g ${g7} -r ${r7} --mode mean --ymax 0.2">> ${out}/script/triple_other_p7.sh


echo "End: Ribosome density around the triplete amino acid (tri-AA) motifs `date`"
}

triplete_Q()
{
echo "Begin: Ribosome density around the triplete amino acid (tri-AA) motifs `date`"
## ribosome density at each tri-AA motif
echohead > ${out}/script/triple_Q.sh
echo "/WORK/app/singularity-2.0-20191204/bin/singularity exec -B /WORK/teaching/project:/WORK/teaching/project /WORK/teaching/project/singularity_images/ribocodeminer /home/test/anaconda3/bin/RiboDensityAroundTripleteAAMotifs -f $meta_out/attributes.txt -c $RiboMiner/longest.transcripts.info.txt -o $out/b8_QQQ -M RPKM -S ${select_gene}  -l 100 -n 10 --table 1 -F $RiboMiner/transcript_cds_sequences.fa --type2 QQQ --type1 QQ" >> ${out}/script/triple_Q.sh
#RiboDensityAroundTripleteAAMotifs -f $meta_out/attributes.txt -c $RiboMiner/longest.transcripts.info.txt -o $out/b8_QQQ_${1} -M RPKM -S ${home_dir}/${1} -l 100 -n 10 --table 1 -F $RiboMiner/transcript_cds_sequences.fa --type2 QQQ --type1 QQ
## plot

echohead > ${out}/script/triple_Q_p1.sh
echo "/WORK/app/singularity-2.0-20191204/bin/singularity exec -B /WORK/teaching/project:/WORK/teaching/project /WORK/teaching/project/singularity_images/ribocodeminer /home/test/anaconda3/bin/PlotRiboDensityAroundTriAAMotifs -i $out/b8_QQQ_motifDensity_dataframe.txt -o $out/b9-QQQ_plot_${o1} -g ${g1} -r ${r1} --mode mean --ymax 0.2"  >> ${out}/script/triple_Q_p1.sh

echohead > ${out}/script/triple_Q_p2.sh
echo "/WORK/app/singularity-2.0-20191204/bin/singularity exec -B /WORK/teaching/project:/WORK/teaching/project /WORK/teaching/project/singularity_images/ribocodeminer /home/test/anaconda3/bin/PlotRiboDensityAroundTriAAMotifs -i $out/b8_QQQ_motifDensity_dataframe.txt -o $out/b9-QQQ_plot_${o2} -g ${g2} -r ${r2} --mode mean --ymax 0.2"  >> ${out}/script/triple_Q_p2.sh

echohead > ${out}/script/triple_Q_p3.sh
echo "PlotRiboDensityAroundTriAAMotifs -i $out/b8_QQQ_motifDensity_dataframe.txt -o $out/b9-QQQ_plot_${o3} -g ${g3} -r ${r3} --mode mean --ymax 0.2"  >> ${out}/script/triple_Q_p3.sh

echohead > ${out}/script/triple_Q_p4.sh
echo "/WORK/app/singularity-2.0-20191204/bin/singularity exec -B /WORK/teaching/project:/WORK/teaching/project /WORK/teaching/project/singularity_images/ribocodeminer /home/test/anaconda3/bin/PlotRiboDensityAroundTriAAMotifs -i $out/b8_QQQ_motifDensity_dataframe.txt -o $out/b9-QQQ_plot_${o4} -g ${g4} -r ${r4} --mode mean --ymax 0.2"  >> ${out}/script/triple_Q_p4.sh

echohead > ${out}/script/triple_Q_p5.sh
echo "/WORK/app/singularity-2.0-20191204/bin/singularity exec -B /WORK/teaching/project:/WORK/teaching/project /WORK/teaching/project/singularity_images/ribocodeminer /home/test/anaconda3/bin/PlotRiboDensityAroundTriAAMotifs -i $out/b8_QQQ_motifDensity_dataframe.txt -o $out/b9-QQQ_plot_${o5} -g ${g5} -r ${r5} --mode mean --ymax 0.2"  >> ${out}/script/triple_Q_p5.sh

echohead > ${out}/script/triple_Q_p6.sh
echo "/WORK/app/singularity-2.0-20191204/bin/singularity exec -B /WORK/teaching/project:/WORK/teaching/project /WORK/teaching/project/singularity_images/ribocodeminer /home/test/anaconda3/bin/PlotRiboDensityAroundTriAAMotifs -i $out/b8_QQQ_motifDensity_dataframe.txt -o $out/b9-QQQ_plot_${o6} -g ${g6} -r ${r6} --mode mean --ymax 0.2"  >> ${out}/script/triple_Q_p6.sh

echohead > ${out}/script/triple_Q_p7.sh
echo "/WORK/app/singularity-2.0-20191204/bin/singularity exec -B /WORK/teaching/project:/WORK/teaching/project /WORK/teaching/project/singularity_images/ribocodeminer /home/test/anaconda3/bin/PlotRiboDensityAroundTriAAMotifs -i $out/b8_QQQ_motifDensity_dataframe.txt -o $out/b9-QQQ_plot_${o7} -g ${g7} -r ${r7} --mode mean --ymax 0.2"  >> ${out}/script/triple_Q_p7.sh


echo "End: Ribosome density around the triplete amino acid (tri-AA) motifs `date`"
}


Pausingscore()
{
echo "Begin: Pausing score of each triplete amino acid`date`"
# Pausing score of each triplete amino acid.

echohead > ${out}/script/pausing.sh
echo "/WORK/app/singularity-2.0-20191204/bin/singularity exec -B /WORK/teaching/project:/WORK/teaching/project /WORK/teaching/project/singularity_images/ribocodeminer /home/test/anaconda3/bin/PausingScore -f $meta_out/attributes.txt -c $RiboMiner/longest.transcripts.info.txt -o  $out/b8-PausingScore -M counts -S ${select_gene}   -l 100 -n 10 --table 1 -F  $RiboMiner/transcript_cds_sequences.fa ">> ${out}/script/pausing.sh
#PausingScore -f $meta_out/attributes.txt -c $RiboMiner/longest.transcripts.info.txt -o  $out/b8-PausingScore_${1} -M counts -S ${home_dir}/${1}  -l 100 -n 10 --table 1 -F  $RiboMiner/transcript_cds_sequences.fa

echohead > ${out}/script/pausing_p1.sh
echo "/WORK/app/singularity-2.0-20191204/bin/singularity exec -B /WORK/teaching/project:/WORK/teaching/project /WORK/teaching/project/singularity_images/ribocodeminer /home/test/anaconda3/bin/ProcessPausingScore -i $out/b8-PausingScore_7-7-R.toTranscriptome.sort_pausing_score.txt,$out/b8-PausingScore_7-111-R.toTranscriptome.sort_pausing_score.txt -o $out/b9-ProcessPausingScore${o1} -g ${g1} -r ${r1} --mode raw --ratio_filter 1 --pausing_score_filter 0.01" >> ${out}/script/pausing_p1.sh
#ProcessPausingScore -i $out/b8-PausingScore_${1}_7-7-R.toTranscriptome.sort_pausing_score.txt,$out/b8-PausingScore_${1}_7-111-R.toTranscriptome.sort_pausing_score.txt -o $out/b9-ProcessPausingScore_${1} -g $groupinfo -r $replace --mode raw --ratio_filter 1 --pausing_score_filter 0.01
echo "End: Pausing score of each triplete amino acid`date`"
}



RPFdistf()
{
echo "Begin: RPFdist calculation.`date`"
echohead > ${out}/script/rpddist.sh
# RPFdist calculation.
echo "/WORK/app/singularity-2.0-20191204/bin/singularity exec -B /WORK/teaching/project:/WORK/teaching/project /WORK/teaching/project/singularity_images/ribocodeminer /home/test/anaconda3/bin/RPFdist -f $meta_out/attributes.txt -c $RiboMiner/longest.transcripts.info.txt -o $out/c3-RPFdist -M counts -S ${select_gene}  -l 100 -n 10 -m 1 -e 5">> ${out}/script/rpddist.sh
#RPFdist -f $meta_out/attributes.txt -c $RiboMiner/longest.transcripts.info.txt -o $out/c3-RPFdist -M counts -S $select_gene -l 100 -n 10 -m 1 -e 5

echo "End: RPFdist calculation. `date`"
}

gccontents()
{
echo "Begin:GC contents for sequences with a fasta format `date`"

# GC contents for sequences with a fasta format.
echohead > ${out}/script/gccontent.sh
echo "/WORK/app/singularity-2.0-20191204/bin/singularity exec -B /WORK/teaching/project:/WORK/teaching/project /WORK/teaching/project/singularity_images/ribocodeminer /home/test/anaconda3/bin/GCContent -i $RiboMiner/transcript_cds_sequences.fa -o $out/c4-GCContent-normal --mode normal">> ${out}/script/gccontent.sh
echo "/WORK/app/singularity-2.0-20191204/bin/singularity exec -B /WORK/teaching/project:/WORK/teaching/project /WORK/teaching/project/singularity_images/ribocodeminer /home/test/anaconda3/bin/GCContent -i $RiboMiner/transcript_cds_sequences.fa -o $out/c4-GCContent-frames --mode frames" >> ${out}/script/gccontent.sh
#GCContent -i $RiboMiner/transcript_cds_sequences.fa -o $out/c4-GCContent-normal --mode normal
#GCContent -i $RiboMiner/transcript_cds_sequences.fa -o $out/c4-GCContent-frames --mode frames
echo "/WORK/app/singularity-2.0-20191204/bin/singularity exec -B /WORK/teaching/project:/WORK/teaching/project /WORK/teaching/project/singularity_images/ribocodeminer /home/test/anaconda3/bin/PlotGCContent -i $out/c4-GCContent-normal_GC_content.txt -o $out/c5-PlotGCContent-normal --mode normal" >> ${out}/script/gccontent.sh
#PlotGCContent -i $out/c4-GCContent-normal_GC_content.txt -o $out/c5-PlotGCContent-normal --mode normal
echo "End: GC contents for sequences with a fasta format`date`"
}



#densityspecific
#densitycodon
#triplete
#triplete_Q
#Pausingscore
RPFdistf
gccontents

```

**b4_feature_analysis.sh**

```sh


out='/home/share/riboseq/d_feature_analysis'
meta_out='/home/share/riboseq/bqualitycontrol/metaplot'

RiboMiner='/home/share/riboseq/RiboMiner'
long=$RiboMiner'/longest.transcripts.info.txt'
select_gene='/home/share/riboseq/selec_trans_longest.txt'

home_dir='/home/share/riboseq'
transcript_down_translation_up_fa='select.transcript_down_translation_up.list.gene.fa'
transcript_only_down_fa='select.transcript_only_down.list.gene.fa'
transcript_only_up_fa='select.transcript_only_up.list.gene.fa'
transcript_translation_down_fa='select.transcript_translation_down.list.gene.fa'
transcript_translation_up_fa='select.transcript_translation_up.list.gene.fa'
transcript_up_translation_down_fa='select.transcript_up_translation_down.list.gene.fa'
translation_only_down_fa='select.translation_only_down.list.gene.fa'
translation_only_up_fa='select.translation_only_up.list.gene.fa'

fa_sum=$home_dir/$transcript_down_translation_up_fa,$home_dir/$transcript_only_down_fa,$home_dir/$transcript_only_up_fa,$home_dir/$transcript_translation_down_fa,$home_dir/$transcript_translation_up_fa,$home_dir/$transcript_up_translation_down_fa,$home_dir/$translation_only_down_fa,$home_dir/$translation_only_up_fa

fa_sum_transcript=$home_dir/$transcript_only_down_fa,$home_dir/$transcript_only_up_fa

fa_sum_translation=$home_dir/$translation_only_down_fa,$home_dir/$translation_only_up_fa

fa_sum_homo=$home_dir/$transcript_translation_down_fa,$home_dir/$transcript_translation_up_fa

fa_sum_opposite=$home_dir/$transcript_down_translation_up_fa,$home_dir/$transcript_up_translation_down_fa

fa_name_sum='transcript_down_translation_up,transcript_only_down,transcript_only_up,transcript_translation_down,transcript_translation_up,transcript_up_translation_down,translation_only_down,translation_only_up'

fa_sum_transcript_name='transcript_only_down_fa,transcript_only_up_fa'

fa_sum_translation_name='translation_only_down_fa,translation_only_up_fa'

fa_sum_homo_name='transcript_translation_down_fa,transcript_translation_up_fa'

fa_sum_opposite_name='transcript_down_translation_up_fa,transcript_up_translation_down_fa'

groupinfo='R111,R7'
replace='7-111-R.toTranscriptome.sort__7-7-R.toTranscriptome.sort'

tRNA_confidence='/home/share/riboseq/mouse-tRNAs-confidence.txt'

# tAIf $name $fa_sum_transcrpt $fa_sum_transcrpt_name

tAIf()
{
echo "Local tRNA adaptation index and global tRNA adaptation index `date`"
#Local tRNA adaptation index and global tRNA adaptation index
echo " tAI -i ${2} -t ${3} -o $out/c6-tAI_${1} -u 0 -d 500 --table 1 -N $tRNA_confidence"
tAI -i ${2} -t ${3} -o $out/c6-tAI_${1} -u 0 -d 500 --table 1 -N $tRNA_confidence

echo " tAIPlot -i $out/c6-tAI_${1}_tAI_dataframe.txt -o $out/c7-tAIPlot_${1} -u 0 -d 500 --mode all --start 5 --window 7 --step 1"
tAIPlot -i $out/c6-tAI_${1}_tAI_dataframe.txt -o $out/c7-tAIPlot_${1} -u 0 -d 500 --mode all --start 5 --window 7 --step 1

echo "Finished: Local tRNA adaptation index and global tRNA adaptation index `date`"

}

# cAIf $name $fa_sum_transcrpt $fa_sum_transcrpt_name
cAIf()
{
echo "Local codon adaptation index and global codon adaptation index `date`"
# Local codon adaptation index and global codon adaptation index
echo "cAI -i ${2} -o $out/c8-cAI_${1} -t ${3} -u 0 -d 500 --reference $RiboMiner/transcript_cds_sequences.fa"
cAI -i ${2} -o $out/c8-cAI_${1} -t ${3} -u 0 -d 500 --reference $RiboMiner/transcript_cds_sequences.fa

echo "cAIPlot -i $out/c8-cAI_${1}_local_cAI_dataframe.txt -o $out/c9-cAIPlot_${1} -u 0 -d 500 --mode all --start 5 --window 7 --step 1"
cAIPlot -i $out/c8-cAI_${1}_local_cAI_dataframe.txt -o $out/c9-cAIPlot_${1} -u 0 -d 500 --mode all --start 5 --window 7 --step 1
echo "Finished: Local codon adaptation index and global codon adaptation index `date`"

}

tAIf 'transcrpt' $fa_sum_transcript $fa_sum_transcript_name
tAIf 'translation' $fa_sum_translation $fa_sum_translation_name
tAIf 'homo' $fa_sum_homo $fa_sum_homo_name
tAIf 'opposite' $fa_sum_opposite $fa_sum_opposite_name

cAIf 'transcrpt' $fa_sum_transcript $fa_sum_transcript_name
cAIf 'translation' $fa_sum_translation $fa_sum_translation_name
cAIf 'homo' $fa_sum_homo $fa_sum_homo_name
cAIf 'opposite' $fa_sum_opposite $fa_sum_opposite_name

```

b4_calculate_count.sh

## Calculate counts
```sh

echohead(){
   echo '#!/bin/bash
#SBATCH -J slum
#SBATCH -p CN_BIOT
#SBATCH --nodes=1
#SBATCH --ntasks=3
#SBATCH --output=%slum.out
#SBATH --error=%slum.err

# Get the software
export PATH=/WORK/teaching/bin:$PATH
export SINGULARITY_BINDPATH="/WORK"
' 
}

geneome_ribo=/WORK/teaching/project/xuzhiyu/b1-qualitycontrol/genome.list
out=/WORK/teaching/project/xuzhiyu/b4-count
gtf=/WORK/teaching/project/xuzhiyu/reference/Arabidopsis_thaliana.TAIR10.43.gtf
[ -d ${out} ] || mkdir ${out}
[ -d ${out}/script ] || mkdir ${out}/script


generate(){
dir='/WORK/teaching/project/xuzhiyu/bam_dir';

if [ -f ${out}/script/all.genome.txt ];then
rm -rf ${out}/script/all.genome.txt
fi


for i in ` ls ${dir}|grep Aligned.sortedByCoord.out.bam`;
do 
 echo "${dir}/${i}" >>${out}/script/all.genome.txt;
done;
}

count_all(){
for i in `cat ${out}/script/all.genome.txt`;do 
name=$(echo $i|rev | cut -f 1 -d '/' | rev|cut -f 1 -d '.') ;
echo ${name};
echohead > ${out}/script/$name.all.sh ;
echo "/WORK/app/singularity-2.0-20191204/bin/singularity exec  -B /WORK/teaching/project:/WORK/teaching/project /WORK/teaching/project/singularity_images/htseq /home/test/miniconda2/bin/python /home/test/miniconda2/bin/RPF_count_CDS.py ${i} ${gtf} > ${out}/${name}.all.count" >> ${out}/script/$name.all.sh;

done
}

#ribo
generate
#count_all

```
rename.pl
```pl

open DATA, "<srr_name.txt";
while(<DATA>)
{
chomp;
@data=split "\t",$_;
print "$data[0]_1.all.count\t$data[1].count\n";
system("mv  $data[0]_1.all.count $data[1].count");
}

```

a1.merge.sh

```sh

name=(RNAseq_upf2-Ethylene-Total1 RNAseq_upf2-Air-Total2 RNAseq_upf2-Air-Total1 riboseq_upf2-Ethylene-RFP2 riboseq_upf2-Ethylene-RFP1 riboseq_upf2-Air-RFP2 riboseq_upf2-Air-RFP1 RNAseq_Col-Ethylene-Total2 RNAseq_Col-Ethylene-Total1 RNAseq_Col-Air-Total2 RNAseq_Col-Air-Total1 riboseq_Col-Ethylene-RFP2 riboseq_Col-Ethylene-RFP1 riboseq_Col_Air_RFP2 riboseq_Col-Air-RFP1)
head='gene'
for i in ${name[@]};
do
 head+=" ${i}";
done
echo -e $head >merge.counter;

begin1=${name[0]};
begin2=${name[1]};
name2=("${name[@]:2}");
join ${begin1}.count ${begin2}.count >merge.tmp
commander='join';
for i in ${name2[@]};
do 
echo ${i}.count;
join merge.tmp ${i}.count >>merge.tmp2;
mv merge.tmp2 merge.tmp
done

cat merge.counter merge.tmp > merge2.tmp;
rm merge.tmp
mv merge2.tmp merge.counter
sed -i 's/ \+/\t/g' merge.counter

grep -v '^__' merge.counter > merge.counter2
mv merge.counter2 ethylene.counter 

```

a2.merge.sh

```sh

name=(riboseq_heat RNAseq_heat riboseq_control_heat2 RNAseq_control_heat2 riboseq_control_heat1 RNAseq_control_heat1)
head='gene'
for i in ${name[@]};
do
 head+=" ${i}";
done
echo -e $head >merge.counter;

begin1=${name[0]};
begin2=${name[1]};
name2=("${name[@]:2}");
join ${begin1}.count ${begin2}.count >merge.tmp
commander='join';
for i in ${name2[@]};
do 
echo ${i}.count;
join merge.tmp ${i}.count >>merge.tmp2;
mv merge.tmp2 merge.tmp
done

cat merge.counter merge.tmp > merge2.tmp;
rm merge.tmp
mv merge2.tmp merge.counter
sed -i 's/ \+/\t/g' merge.counter

grep -v '^__' merge.counter > merge.counter2
mv merge.counter2 heat.counter 

```


a3.merge.sh

```sh

name=(RNAseq_shoot_3 RNAseq_shoot_2 RNAseq_shoot_1 RNAseq_root_3 RNAseq_root_2 RNAseq_root_1 riboseq_shoot_3 riboseq_shoot_2 riboseq_shoot_1 riboseq_root_3 riboseq_root_2 riboseq_root_1)
head='gene'
for i in ${name[@]};
do
 head+=" ${i}";
done
echo -e $head >merge.counter;

begin1=${name[0]};
begin2=${name[1]};
name2=("${name[@]:2}");
join ${begin1}.count ${begin2}.count >merge.tmp
commander='join';
for i in ${name2[@]};
do 
echo ${i}.count;
join merge.tmp ${i}.count >>merge.tmp2;
mv merge.tmp2 merge.tmp
done

cat merge.counter merge.tmp > merge2.tmp;
rm merge.tmp
mv merge2.tmp merge.counter
sed -i 's/ \+/\t/g' merge.counter

grep -v '^__' merge.counter > merge.counter2
mv merge.counter2 shoot_root.counter 

```


a4.merge.sh

```sh

name=(riboseq_L4h riboseq_Dark RNAseq_L4h RNAseq_Dark riboseq_L4h_2 riboseq_Dark_2 RNAseq_L4h_2 RNAseq_Dark_2)
head='gene'
for i in ${name[@]};
do
 head+=" ${i}";
done
echo -e $head >merge.counter;

begin1=${name[0]};
begin2=${name[1]};
name2=("${name[@]:2}");
join ${begin1}.count ${begin2}.count >merge.tmp
commander='join';
for i in ${name2[@]};
do 
echo ${i}.count;
join merge.tmp ${i}.count >>merge.tmp2;
mv merge.tmp2 merge.tmp
done

cat merge.counter merge.tmp > merge2.tmp;
rm merge.tmp
mv merge2.tmp merge.counter
sed -i 's/ \+/\t/g' merge.counter

grep -v '^__' merge.counter > merge.counter2
mv merge.counter2 light.counter 

```
**control_heat.sh**
```sh
#!/bin/bash
#SBATCH -J tophat_test
#SBATCH -p CN_BIOT
#SBATCH --nodes=1
#SBATCH --ntasks=10
#SBATCH --output=%j.out
#SBATH --error=%j.err

# Get the software
export PATH=/WORK/teaching/bin:$PATH

/WORK/app/singularity-2.0-20191204/bin/singularity exec  -B /WORK/teaching/project:/WORK/teaching/project /WORK/teaching/project/singularity_images/xtail Rscript /WORK/teaching/project/xuzhiyu/b6_xtail/control_heat.r
```

**control_heat.r**

name=(riboseq_heat RNAseq_heat riboseq_control_heat2 RNAseq_control_heat2 riboseq_control_heat1 RNAseq_control_heat1)



```R
library(xtail)
lbxd=read.table('/WORK/teaching/project/xuzhiyu/b4-count/heat.counter',header=T,row.name=1)
mrna=lbxd[,c(4,6,2)]
rpf=lbxd[,c(3,5,1)]
condition=c("control","control","heat")
test.results=xtail(mrna,rpf,condition,bins=1000,threads=2)
summary(test.results)

#
test.tab=resultsTable(test.results);
head(test.tab,5)

write.table(test.tab,"/WORK/teaching/project/xuzhiyu/b6_xtail/conrol_heat_results.txt",quote=F,sep="\t");

# Visualization
pdf('/WORK/teaching/project/xuzhiyu/b6_xtail/conrol_heatFC.pdf',width=6,height=4,paper='special')
lbxfc=plotFCs(test.results)
dev.off()
write.table(lbxfc$resultsTable,"/WORK/teaching/project/xuzhiyu/b6_xtail/conrol_heatfc_results.txt",quote=F,sep="\t");

pdf('/WORK/teaching/project/xuzhiyu/b6_xtail/conrol_heatRs.pdf',width=6,height=4,paper='special')
lbxrs=plotRs(test.results)
dev.off()
write.table(lbxrs$resultsTable,"/WORK/teaching/project/xuzhiyu/b6_xtail/conrol_heatrs_results.txt",quote=F,sep="\t");

pdf('/WORK/teaching/project/xuzhiyu/b6_xtail/conrol_heatvolcano.pdf',width=6,height=4,paper='special')
volcanoPlot(test.results)
dev.off()
```

**dark_light.sh**
```sh
#!/bin/bash
#SBATCH -J tophat_test
#SBATCH -p CN_BIOT
#SBATCH --nodes=1
#SBATCH --ntasks=10
#SBATCH --output=%j.out
#SBATH --error=%j.err

# Get the software
export PATH=/WORK/teaching/bin:$PATH

/WORK/app/singularity-2.0-20191204/bin/singularity exec  -B /WORK/teaching/project:/WORK/teaching/project /WORK/teaching/project/singularity_images/xtail Rscript /WORK/teaching/project/xuzhiyu/b6_xtail/dark_light.r
```

**dark_light.r**

name=(riboseq_L4h riboseq_Dark RNAseq_L4h RNAseq_Dark riboseq_L4h_2 riboseq_Dark_2 RNAseq_L4h_2 RNAseq_Dark_2)


```R
library(xtail)
lbxd=read.table('/WORK/teaching/project/xuzhiyu/b4-count/light.counter',header=T,row.name=1)
mrna=lbxd[,c(4,8,3,7)]
rpf=lbxd[,c(2,6,1,5)]
condition=c("dark","dark","light","light")
test.results=xtail(mrna,rpf,condition,bins=1000,threads=2)
summary(test.results)

#
test.tab=resultsTable(test.results);
head(test.tab,5)

write.table(test.tab,"/WORK/teaching/project/xuzhiyu/b6_xtail/dark_light_results.txt",quote=F,sep="\t");

# Visualization
pdf('/WORK/teaching/project/xuzhiyu/b6_xtail/dark_lightFC.pdf',width=6,height=4,paper='special')
lbxfc=plotFCs(test.results)
dev.off()
write.table(lbxfc$resultsTable,"/WORK/teaching/project/xuzhiyu/b6_xtail/dark_lightfc_results.txt",quote=F,sep="\t");

pdf('/WORK/teaching/project/xuzhiyu/b6_xtail/dark_lightRs.pdf',width=6,height=4,paper='special')
lbxrs=plotRs(test.results)
dev.off()
write.table(lbxrs$resultsTable,"/WORK/teaching/project/xuzhiyu/b6_xtail/dark_lightrs_results.txt",quote=F,sep="\t");

pdf('/WORK/teaching/project/xuzhiyu/b6_xtail/dark_lightvolcano.pdf',width=6,height=4,paper='special')
volcanoPlot(test.results)
dev.off()
```


**shoot_root.sh**
```sh
#!/bin/bash
#SBATCH -J tophat_test
#SBATCH -p CN_BIOT
#SBATCH --nodes=1
#SBATCH --ntasks=10
#SBATCH --output=%j.out
#SBATH --error=%j.err

# Get the software
export PATH=/WORK/teaching/bin:$PATH

/WORK/app/singularity-2.0-20191204/bin/singularity exec  -B /WORK/teaching/project:/WORK/teaching/project /WORK/teaching/project/singularity_images/xtail Rscript /WORK/teaching/project/xuzhiyu/b6_xtail/shoot_root.r
```

**shoot_root.r**

(RNAseq_shoot_3 RNAseq_shoot_2 RNAseq_shoot_1 RNAseq_root_3 RNAseq_root_2 RNAseq_root_1 riboseq_shoot_3 riboseq_shoot_2 riboseq_shoot_1 riboseq_root_3 riboseq_root_2 riboseq_root_1)


```R
library(xtail)
lbxd=read.table('/WORK/teaching/project/xuzhiyu/b4-count/shoot_root.counter',header=T,row.name=1)
mrna=lbxd[,c(1,2,3,4,5,6)]
rpf=lbxd[,c(7,8,9,10,11,12)]
condition=c("shoot","shoot","shoot","root","root","root")
test.results=xtail(mrna,rpf,condition,bins=1000,threads=2)
summary(test.results)

#
test.tab=resultsTable(test.results);
head(test.tab,5)

write.table(test.tab,"/WORK/teaching/project/xuzhiyu/b6_xtail/shoot_root_results.txt",quote=F,sep="\t");

# Visualization
pdf('/WORK/teaching/project/xuzhiyu/b6_xtail/shoot_rootFC.pdf',width=6,height=4,paper='special')
lbxfc=plotFCs(test.results)
dev.off()
write.table(lbxfc$resultsTable,"/WORK/teaching/project/xuzhiyu/b6_xtail/shoot_rootfc_results.txt",quote=F,sep="\t");

pdf('/WORK/teaching/project/xuzhiyu/b6_xtail/shoot_rootRs.pdf',width=6,height=4,paper='special')
lbxrs=plotRs(test.results)
dev.off()
write.table(lbxrs$resultsTable,"/WORK/teaching/project/xuzhiyu/b6_xtail/shoot_rootrs_results.txt",quote=F,sep="\t");

pdf('/WORK/teaching/project/xuzhiyu/b6_xtail/shoot_rootvolcano.pdf',width=6,height=4,paper='special')
volcanoPlot(test.results)
dev.off()
```

**upf2_air__upf2_eth.sh**
```sh
#!/bin/bash
#SBATCH -J tophat_test
#SBATCH -p CN_BIOT
#SBATCH --nodes=1
#SBATCH --ntasks=10
#SBATCH --output=%j.out
#SBATH --error=%j.err

# Get the software
export PATH=/WORK/teaching/bin:$PATH

/WORK/app/singularity-2.0-20191204/bin/singularity exec  -B /WORK/teaching/project:/WORK/teaching/project /WORK/teaching/project/singularity_images/xtail Rscript /WORK/teaching/project/xuzhiyu/b6_xtail/upf2_air__upf2_eth.r
```

**upf2_air__upf2_eth.r**

name=(RNAseq_upf2-Ethylene-Total1 RNAseq_upf2-Air-Total2 RNAseq_upf2-Air-Total1 riboseq_upf2-Ethylene-RFP2 riboseq_upf2-Ethylene-RFP1 riboseq_upf2-Air-RFP2 riboseq_upf2-Air-RFP1 RNAseq_Col-Ethylene-Total2 RNAseq_Col-Ethylene-Total1 RNAseq_Col-Air-Total2 RNAseq_Col-Air-Total1 riboseq_Col-Ethylene-RFP2 riboseq_Col-Ethylene-RFP1 riboseq_Col_Air_RFP2 riboseq_Col-Air-RFP1)

```R
library(xtail)
lbxd=read.table('/WORK/teaching/project/xuzhiyu/b4-count/ethylene.counter',header=T,row.name=1)
mrna=lbxd[,c(2,3,1,1)]
rpf=lbxd[,c(6,7,4,5)]
condition=c("upf2_air","upf2_air","upf2_eth","upf2_eth")
test.results=xtail(mrna,rpf,condition,bins=1000,threads=2)
summary(test.results)

#
test.tab=resultsTable(test.results);
head(test.tab,5)

write.table(test.tab,"/WORK/teaching/project/xuzhiyu/b6_xtail/upf2_air__upf2_eth_results.txt",quote=F,sep="\t");

# Visualization
pdf('/WORK/teaching/project/xuzhiyu/b6_xtail/upf2_air__upf2_ethFC.pdf',width=6,height=4,paper='special')
lbxfc=plotFCs(test.results)
dev.off()
write.table(lbxfc$resultsTable,"/WORK/teaching/project/xuzhiyu/b6_xtail/upf2_air__upf2_ethfc_results.txt",quote=F,sep="\t");

pdf('/WORK/teaching/project/xuzhiyu/b6_xtail/upf2_air__upf2_ethRs.pdf',width=6,height=4,paper='special')
lbxrs=plotRs(test.results)
dev.off()
write.table(lbxrs$resultsTable,"/WORK/teaching/project/xuzhiyu/b6_xtail/upf2_air__upf2_ethrs_results.txt",quote=F,sep="\t");

pdf('/WORK/teaching/project/xuzhiyu/b6_xtail/upf2_air__upf2_ethvolcano.pdf',width=6,height=4,paper='special')
volcanoPlot(test.results)
dev.off()
```



**col_air__col_eth.sh**
```sh
#!/bin/bash
#SBATCH -J tophat_test
#SBATCH -p CN_BIOT
#SBATCH --nodes=1
#SBATCH --ntasks=10
#SBATCH --output=%j.out
#SBATH --error=%j.err

# Get the software
export PATH=/WORK/teaching/bin:$PATH

/WORK/app/singularity-2.0-20191204/bin/singularity exec  -B /WORK/teaching/project:/WORK/teaching/project /WORK/teaching/project/singularity_images/xtail Rscript /WORK/teaching/project/xuzhiyu/b6_xtail/col_air__col_eth.r
```
**col_air__col_eth.r**

name=(RNAseq_upf2-Ethylene-Total1 RNAseq_upf2-Air-Total2 RNAseq_upf2-Air-Total1 riboseq_upf2-Ethylene-RFP2 riboseq_upf2-Ethylene-RFP1 riboseq_upf2-Air-RFP2 riboseq_upf2-Air-RFP1 RNAseq_Col-Ethylene-Total2 RNAseq_Col-Ethylene-Total1 RNAseq_Col-Air-Total2 RNAseq_Col-Air-Total1 riboseq_Col-Ethylene-RFP2 riboseq_Col-Ethylene-RFP1 riboseq_Col_Air_RFP2 riboseq_Col-Air-RFP1)
```R
library(xtail)
lbxd=read.table('/WORK/teaching/project/xuzhiyu/b4-count/ethylene.counter',header=T,row.name=1)
mrna=lbxd[,c(10,11,8,9)]
rpf=lbxd[,c(14,15,12,13)]
condition=c("col_air","col_air","col_eth","col_eth")
test.results=xtail(mrna,rpf,condition,bins=1000,threads=2)
summary(test.results)

#
test.tab=resultsTable(test.results);
head(test.tab,5)

write.table(test.tab,"/WORK/teaching/project/xuzhiyu/b6_xtail/col_air__col_eth_results.txt",quote=F,sep="\t");

# Visualization
pdf('/WORK/teaching/project/xuzhiyu/b6_xtail/col_air__col_ethFC.pdf',width=6,height=4,paper='special')
lbxfc=plotFCs(test.results)
dev.off()
write.table(lbxfc$resultsTable,"/WORK/teaching/project/xuzhiyu/b6_xtail/col_air__col_ethfc_results.txt",quote=F,sep="\t");

pdf('/WORK/teaching/project/xuzhiyu/b6_xtail/col_air__col_ethRs.pdf',width=6,height=4,paper='special')
lbxrs=plotRs(test.results)
dev.off()
write.table(lbxrs$resultsTable,"/WORK/teaching/project/xuzhiyu/b6_xtail/col_air__col_ethrs_results.txt",quote=F,sep="\t");

pdf('/WORK/teaching/project/xuzhiyu/b6_xtail/col_air__col_ethvolcano.pdf',width=6,height=4,paper='special')
volcanoPlot(test.results)
dev.off()
```


**col_eth__upf2_eth.sh**
```sh
#!/bin/bash
#SBATCH -J tophat_test
#SBATCH -p CN_BIOT
#SBATCH --nodes=1
#SBATCH --ntasks=10
#SBATCH --output=%j.out
#SBATH --error=%j.err

# Get the software
export PATH=/WORK/teaching/bin:$PATH

/WORK/app/singularity-2.0-20191204/bin/singularity exec  -B /WORK/teaching/project:/WORK/teaching/project /WORK/teaching/project/singularity_images/xtail Rscript /WORK/teaching/project/xuzhiyu/b6_xtail/col_eth__upf2_eth.r
```
**col_eth__upf2_eth.r**

name=(RNAseq_upf2-Ethylene-Total1 RNAseq_upf2-Air-Total2 RNAseq_upf2-Air-Total1 riboseq_upf2-Ethylene-RFP2 riboseq_upf2-Ethylene-RFP1 riboseq_upf2-Air-RFP2 riboseq_upf2-Air-RFP1 RNAseq_Col-Ethylene-Total2 RNAseq_Col-Ethylene-Total1 RNAseq_Col-Air-Total2 RNAseq_Col-Air-Total1 riboseq_Col-Ethylene-RFP2 riboseq_Col-Ethylene-RFP1 riboseq_Col_Air_RFP2 riboseq_Col-Air-RFP1)
```R
library(xtail)
lbxd=read.table('/WORK/teaching/project/xuzhiyu/b4-count/ethylene.counter',header=T,row.name=1)
mrna=lbxd[,c(8,9,1,1)]
rpf=lbxd[,c(12,13,4,5)]
condition=c("col_eth","col_eth","upf2_eth","upf2_eth")
test.results=xtail(mrna,rpf,condition,bins=1000,threads=2)
summary(test.results)

#
test.tab=resultsTable(test.results);
head(test.tab,5)

write.table(test.tab,"/WORK/teaching/project/xuzhiyu/b6_xtail/col_eth__upf2_eth_results.txt",quote=F,sep="\t");

# Visualization
pdf('/WORK/teaching/project/xuzhiyu/b6_xtail/col_eth__upf2_ethFC.pdf',width=6,height=4,paper='special')
lbxfc=plotFCs(test.results)
dev.off()
write.table(lbxfc$resultsTable,"/WORK/teaching/project/xuzhiyu/b6_xtail/col_eth__upf2_ethfc_results.txt",quote=F,sep="\t");

pdf('/WORK/teaching/project/xuzhiyu/b6_xtail/col_eth__upf2_ethRs.pdf',width=6,height=4,paper='special')
lbxrs=plotRs(test.results)
dev.off()
write.table(lbxrs$resultsTable,"/WORK/teaching/project/xuzhiyu/b6_xtail/col_eth__upf2_ethrs_results.txt",quote=F,sep="\t");

pdf('/WORK/teaching/project/xuzhiyu/b6_xtail/col_eth__upf2_ethvolcano.pdf',width=6,height=4,paper='special')
volcanoPlot(test.results)
dev.off()
```


**col_air__upf2_air.sh**
```sh
#!/bin/bash
#SBATCH -J tophat_test
#SBATCH -p CN_BIOT
#SBATCH --nodes=1
#SBATCH --ntasks=10
#SBATCH --output=%j.out
#SBATH --error=%j.err

# Get the software
export PATH=/WORK/teaching/bin:$PATH

/WORK/app/singularity-2.0-20191204/bin/singularity exec  -B /WORK/teaching/project:/WORK/teaching/project /WORK/teaching/project/singularity_images/xtail Rscript /WORK/teaching/project/xuzhiyu/b6_xtail/col_air__upf2_air.r
```
**col_air__upf2_air.r**

name=(RNAseq_upf2-Ethylene-Total1 RNAseq_upf2-Air-Total2 RNAseq_upf2-Air-Total1 riboseq_upf2-Ethylene-RFP2 riboseq_upf2-Ethylene-RFP1 riboseq_upf2-Air-RFP2 riboseq_upf2-Air-RFP1 RNAseq_Col-Ethylene-Total2 RNAseq_Col-Ethylene-Total1 RNAseq_Col-Air-Total2 RNAseq_Col-Air-Total1 riboseq_Col-Ethylene-RFP2 riboseq_Col-Ethylene-RFP1 riboseq_Col_Air_RFP2 riboseq_Col-Air-RFP1)
```R
library(xtail)
lbxd=read.table('/WORK/teaching/project/xuzhiyu/b4-count/ethylene.counter',header=T,row.name=1)
mrna=lbxd[,c(10,11,2,3)]
rpf=lbxd[,c(14,15,6,7)]
condition=c("col_air","col_air","upf2_air","upf2_air")
test.results=xtail(mrna,rpf,condition,bins=1000,threads=2)
summary(test.results)

#
test.tab=resultsTable(test.results);
head(test.tab,5)

write.table(test.tab,"/WORK/teaching/project/xuzhiyu/b6_xtail/col_air__upf2_air_results.txt",quote=F,sep="\t");

# Visualization
pdf('/WORK/teaching/project/xuzhiyu/b6_xtail/col_air__upf2_airFC.pdf',width=6,height=4,paper='special')
lbxfc=plotFCs(test.results)
dev.off()
write.table(lbxfc$resultsTable,"/WORK/teaching/project/xuzhiyu/b6_xtail/col_air__upf2_airfc_results.txt",quote=F,sep="\t");

pdf('/WORK/teaching/project/xuzhiyu/b6_xtail/col_air__upf2_airRs.pdf',width=6,height=4,paper='special')
lbxrs=plotRs(test.results)
dev.off()
write.table(lbxrs$resultsTable,"/WORK/teaching/project/xuzhiyu/b6_xtail/col_air__upf2_airrs_results.txt",quote=F,sep="\t");

pdf('/WORK/teaching/project/xuzhiyu/b6_xtail/col_air__upf2_airvolcano.pdf',width=6,height=4,paper='special')
volcanoPlot(test.results)
dev.off()
```
**extrac_fc.pl**
```pl
open DATA, "<$ARGV[0]";
$name=$ARGV[0];
$name=~s/.txt//g;
$name="c-out".$name;
system("mkdir $name");
open OUT1, ">$name/$ARGV[0].transcript_up.txt";
open OUT2, ">$name/$ARGV[0].transcript_do.txt";
open OUT3, ">$name/$ARGV[0].translatio_up.txt";
open OUT4, ">$name/$ARGV[0].translatio_do.txt";
$head=<DATA>;
print OUT1 "$head";
print OUT2 "$head";
print OUT3 "$head";
print OUT4 "$head";
while(<DATA>)
{
chomp;
@data=split /\t/,$_;
if($data[13]=~/transcription_only/)
{
if($data[1]>0)
{
   print OUT1 "$_\n";
}else{
   print OUT2 "$_\n";
}
}
if($data[13]=~/translation_only/)
{
if($data[2]>0)
{
   print OUT3 "$_\n";
}else{
   print OUT4 "$_\n";
}
}

}


```
```sh
cut -f 14 col_air__col_ethfc_results.txt|sort -u
vi extrac_fc.pl
perl extrac_fc.pl col_air__col_ethfc_results.txt

for i in `ls|grep fc_results.txt`;
do 
echo $i;
perl extrac_fc.pl $i;
done

```
**extrac_rs.pl**
```pl

open DATA, "<$ARGV[0]";
$name=$ARGV[0];
$name=~s/.txt//g;
$name="c-out".$name;
system("mkdir $name");


while(<DATA>)
{
   chomp;
   @data=split /\t/,$_;
   if($data[13]=~/homodirectional/){}
   elsif($data[13]=~/opposite_change/){}
   elsif($data[13]=~/^$/){}
   elsif($data[13]=~/stable/){}else
   {
      $hash{$data[13]}='';
   }
}
close DATA;
@name=keys(%hash);
print "@name\n";
open DATA, "<$ARGV[0]";
open OUT1, ">$name/$name[0].up.txt";
open OUT2, ">$name/$name[0].do.txt";
open OUT3, ">$name/$name[1].up.txt";
open OUT4, ">$name/$name[1].do.txt";
$head=<DATA>;
print OUT1 "$head";
print OUT2 "$head";
print OUT3 "$head";
print OUT4 "$head";
while(<DATA>)
{
   chomp;
   @data=split /\t/,$_;
   if($data[13]=~/$name[0]/)
   {
      if(abs($data[5])>1)
      {
         if($data[5]>0){
            print OUT1 "$_\n";
         }else{
            print OUT2 "$_\n";
         }
      }
      if(abs($data[6])>1)
      {
         if($data[6]>0)
         {
            print OUT1 "$_\n";
         }else{
            print OUT2 "$_\n";
         }
      }
   }
   if($data[13]=~/$name[1]/)
   {
            if(abs($data[5])>1)
      {
         if($data[5]>0){
            print OUT3 "$_\n";
         }else{
            print OUT4 "$_\n";
         }
      }
      if(abs($data[6])>1)
      {
         if($data[6]>0)
         {
            print OUT3 "$_\n";
         }else{
            print OUT4 "$_\n";
         }
      }
   }  

}

```
```sh
cut -f 14 col_air__col_ethrs_results.txt | sort -u

vi extrac_rs.pl

perl extrac_rs.pl  col_air__col_ethrs_results.txt
for i in `ls|grep rs_results.txt`;
do echo $i;
perl extrac_rs.pl $i;
done


head -n 1 col_air__col_ethrs_results.txt
grep col_air_only  col_air__col_ethrs_results.txt | head -n 2

cd c-outcol_air__col_ethrs_results
```
**te_result.pl**
```pl

open DATA, "<$ARGV[0]";
$name=$ARGV[0];
$name=~s/.txt//g;
$name="c-out".$name;
system("mkdir $name");
open OUT1, ">$name/$ARGV[0].up.txt";
open OUT2, ">$name/$ARGV[0].down.txt";
$head=<DATA>;
print OUT1 $head;
print OUT2 $head;
while(<DATA>)
{
chomp;
@data=split /\t/,$_;
   if($data[6]<0.01)
   {
      if($data[5]>1)
      {
         print OUT1 "$_\n";
      }
      if($data[5]<-1)
      {
         print OUT2 "$_\n";
      }
   }
}

```

```sh
head col_air__col_eth_results.txt
vi te_result.pl
perl te_result.pl col_air__col_eth_results.txt

for i in `ls|grep results.txt$|grep -v fc_results.txt|grep -v rs_results.txt`;
do echo $i;
perl te_result.pl $i;
done
cd c-outcol_air__col_eth_results
```
**b7_count_ribocode.sh**
```sh

echohead(){
   echo '#!/bin/bash
#SBATCH -J slum
#SBATCH -p CN_BIOT
#SBATCH --nodes=1
#SBATCH --ntasks=3
#SBATCH --output=%slum.out
#SBATH --error=%slum.err

# Get the software
export PATH=/WORK/teaching/bin:$PATH
export SINGULARITY_BINDPATH="/WORK"
'
}

out=/WORK/teaching/project/xuzhiyu/b7-count-ribocode
gtf=/WORK/teaching/project/xuzhiyu/a6-ribocode/RiboCode_ORFs_result.gtf
[ -d ${out} ] || mkdir ${out}
[ -d ${out}/script ] || mkdir ${out}/script


generate(){
dir='/WORK/teaching/project/xuzhiyu/bam_dir';

if [ -f ${out}/script/all.genome.txt ];then
rm -rf ${out}/script/all.genome.txt
fi


for i in ` ls ${dir}|grep Aligned.sortedByCoord.out.bam`;
do
 echo "${dir}/${i}" >>${out}/script/all.genome.txt;
done;
}

count_all(){
for i in `cat ${out}/script/all.genome.txt`;do
name=$(echo $i|rev | cut -f 1 -d '/' | rev|cut -f 1 -d '.') ;
echo ${name};
echohead > ${out}/script/$name.all.sh ;
echo "/WORK/app/singularity-2.0-20191204/bin/singularity exec  -B /WORK/teaching/project:/WORK/teaching/project /WORK/teaching/project/singularity_images/ribocodeminer /home/test/anaconda3/bin/ORFcount -g ${gtf} -r ${i} -f 15 -l 5 -e 100 -m 26 -M 34 -o ${out}/${name}.all.count " >> ${out}/script/$name.all.sh;

done
}

#generate
count_all

```


