# snp

## remove adapter
```sh
cutadapt_function(){
echo cutadapt -m 20 -a ${adapt1} -A ${adapt2} -o ${data_path}/out1.fastq -p ${data_path}/out2.fastq ${data_path}/${file1} ${data_path}/${file2}
cutadapt -m 20 -a ${adapt1} -A ${adapt2} -o ${data_path}/out1.fastq -p ${data_path}/out2.fastq ${data_path}/${file1} ${data_path}/${file2}

}


# sample1
data_path=/home/xugang/data/wangjydata/1.rawdata/0_FKDL202596126-1a/
adapt1=GGACTCC
adapt2=TATCCTCT
file1=0_FKDL202596126-1a_1.fq.gz
file2=0_FKDL202596126-1a_2.fq.gz
cutadapt_function

#sample2
data_path=/home/xugang/data/wangjydata/1.rawdata/0-jia_FKDL202596128-1a/
adapt1=TAGGCATG
adapt2=TATCCTCT
file1=0-jia_FKDL202596128-1a_1.fq.gz
file2=0-jia_FKDL202596128-1a_2.fq.gz
cutadapt_function
#sample3
data_path=/home/xugang/data/wangjydata/1.rawdata/1_FKDL202596127-1a/
adapt1=TAAGGCGA
adapt2=AGAGTAGA
file1=1_FKDL202596127-1a_1.fq.gz
file2=1_FKDL202596127-1a_2.fq.gz
cutadapt_function
#sample4
data_path=/home/xugang/data/wangjydata/1.rawdata/1-jia_FKDL202596129-1a/
adapt1=CGTACTAG
adapt2=AGAGTAGA
file1=1-jia_FKDL202596129-1a_1.fq.gz
file2=1-jia_FKDL202596129-1a_2.fq.gz
cutadapt_function

```

## bowtie map
```sh
indexp=/home/xugang/data/wangjydata/ref/bowtie_index/emx1
indexp2=/home/xugang/data/wangjydata/ref/bowtie2_index/emx1


bowtief2(){
echo "bowtie2 $name";
bowtie2 -p 6 -N 1 -x ${indexp2} -1 ${datap}/out1.fastq -2 ${datap}/out2.fastq --seed 0 --very-sensitive -S ${datap}/${name}.sam > ${datap}/log.${name}.log

}

bowtief(){
echo "bowtie $name";
bowtie -p 6  ${indexp} -1 ${datap}/out1.fastq -2 ${datap}/out2.fastq -n 3 -S ${datap}/${name}.bowtie1.sam
}

datap=/home/xugang/data/wangjydata/1.rawdata/0_FKDL202596126-1a
name=0_fkd
bowtief
bowtief2

datap=/home/xugang/data/wangjydata/1.rawdata/0-jia_FKDL202596128-1a/
name=0-jia
bowtief
bowtief2

datap=/home/xugang/data/wangjydata/1.rawdata/1_FKDL202596127-1a/
name=1_fkd
bowtief
bowtief2

datap=/home/xugang/data/wangjydata/1.rawdata/1-jia_FKDL202596129-1a/
name=1-jia
bowtief
bowtief2
```

## bam sort and index
```sh
sort(){
samtools view -S -b ${datap}/${name} > ${datap}/${name}.bam
samtools sort ${datap}/${name}.bam > ${datap}/${name}.sort.bam
rm ${datap}/${name}.bam
samtools index ${datap}/${name}.sort.bam
}

# file path
datap=/home/xugang/data/wangjydata/1.rawdata/0_FKDL202596126-1a/
# file name
name=0_fkd.sam
# function
sort
name=0_fkd.bowtie1.sam
# function
sort

# file path
datap=/home/xugang/data/wangjydata/1.rawdata/0-jia_FKDL202596128-1a
# file name
name=0-jia.sam
# function
sort
# file name
name=0-jia.bowtie1.sam
# function
sort

# file path
datap=/home/xugang/data/wangjydata/1.rawdata/1_FKDL202596127-1a
# file name
name=1_fkd.sam
# function
sort
# file name
name=1_fkd.bowtie1.sam
# function
sort

# file path
datap=/home/xugang/data/wangjydata/1.rawdata/1-jia_FKDL202596129-1a
# file name
name=1-jia.sam
# function
sort
# file name
name=1-jia.bowtie1.sam
# function
sort
```

## call snp
```sh
snp(){
[ -d ${outp}/ ] || mkdir -p ${outp}/
freebayes -f ${genome} ${datap}/${name} >${outp}/${name}.vcf
}


# do not change
genome=/home/xugang/data/wangjydata/ref/emx1.fa
outp=/home/xugang/data/wangjydata/3.var.new

# file path
datap=/home/xugang/data/wangjydata/1.rawdata/0_FKDL202596126-1a/
# file name
name=0_fkd.sam.sort.bam
# function
snp
name=0_fkd.bowtie1.sam.sort.bam
# function
snp

# file path
datap=/home/xugang/data/wangjydata/1.rawdata/0-jia_FKDL202596128-1a
# file name
name=0-jia.sam.sort.bam
# function
snp
# file name
name=0-jia.bowtie1.sam.sort.bam
# function
snp

# file path
datap=/home/xugang/data/wangjydata/1.rawdata/1_FKDL202596127-1a
# file name
name=1_fkd.sam.sort.bam
# function
snp
# file name
name=1_fkd.bowtie1.sam.sort.bam
# function
snp

# file path
datap=/home/xugang/data/wangjydata/1.rawdata/1-jia_FKDL202596129-1a
# file name
name=1-jia.sam.sort.bam
# function
snp
# file name
name=1-jia.bowtie1.sam.sort.bam
# function
snp
```
