# Ribo-seq : Raw data process pipeline
[Back home](../README.md)


## Ribo-seq : Raw data process pipeline

![pipeline](../image/a1.pipeline.png)

## Ribo-seq : Raw data process pipeline

* Step1 : cutadapt ( RPF : [25,35] , RNA: [17, ] )
* Step2 : beforeQC
* Step3 : filter
* Step4 : contam (remove ribosome sequence)
* Step5 : afterQC (overrepresent reads)
* Step6 : Tophat + readsNumCal_intron_v3.py
* Step7 : STAR + periodicity
* Step8 : readLengthDistribution
* Step9 : report


## Step1 : cutadapt ( RPF : [25,35] , RNA: [17, ] )

```sh
RPF : cutadaptRPF.sh      –m 25 –M 35
RNA : cutadaptRNA.sh    -m 17 (19)
```
**cutadaptRPF.sh**
```sh
#!/bin/zsh

export PATH=$PATH:/Share/home/tiangeng/software/cap-miRNA/bin
rpf=(RW3 RC3 RC15 RW15)
adapt=CTGTAGGCACCATCAAT

for i in ${rpf[@]}
do
        cutadapt -m 25 -M 35 --match-read-wildcards -a $adapt -o ${i}_trimmed.fastq ../rawdata/${i}_R1.fastq > ${i}_trimmed.log
done
```

|command |mean |
|:-|:-|
|-m |Discard trimmed reads that are shorter than LENGTH.Reads that are too short even before adapter removal are also discarded. In colorspace, an initial primer is not counted (default: 0). |
|-M |Discard trimmed reads that are longer than LENGTH. Reads that are too long even before adapter removal are also discarded. In colorspace, an initial primer is not counted (default: no limit). |
|--match-read-wildcards |Allow IUPAC wildcards in reads (default: False). |
|-a |Sequence of an adapter that was ligated to the 3' end. The adapter itself and anything that follows is trimmed. If the adapter sequence ends with the '$' character, the adapter is anchored to the end of the read and only found if it is a suffix of the read. |


