# Ribo-seq : Raw data process pipeline
[Back home](../README.md)


## Ribo-seq : Raw data process pipeline

![pipeline](../image/a1.pipeline.png)

## Ribo-seq : Raw data process pipeline

* Step1 : cutadapt ( RPF : [25,35] , RNA: [17, ] )
* Step2 : beforeQC
* Step3 : filter
* Step4 : contam (remove ribosome sequence)
* Step5 : afterQC (overrepresent reads)
* Step6 : Tophat + readsNumCal_intron_v3.py
* Step7 : STAR + periodicity
* Step8 : readLengthDistribution
* Step9 : report


## Step1 : cutadapt ( RPF : [25,35] , RNA: [17, ] )

```sh
RPF : cutadaptRPF.sh      –m 25 –M 35
RNA : cutadaptRNA.sh    -m 17 (19)
```
**cutadaptRPF.sh**
```sh
#!/bin/zsh

export PATH=$PATH:/Share/home/tiangeng/software/cap-miRNA/bin
rpf=(RW3 RC3 RC15 RW15)
adapt=CTGTAGGCACCATCAAT

for i in ${rpf[@]}
do
        cutadapt -m 25 -M 35 --match-read-wildcards -a $adapt -o ${i}_trimmed.fastq ../rawdata/${i}_R1.fastq > ${i}_trimmed.log
done
```

|command |mean |
|:-|:-|
|-m |Discard trimmed reads that are shorter than LENGTH.Reads that are too short even before adapter removal are also discarded. In colorspace, an initial primer is not counted (default: 0). |
|-M |Discard trimmed reads that are longer than LENGTH. Reads that are too long even before adapter removal are also discarded. In colorspace, an initial primer is not counted (default: no limit). |
|--match-read-wildcards |Allow IUPAC wildcards in reads (default: False). |
|-a |Sequence of an adapter that was ligated to the 3' end. The adapter itself and anything that follows is trimmed. If the adapter sequence ends with the '$' character, the adapter is anchored to the end of the read and only found if it is a suffix of the read. |

**cutadapt_totalRNA.sh**
```sh
#!/bin/zsh

export PATH=$PATH:/Share/home/tiangeng/software/cap-miRNA/bin
rpf=(TC15  TC3  TW15  TW3)
adapt=CTGTAGGCACCATCAAT

for i in ${rpf[@]}
do
        cutadapt -m 17 --match-read-wildcards -a $adapt -o ${i}_trimmed.fastq ../rawdata/${i}_R1.fastq > ${i}_trimmed.log
done
```

## Step2 : beforeQC

```sh
#!/bin/bash

rpf=(RC15 RC3 RW15 RW3  TC15 TC3 TW15 TW3)
for i in ${rpf[@]}
do
        /Share/home/tiangeng/software/FastQC/fastqc ../cutadapt/${i}_trimmed.fastq -o ./
done
```

## Step3 : filter
```sh
#!/bin/bash

rpf=(RC15  RC3  RW15  RW3 TC15  TC3  TW15  TW3)
for i in ${rpf[@]}
do
        /Share/app/fastx_toolkit0.14/bin/fastq_quality_filter -Q33 -v -q 25 -p 75 -i ../cutadapt/${i}_trimmed.fastq -o ${i}_trimmedQfilter.fastq > ${i}_Qfilter.log
done
```

|command | mean|
|:- |:- |
|[-h]|         = This helpful help screen.| 
|[-q N]|        = Minimum quality score to keep.| 
|[-p N]|        = Minimum percent of bases that must have [-q] quality.| 
|[-z] |         = Compress output with GZIP.| 
|[-i INFILE]|   = FASTA/Q input file. default is STDIN.| 
|[-o OUTFILE]|  = FASTA/Q output file. default is STDOUT.| 
|[-v]|          = Verbose - report number of sequences. If [-o] is specified,  report will be printed to STDOUT. If [-o] is not specified (and output goes to STDOUT), report will be printed to STDERR.| 

##  Step4 : contam (remove ribosome sequence)
```sh
#!/bin/bash

bowtieindex=/Share/home/tiangeng/Database/Reference_genome/Mus-musculus_rRNA_bowtie-index/musRibosomal
name=(RC15 RC3 RW15 RW3  TC15 TC3 TW15 TW3)

for i in ${name[@]}
do
         bowtie -n 0 -norc --best -l 15 -p 8 --un=nocontam_${i}.fastq $bowtieindex -q ../filter/${i}_trimmedQfilter.fastq ${i}.alin > ${i}.err
done
```

|command | mean|
|:- |:- |
|-n/--seedmms <int>| max mismatches in seed (can be 0-3, default: -n 2)|
|--nofw/--norc|do not align to forward/reverse-complement reference strand|
|--best|hits guaranteed best stratum; ties broken by quality|
|-l/--seedlen <int>| seed length for -n (default: 28)|
|-p/--threads <int>| number of alignment threads to launch (default: 1)|
|--un <fname>|write unaligned reads/pairs to file(s) <fname>|
|-q|query input files are FASTQ .fq/.fastq (default)|

## Step5 : afterQC (overrepresent reads)

```sh
#!/bin/bash

rpf=(RC15 RC3 RW15 RW3  TC15 TC3 TW15 TW3)
for i in ${rpf[@]}
do
        /Share/home/tiangeng/software/FastQC/fastqc ../contam/nocontam_${i}.fastq -o ./
done
```

## Step6 : Tophat + readsNumCal_intron_v3.py

```sh
-rw-rw-r--  1 tiangeng tiangeng  22K Jun 28 19:37 nocontam_TC15_Intron.png
gtf=/Share/home/tiangeng/Database/Reference_genome/Mus_musculus_Ensembl_GRCm38_star_genome-index/Mus_musculus.GRCm38.95.gtf
bowtieIndex=/Share/home/tiangeng/Database/Reference_genome/Mus_musculus_Ensembl_GRCm38_bowtie_genome-index/Mus_musculus.GRCm38.dna.primary_assembly
transIndex=/Share/home/tiangeng/Database/Reference_genome/Mus_musculus_Ensembl_GRCm38_tophat_trans-index/Mus_musculus.GRCm38.95
export PATH=/Share/home/tiangeng/software/bowtie-1.1.2:$PATH
export PATH=/Share/home/tiangeng/anaconda2/bin:$PATH
fileName=(  TC15 TC3 TW15 TW3)
for i in ${fileName[@]}
do
     tophat  -o ${i}_out -g 10 --bowtie1 --read-realign-edit-dist 0 --library-type fr-secondstrand -G $gtf --transcriptome-index=$transIndex --no-novel-juncs --segment-length=15 $bowtieIndex ../contam/nocontam_${i}.fastq
done
```
|command |type| mean|
|:- |:- |:- |
|-o/--output-dir |<string>| [ default: ./tophat_out]|
|-g/--max-multihits|<int>| [ default: 20]
|--bowtie-n| |[ default: bowtie -v]|
|--read-realign-edit-dist|<int>|[ default: "read-edit-dist" + 1 ]|
|--library-type| <string> |   (fr-unstranded, fr-firststrand,  fr-secondstrand)|
|--transcriptome-index|<bwtidx> |(transcriptome bowtie index)|
|--segment-length |<int>|[ default: 25 ]|

**run.sh**
```sh
#!/bin/bash
export PATH=/Share/home/tiangeng/anaconda2/bin:$PATH
name=(RC15  RC3  RW15  RW3 TC15  TC3  TW15  TW3 )
gtf=/Share/home/tiangeng/Database/Reference_genome/Mus_musculus_Ensembl_GRCm38_star_genome-index/Mus_musculus.GRCm38.95.gtf
for i in ${name[@]}
do
        python ./readsNumCal_intron_v3.py ./${i}_out/accepted_hits.bam $gtf nocontam_${i}_mappedNum_intron.txt nocontam_${i}
done
```
[readsNumCal_intron_v3.py](readsNumCal_intron_v3.py)

## Step7 : STAR + periodicity
**STAR.sh**
```sh
#!/bin/bash
export PATH=$PATH:~/software/STAR-master/bin/Linux_x86_64_static
genomeFile=/Share/home/tiangeng/Database/Reference_genome/Mus_musculus_Ensembl_GRCm38_star_genome-index
fileName=(RC15 RC3 RW15 RW3  TC15 TC3 TW15 TW3)
for i in ${fileName[@]}
do
        mkdir -p ${i}_STAR
        cd ${i}_STAR
        STAR --runThreadN 8 --alignEndsType EndToEnd --outFilterMismatchNmax 1 --outFilterMultimapNmax 1 --genomeDir $genomeFile --readFilesIn ../../contam/nocontam_${i}.fastq --outFileNamePrefix $i --outSAMtype BAM SortedByCoordinate --quantMode TranscriptomeSAM GeneCounts
        cd ../
done
```

|command |type| mean|
|:- |:- |:- |
|runThreadN |int:| number of threads to run STAR| 
|alignEndsType |string: |see below|
| |Local|standard local alignment with soft-clipping allowed|
| |EndToEnd| force end-to-end read alignment, do not soft-clip |
| | Extend5pOfRead1|fully extend only the 5p of the read1, all other ends: local alignment|
| |Extend5pOfReads12 |fully extend only the 5p of the both read1 and read2, all other ends: local alignment|
|alignEndsType |  string: |type of read ends alignment|
| |Local|standard local alignment with soft-clipping allowed|
| |EndToEnd| force end-to-end read alignment, do not soft-clip|
| |Extend5pOfRead1|fully extend only the 5p of the read1, all other ends: local alignment|
| |Extend5pOfReads12| fully extend only the 5p of the both read1 and read2, all other ends: local alignment|
|outFilterMismatchNmax|int:| alignment will be output only if it has no more mismatches than this value.|
|outSAMtype|strings: |type of SAM/BAM output|
| |1st word: | |
| |BAM|output BAM without sorting|
| |SAM|output SAM without sorting|
| |None|no SAM/BAM output|
| |2nd, 3rd| |
| |Unsorted| standard unsorted|
| |SortedByCoordinate|sorted by coordinate. This option will allocate extra memory for sorting which can be specified by --limitBAMsortRAM.|
|quantMode |string(s): |types of quantification requested|
| | - |none|
| |TranscriptomeSAM|output SAM/BAM alignments to transcriptome into a separate file|
| |GeneCounts| count reads per gene|

## Step8 : readLengthDistribution
**run.sh**
```sh
#!/bin/bash
export PATH=/Share/home/tiangeng/anaconda2/bin:$PATH
name=(19R129_combined_R1 19R131_combined_R1 19R130_combined_R1 19R132_combined_R1 19R123_combined_R1 19R125_combined_R1 19R127_combined_R1 19R124_combined_R1 19R126_combined_R1 19R128_combined_R1)
gtf=/Share/home/tiangeng/Database/Reference_genome/Mus_musculus_Ensembl_GRCm38_star_genome-index/Mus_musculus.GRCm38.95.gtf
for i in ${name[@]}
do
        python ./readsNumCal_intron_v3.py ./${i}_out/accepted_hits.bam $gtf nocontam_${i}_mappedNum_intron.txt nocontam_${i}
done
```
[readsNumCal_intron_v3.py](readsNumCal_intron_v3.py)

## Step9 : report




