# Ribo-seq : Differential translations ( Xtail )
[Back home](../README.md)

## Workflow
![Workflow](../image/a2.te.png)

TE (Translation Efficiency) = RPF / RNA

logTE = log (TE treat / TE control) 

          = log( (RPF2/RNA2)  / (RPF1/RNA1) )               

          = log( (RPF2/RPF1)   / (RNA2/RNA1) )

## Step:

* Step1 :  combined remove Ribosome RNA fastq files
* Step2 :  STAR mapping
* Step3 :  calculate counts
* Step4 :  run Xtail

## Step3 :  calculate counts   
RNA:  exon  ( mRNA_count.py  or HTSeq)  
RPF:  CDS, exclude first15 codons last5 codons  ( RPF_count_CDS.py or excludeStartStopModifHTSeq.py)  

## Step4 :  run Xtail  (refer to xtail.pdf )   
Attention : test.results <- xtail(mrna,rpf,condition,bins=1000)  

bins=1000 for example , should change to bins=10000  







