# Singularity

## Setup

First, install brew if you do not have it already.
```sh
/usr/bin/ruby -e "$(curl -fsSL https://raw.githubusercontent.com/Homebrew/install/master/install)"
```

Next, install Vagrant and the necessary bits.
```sh
brew cask install virtualbox
brew cask install vagrant
brew cask install vagrant-manager
```

## Option 1: Singularityware Vagrant Box
We are maintaining a set of Vagrant Boxes via Vagrant Cloud, one of Hashicorp many tools that likely you’ve used and haven’t known it. The current stable version of Singularity is available here:
[singularityware/singularity-2.4](https://app.vagrantup.com/singularityware/boxes/singularity-2.4/versions/2.4)

For other versions of Singularity see our Vagrant Cloud repository.
```sh
mkdir singularity-vm
cd singularity-vm
```

Then issue the following commands to bring up the Virtual Machine:

```sh
vagrant init singularityware/singularity-2.4
vagrant up
vagrant ssh
```
You are then ready to go with Singularity 2.4!

```sh
vagrant@vagrant:~$ which singularity
/usr/local/bin/singularity
vagrant@vagrant:~$ singularity --version
2.4-dist

vagrant@vagrant:~$ sudo singularity build growl-llo-world.simg shub://vsoch/hello-world
Cache folder set to /root/.singularity/shub
Progress |===================================| 100.0% 
Building from local image: /root/.singularity/shub/vsoch-hello-world-master.simg
Building Singularity image...
Singularity container built: growl-llo-world.simg
Cleaning up...
vagrant@vagrant:~$ ./growl-llo-world.simg
RaawwWWWWWRRRR!!
```

Note that when you do vagrant up you can also select the provider, if you use vagrant for multiple providers. For example:
```sh
vagrant up --provider virtualbox
```
although this isn’t entirely necessary if you only have it configured for virtualbox.
```sh
vagrant ssh
```

## Docker
```sh
singularity --help
USAGE: singularity [global options...] <command> [command options...] ...

GLOBAL OPTIONS:
    -d|--debug    Print debugging information
    -h|--help     Display usage summary
    -s|--silent   Only print errors
    -q|--quiet    Suppress all normal output
       --version  Show application version
    -v|--verbose  Increase verbosity +1
    -x|--sh-debug Print shell wrapper debugging information

GENERAL COMMANDS:
    help       Show additional help for a command or container                  
    selftest   Run some self tests for singularity install                      

CONTAINER USAGE COMMANDS:
    exec       Execute a command within container                               
    run        Launch a runscript within container                              
    shell      Run a Bourne shell within container                              
    test       Launch a testscript within container                             

CONTAINER MANAGEMENT COMMANDS:
    apps       List available apps within a container                           
    bootstrap  *Deprecated* use build instead                                   
    build      Build a new Singularity container                                
    check      Perform container lint checks                                    
    inspect    Display a container's metadata                                   
    mount      Mount a Singularity container image                              
    pull       Pull a Singularity/Docker container to $PWD                      

COMMAND GROUPS:
    image      Container image command group                                    
    instance   Persistent instance command group                                


CONTAINER USAGE OPTIONS:
    see singularity help <command>

For any additional help or support visit the Singularity
website: http://singularity.lbl.gov/
```

For example, to pass the --debug option to the main singularity command and run Singularity with debugging messages on:
```sh
$ singularity --debug run shub://GodloveD/lolcow
```
And to pass the --containall option to the run command and run a Singularity image in an isolated manner:
```sh
$ singularity run --containall shub://GodloveD/lolcow
```
To learn more about a specific Singularity command, type one of the following:
```sh
$ singularity help <command>
$ singularity --help <command>
$ singularity -h <command>
$ singularity <command> --help
$ singularity <command> -h
```
Users can also write help docs specific to a container or for an internal module called an app. If those help docs exist for a particular container, you can view them like so.
```sh
$ singularity help container.simg            # See the container's help, if provided
$ singularity help --app foo container.simg  # See the help for foo, if provided
```

## Download pre-built images
```sh
# pull docker image from docker hub.
singularity build lolcow.simg docker://godlovedc/lolcow
# open one shell into the image to processing
singularity shell lolcow.simg
# 
singularity shell -B /home/vagrant/share:/home lolcow.simg
#
singularity exec lolcow.simg ls /

singularity exec lolcow.simg bash

singularity exec shub://singularityhub/ubuntu cat /etc/os-release

singularity run lolcow.simg

sudo singularity build --writable ubuntu.img docker://ubuntu

sudo singularity shell --writable ubuntu.img

```

```sh
singularity build bioinfo_tsinghua.simg docker://gangxu/bioinfo_tsinghua
singularity shell -B /home/vagrant/share:/home/share bioinfo_tsinghua.simg
singularity exec bioinfo_tsinghua.simg bash
singularity run bioinfo_tsinghua.simg
singularity exec /home/vagrant/bioinfo_tsinghua.simg tophat -p 4 -G /home/vagrant/share/yeast_annotation.gff --no-coverage-search -o /home/vagrant/mapping/wt1_thout \
    /home/vagrant/share/bowtie_index/YeastGenome /home/vagrant/share/Raw_reads_10k/wt1.fq 

singularity exec /home/vagrant/bioinfo_tsinghua.simg tophat -p 4 -G /home/vagrant/share/yeast_annotation.gff --no-coverage-search -o /home/vagrant/mapping/wt2_thout \
    /home/vagrant/share/bowtie_index/YeastGenome /home/vagrant/share/Raw_reads_10k/wt2.fq 

singularity exec /home/vagrant/bioinfo_tsinghua.simg tophat -p 4 -G /home/vagrant/share/yeast_annotation.gff --no-coverage-search -o /home/vagrant/mapping/wt1X_thout \
    /home/vagrant/share/bowtie_index/YeastGenome /home/vagrant/share/Raw_reads_10k/wt1X.fq 

singularity exec /home/vagrant/bioinfo_tsinghua.simg tophat -p 4 -G /home/vagrant/share/yeast_annotation.gff --no-coverage-search -o /home/vagrant/mapping/wt2X_thout \
    /home/vagrant/share/bowtie_index/YeastGenome /home/vagrant/share/Raw_reads_10k/wt2X.fq

singularity exec /home/vagrant/bioinfo_tsinghua.simg bamtools index -in /home/vagrant/mapping/wt1_thout/accepted_hits.bam 

singularity exec /home/vagrant/bioinfo_tsinghua.simg bamtools index -in /home/vagrant/mapping/wt2_thout/accepted_hits.bam 

singularity exec /home/vagrant/bioinfo_tsinghua.simg bamtools index -in /home/vagrant/mapping/wt1X_thout/accepted_hits.bam 

singularity exec /home/vagrant/bioinfo_tsinghua.simg bamtools index -in /home/vagrant/mapping/wt2X_thout/accepted_hits.bam

singularity exec /home/vagrant/bioinfo_tsinghua.simg bamtools filter -region chrI -in /home/vagrant/mapping/wt1_thout/accepted_hits.bam \
    -out /home/vagrant/mapping/wt1_thout/chrI.bam

singularity exec /home/vagrant/bioinfo_tsinghua.simg bamtools filter -region chrI -in /home/vagrant/mapping/wt2_thout/accepted_hits.bam \
    -out /home/vagrant/mapping/wt2_thout/chrI.bam

singularity exec /home/vagrant/bioinfo_tsinghua.simg bamtools filter -region chrI -in /home/vagrant/mapping/wt1X_thout/accepted_hits.bam \
    -out /home/vagrant/mapping/wt1X_thout/chrI.bam

singularity exec /home/vagrant/bioinfo_tsinghua.simg bamtools filter -region chrI -in /home/vagrant/mapping/wt2X_thout/accepted_hits.bam \
    -out /home/vagrant/mapping/wt2X_thout/chrI.bam

singularity exec /home/vagrant/bioinfo_tsinghua.simg cufflinks -p 4 -o /home/vagrant/assembly/wt1_clout  /home/vagrant/mapping/wt1_thout/chrI.bam 

singularity exec /home/vagrant/bioinfo_tsinghua.simg cufflinks -p 4 -o /home/vagrant/assembly/wt2_clout  /home/vagrant/mapping/wt2_thout/chrI.bam 

singularity exec /home/vagrant/bioinfo_tsinghua.simg cufflinks -p 4 -o /home/vagrant/assembly/wt1X_clout /home/vagrant/mapping/wt1X_thout/chrI.bam 

singularity exec /home/vagrant/bioinfo_tsinghua.simg cufflinks -p 4 -o /home/vagrant/assembly/wt2X_clout /home/vagrant/mapping/wt2X_thout/chrI.bam

ls /home/vagrant/assembly/*/transcripts.gtf > assembly/assemblies.txt

singularity exec /home/vagrant/bioinfo_tsinghua.simg cuffmerge -g /home/vagrant/share/yeast_chrI_annotation.gff -s /home/vagrant/share/bowtie_index/YeastGenome.fa \
    -p 4 -o /home/vagrant/share/assembly/merged /home/vagrant/assembly/assemblies.txt

singularity exec /home/vagrant/bioinfo_tsinghua.simg cuffdiff -o /home/vagrant/diff_expr -b /home/vagrant/share/bowtie_index/YeastGenome.fa \
    -p 4 -u /home/vagrant/share/assembly/merged/merged.gtf \
    /home/vagrant/mapping/wt1_thout/chrI.bam,/home/vagrant/mapping/wt2_thout/chrI.bam \
    /home/vagrant/mapping/wt1X_thout/chrI.bam,/home/vagrant/mapping/wt2X_thout/chrI.bam
singularity exec /home/vagrant/bioinfo_tsinghua.simg Rscript  plot_DE_chart.R
```

```sh
SINGULARITY_LOCALCACHEDIR=/home/xugang/
export SINGULARITY_LOCALCACHEDIR

#sanbox build dir
sudo singularity build --sandbox ubuntu/ docker://ubuntu
#write
sudo singularity shell --writable ubuntu/
#singularity shell ubuntu
#sudo singularity build --writable ubuntu.img docker://ubuntu



sudo singularity build --sandbox /home/xugang/singularity_image/ribocodeminer docker://gangxu/base:2.0
sudo singularity exec -w /home/xugang/singularity_image/ribocodeminer/ bash
cd /home
mkdir share
sudo singularity exec -w -B /home/xugang/singularity_image:/home/share /home/xugang/singularity_image/ribocodeminer/ bash
cd /home/test
cp ../share/Anaconda3-2019.07-Linux-x86_64.sh ./

# install in /home/test/anaconda3
# install anaconda
# pip install RiboMiner
# normal user
pip install RiboCode

pip install RiboMiner
pip install --upgrade RiboMiner


singularity exec /home/xugang/singularity_image/debian/ bash


```
add this into /home/test/.bashrc
```sh
# >>> conda initialize >>>
# !! Contents within this block are managed by 'conda init' !!
__conda_setup="$('/home/test/anaconda3/bin/conda' 'shell.bash' 'hook' 2> /dev/null)"
if [ $? -eq 0 ]; then
    eval "$__conda_setup"
else
    if [ -f "/home/test/anaconda3/etc/profile.d/conda.sh" ]; then
        . "/home/test/anaconda3/etc/profile.d/conda.sh"
    else
        export PATH="/home/test/anaconda3/bin:$PATH"
    fi
fi
unset __conda_setup
# <<< conda initialize <<<
```




