# Human emc cell lines.

----

## Content

* [Prepare data](#First-set-the-reference-files.)  
* [Statistic dwell data](#Using-nanopolish-calculate-each-read-dwell-time)  
* [Predict structure](#predict-structure.)
* [Calculate entropy and GC.](#Calculate-entropy-and-GC.)  
* [nanoplish dwell time process](#nanoplish-dwell-time-process)
* [R script](#R-script)


---

## First, set the reference files.
```sh
gtf=/home/xugang/data/reference/hg38/Homo_sapiens.GRCh38.100.chr.gtf
genome=/home/xugang/data/reference/hg38/Homo_sapiens.GRCh38.dna.primary_assembly.fa.clean.fa
thread=50
output=output
ref=${output}/a0_annotation/transcripts_sequence.fa
localpth=`pwd`
```


```sh
for i in `ls rawdata|grep fast5`;
do 
	#echo $i;
	name="${i/_fast5/}"
	name="${name/.tar.gz/}"
	echo $name
	f5=rawdata/$i
	echo $f5
done
```
* a3.extrace.pl
```pl
sub entropy {
    my ($entropy, $len, $p, %t) = (0, length($_[0]));
    my @chars = split '', $_[0];
    $t{$_}++ foreach @chars;

    foreach (values %t) {
        $p = $_/$len;
        $entropy -= $p * log $p ;
    }

    return $entropy / log 2;
}

sub calcgc {
 my $seq = $_[0];
 my @seqarray = split('',$seq);
 my $count = 0;
 foreach my $base (@seqarray) {
   $count++ if $base =~ /[G|C|g|c]/i;
 }
 my $len = $#seqarray+1;
 my $num=$count / $len;
 #my ($dec)=$num =~ /(\S{6})/;
 return $num;
}


open I, "<$ARGV[0]";
open O, ">$ARGV[0].mfg.tsv";



print O "name\tposition\tmfe\tbin\tentropy\tgc\n";
$bin;
while(<I>){
chomp;
if($_=~/^>/)
{
$_=~/>(.*?)\t(\d+)/;
$name=$1;
$position=$2;
#$factor=$length/100;
print O "$name\t$position\t";
}

$seq=<I>;
$bin=length($seq)-1;
$entropy=entropy($seq);
$gc=calcgc($seq);
$st1=<I>;
$st1=~/\(((\s)?.\d+.*?)\)/;
$mfe=$1;
$st2=<I>;
$st3=<I>;
$fre=<I>;
#$mfe_ave=$mfe/$factor;
#$mfe_ave=$mfe;
print O "$mfe\t$bin\t$entropy\t$gc\n";
}

close I;
close O;

```
* $bin=100;
$id='5272';
open I, "<select_trans_mfe.".$bin.".tsv.mfg.tsv";
open I2, "<faj".$id.".time.pos.tsv";

open O, ">o1.faj".$id.".time.mfe.".$bin.".tsv";
while(<I>){
chomp;
@data=split/\t/,$_;
$name=$data[0]."\t".$data[1];
$mfe{$name}=$data[2]."\t".$data[3]."\t".$data[4]."\t".$data[5];
}
close I;

$head=<I2>;
chomp($head);
print O "$head\tmfe\tbin\tentropy\tgc\n";
while(<I2>){
chomp;

@data=split/\t/,$_;
$data[1]=1+$data[1];
$name=$data[0]."\t".$data[1];
$info=join "\t",@data;
if(exists($mfe{$name})){

print O "$info\t$mfe{$name}\n";
}else{
print O "$info\tNA\tNA\tNA\tNA\n";
}

}

close O;

```sh
$bin=100;
$id='5272';
open I, "<select_trans_mfe.".$bin.".tsv.mfg.tsv";
open I2, "<faj".$id.".time.pos.tsv";

open O, ">o1.faj".$id.".time.mfe.".$bin.".tsv";
while(<I>){
chomp;
@data=split/\t/,$_;
$name=$data[0]."\t".$data[1];
$mfe{$name}=$data[2]."\t".$data[3]."\t".$data[4]."\t".$data[5];
}
close I;

$head=<I2>;
chomp($head);
print O "$head\tmfe\tbin\tentropy\tgc\n";
while(<I2>){
chomp;

@data=split/\t/,$_;
$data[1]=1+$data[1];
$name=$data[0]."\t".$data[1];
$info=join "\t",@data;
if(exists($mfe{$name})){

print O "$info\t$mfe{$name}\n";
}else{
print O "$info\tNA\tNA\tNA\tNA\n";
}

}

close O;

```


## second, using albacore software to transfer current signal into fastq files.
```sh
[[ -d $output ]] || mkdir -p $output
basecaller(){

[[ -d $output/a1-basecaller ]] || mkdir -p $output/a1-basecaller
echo -e "tar -xvzf $f5 -C $output/a1-basecaller/"
tar -xvzf $f5 -C $output/a1-basecaller/ > $output/a1-basecaller/tmp
echo -e "mv $output/a1-basecaller/fast5/ $output/a1-basecaller/$name.f5 "
mv $output/a1-basecaller/fast5/ $output/a1-basecaller/$name.f5
echo -e "conda run -n albacore read_fast5_basecaller.py -i $output/a1-basecaller/$name.f5 -s $output/a1-basecaller/$name -t 30 -r -k SQK-RNA001 -f FLO-MIN106 -o fast5,fastq --disable_filtering"
conda run -n albacore read_fast5_basecaller.py -i $output/a1-basecaller/$name.f5 -s $output/a1-basecaller/$name -t 30 -r -k SQK-RNA001 -f FLO-MIN106 -o fast5,fastq --disable_filtering
cat $output/a1-basecaller/$name/workspace/*.fastq | sed 's/U/T/g' > $output/a1-basecaller/$name/$name.fq
rm -rf $output/a1-basecaller/$name.f5 $output/a1-basecaller/tmp
}
```
## Third, remove the tmp files.
```sh
baserm(){
rm -rf $output/a1-basecaller/$name/workspace/

}
```
## Fourth, generating transcript sequence
```sh
ribocodeannf(){
[[ -d $output/a0_annotation ]] || mkdir -p $output/a0_annotation

 prepare_transcripts -g ${gtf} -f ${genome} -o ${output}/a0_annotation
}
```
## Fivth, useing graphmap to aligment 
We use two stratge of alignment, one was transcriptome as reference, and the other is use genome sequence.
```sh
#Using the transcript as the reference to alignment.
graphmapf(){
[[ -d $output/a2-graphmap ]] || mkdir -p $output/a2-graphmap

echo -e "graphmap align -r ${output}/a0_annotation/transcripts_sequence.fa -d $output/a1-basecaller/$name/$name.fq -o $output/a2-graphmap/$name.sam  --double-index"
graphmap align -r ${output}/a0_annotation/transcripts_sequence.fa -d $output/a1-basecaller/$name/$name.fq -o $output/a2-graphmap/$name.sam  --double-index

}
# Using the genome as the reference.

graphmapgf(){
[[ -d $output/a2-graphmap ]] || mkdir -p $output/a2-graphmap

echo -e "graphmap align -r ${genome} -d $output/a1-basecaller/$name/$name.fq -o $output/a2-graphmap/$name.genome.sam  --double-index"
graphmap align -r ${genome} -d $output/a1-basecaller/$name/$name.fq -o $output/a2-graphmap/$name.genome.sam  --double-index

}
```
## Transform file type.
```sh
samf(){


samtools view -bT ${output}/a0_annotation/transcripts_sequence.fa -F 16 $output/a2-graphmap/$name.sam > $output/a2-graphmap/$name.gene.bam
samtools sort $output/a2-graphmap/$name.gene.bam > $output/a2-graphmap/$name.gene.s.bam
samtools index $output/a2-graphmap/$name.gene.s.bam
}

samgf(){

samtools view -bT ${genome} -F 16 $output/a2-graphmap/$name.genome.sam > $output/a2-graphmap/$name.gene.genome.bam
samtools sort $output/a2-graphmap/$name.gene.genome.bam > $output/a2-graphmap/$name.gene.s.genome.bam
samtools index $output/a2-graphmap/$name.gene.s.genome.bam


}

```
## Using nanopolish index fastq file.

```sh
nanopolishindex(){
nanopolish index -d $output/a1-basecaller/$name.f5 $output/a1-basecaller/$name/$name.fq
}
```
## Using nanopolish calculate each read dwell time.

```sh
nanopolishalign(){
[[ -d $output/a3-nanopolish ]] || mkdir $output/a3-nanopolish 
echo "nanopolish eventalign -t 20  --reads $output/a1-basecaller/$name/$name.fq --bam $output/a2-graphmap/$name.gene.s.bam --genome $ref --print-read-names --scale-events > $output/a3-nanopolish/$name.gene.event"
nanopolish eventalign -t 20  --reads $output/a1-basecaller/$name/$name.fq --bam $output/a2-graphmap/$name.gene.s.bam --genome $ref --print-read-names --scale-events > $output/a3-nanopolish/$name.gene.event
}

nanopolishaligng(){
[[ -d $output/a3-nanopolish ]] || mkdir $output/a3-nanopolish 
echo "nanopolish eventalign -t 20  --reads $output/a1-basecaller/$name/$name.fq --bam $output/a2-graphmap/$name.gene.s.bam --genome $ref --print-read-names --scale-events > $output/a3-nanopolish/$name.gene.event"
nanopolish eventalign -t 20  --reads $output/a1-basecaller/$name/$name.fq --bam $output/a2-graphmap/$name.gene.s.genome.bam --genome $genome --print-read-names --scale-events > $output/a3-nanopolish/$name.gene.genome.event
}
```

* The output file.

|contig | position  |      reference_kmer | read_name  |     strand | event_index   |  event_level_mean   |     event_stdv |event_length  |  model_kmer  |    model_mean  |    model_stdv    |  standardized_level |
|-|-|-|-|-|-|-|-|-|-|-|-|-|
|8  |     4787263| TCACC  | 34aaf652-68be-4abf-a8d3-bc445f8670fa |   t   |    1    |   72.76 |  1.181 |  0.00598 |TCACC |  74.44  | 1.88 |   -0.78|
|8 |      4787264 |CACCC |  34aaf652-68be-4abf-a8d3-bc445f8670fa |   t  |     2   |    77.37 |  3.248  | 0.00232| CACCC |  75.11   |2.81 |  0.70|



## Extract current dwell time gene current.
The a2.current.pl script file will calculate the average of event. One file is name.gene.event.txt. The file contains chromosome, position, current,sd, time, ref for each reads. The other output file is name.gene.event.sta.tsv, which contains chromosome, position,current, sd, and time. There are the average of current, sd and time.  

```sh
statisevent(){
echo "perl a2.current.pl $output/a3-nanopolish/$name.gene.event"
perl a2.current.pl $output/a3-nanopolish/$name.gene.event 
}
statiseventg(){

echo "perl a2.current.pl $output/a3-nanopolish/$name.gene.genome.event"
perl a2.current.pl $output/a3-nanopolish/$name.gene.genome.event 

}
```
## split file. 
We split the large file into sever small files. Using command split with parameter -l 185069885 means the number of lines each file.

Next, each transcript put into single file with script ab.split.pl. With ac.check.pl script checking header for file, if there are not header, give header for the file.

After, use the Rscript Read_events_tsv.R generate 3 files. One is dat.f.combinded.transcript.RData, the second one is the mod.tsv, the third is mod.sum.tsv.

```sh
readevent(){
[[ -d $output/a4-readevent ]] || mkdir -p $output/a4-readevent
[[ -d $output/a3-nanopolish/$name ]] || mkdir -p $output/a3-nanopolish/$name
[[ -d $output/a3-nanopolish/$name/result ]] || mkdir -p $output/a3-nanopolish/$name/result

echo -e "split -l 185069885 $output/a3-nanopolish/$name.gene.event $output/a3-nanopolish/$name/$name"
#split -l 185069885 $output/a3-nanopolish/$name.gene.event $output/a3-nanopolish/$name/$name

cp ab.split.pl $output/a3-nanopolish/$name
cp ac.check.pl $output/a3-nanopolish/$name

## 

#rm -rf tmp
cd $output/a3-nanopolish/$name/

for i in `ls | grep $name`;
do
	echo $i;
echo -e "perl ab.split.pl $i"
#perl ab.split.pl $i
done
cd -

#ls $output/a3-nanopolish/$name/tmp|grep -v ch$ | grep -v cpp$|grep -v R$|grep -v bk$ | xargs -I "{tr}" -P ${thread} perl ac.check.pl $output/a3-nanopolish/$name/tmp/{tr}

cp /home/app/PORE-cupine-master/for_single_gene/* $output/a3-nanopolish/$name/tmp
cd $output/a3-nanopolish/$name/tmp
echo "/usr/bin/Rscript /home/app/PORE-cupine-master/for_single_gene/Read_events_tsv.R -f $localpth/$output/a3-nanopolish/$name.gene.event -o $localpth/$output/a4-readevent/$name.combined.RData"
echo -e "ls |grep -v cpp$|grep -v R$|grep -v bk$ grep -v 'ch.ch'| xargs -I "{tr}" -P ${thread} /usr/bin/Rscript Read_events_tsv.R -f {tr} -o {tr}.combined.RData"
ls |grep -v cpp$|grep -v R$|grep -v bk$ | grep -v 'ch.ch'| grep ch | xargs -I "{tr}" -P ${thread} /usr/bin/Rscript Read_events_tsv.R -f {tr} -o {tr}.combined.RData

#/usr/bin/Rscript Read_events_tsv.R -f $name.gene.event -o $name.combined.RData
cd -



}
```

```r
#!/usr/bin/env Rscript-3.4.1
suppressMessages(library(optparse))

#for command line parsing
args = commandArgs(trailingOnly=TRUE)
option_list = list(
  make_option(c("-f", "--file"), type="character", default=NULL,
              help="Event file name from Nanopolish", metavar="character"),
        make_option(c("-o", "--out"), type="character", default="out.txt",
              help="output file name", metavar="character")
)

opt_parser = OptionParser(option_list=option_list)
opt = parse_args(opt_parser)

if (is.null(opt$file) | is.null(opt$out)){
  print_help(opt_parser)
  stop("Input file and output names must be supplied.", call.=FALSE)
}

suppressMessages(library(dplyr))
suppressMessages(library(Rcpp))
suppressMessages(library(pracma))
suppressMessages(library(data.table))

#loading c++ script
Rcpp::sourceCpp("./for_r.cpp")

#loading of event files
dat= fread(paste(opt$file))
print("Done loading")
mod=opt$out

#combine events of same strands and positions
dat.com= dat %>%
		  mutate(count=round(3012*event_length)) %>%
		  group_by(contig,read_name,position) %>%
		  summarise(event_stdv=sd_combine(event_stdv,event_level_mean,count),
					event_level_mean=mean_combine(event_level_mean,count),
					count=sum(count),reference_kmer=unique(reference_kmer))  %>%
		  ungroup() %>%
		  mutate(event_stdv=ifelse(event_stdv==0,0.01,event_stdv))

dat_out= dat %>%
                  mutate(count=round(3012*event_length)) %>%
                  group_by(contig,read_name,position) %>%
                  summarise(event_stdv=sd_combine(event_stdv,event_level_mean,count),
                                        event_level_mean=mean_combine(event_level_mean,count),
                                        count=mean(count),reference_kmer=unique(reference_kmer))  %>%
                  ungroup() %>%
                  mutate(event_stdv=ifelse(event_stdv==0,0.01,event_stdv))
#saving the results
assign(paste("dat.f.combined.",mod, sep=""),dat.com)
tmp2=(paste("dat.f.combined.",mod, sep=""))
save(list=(tmp2),file=paste("./dat.f.combined.",mod,".RData", sep=""))

write.table(dat_out,file=paste("./",mod,".tsv",sep=""),sep="\t",quote=F,row.names=F)
write.table(dat.com,file=paste("./",mod,".sum.tsv",sep=""),sep="\t",quote=F,row.names=F)

rm(list=ls(pattern="dat.f"))

print("script ran successfully")
```
The file with sum.tsv suffix have event_length multiplied 3012, and group by contig, read, position, then calculate

---

count=round(3012*event_length)
group_by(contig,read_name,position)
event_stdv=sd_combine(event_stdv,event_level_mean,count). 
event_level_mean=mean_combine(event_level_mean,count),
count=sum(count),
reference_kmer=unique(reference_kmer) 

## move data.

The dat.f.combined.files were moved into folder output/a4-readevent/$name。   

The name of file is "dat.f.combined.ENST00000523524.ch.combined.RData.RData", which is contain binary formate file.

The combined.RData files were moved into folder output/a6-time/a1-readevent/$name. "ENST00000674460.ch.combined.RData.sum.tsv", "ENST00000674460.ch.combined.RData.tsv" the files are the same.

"ENST00000674460.ch.combined.RData.sum.tsv"  

|contig | read_name  |     position  |      event_stdv   |   event_level_mean   |     count |  reference_kmer |
| -|- |-|-|-|-|-|
|ENST00000674460| 034aec2f-35f6-41e8-8179-0a26da4e2326 |   502 |    8.99051474355221 |       124.262099700533 |       80  |    AGACC|
|ENST00000674460| 034aec2f-35f6-41e8-8179-0a26da4e2326  |  503 |    1.6  |   82.07  | 9  |     GACCC |
"ENST00000674460.ch.combined.RData.tsv"

|contig | read_name  |     position  |      event_stdv   |   event_level_mean   |     count |  reference_kmer |
| -|- |-|-|-|-|-|
|ENST00000674460| 034aec2f-35f6-41e8-8179-0a26da4e2326 |   502 |    8.99051474355221 |       124.262099700533 |       80  |    AGACC|
|ENST00000674460| 034aec2f-35f6-41e8-8179-0a26da4e2326  |  503 |    1.6  |   82.07  | 9  |     GACCC |

```sh

mvRdata(){
[[ -d $output/a4-readevent/$name ]] || mkdir -p $output/a4-readevent/$name
for i in `ls  $output/a3-nanopolish/$name/tmp/|grep dat.f.combined`;do 
mv $output/a3-nanopolish/$name/tmp/$i $output/a4-readevent/$name
done
# mv $output/a3-nanopolish/$name/tmp/dat.f.combined.*.combined.RData.RData $output/a4-readevent/$name
[[ -d $output/a6-time/a1-readevent/$name ]] || mkdir -p $output/a6-time/a1-readevent/$name
for i in `ls  $output/a3-nanopolish/$name/tmp/|grep combined.RData`;do
mv $output/a3-nanopolish/$name/tmp/$i $output/a6-time/a1-readevent/$name

done

}
```
## extract 
This process use script "a14.extracttime.pl" extract fasta file and 
"$ARGV[0].tab.tsv";

|read| time in pos1| time in pos2 | time in pos3 | time in pos4 |
|- | -| -| -| -|
|43fbb308-ebba-4a62-af45-250266b69e69 |   NA    |  NA   |   NA  |    NA   |


"$ARGV[0].fa.tsv";
|read| seq in pos1| seq in pos2 | seq in pos3 | seq in pos4 |
|- | -| -| -| -|
|ENST00000674500| N   |    N  |     N   |    N   |  

```sh
extracttime(){
[[ -d $output/a6-time/$name ]] || mkdir -p $output/a6-time/$name

cp a14.extracttime.pl $output/a6-time/a1-readevent/$name
# cd path
cd $output/a6-time/a1-readevent/$name
ls |grep RData.sum.tsv | xargs -I "{tr}" -P ${thread} perl a14.extracttime.pl {tr}
cd -
#cd back place
for i in `ls $output/a6-time/a1-readevent/$name|grep tsv.tab.tsv`;do 
	mv $output/a6-time/a1-readevent/$name/$i $output/a6-time/$name
done
for i in `ls $output/a6-time/a1-readevent/$name|grep tsv.fa.tsv`;do
        mv $output/a6-time/a1-readevent/$name/$i $output/a6-time/$name
done

#[[ -d $output/a6-time-sum ]] || mkdir -p $output/a6-time-sum
#perl a3.current.combined.pl $output/a6-time/a1-readevent/$name.combined.RData.tsv
#perl a3.current.combined.pl $output/a6-time/a1-readevent/$name.combined.RData.sum.tsv
#mv $output/a6-time/a1-readevent/$name.combined.RData.tsv.sta.tsv $output/a6-time/$name.time.tsv
#mv $output/a6-time/a1-readevent/$name.combined.RData.sum.tsv.sta.tsv $output/a6-time-sum/$name.sum.time.tsv

}
```
## transform data.
1. Using the name of a15.ks.R R programe to calculate ks test result.


```sh
transform(){

# ls |grep RData.sum.tsv.fa.tsv$| sed 's/.fa.tsv//' | xargs -I "{tr}" -P 1 ls {tr}.fa.tsv {tr}.tab.tsv 
echo "ls $output/a6-time/$name |grep RData.sum.tsv.fa.tsv$| sed 's/.fa.tsv//' | xargs -I "{tr}" -P ${thread} /usr/bin/Rscript a15.ks.R $output/a6-time/$name/{tr}.tab.tsv $output/a6-time/$name/{tr}.fa.tsv"
ls $output/a6-time/$name |grep RData.sum.tsv.fa.tsv$| sed 's/.fa.tsv//' | xargs -I "{tr}" -P ${thread} /usr/bin/Rscript a15.ks.R $output/a6-time/$name/{tr}.tab.tsv $output/a6-time/$name/{tr}.fa.tsv 

}
```

* a15.ks.R

```r
##
## load function
transfor=function(denature_time,denature_name){
denature_time_t=t(denature_time)
denature_name_t=t(denature_name)

denature_t=cbind(denature_name_t,denature_time_t)

coln=dim(denature_t)[2]-1
colnames(denature_t)=c('ref',1:coln)
denature_t=denature_t[-1,]
denature_t[denature_t[,1]=="TRUE",1]="T"
return(denature_t)
}
### function for Z-score standardization
zscore_nor=function(x){
  (x-mean(x))/sd(x)
}

#load file with Rdata.
library(dgof)
load("denature_time.RData")
args <- commandArgs(trailingOnly = TRUE)
#tab
filename1=args[1]
#filename1='ENST00000361624.tab.tsv'
#ref
filename2=args[2]
#filename2='ENST00000361624.fa.tsv'
#
rna_time=read.table(filename1)
rna_name=read.table(filename2)
rna_t=transfor(rna_time,rna_name)


#
position=dim(rna_t)[1]
totalnum=dim(rna_t)[2]
#
shiftt2= data.frame(pos=c(),DF=c(),pv=c())
#loop
for (b in 1:position){
#b=1
content=rna_t[b,1]
rna_pos_time=rna_t[b,2:totalnum]
rna_pos_time2=zscore_nor(as.numeric(rna_pos_time[!is.na(rna_pos_time)]))
if(content=='A'){
s2=ks.test(timeA,rna_pos_time2,alternative = c( "greater"))
}
if(content=='T'){
s2=ks.test(timeT,rna_pos_time2,alternative = c( "greater"))
}
if(content=='C'){
s2=ks.test(timeC,rna_pos_time2,alternative = c( "greater"))
}
if(content=='G'){
s2=ks.test(timeG,rna_pos_time2,alternative = c( "greater"))
}
  # get info unmod vs denature
  #
  div2=s2['statistic'][[1]][[1]]
  dip2=s2['p.value'][[1]]
  shiftt2=rbind(shiftt2,c(b,div2,dip2))
}

colnames(shiftt2)=c("position","df","pvalue")
## set sig
shiftt2$sig=0
shiftt2$sig[shiftt2$df>0.15 & shiftt2$pvalue<0.01]=1
## set df and -log10 pvalue
shiftt2$dfpv=shiftt2$df*(-log10(shiftt2$pvalue+10^-270))

## smooth dfpv
shiftt2$dfpv_sm=filter(shiftt2$dfpv,filter=c(rep(1/4,4)))

#smooth
shiftt2$lin<-filter(shiftt2$df,filter=c(rep(1/4,4)))

#trans data to ksdata.
ksdata2=shiftt2

#save file
write.table(ksdata2,paste(filename1,'ks.tsv',sep=''),sep="\t")
```
* The file with the name of "denature_time.RData" contain simochnong denature dwell time.

The output file with name of "0.ENST00000447282.ch.combined.RData.sum.tsv.tab.tsvks.tsv".

|"position"  |    "df" |   "pvalue"   |     "sig" |  "dfpv"  |"dfpv_sm"    |   "lin" |
|- | -|- |- |- |- |-|
|  12  |    0.443334881560613  |     0.|455646844598687|       0   |    0.151341952446114   |    0.148465775308584    |   0.440451722180277|
|   13    |  0.443334881560613    |   0.455646844598687 |      0    |   0.151341952446114   |    0.151341952446114   |    0.443334881560613|



## check data.

This process select structure data from data. Becuase most of transcript were no significantly different using KS test. If there are differential longer dwell time than denature time, there are p-values were 0, or smaller than 0.01. Otherwise, all the position p-value is 1.

```sh
timeckeck(){
[[ -d $output/a7-checktime/$name ]] || mkdir -p $output/a7-checktime/$name
for i in `ls $output/a6-time/$name| grep tab.tsvks.tsv$`;
do 
count=`cut -f 3 $output/a6-time/$name/$i|sort -u|wc -l`
#echo $i ${count}
if [ $count -gt 10 ]  ;then

filename="${i/.ch.combined.RData.sum.tsv.tab.tsvks.tsv/.ks.tsv}"
#echo $i ${count} $filename
cp $output/a6-time/$name/$i $output/a7-checktime/$name/$filename
fi
done


}
```

--- 

## predict structure.
1. Using have significant structure data, get the transcript name. tran="${i/.ks.tsv/}"
2. Using grep get transcript sequence and length from a0 annotation, transcript_sequence.fa.
3. Using a7.ks.fa.pos.pl script combine ks test result and sequence, and threshold was 0.4 to filter low value results.


```sh
predict_structure(){
[[ -d $output/a8-ks/$name ]] || mkdir -p $output/a8-ks/$name
[[ -d $output/a9-predict/$name ]] || mkdir -p $output/a9-predict/$name
for i in {1..33};do
filen=file$i
[[ -d $output/a7-checktime/$name/$filen  ]] || mkdir -p $output/a7-checktime/$name/$filen
done

for j in  na p0 p1 ;
do 
	filen=file$j
	[[ -d $output/a7-checktime/$name/$filen  ]] || mkdir -p $output/a7-checktime/$name/$filen
done


for i in `ls $output/a7-checktime/$name|grep .ks.tsv$`;
do 
tran="${i/.ks.tsv/}"
#echo $i $tran
grep -A 1 $tran $output/a0_annotation/transcripts_sequence.fa > $output/a8-ks/$name/$tran.fa
length=`head -n 1 $output/a8-ks/$name/$tran.fa|cut -f 2 -d ' '`
perl a7.ks.fa.pos.pl $output/a7-checktime/$name/$i $output/a8-ks/$name/$tran.fa 0.4

#echo -e "RME -d SHAPE -p 38 $output/a7-checktime/$name/$i.12.range.tsv $output/a9-predict/$name/$i"
#RME -d SHAPE -p 38 $output/a7-checktime/$name/$i.12.range.tsv $output/a9-predict/$name/$i
for j in {0..29};
do
mv $output/a7-checktime/$name/$i.$j.range.tsv $output/a7-checktime/$name/file$j
done
for j in  na p0 p1  ;
do mv $output/a7-checktime/$name/$i.$j.range.tsv $output/a7-checktime/$name/file$j
done


# echo -e " ls $output/a7-checktime/$name/ |grep 12.range.tsv$|sed 's/.12.range.tsv//' | xargs -I "{tr}" -P ${thread} RME -p 1 -d SHAPE $output/a7-checktime/$name/{tr}.12.range.tsv $output/a9-predict/$name/{tr} "

#for i in `ls $output/a7-checktime/$name/file12`;do
#	echo $i;
#	RME -p 38 -d SHAPE $output/a7-checktime/$name/file12/$i $output/a9-predict/$name/$i
#done

}
```
## predict_structure2
1. Using command to calculate the length of sequence that have the size smaller than 5000bp. Then sequences were copied to $output/a7-checktime/$name/file12/length5000.


```sh
predict_structure2(){
[[ -d $output/a7-checktime/$name/file12/length5000 ]] || mkdir -p $output/a7-checktime/$name/file12/length5000
for i in `ls $output/a7-checktime/$name/file12|grep tsv$`;
do 
count=`wc -l $output/a7-checktime/$name/file12/$i | cut -f 1 -d ' '`
if [ "$count"  -lt 5000  ] ; then
cp $output/a7-checktime/$name/file12/$i $output/a7-checktime/$name/file12/length5000	
fi
done
}
```
## predict_structure3
1. Using RME software to predict RNA structures.

```sh
predict_structure3(){

[[ -d $output/a9-predict/$name/ ]] || mkdir -p $output/a9-predict/$name/
ls $output/a7-checktime/$name/file12/length5000 |grep 12.range.tsv$|sed 's/.ks.tsv.12.range.tsv//' | xargs -I "{tr}" -P ${thread} RME -p 2 -d SHAPE $output/a7-checktime/$name/file12/length5000/{tr}.ks.tsv.12.range.tsv $output/a9-predict/$name/{tr}

#for i in `ls $output/a7-checktime/$name/file12`;
#do
#fn="${i/.ks.tsv.12.range.tsv/}"
#echo $i $fn
#RME -p 38 -d SHAPE $output/a7-checktime/$name/file12/$i $output/a9-predict/$name/$fn
#done

}
```

## reactivity
Uisng script SVM_plus.R to calculate the reactivity score.


```sh

reactivity(){
modified=$1
unmodified=$2
length=$3
[[ -d $output/a5-reactivity ]] || mkdir -p $output/a5-reactivity
[[ -d $output/a7-mod ]] || mkdir -p $output/a7-mod
cp /home/app/PORE-cupine-master/for_single_gene/* $output/a4-readevent
cd $output/a4-readevent
echo "/usr/bin/Rscript ./SVM_plus.R -m $modified -u $unmodified -o $modified.csv -l $length"
/usr/bin/Rscript ./SVM_plus.R -m $modified -u $unmodified -o $modified -l $length
#ls
echo "mv *.csv ../a5-reactivity/$modified.csv"

mv *csv ../a5-reactivity/
mv *mod_mat.tsv ../a7-mod
cd ../..

}
```

```sh
split_time(){

[[ -d $output/a7-mod-sum ]] || mkdir -p $output/a7-mod-sum	
cp $output/a6-time/a1-readevent/*combined.RData.tsv $output/a7-mod
cp $output/a6-time/a1-readevent/*combined.RData.sum.tsv $output/a7-mod-sum
cp $output/a7-mod/*RDatamod_mat.tsv $output/a7-mod-sum

for i in `ls $output/a7-mod/|grep RDatamod_mat.tsv$`;
do echo $i;
readinfo="${i/RDatamod_mat.tsv/RData.tsv}";
echo $readinfo
echo -e "perl a4.unmod.event.pl $output/a7-mod/$i $output/a7-mod/$readinfo 440"
perl a4.unmod.event.pl $output/a7-mod/$i $output/a7-mod/$readinfo 440
echo -e "perl a5.mod.event.pl $output/a7-mod/$i $output/a7-mod/$readinfo 440"
perl a5.mod.event.pl $output/a7-mod/$i $output/a7-mod/$readinfo 440
done

for i in `ls $output/a7-mod-sum/|grep RDatamod_mat.tsv$`;
do 
readinfo="${i/RDatamod_mat.tsv/RData.sum.tsv}";
echo $readinfo
echo $i
echo -e "perl a4.unmod.event.pl $output/a7-mod-sum/$i $output/a7-mod-sum/$readinfo 440"
perl a4.unmod.event.pl $output/a7-mod-sum/$i $output/a7-mod-sum/$readinfo 440
echo -e "perl a5.mod.event.pl $output/a7-mod-sum/$i $output/a7-mod-sum/$readinfo 440"
perl a5.mod.event.pl $output/a7-mod-sum/$i $output/a7-mod-sum/$readinfo 440
done



[[ -d $output/a8-pos-time ]] || mkdir -p $output/a8-pos-time
for i in `ls $output/a7-mod/|grep RDatamod_mat.tsv$`;
do
readinfo="${i/.combined.RDatamod_mat.tsv/}";
echo $readinfo
echo -e "perl a6.forposition.pl $output/a7-mod/$readinfo.mod.tsv"
perl a6.forposition.pl $output/a7-mod/$readinfo.mod.tsv
echo -e "perl a6.forposition.pl $output/a7-mod/$readinfo.unmod.tsv"
perl a6.forposition.pl $output/a7-mod/$readinfo.unmod.tsv

mv $output/a7-mod/$readinfo.mod.pos.tsv $output/a8-pos-time
mv $output/a7-mod/$readinfo.unmod.pos.tsv $output/a8-pos-time

done

[[ -d $output/a8-pos-time-sum ]] || mkdir -p $output/a8-pos-time-sum
for i in `ls $output/a7-mod-sum/|grep RDatamod_mat.tsv$`;
do
readinfo="${i/.combined.RDatamod_mat.tsv/}";
echo $readinfo
echo -e "perl a6.forposition.pl $output/a7-mod-sum/$readinfo.mod.tsv"
perl a6.forposition.pl $output/a7-mod-sum/$readinfo.mod.tsv

echo -e "perl a6.forposition.pl $output/a7-mod-sum/$readinfo.unmod.tsv"
perl a6.forposition.pl $output/a7-mod-sum/$readinfo.unmod.tsv

mv $output/a7-mod-sum/$readinfo.mod.pos.tsv $output/a8-pos-time-sum
mv $output/a7-mod-sum/$readinfo.unmod.pos.tsv $output/a8-pos-time-sum
done

}
```

## run loop
```
run_loop(){
#ribocodeannf
for i in `ls rawdata|grep fast5`;
do 
	#echo $i;
	name="${i/_fast5/}"
	name="${name/.tar.gz/}"
	echo $name
	f5=rawdata/$i
	echo $f5
#basecaller
#baserm
#graphmapf
#graphmapgf
#samf
#samgf
#nanopolishindex
#nanopolishalign
#nanopolishaligng
#statisevent
statiseventg
#readevent
#mvRdata
#extracttime
#transform
#timeckeck
#predict_structure
#predict_structure2
#predict_structure3
done
}
```
## run command.
```sh
run_loop
#name=test
#transform
#timeckeck
#predict_structure
#predict_structure2

#reactivity Tetra_NAI_N3_1.combined.RData Tetra_unmod.combined.RData 421
#reactivity Tetra_NAI_N3_25min.combined.RData Tetra_unmod.combined.RData 421
#reactivity Tetra_NAI_N3_2.combined.RData Tetra_unmod.combined.RData 421
#reactivity Tetra_NAI_N3_denatured.combined.RData Tetra_unmod.combined.RData 421
#reactivity Tetra_unmod2.combined.RData Tetra_unmod.combined.RData 421

#split_time

```
## predict
```sh
predict(){
val=$1
echo -e "perl a7.ks.fa.pos.pl output/a10-ks/nai_n3_ks.tsv tetra_ribozyme.fa $val"
perl a7.ks.fa.pos.pl output/a10-ks/nai_n3_ks.tsv tetra_ribozyme.fa $val
[[ -d $output/a12-predict ]] || mkdir -p $output/a12-predict
for i in `ls output/a10-ks/|grep range.tsv`;do echo $i;
RME -d SHAPE -p 38 $output/a10-ks/$i $output/a12-predict/$i

done
[[ -d $output/a13-jpg ]] || mkdir -p $output/a13-jpg
for i in `ls output/a12-predict/|grep range.tsv`;do echo $i;
for j in `ls output/a12-predict/$i`;do
echo $i/$j;	
cut -f 3 output/a10-ks/$i |sed s/NA/0/ > tmp ;
sed 1d tmp |tr '\n' ';' > tmp2
color=`cat tmp2`;
rm tmp tmp2;
echo -e "java -cp /Users/xugang/Desktop/sequencing_center_desktop/2020/rna_structure/VARNAv3-93-src.jar fr.orsay.lri.varna.applications.VARNAcmd -i a12-predict/$i/$j -colorMapStyle \"0:#FFFFFF;0.65:#d0c9c9;1:#d6341e\" -colorMap \"$color\" "> $output/a13-jpg/code_${i}_${j}.sh
java -cp /home/app/varna/VARNAv3-93-src.jar fr.orsay.lri.varna.applications.VARNAcmd -i $output/a12-predict/$i/$j -o $output/a13-jpg/${i}_${j}.svg -colorMapStyle "0:#FFFFFF;0.65:#d0c9c9;1:#d6341e" -colorMap \"${color}\"

#java -cp /home/app/varna/VARNAv3-93-src.jar fr.orsay.lri.varna.applications.VARNAcmd -i $output/a12-predict/$i/$j -o $output/a13-jpg/${i}_${j}.svg -colorMapStyle heat
# -colorMapStyle "-0.38:#FFFFFF;3.13:#d6341e" -colorMap "0.05;3.7;0.4"
done
done
[[ -d $output/a15-history/$val ]] || mkdir -p $output/a15-history/$val
cp -r $output/a13-jpg/ $output/a15-history/$val
cp -r $output/a12-predict $output/a15-history/$val

 
}

```

```sh
#split dot parent file into ct file.
#perl a10.dp2ct.pl pdb_00805.dp

#assest the ct

assestf(){
echo -e "\n$1" >> list.txt

for i in `ls $output/a12-predict|grep tsv`;
do 

for j in `ls $output/a12-predict/$i`;do 

	echo $i;
	echo $i >> list.txt;

perl a9.assuse.pl $output/a12-predict/$i/$j PDB_00805_part_1.ct 103 241 1 139 >> list.txt
perl a9.assuse.pl $output/a12-predict/$i/$j PDB_00805_part_2.ct 103 241 1 139 >> list.txt
perl a9.assuse.pl $output/a12-predict/$i/$j PDB_00805_part_3.ct 103 241 1 139 >> list.txt
perl a9.assuse.pl $output/a12-predict/$i/$j PDB_00805_part_4.ct 103 241 1 139 >> list.txt
done

done
}
```

```sh
loop_test(){
rm list.txt
#for k in 0.1 0.15 0.2 0.25 0.3 0.35 0.4 0.45 0.5 0.55 0.6 0.65 0.7 0.75 0.8 0.85 0.9 0.95 ;
for k in 0.45 ;
do
echo $k
predict $k
#assestf $k
done;
}
```

```sh
makeclean(){
rm -rf $output/a1-basecaller/*
}
#makeclean
#Rscript a11.ks.R 
#loop_test

#perl a7.ks.fa.pos.pl output/a10-ks/nai_n3_ks.tsv tetra_ribozyme.fa 0.95
```

---

## Calculate entropy and GC.
 The name of a21.100bin.pl perl script can split transcript with setting length bin.

* a21.100bin.pl 

```pl
open I, "<transcripts_sequence.fa";
open I2, "<select_struture.tsv";
$bin=12;
open O, ">select_trans.".$bin.".bin.fa";
while(<I>){
if($_=~/>(\w+)/){
$name=$1;
$seq=<I>;
chomp($seq);
$seq{$name}=$seq;
}

}
close I;

while(<I2>){
chomp;
@data=split/\t/,$_;
if($data[1]=~/\w/){
$name{$data[1]}='';
}

}

foreach(keys(%seq)){
$name=$_;
$seq=$seq{$_};
$length=length($seq);
for ($i=1;$i<$length-$bin+1;$i++){
my $seq2  = substr $seq, $i, $bin;
print O ">$name\t$i\n$seq2\n";
}
#print "$_\n$length\t$seq{$_}\n";
}

```

### 
* a1.rnafold.sh
Using RNAfold to calculate RNA structure.
```sh
#docker run -v /home/xugang/:/home/xugang/ fjossinet/assemble2 bash -c 'RNAfold -p < /home/xugang/data/rnastructure/a4-human/output/a0_annotation/a2.bin_mfe/select_trans_100bin.fa > /home/xugang/data/rnastructure/a4-human/output/a0_annotation/a2.bin_mfe/select_trans_mfe.tsv' > c0.log



rnafoldf(){
len=$1
docker run -v /home/xugang/:/home/xugang/ fjossinet/assemble2 bash -c 'RNAfold -p < /home/xugang/data/rnastructure/a4-human/output/a0_annotation/a2.bin_mfe/select_trans.'${len}'.bin.fa > /home/xugang/data/rnastructure/a4-human/output/a0_annotation/a2.bin_mfe/select_trans_mfe.'${len}'.tsv' > c0.log

}


#rnafoldf 100
#rnafoldf 50
#rnafoldf 30
rnafoldf 12
#rnafoldf 10
#rnafoldf 5
```
* a2.getpostion.sh

```sh
#rm faj5033.time.pos.tsv
#rm faj5273.time.pos.tsv
#rm faj5272.time.pos.tsv
#rm faj5272_2.time.pos.tsv


f1(){
for i in `cut -f 2 select_struture.tsv`;
do
echo $i;
grep $i /home/xugang/data/rnastructure/a4-human/output/a3-nanopolish/FAJ05033.gene.event.txt.neam.pos.tsv >> faj5033.time.pos.tsv
grep $i /home/xugang/data/rnastructure/a4-human/output/a3-nanopolish/FAJ05273.gene.event.txt.neam.pos.tsv >> faj5273.time.pos.tsv
grep $i /home/xugang/data/rnastructure/a4-human/output/a3-nanopolish/FAJ05272.gene.event.txt.neam.pos.tsv >> faj5272.time.pos.tsv
grep $i /home/xugang/data/rnastructure/a4-human/output/a3-nanopolish/FAJ05272_2.gene.event.txt.neam.pos.tsv >> faj5272_2.time.pos.tsv

done;
}

f2(){
#rm faj5033.time.pos.tsv
for i in `cut -f 1 Human_undify5033expressed.txt`;
do
echo $i;
#grep $i /home/xugang/data/rnastructure/a4-human/output/a3-nanopolish/FAJ05033.gene.event.txt.neam.pos.tsv >> faj5033.time.pos.tsv
done
echo -e "name\tpos\ttime\tcurrent\tcount" > head.txt
cat head.txt faj5033.time.pos.tsv > faj5033.time.pos.tsv2
mv faj5033.time.pos.tsv2 faj5033.time.pos.tsv
}
#f2
f3(){
#rm faj5273.time.pos.tsv
for i in `cut -f 1 Human_undify5273expressed.txt`;
do
echo $i;
#grep $i /home/xugang/data/rnastructure/a4-human/output/a3-nanopolish/FAJ05273.gene.event.txt.neam.pos.tsv >> faj5273.time.pos.tsv
done
echo -e "name\tpos\ttime\tcurrent\tcount" > head.txt
cat head.txt faj5273.time.pos.tsv > faj5273.time.pos.tsv2
mv faj5273.time.pos.tsv2 faj5273.time.pos.tsv
}
#f3
f4(){

#rm faj5272.time.pos.tsv
for i in `cut -f 1 Human_nain35272expressed.txt`;
do
echo $i;
#grep $i /home/xugang/data/rnastructure/a4-human/output/a3-nanopolish/FAJ05272.gene.event.txt.neam.pos.tsv >> faj5272.time.pos.tsv
done
echo -e "name\tpos\ttime\tcurrent\tcount" > head.txt
cat head.txt faj5272.time.pos.tsv > faj5272.time.pos.tsv2
mv faj5272.time.pos.tsv2 faj5272.time.pos.tsv

}
#f4
f5(){

rm faj5272_2.time.pos.tsv
for i in `cut -f 1 Human_nain35272_2expressed.txt`;
do
echo $i;
grep $i /home/xugang/data/rnastructure/a4-human/output/a3-nanopolish/FAJ05272_2.gene.event.txt.neam.pos.tsv >> faj5272_2.time.pos.tsv
done
echo -e "name\tpos\ttime\tcurrent\tcount" > head.txt
cat head.txt faj5272_2.time.pos.tsv > faj5272_2.time.pos.tsv2
mv faj5272_2.time.pos.tsv2 faj5272_2.time.pos.tsv
}
#f5



f3(){
#rm faj5273.time.pos.tsv
for i in `cut -f 1 Human_undify5273expressed.txt`;
do
echo $i;
#grep $i /home/xugang/data/rnastructure/a4-human/output/a3-nanopolish/FAJ05273.gene.event.txt.neam.pos.tsv >> faj5273.time.pos.tsv
done
echo -e "name\tpos\ttime\tcurrent\tcount" > head.txt
cat head.txt faj5273.time.pos.tsv > faj5273.time.pos.tsv2
mv faj5273.time.pos.tsv2 faj5273.time.pos.tsv
}

```
* a4.merge.pl
```pl
$bin=100;
$id='5272';
open I, "<select_trans_mfe.".$bin.".tsv.mfg.tsv";
open I2, "<faj".$id.".time.pos.tsv";

open O, ">o1.faj".$id.".time.mfe.".$bin.".tsv";
while(<I>){
chomp;
@data=split/\t/,$_;
$name=$data[0]."\t".$data[1];
$mfe{$name}=$data[2]."\t".$data[3]."\t".$data[4]."\t".$data[5];
}
close I;

$head=<I2>;
chomp($head);
print O "$head\tmfe\tbin\tentropy\tgc\n";
while(<I2>){
chomp;

@data=split/\t/,$_;
$data[1]=1+$data[1];
$name=$data[0]."\t".$data[1];
$info=join "\t",@data;
if(exists($mfe{$name})){

print O "$info\t$mfe{$name}\n";
}else{
print O "$info\tNA\tNA\tNA\tNA\n";
}

}

close O;
```

* a4.merge.all.pl  

```pl
$bin=12;
$id='5272_2';
open I, "<trans.mfe.".$bin.".tsv";
open I2, "<faj".$id.".time.pos.tsv";
#trans.mfe.12.tsv
open O, ">o1.faj".$id.".time.mfe.all.".$bin.".tsv";
while(<I>){
chomp;
@data=split/\t/,$_;
$name=$data[0]."\t".$data[1];
$mfe{$name}=$data[2]."\t".$data[3]."\t".$data[4]."\t".$data[5];
}
close I;

$head=<I2>;
chomp($head);
print O "$head\tmfe\tbin\tentropy\tgc\n";
while(<I2>){
chomp;

@data=split/\t/,$_;
$data[1]=1+$data[1];
$name=$data[0]."\t".$data[1];
$info=join "\t",@data;
if(exists($mfe{$name})){

print O "$info\t$mfe{$name}\n";
}else{
print O "$info\tNA\tNA\tNA\tNA\n";
}

}

close O;
```
* a4.merge.all.95th.pl  
```pl
$bin=12;
$id='05033';
print "/home/xugang/data/rnastructure/a4-human/output/a3-nanopolish/FAJ".$id.".gene.event.txt.neam.95.pos.tsv\n";

open I, "<trans.mfe.".$bin.".tsv";
open I2, "</home/xugang/data/rnastructure/a4-human/output/a3-nanopolish/FAJ".$id.".gene.event.txt.neam.95.pos.tsv";

open O, ">o1.faj".$id.".time.mfe.all.95th.".$bin.".tsv";
while(<I>){
chomp;
@data=split/\t/,$_;
$name=$data[0]."\t".$data[1];
$mfe{$name}=$data[2]."\t".$data[3]."\t".$data[4]."\t".$data[5];
}
close I;

$head=<I2>;
chomp($head);
print O "$head\tmfe\tbin\tentropy\tgc\n";
while(<I2>){
chomp;

@data=split/\t/,$_;
$data[1]=1+$data[1];
$name=$data[0]."\t".$data[1];
$info=join "\t",@data;
if(exists($mfe{$name})){

print O "$info\t$mfe{$name}\n";
}else{
print O "$info\tNA\tNA\tNA\tNA\n";
}

}

close O;
```

* a4.merge.all.95th.pl
```pl
$bin=12;
$id='05033';
print "/home/xugang/data/rnastructure/a4-human/output/a3-nanopolish/FAJ".$id.".gene.event.txt.neam.95.pos.tsv\n";

open I, "<trans.mfe.".$bin.".tsv";
open I2, "</home/xugang/data/rnastructure/a4-human/output/a3-nanopolish/FAJ".$id.".gene.event.txt.neam.95.pos.tsv";

open O, ">o1.faj".$id.".time.mfe.all.95th.".$bin.".tsv";
while(<I>){
chomp;
@data=split/\t/,$_;
$name=$data[0]."\t".$data[1];
$mfe{$name}=$data[2]."\t".$data[3]."\t".$data[4]."\t".$data[5];
}
close I;

$head=<I2>;
chomp($head);
print O "$head\tmfe\tbin\tentropy\tgc\n";
while(<I2>){
chomp;

@data=split/\t/,$_;
$data[1]=1+$data[1];
$name=$data[0]."\t".$data[1];
$info=join "\t",@data;
if(exists($mfe{$name})){

print O "$info\t$mfe{$name}\n";
}else{
print O "$info\tNA\tNA\tNA\tNA\n";
}

}

close O;

```
* a5.entropy.pl

```pl
sub entropy {
    my ($entropy, $len, $p, %t) = (0, length($_[0]));
    my @chars = split '', $_[0];
    $t{$_}++ foreach @chars;

    foreach (values %t) {
        $p = $_/$len;
        $entropy -= $p * log $p ;
    }

    return $entropy / log 2;
}

$seq='CGT';
$a1=entropy($seq);
print "seq: $seq\nEntroy: $a1\n";

$seq='AAA';
$a1=entropy($seq);
print "seq: $seq\nEntroy: $a1\n";

$seq='CACGT';
$a1=entropy($seq);
print "seq: $seq\nEntroy: $a1\n";

$seq='ACGTCA';
$a1=entropy($seq);
print "seq: $seq\nEntroy: $a1\n";
```

* a1.split.pl
```pl

open I, "<select_trans.12.bin.fa";# select_trans.12.bin.fa
open O2, ">a2.command.sh";
$line=0;
while(<I>){
$name=$_;
$seq=<I>;
$line++;
}


print "file has $line lines.";
close I;

$split_num=40;

$reminder=$line%($split_num);
$mean2=int($line/($split_num));

print "$line\t$mean2\t$split_num\t$reminder\n";

open I, "<select_trans.12.bin.fa";
$n=0;
while(<I>){
$name=$_;
$seq=<I>;
$n++;
$reminder=$n%$mean2;
$mean=int($n/$mean2);
if($reminder==1){
open O, ">b$mean.trans.fa";
print O2 "nohup docker run -v /home/xugang/:/home/xugang/ fjossinet/assemble2 bash -c 'RNAfold -p < /home/xugang/data/rnastructure/a4-human/output/a0_annotation/a2.bin_mfe/b1.split/b$mean.trans.fa > /home/xugang/data/rnastructure/a4-human/output/a0_annotation/a2.bin_mfe/b1.split/b$mean.mfv.tsv' > c$mean.log 2>&1 & \n";
}
print O $name;
print O $seq;
if($reminder==0){
close O;
}

}
close O;

```
* a2.command.sh
```sh
nohup docker run -v /home/xugang/:/home/xugang/ fjossinet/assemble2 bash -c 'RNAfold -p < /home/xugang/data/rnastructure/a4-human/output/a0_annotation/a2.bin_mfe/b1.split/b0.trans.fa > /home/xugang/data/rnastructure/a4-human/output/a0_annotation/a2.bin_mfe/b1.split/b0.mfv.tsv' > c0.log 2>&1 &
nohup docker run -v /home/xugang/:/home/xugang/ fjossinet/assemble2 bash -c 'RNAfold -p < /home/xugang/data/rnastructure/a4-human/output/a0_annotation/a2.bin_mfe/b1.split/b1.trans.fa > /home/xugang/data/rnastructure/a4-human/output/a0_annotation/a2.bin_mfe/b1.split/b1.mfv.tsv' > c1.log 2>&1 &
nohup docker run -v /home/xugang/:/home/xugang/ fjossinet/assemble2 bash -c 'RNAfold -p < /home/xugang/data/rnastructure/a4-human/output/a0_annotation/a2.bin_mfe/b1.split/b2.trans.fa > /home/xugang/data/rnastructure/a4-human/output/a0_annotation/a2.bin_mfe/b1.split/b2.mfv.tsv' > c2.log 2>&1 &
nohup docker run -v /home/xugang/:/home/xugang/ fjossinet/assemble2 bash -c 'RNAfold -p < /home/xugang/data/rnastructure/a4-human/output/a0_annotation/a2.bin_mfe/b1.split/b3.trans.fa > /home/xugang/data/rnastructure/a4-human/output/a0_annotation/a2.bin_mfe/b1.split/b3.mfv.tsv' > c3.log 2>&1 &
nohup docker run -v /home/xugang/:/home/xugang/ fjossinet/assemble2 bash -c 'RNAfold -p < /home/xugang/data/rnastructure/a4-human/output/a0_annotation/a2.bin_mfe/b1.split/b4.trans.fa > /home/xugang/data/rnastructure/a4-human/output/a0_annotation/a2.bin_mfe/b1.split/b4.mfv.tsv' > c4.log 2>&1 &
nohup docker run -v /home/xugang/:/home/xugang/ fjossinet/assemble2 bash -c 'RNAfold -p < /home/xugang/data/rnastructure/a4-human/output/a0_annotation/a2.bin_mfe/b1.split/b5.trans.fa > /home/xugang/data/rnastructure/a4-human/output/a0_annotation/a2.bin_mfe/b1.split/b5.mfv.tsv' > c5.log 2>&1 &
nohup docker run -v /home/xugang/:/home/xugang/ fjossinet/assemble2 bash -c 'RNAfold -p < /home/xugang/data/rnastructure/a4-human/output/a0_annotation/a2.bin_mfe/b1.split/b6.trans.fa > /home/xugang/data/rnastructure/a4-human/output/a0_annotation/a2.bin_mfe/b1.split/b6.mfv.tsv' > c6.log 2>&1 &
nohup docker run -v /home/xugang/:/home/xugang/ fjossinet/assemble2 bash -c 'RNAfold -p < /home/xugang/data/rnastructure/a4-human/output/a0_annotation/a2.bin_mfe/b1.split/b7.trans.fa > /home/xugang/data/rnastructure/a4-human/output/a0_annotation/a2.bin_mfe/b1.split/b7.mfv.tsv' > c7.log 2>&1 &
nohup docker run -v /home/xugang/:/home/xugang/ fjossinet/assemble2 bash -c 'RNAfold -p < /home/xugang/data/rnastructure/a4-human/output/a0_annotation/a2.bin_mfe/b1.split/b8.trans.fa > /home/xugang/data/rnastructure/a4-human/output/a0_annotation/a2.bin_mfe/b1.split/b8.mfv.tsv' > c8.log 2>&1 &
nohup docker run -v /home/xugang/:/home/xugang/ fjossinet/assemble2 bash -c 'RNAfold -p < /home/xugang/data/rnastructure/a4-human/output/a0_annotation/a2.bin_mfe/b1.split/b9.trans.fa > /home/xugang/data/rnastructure/a4-human/output/a0_annotation/a2.bin_mfe/b1.split/b9.mfv.tsv' > c9.log 2>&1 &
nohup docker run -v /home/xugang/:/home/xugang/ fjossinet/assemble2 bash -c 'RNAfold -p < /home/xugang/data/rnastructure/a4-human/output/a0_annotation/a2.bin_mfe/b1.split/b10.trans.fa > /home/xugang/data/rnastructure/a4-human/output/a0_annotation/a2.bin_mfe/b1.split/b10.mfv.tsv' > c10.log 2>&1 &
nohup docker run -v /home/xugang/:/home/xugang/ fjossinet/assemble2 bash -c 'RNAfold -p < /home/xugang/data/rnastructure/a4-human/output/a0_annotation/a2.bin_mfe/b1.split/b11.trans.fa > /home/xugang/data/rnastructure/a4-human/output/a0_annotation/a2.bin_mfe/b1.split/b11.mfv.tsv' > c11.log 2>&1 &
nohup docker run -v /home/xugang/:/home/xugang/ fjossinet/assemble2 bash -c 'RNAfold -p < /home/xugang/data/rnastructure/a4-human/output/a0_annotation/a2.bin_mfe/b1.split/b12.trans.fa > /home/xugang/data/rnastructure/a4-human/output/a0_annotation/a2.bin_mfe/b1.split/b12.mfv.tsv' > c12.log 2>&1 &
nohup docker run -v /home/xugang/:/home/xugang/ fjossinet/assemble2 bash -c 'RNAfold -p < /home/xugang/data/rnastructure/a4-human/output/a0_annotation/a2.bin_mfe/b1.split/b13.trans.fa > /home/xugang/data/rnastructure/a4-human/output/a0_annotation/a2.bin_mfe/b1.split/b13.mfv.tsv' > c13.log 2>&1 &
nohup docker run -v /home/xugang/:/home/xugang/ fjossinet/assemble2 bash -c 'RNAfold -p < /home/xugang/data/rnastructure/a4-human/output/a0_annotation/a2.bin_mfe/b1.split/b14.trans.fa > /home/xugang/data/rnastructure/a4-human/output/a0_annotation/a2.bin_mfe/b1.split/b14.mfv.tsv' > c14.log 2>&1 &
nohup docker run -v /home/xugang/:/home/xugang/ fjossinet/assemble2 bash -c 'RNAfold -p < /home/xugang/data/rnastructure/a4-human/output/a0_annotation/a2.bin_mfe/b1.split/b15.trans.fa > /home/xugang/data/rnastructure/a4-human/output/a0_annotation/a2.bin_mfe/b1.split/b15.mfv.tsv' > c15.log 2>&1 &
nohup docker run -v /home/xugang/:/home/xugang/ fjossinet/assemble2 bash -c 'RNAfold -p < /home/xugang/data/rnastructure/a4-human/output/a0_annotation/a2.bin_mfe/b1.split/b16.trans.fa > /home/xugang/data/rnastructure/a4-human/output/a0_annotation/a2.bin_mfe/b1.split/b16.mfv.tsv' > c16.log 2>&1 &
nohup docker run -v /home/xugang/:/home/xugang/ fjossinet/assemble2 bash -c 'RNAfold -p < /home/xugang/data/rnastructure/a4-human/output/a0_annotation/a2.bin_mfe/b1.split/b17.trans.fa > /home/xugang/data/rnastructure/a4-human/output/a0_annotation/a2.bin_mfe/b1.split/b17.mfv.tsv' > c17.log 2>&1 &
nohup docker run -v /home/xugang/:/home/xugang/ fjossinet/assemble2 bash -c 'RNAfold -p < /home/xugang/data/rnastructure/a4-human/output/a0_annotation/a2.bin_mfe/b1.split/b18.trans.fa > /home/xugang/data/rnastructure/a4-human/output/a0_annotation/a2.bin_mfe/b1.split/b18.mfv.tsv' > c18.log 2>&1 &
nohup docker run -v /home/xugang/:/home/xugang/ fjossinet/assemble2 bash -c 'RNAfold -p < /home/xugang/data/rnastructure/a4-human/output/a0_annotation/a2.bin_mfe/b1.split/b19.trans.fa > /home/xugang/data/rnastructure/a4-human/output/a0_annotation/a2.bin_mfe/b1.split/b19.mfv.tsv' > c19.log 2>&1 &
nohup docker run -v /home/xugang/:/home/xugang/ fjossinet/assemble2 bash -c 'RNAfold -p < /home/xugang/data/rnastructure/a4-human/output/a0_annotation/a2.bin_mfe/b1.split/b20.trans.fa > /home/xugang/data/rnastructure/a4-human/output/a0_annotation/a2.bin_mfe/b1.split/b20.mfv.tsv' > c20.log 2>&1 &
nohup docker run -v /home/xugang/:/home/xugang/ fjossinet/assemble2 bash -c 'RNAfold -p < /home/xugang/data/rnastructure/a4-human/output/a0_annotation/a2.bin_mfe/b1.split/b21.trans.fa > /home/xugang/data/rnastructure/a4-human/output/a0_annotation/a2.bin_mfe/b1.split/b21.mfv.tsv' > c21.log 2>&1 &
nohup docker run -v /home/xugang/:/home/xugang/ fjossinet/assemble2 bash -c 'RNAfold -p < /home/xugang/data/rnastructure/a4-human/output/a0_annotation/a2.bin_mfe/b1.split/b22.trans.fa > /home/xugang/data/rnastructure/a4-human/output/a0_annotation/a2.bin_mfe/b1.split/b22.mfv.tsv' > c22.log 2>&1 &
nohup docker run -v /home/xugang/:/home/xugang/ fjossinet/assemble2 bash -c 'RNAfold -p < /home/xugang/data/rnastructure/a4-human/output/a0_annotation/a2.bin_mfe/b1.split/b23.trans.fa > /home/xugang/data/rnastructure/a4-human/output/a0_annotation/a2.bin_mfe/b1.split/b23.mfv.tsv' > c23.log 2>&1 &
nohup docker run -v /home/xugang/:/home/xugang/ fjossinet/assemble2 bash -c 'RNAfold -p < /home/xugang/data/rnastructure/a4-human/output/a0_annotation/a2.bin_mfe/b1.split/b24.trans.fa > /home/xugang/data/rnastructure/a4-human/output/a0_annotation/a2.bin_mfe/b1.split/b24.mfv.tsv' > c24.log 2>&1 &
nohup docker run -v /home/xugang/:/home/xugang/ fjossinet/assemble2 bash -c 'RNAfold -p < /home/xugang/data/rnastructure/a4-human/output/a0_annotation/a2.bin_mfe/b1.split/b25.trans.fa > /home/xugang/data/rnastructure/a4-human/output/a0_annotation/a2.bin_mfe/b1.split/b25.mfv.tsv' > c25.log 2>&1 &
nohup docker run -v /home/xugang/:/home/xugang/ fjossinet/assemble2 bash -c 'RNAfold -p < /home/xugang/data/rnastructure/a4-human/output/a0_annotation/a2.bin_mfe/b1.split/b26.trans.fa > /home/xugang/data/rnastructure/a4-human/output/a0_annotation/a2.bin_mfe/b1.split/b26.mfv.tsv' > c26.log 2>&1 &
nohup docker run -v /home/xugang/:/home/xugang/ fjossinet/assemble2 bash -c 'RNAfold -p < /home/xugang/data/rnastructure/a4-human/output/a0_annotation/a2.bin_mfe/b1.split/b27.trans.fa > /home/xugang/data/rnastructure/a4-human/output/a0_annotation/a2.bin_mfe/b1.split/b27.mfv.tsv' > c27.log 2>&1 &
nohup docker run -v /home/xugang/:/home/xugang/ fjossinet/assemble2 bash -c 'RNAfold -p < /home/xugang/data/rnastructure/a4-human/output/a0_annotation/a2.bin_mfe/b1.split/b28.trans.fa > /home/xugang/data/rnastructure/a4-human/output/a0_annotation/a2.bin_mfe/b1.split/b28.mfv.tsv' > c28.log 2>&1 &
nohup docker run -v /home/xugang/:/home/xugang/ fjossinet/assemble2 bash -c 'RNAfold -p < /home/xugang/data/rnastructure/a4-human/output/a0_annotation/a2.bin_mfe/b1.split/b29.trans.fa > /home/xugang/data/rnastructure/a4-human/output/a0_annotation/a2.bin_mfe/b1.split/b29.mfv.tsv' > c29.log 2>&1 &
nohup docker run -v /home/xugang/:/home/xugang/ fjossinet/assemble2 bash -c 'RNAfold -p < /home/xugang/data/rnastructure/a4-human/output/a0_annotation/a2.bin_mfe/b1.split/b30.trans.fa > /home/xugang/data/rnastructure/a4-human/output/a0_annotation/a2.bin_mfe/b1.split/b30.mfv.tsv' > c30.log 2>&1 &
nohup docker run -v /home/xugang/:/home/xugang/ fjossinet/assemble2 bash -c 'RNAfold -p < /home/xugang/data/rnastructure/a4-human/output/a0_annotation/a2.bin_mfe/b1.split/b31.trans.fa > /home/xugang/data/rnastructure/a4-human/output/a0_annotation/a2.bin_mfe/b1.split/b31.mfv.tsv' > c31.log 2>&1 &
nohup docker run -v /home/xugang/:/home/xugang/ fjossinet/assemble2 bash -c 'RNAfold -p < /home/xugang/data/rnastructure/a4-human/output/a0_annotation/a2.bin_mfe/b1.split/b32.trans.fa > /home/xugang/data/rnastructure/a4-human/output/a0_annotation/a2.bin_mfe/b1.split/b32.mfv.tsv' > c32.log 2>&1 &
nohup docker run -v /home/xugang/:/home/xugang/ fjossinet/assemble2 bash -c 'RNAfold -p < /home/xugang/data/rnastructure/a4-human/output/a0_annotation/a2.bin_mfe/b1.split/b33.trans.fa > /home/xugang/data/rnastructure/a4-human/output/a0_annotation/a2.bin_mfe/b1.split/b33.mfv.tsv' > c33.log 2>&1 &
nohup docker run -v /home/xugang/:/home/xugang/ fjossinet/assemble2 bash -c 'RNAfold -p < /home/xugang/data/rnastructure/a4-human/output/a0_annotation/a2.bin_mfe/b1.split/b34.trans.fa > /home/xugang/data/rnastructure/a4-human/output/a0_annotation/a2.bin_mfe/b1.split/b34.mfv.tsv' > c34.log 2>&1 &
nohup docker run -v /home/xugang/:/home/xugang/ fjossinet/assemble2 bash -c 'RNAfold -p < /home/xugang/data/rnastructure/a4-human/output/a0_annotation/a2.bin_mfe/b1.split/b35.trans.fa > /home/xugang/data/rnastructure/a4-human/output/a0_annotation/a2.bin_mfe/b1.split/b35.mfv.tsv' > c35.log 2>&1 &
nohup docker run -v /home/xugang/:/home/xugang/ fjossinet/assemble2 bash -c 'RNAfold -p < /home/xugang/data/rnastructure/a4-human/output/a0_annotation/a2.bin_mfe/b1.split/b36.trans.fa > /home/xugang/data/rnastructure/a4-human/output/a0_annotation/a2.bin_mfe/b1.split/b36.mfv.tsv' > c36.log 2>&1 &
nohup docker run -v /home/xugang/:/home/xugang/ fjossinet/assemble2 bash -c 'RNAfold -p < /home/xugang/data/rnastructure/a4-human/output/a0_annotation/a2.bin_mfe/b1.split/b37.trans.fa > /home/xugang/data/rnastructure/a4-human/output/a0_annotation/a2.bin_mfe/b1.split/b37.mfv.tsv' > c37.log 2>&1 &
nohup docker run -v /home/xugang/:/home/xugang/ fjossinet/assemble2 bash -c 'RNAfold -p < /home/xugang/data/rnastructure/a4-human/output/a0_annotation/a2.bin_mfe/b1.split/b38.trans.fa > /home/xugang/data/rnastructure/a4-human/output/a0_annotation/a2.bin_mfe/b1.split/b38.mfv.tsv' > c38.log 2>&1 &
nohup docker run -v /home/xugang/:/home/xugang/ fjossinet/assemble2 bash -c 'RNAfold -p < /home/xugang/data/rnastructure/a4-human/output/a0_annotation/a2.bin_mfe/b1.split/b39.trans.fa > /home/xugang/data/rnastructure/a4-human/output/a0_annotation/a2.bin_mfe/b1.split/b39.mfv.tsv' > c39.log 2>&1 &
nohup docker run -v /home/xugang/:/home/xugang/ fjossinet/assemble2 bash -c 'RNAfold -p < /home/xugang/data/rnastructure/a4-human/output/a0_annotation/a2.bin_mfe/b1.split/b40.trans.fa > /home/xugang/data/rnastructure/a4-human/output/a0_annotation/a2.bin_mfe/b1.split/b40.mfv.tsv' > c40.log 2>&1 &
```
* a3.extrace.pl
```pl
sub entropy {
    my ($entropy, $len, $p, %t) = (0, length($_[0]));
    my @chars = split '', $_[0];
    $t{$_}++ foreach @chars;

    foreach (values %t) {
        $p = $_/$len;
        $entropy -= $p * log $p ;
    }

    return $entropy / log 2;
}

sub calcgc {
 my $seq = $_[0];
 my @seqarray = split('',$seq);
 my $count = 0;
 foreach my $base (@seqarray) {
   $count++ if $base =~ /[G|C|g|c]/i;
 }
 my $len = $#seqarray+1;
 my $num=$count / $len;
 #my ($dec)=$num =~ /(\S{6})/;
 return $num;
}


open I, "<$ARGV[0]";
open O, ">$ARGV[0].mfg.tsv";



print O "name\tposition\tmfe\tbin\tentropy\tgc\n";
$bin;
while(<I>){
chomp;
if($_=~/^>/)
{
$_=~/>(.*?)\t(\d+)/;
$name=$1;
$position=$2;
#$factor=$length/100;
print O "$name\t$position\t";
}

$seq=<I>;
$bin=length($seq)-1;
$entropy=entropy($seq);
$gc=calcgc($seq);
$st1=<I>;
$st1=~/\(((\s)?.\d+.*?)\)/;
$mfe=$1;
$st2=<I>;
$st3=<I>;
$fre=<I>;
#$mfe_ave=$mfe/$factor;
#$mfe_ave=$mfe;
print O "$mfe\t$bin\t$entropy\t$gc\n";
}

close I;
close O;
```

---

## nanoplish dwell time process

* a14.mean.time.pl

```pl
open I, "<$ARGV[0]";
open O, ">$ARGV[0].neam.tsv";
<I>;
while(<I>){
chomp;
@data=split/\t/,$_;
if($#data==4){
#print "$_\n";
$time{$data[0]}+=$data[4];
$current{$data[0]}+=$data[2];
$count{$data[0]}++;
}
}
print O "name\ttime\tcurrent\n";
foreach(keys(%time)){
$timem=$time{$_}/$count{$_};
$currentm=$current{$_}/$count{$_};
	print O "$_\t$timem\t$currentm\n";

}
```
* a15.postion.mean.pl
```pl
open I, "<$ARGV[0]";
open O, ">$ARGV[0].neam.pos.tsv";
<I>;
while(<I>){
chomp;
@data=split/\t/,$_;
#print "$_\n";
$time{$data[0]."\t".$data[1]}+=$data[4];
$current{$data[0]."\t".$data[1]}+=$data[2];
$count{$data[0]."\t".$data[1]}++;
}
#print O "name\tpos\ttime\tcurrent\tcount\n";
foreach(keys(%time)){
$timem=$time{$_}/$count{$_};
$currentm=$current{$_}/$count{$_};
	print O "$_\t$timem\t$currentm\t$count{$_}\n";

}
close I;
close O;
system("sort -k1,1 -k2,2n ".$ARGV[0].".neam.pos.tsv > ".$ARGV[0].".neam.pos2.tsv");
#system('echo -e "chrosome\tpos\ttime\tcurrent\tcount\n"> tmp.head ');
open I, "<$ARGV[0].neam.pos2.tsv";
open O, ">$ARGV[0].neam.pos.tsv";
print O "name\tpos\ttime\tcurrent\tcount\n";
while(<I>){
print O "$_";
}
close O;
close I;
system("rm  ".$ARGV[0].".neam.pos2.tsv")
```

* a15.postion.mean.pl
```pl
open I, "<$ARGV[0]";
open O, ">$ARGV[0].neam.pos.tsv";
<I>;
while(<I>){
chomp;
@data=split/\t/,$_;
#print "$_\n";
$time{$data[0]."\t".$data[1]}+=$data[4];
$current{$data[0]."\t".$data[1]}+=$data[2];
$count{$data[0]."\t".$data[1]}++;
}
#print O "name\tpos\ttime\tcurrent\tcount\n";
foreach(keys(%time)){
$timem=$time{$_}/$count{$_};
$currentm=$current{$_}/$count{$_};
	print O "$_\t$timem\t$currentm\t$count{$_}\n";

}
close I;
close O;
system("sort -k1,1 -k2,2n ".$ARGV[0].".neam.pos.tsv > ".$ARGV[0].".neam.pos2.tsv");
#system('echo -e "chrosome\tpos\ttime\tcurrent\tcount\n"> tmp.head ');
open I, "<$ARGV[0].neam.pos2.tsv";
open O, ">$ARGV[0].neam.pos.tsv";
print O "name\tpos\ttime\tcurrent\tcount\n";
while(<I>){
print O "$_";
}
close O;
close I;
system("rm  ".$ARGV[0].".neam.pos2.tsv")
```
* a16.trantogenome.pl
```pl
open I, "</home/xugang/data/reference/hg38/Homo_sapiens.GRCh38.100.chr.gtf.exon";
open I2, "<$ARGV[0]";
open O, ">$ARGV[0].chrome.tsv";
$old='oldna';
while(<I>){
chomp;
@data=split/\t/,$_;
$_=~/transcript_id "(.*?)";/;
$trans=$1;
$_=~/exon_number "(.*?)"/;
$exonnumer=$1;
$length=$data[4]-$data[3]+1;
if($old=~/^$trans$/){
$totallength+=$length;
}else{
$old=$trans;
$totallength=$length;
}

#print "$data[0]\t$data[3]\t$data[4]\t$length\t$totallength\t$data[6]\t$exonnumer\t$trans\n";
#$info{$trans."\t".$totallength}="$data[0]\t$data[3]\t$data[4]\t$length\t$totallength\t$data[6]\t$exonnumer\t$trans\n";

$name{$trans}.=$totallength."\t";

$chrom{$trans}=$data[0];
$start{$trans}.=$data[3]."\t";
$end{$trans}.=$data[4]."\t";
$length{$trans}.=$length."\t";
$totallength{$trans}.=$totallength."\t";
$strand{$trans}=$data[6];
$exonnum{$trans}.=$exonnumer."\t";


}
close I;

foreach(keys(%name)){
#print "$_\t$name{$_}\n";
@name=split/\t/,$name{$_};
#print "$#name\t$name[0]\t$name[1]\n";

}

$name='ENST00000666741';
$name='ENST00000341290';
$p1=100;
$p2=200;
$p3=700;
$p=$p3;
#print "$name\t$p\n";

while(<I2>){
chomp;
@d2=split /\t/,$_;
$name=$d2[0];
$p=$d2[1]+1;
$infoa=$_;
if(exists($totallength{$name})){
$totallength=$totallength{$name};
#print "$totallength\n";
@totallength=split/\t/,$totallength; # get total length.
$chrom=$chrom{$name};# get chromsome.
@start=split/\t/,$start{$name};
@end=split/\t/,$end{$name};
@length=split/\t/,$length{$name};
$strand=$strand{$name};
@exonnum=split /\t/,$exonnum{$name};
#set indexes.
$n=0;
#set distance minier 0 and next length...
$distance_minier=0;
if($strand=~/\+/){
foreach(@totallength){
#print "$_\n";
$totall=$_;
# set the last position, and the next will subtrace.
#print "$chrom\t$start[$n]\t$end[$n]\t$length[$n]\t$strand\t$exonnum[$n]\t$totall\n";
if($p<=$totall){ # ditermine the position is smaller than totall length. Then it will be use for next
$distance=$p-$distance_minier-1;
$start_f=$start[$n]+$distance;
#print "p:$p\t distance_miner: $distance_minier\t distance: $distance\n";
print O "$infoa\t$chrom\t$start_f\t$strand\n";

last;
}else{
$distance_minier=$totall;
}
$n++; ## $n as index, and it will be useful to get strand, chrome, start, end,
}


}
if($strand=~/\-/){
foreach(@totallength){
#print "$_\n";
$totall=$_;
# set the last position, and the next will subtrace.
#print "$chrom\t$start[$n]\t$end[$n]\t$length[$n]\t$strand\t$exonnum[$n]\t$totall\n";
if($p<=$totall){ # ditermine the position is smaller than totall length. Then it will be use for next
$distance=$p-$distance_minier-1;
$start_f=$end[$n]-$distance;
#print "p:$p\t distance_miner: $distance_minier\t distance: $distance\n";
print O "$infoa\t$chrom\t$start_f\t$strand\n";

last;
}else{
$distance_minier=$totall;
}
$n++; ## $n as index, and it will be useful to get strand, chrome, start, end,
}
}
}


}
close O;
```
* a17.mfg.pl
```pl
open I1, "</home/xugang/data/rnastructure_database/all_chr_metrics_forward.out_.gff3";
open I2, "</home/xugang/data/rnastructure_database/all_chr_metrics_reverse.out_.gff3";
open O, ">genome.mef.gc.tsv";
print O "chr\tstart\tend\tstrand\tmfe\tzscore\tpvalue\tdnsDiv\tFreqmef\tgc\n";
sub calcgc {
 my $seq = $_[0];
 my @seqarray = split('',$seq);
 my $count = 0;
 foreach my $base (@seqarray) {
   $count++ if $base =~ /[G|C|g|c]/i;
 }
 my $len = $#seqarray+1;
 my $num=$count / $len;
 #my ($dec)=$num =~ /(\S{6})/;
 return $num;
}




while(<I1>){
chomp;
# use split array
@data=split/\t/,$_;
$data[0]=~s/^chr//g;
$data[8]=~/MFE=(.*?);Z-Score=(.*?);P-value=(.*?);EnsDiv=(.*?);freqMFE=(.*?);Sequence=(.*?);/;
$mfe=$1;
$zscore=$2;
$pv=$3;
$dnsdiv=$4;
$freqmef=$5;
$seq=$6;
$gc=calcgc($seq);
#output
#print
print O "$data[0]\t$data[3]\t$data[4]\t$data[6]\t$mfe\t$zscore\t$pv\t$dnsdiv\t$freqmef\t$gc\n";

}


while(<I2>){
chomp;
# use split array
@data=split/\t/,$_;
$data[0]=~s/^chr//g;
$data[8]=~/MFE=(.*?);Z-Score=(.*?);P-value=(.*?);EnsDiv=(.*?);freqMFE=(.*?);Sequence=(.*?);/;
$mfe=$1;
$zscore=$2;
$pv=$3;
$dnsdiv=$4;
$freqmef=$5;
$seq=$6;
$gc=calcgc($seq);
#output
#print
print O "$data[0]\t$data[3]\t$data[4]\t$data[6]\t$mfe\t$zscore\t$pv\t$dnsdiv\t$freqmef\t$gc\n";

}



close I1;
close I2;
```
* a18.time.mfe.genome.pl
```pl
open I, "<$ARGV[0]";
open I2, "<genome.mef.gc.tsv";
open O, ">$ARGV[0].merge.tsv";
while(<I2>){
chomp;
@data=split/\t/,$_;
# name is chromesome, strand, start.
$name=$data[0]."\t".$data[3]."\t".$data[1];
$hash{$name}=$data[4]."\t".$data[5]."\t".$data[6]."\t".$data[7]."\t".$data[8]."\t".$data[9];
#print "$name\n";
}
close I2;

while(<I>){
chomp;
@data=split /\t/,$_;
#$reminder=$data[6]%40;
#$transform=$data[6]-$reminder+1;
#print "$transform\n";
#print "$_\t$transform\n";
#$name=$data[5]."\t".$data[7]."\t".$transform;
$name=$data[0]."\t+\t".$data[1];
#print "2:$name\n";
if(exists($hash{$name})){
print O "$_\t$hash{$name}\n";
}
}

close I;
close O;
```
* a18.time.mfe.pl
```pl
open I, "<$ARGV[0]";
open I2, "<genome.mef.tsv";
open O, ">$ARGV[0].merge.tsv";
while(<I2>){
chomp;
@data=split/\t/,$_;
$name=$data[0]."\t".$data[3]."\t".$data[1];
$hash{$name}=$data[4]."\t".$data[5]."\t".$data[6]."\t".$data[7]."\t".$data[8];
#print "$name\n";
}
close I2;

while(<I>){
chomp;
@data=split /\t/,$_;
$reminder=$data[6]%40;
$transform=$data[6]-$reminder+1;
#print "$transform\n";
#print "$_\t$transform\n";
$name=$data[5]."\t".$data[7]."\t".$transform;
#print "2:$name\n";
if(exists($hash{$name})){
print O "$_\t$transform\t$hash{$name}\n";
}
}

close I;
close O;
```
* a20.mtg.plaa

|contig|	position|	reference_kmer|	read_name	|strand	|event_index|	event_level_mean|	event_stdvevent_length	|model_kmer|	model_mean|	model_stdv	|standardized_level|
|-|-|-|-|-|-|-|-|-|-|-|-|
|8|	4787165|	GTAGC|	03b065a1-4e19-4789-8598-8d97dc5079b4|	t|	100	|85.18|	1.738|	0.00332	|GTAGC|	92.91|	3.78|	-1.57|
|8	|4787165	|GTAGC|	03b065a1-4e19-4789-8598-8d97dc5079b4|	t|	101	97.24	|3.180|	0.00432	|GTAGC	|92.91|	3.78|	0.88|


* ab.split.pl
```pl
open I, "<$ARGV[0]";
#open O, ">";
$head=<I>;
while(<I>){
#chomp;
@data=split/\t/,$_;
$hash{$data[0]}.=$_;

}
close I;
system("mkdir tmp");
foreach(keys(%hash)){
$file=$_;
	open O, ">>tmp/$file";
print O $head;
print O "$hash{$_}";
	close O;
}

```

* ac.check.pl
```pl
open I, "<$ARGV[0]";
open O, ">$ARGV[0].ch";
$head=<I>;
if($head=~/^contig/){
print O $head;
}else{
print O "contig\tposition\treference_kmer\tread_name\tstrand\tevent_index\tevent_level_mean\tevent_stdv\tevent_length\tmodel_kmer\tmodel_mean\tmodel_stdv\tstandardized_level\n";
}
while(<I>){

print O "$_";
}
close I;
close O;

```

## R script
* current.R
```R

library(ggplot2)
#simulate
simulate=read.table('/Users/xugang/Desktop/sequencing_center_desktop/2020/rna_structure/data_plot/tetra_rna-time.csv',sep="\t",header=T)
head(simulate)

ggplot(simulate,aes(x=position,y=nanopore))+ geom_line()+ scale_x_continuous(breaks = seq(0, 425, by = 10))

#
unmod2_time=read.table('/Users/xugang/Desktop/sequencing_center_desktop/2020/rna_structure/data_plot/Tetra_unmod2.gene.event.sta.tsv',sep="\t",header=T)
head(unmod2_time)

ggplot(unmod2_time,aes(x=position,y=time))+ geom_line()+ scale_x_continuous(breaks = seq(0, 425, by = 10))

unmod2_reactivity=read.table('/Users/xugang/Desktop/sequencing_center_desktop/2020/rna_structure/data_plot/Output_TeTrahymenaTetra_unmod2.combined.RData.csv',sep=",",header=T)
head(unmod2_reactivity)

ggplot(unmod2_reactivity,aes(x=Position,y=Mod_percentage))+ geom_line()+ scale_x_continuous(breaks = seq(0, 425, by = 10))
# 2
Tetra_NAI_N3_1=read.table('/Users/xugang/Desktop/sequencing_center_desktop/2020/rna_structure/data_plot/Tetra_NAI_N3_1.gene.event.sta.tsv',sep="\t",header=T)
head(Tetra_NAI_N3_1)

ggplot(Tetra_NAI_N3_1,aes(x=position,y=time))+ geom_line()+ scale_x_continuous(breaks = seq(0, 425, by = 10))

NAI_N3_1_reactivity=read.table('/Users/xugang/Desktop/sequencing_center_desktop/2020/rna_structure/data_plot/Output_TeTrahymenaTetra_NAI_N3_1.combined.RData.csv',sep=",",header=T)
head(NAI_N3_1_reactivity)

ggplot(NAI_N3_1_reactivity,aes(x=Position,y=Mod_percentage))+ geom_line()+ scale_x_continuous(breaks = seq(0, 425, by = 10))
#
par(mar = c(5, 4, 4, 4) + 0.3)              # Additional space for second y-axis
plot(Tetra_NAI_N3_1$position[110:150], Tetra_NAI_N3_1$time[110:150], pch = 16, col = 2, type = "l")   # Basic line plot in R) 
axis(side=1, at=seq(110, 150, by=1))
# Create first plot
par(new = TRUE)                             # Add new plot
# Create second plot without axes
plot(NAI_N3_1_reactivity$Position[110:150], NAI_N3_1_reactivity$Mod_percentage[110:150], pch = 17, col = 3,              
     axes = FALSE, xlab = "", ylab = "",type="l")
axis(side = 4, at = pretty(range(NAI_N3_1_reactivity$Mod_percentage)))      # Add second axis
mtext("reactivity", side = 4, line = 3)             # Add second axis label
dev.off()
#

#
par(mar = c(5, 4, 4, 4) + 0.3)              # Additional space for second y-axis
plot(Tetra_NAI_N3_1$position[100:200], Tetra_NAI_N3_1$time[100:200], pch = 16, col = 2, type = "l")   # Basic line plot in R) 
axis(side=1, at=seq(100, 200, by=5))
# Create first plot
par(new = TRUE)                             # Add new plot
# Create second plot without axes
plot(NAI_N3_1_reactivity$Position[100:200], NAI_N3_1_reactivity$Mod_percentage[100:200], pch = 17, col = 3,              
     axes = FALSE, xlab = "", ylab = "",type="l")
axis(side = 4, at = pretty(range(NAI_N3_1_reactivity$Mod_percentage)))      # Add second axis
mtext("reactivity", side = 4, line = 3)             # Add second axis label
dev.off()
#
par(mar = c(5, 4, 4, 4) + 0.3)              # Additional space for second y-axis
plot(Tetra_NAI_N3_1$position, Tetra_NAI_N3_1$time, pch = 16, col = 2, type = "l")   # Basic line plot in R) 
axis(side=1, at=seq(0, 430, by=10))
# Create first plot
par(new = TRUE)                             # Add new plot
# Create second plot without axes
plot(NAI_N3_1_reactivity$Position, NAI_N3_1_reactivity$Mod_percentage, pch = 17, col = 3,              
     axes = FALSE, xlab = "", ylab = "",type="l")
axis(side = 4, at = pretty(range(NAI_N3_1_reactivity$Mod_percentage)))      # Add second axis
mtext("reactivity", side = 4, line = 3)             # Add second axis label
dev.off()

#3
Tetra_NAI_N3_2=read.table('/Users/xugang/Desktop/sequencing_center_desktop/2020/rna_structure/data_plot/Tetra_NAI_N3_2.gene.event.sta.tsv',sep="\t",header=T)
head(Tetra_NAI_N3_2)

ggplot(Tetra_NAI_N3_2,aes(x=position,y=time))+ geom_line()+ scale_x_continuous(breaks = seq(0, 425, by = 10))

NAI_N3_2_reactivity=read.table('/Users/xugang/Desktop/sequencing_center_desktop/2020/rna_structure/data_plot/Output_TeTrahymenaTetra_NAI_N3_2.combined.RData.csv',sep=",",header=T)
head(NAI_N3_2_reactivity)

ggplot(NAI_N3_2_reactivity,aes(x=Position,y=Mod_percentage))+ geom_line()+ scale_x_continuous(breaks = seq(0, 425, by = 10))

#4
Tetra_NAI_N3_25=read.table('/Users/xugang/Desktop/sequencing_center_desktop/2020/rna_structure/data_plot/Tetra_NAI_N3_25min.gene.event.sta.tsv',sep="\t",header=T)
head(Tetra_NAI_N3_25)

ggplot(Tetra_NAI_N3_25,aes(x=position,y=time))+ geom_line()+ scale_x_continuous(breaks = seq(0, 425, by = 10))

NAI_N3_25_reactivity=read.table('/Users/xugang/Desktop/sequencing_center_desktop/2020/rna_structure/data_plot/Output_TeTrahymenaTetra_NAI_N3_25min.combined.RData.csv',sep=",",header=T)
head(NAI_N3_25_reactivity)

ggplot(NAI_N3_25_reactivity,aes(x=Position,y=Mod_percentage))+ geom_line()+ scale_x_continuous(breaks = seq(0, 425, by = 10))


#5
Tetra_denature=read.table('/Users/xugang/Desktop/sequencing_center_desktop/2020/rna_structure/data_plot/Tetra_NAI_N3_denatured.gene.event.sta.tsv',sep="\t",header=T)
head(Tetra_denature)

ggplot(Tetra_denature,aes(x=position,y=time))+ geom_line()+ scale_x_continuous(breaks = seq(0, 425, by = 10))

Tetra_denature_reactivity=read.table('/Users/xugang/Desktop/sequencing_center_desktop/2020/rna_structure/data_plot/Output_TeTrahymenaTetra_NAI_N3_denatured.combined.RData.csv',sep=",",header=T)
head(Tetra_denature_reactivity)

ggplot(Tetra_denature_reactivity,aes(x=Position,y=Mod_percentage))+ geom_line()+ scale_x_continuous(breaks = seq(0, 425, by = 10))

#
#
head(unmod2_time)
# unmodifited time and NAI-N3-I reactivity score.
par(mar = c(5, 4, 4, 4) + 0.3)              # Additional space for second y-axis
plot(unmod2_time$position, unmod2_time$time, pch = 16, col = 2, type = "l")   # Basic line plot in R) 
axis(side=1, at=seq(10, 420, by=10))
# Create first plot
par(new = TRUE)                             # Add new plot
# Create second plot without axes
plot(NAI_N3_1_reactivity$Position, NAI_N3_1_reactivity$Mod_percentage, pch = 17, col = 3,              
     axes = FALSE, xlab = "", ylab = "",type="l")
axis(side = 4, at = pretty(range(NAI_N3_1_reactivity$Mod_percentage)))      # Add second axis
mtext("NAI_N3_1 reactivity", side = 4, line = 3)             # Add second axis label
dev.off()
#
# unmodifited time and NAI-N3-I reactivity score. part
par(mar = c(5, 4, 4, 4) + 0.3)              # Additional space for second y-axis
plot(unmod2_time$position[280:330], unmod2_time$time[280:330], pch = 16, col = 2, type = "l")   # Basic line plot in R) 
axis(side=1, at=seq(280, 330, by=2))
# Create first plot
par(new = TRUE)                             # Add new plot
# Create second plot without axes
plot(NAI_N3_1_reactivity$Position[280:330], NAI_N3_1_reactivity$Mod_percentage[280:330], pch = 17, col = 3,              
     axes = FALSE, xlab = "", ylab = "",type="l")
axis(side = 4, at = pretty(range(NAI_N3_1_reactivity$Mod_percentage)))      # Add second axis
mtext("NAI_N3_1 reactivity", side = 4, line = 3)             # Add second axis label
dev.off()
#
# 
ggplot(simulate,aes(x=position,y=nanopore))+ geom_line()+ scale_x_continuous(breaks = seq(0, 425, by = 10))

# NAI simulate and time. full region
par(mar = c(5, 4, 4, 4) + 0.3)              # Additional space for second y-axis
plot(Tetra_NAI_N3_1$position, Tetra_NAI_N3_1$time, pch = 16, col = 2, type = "l")   # Basic line plot in R) 
axis(side=1, at=seq(0, 430, by=10))
par(new = TRUE)
plot(simulate$position, simulate$nanopore, pch = 17, col = 4,              
     axes = FALSE, xlab = "", ylab = "",type="h")
axis(side = 4, at = pretty(range(simulate$nanopore)))      # Add second axis
mtext("simulate", side = 4, line = 3)  


simulate$position[1:385]
simulate$nanopore[1:385]
Tetra_NAI_N3_1$position[30:414]
Tetra_NAI_N3_1$time[30:414]
# NAI simulate and time. 378-414 region
par(mar = c(5, 4, 4, 4) + 0.3)              # Additional space for second y-axis
plot(Tetra_NAI_N3_1$position[378:414], Tetra_NAI_N3_1$time[378:414], pch = 16, col = 2, type = "l")   # Basic line plot in R) 
axis(side=1, at=seq(0, 430, by=10))
par(new = TRUE)
plot(simulate$position[350:385], simulate$nanopore[350:385], pch = 17, col = 4,              
     axes = FALSE, xlab = "", ylab = "",type="h")
axis(side = 4, at = pretty(range(simulate$nanopore)))      # Add second axis
mtext("simulate", side = 4, line = 3)  
#
simulate$position[326:365]
simulate$nanopore[326:365]
Tetra_NAI_N3_1$position[354:393]
Tetra_NAI_N3_1$time[354:393]
# NAI simulate and time. 344:383 region
par(mar = c(5, 4, 4, 4) + 0.3)              # Additional space for second y-axis
plot(Tetra_NAI_N3_1$position[354:393], Tetra_NAI_N3_1$time[354:393], pch = 16, col = 2, type = "l")   # Basic line plot in R) 
axis(side=1, at=seq(0, 430, by=10))
par(new = TRUE)
plot(simulate$position[326:365], simulate$nanopore[326:365], pch = 17, col = 4,              
     axes = FALSE, xlab = "", ylab = "",type="h")
axis(side = 4, at = pretty(range(simulate$nanopore)))      # Add second axis
mtext("simulate", side = 4, line = 3)  
#
simulate$position[1:385]
simulate$nanopore[1:385]
Tetra_NAI_N3_1$position[30:414]
Tetra_NAI_N3_1$time[30:414]
#shift 0
plot(simulate$nanopore[1:385],Tetra_NAI_N3_1$time[30:414])
cor.test(simulate$nanopore[1:385],Tetra_NAI_N3_1$time[30:414])
#-0.03565819 
#shift 1
plot(simulate$nanopore[1:385],Tetra_NAI_N3_1$time[29:413])
cor.test(simulate$nanopore[1:385],Tetra_NAI_N3_1$time[29:413])
#-0.03086573
#shift 2
plot(simulate$nanopore[1:385],Tetra_NAI_N3_1$time[28:412])
cor.test(simulate$nanopore[1:385],Tetra_NAI_N3_1$time[28:412])
#0.002251606
#shift 3
plot(simulate$nanopore[1:385],Tetra_NAI_N3_1$time[27:411])
cor.test(simulate$nanopore[1:385],Tetra_NAI_N3_1$time[27:411])
#-0.05029425
#shift 4
plot(simulate$nanopore[1:385],Tetra_NAI_N3_1$time[26:410])
cor.test(simulate$nanopore[1:385],Tetra_NAI_N3_1$time[26:410])
#-0.04943732
#shift 5
plot(simulate$nanopore[1:385],Tetra_NAI_N3_1$time[25:409])
cor.test(simulate$nanopore[1:385],Tetra_NAI_N3_1$time[25:409])
#-0.1108445
#shift 6
plot(simulate$nanopore[1:385],Tetra_NAI_N3_1$time[24:408])
cor.test(simulate$nanopore[1:385],Tetra_NAI_N3_1$time[24:408])
#-0.1303932
#shift 7
plot(simulate$nanopore[1:385],Tetra_NAI_N3_1$time[23:407])
cor.test(simulate$nanopore[1:385],Tetra_NAI_N3_1$time[23:407])
#-0.08999292
#shift 8
plot(simulate$nanopore[1:385],Tetra_NAI_N3_1$time[22:406])
cor.test(simulate$nanopore[1:385],Tetra_NAI_N3_1$time[22:406])
#-0.03313279 
#shift 9
plot(simulate$nanopore[1:385],Tetra_NAI_N3_1$time[21:405])
cor.test(simulate$nanopore[1:385],Tetra_NAI_N3_1$time[21:405])
#-0.02798022
#shift 10
plot(simulate$nanopore[1:385],Tetra_NAI_N3_1$time[20:404])
cor.test(simulate$nanopore[1:385],Tetra_NAI_N3_1$time[20:404])
#-0.06321941
#shift 11
plot(simulate$nanopore[1:385],Tetra_NAI_N3_1$time[19:403])
cor.test(simulate$nanopore[1:385],Tetra_NAI_N3_1$time[19:403])
#-0.07048435
#shift 12
plot(simulate$nanopore[1:385],Tetra_NAI_N3_1$time[18:402])
cor.test(simulate$nanopore[1:385],Tetra_NAI_N3_1$time[18:402])
#-0.02758726
#shift 13
plot(simulate$nanopore[1:385],Tetra_NAI_N3_1$time[17:401])
cor.test(simulate$nanopore[1:385],Tetra_NAI_N3_1$time[17:401])
#0.05730962
#shift 14
plot(simulate$nanopore[1:385],Tetra_NAI_N3_1$time[16:400])
cor.test(simulate$nanopore[1:385],Tetra_NAI_N3_1$time[16:400])
#0.05457758
#shift 15
plot(simulate$nanopore[1:385],Tetra_NAI_N3_1$time[15:399])
cor.test(simulate$nanopore[1:385],Tetra_NAI_N3_1$time[15:399])
#0.06964819
#shift 16
plot(simulate$nanopore[1:385],Tetra_NAI_N3_1$time[14:398])
cor.test(simulate$nanopore[1:385],Tetra_NAI_N3_1$time[14:398])
#0.1045263
#shift 17
plot(simulate$nanopore[1:385],Tetra_NAI_N3_1$time[13:397])
cor.test(simulate$nanopore[1:385],Tetra_NAI_N3_1$time[13:397])
#0.132324
#shift 18
plot(simulate$nanopore[1:385],Tetra_NAI_N3_1$time[12:396])
cor.test(simulate$nanopore[1:385],Tetra_NAI_N3_1$time[12:396])
#0.09978341
#shift 19
plot(simulate$nanopore[1:385],Tetra_NAI_N3_1$time[11:395])
cor.test(simulate$nanopore[1:385],Tetra_NAI_N3_1$time[11:395])
#0.07058782
# time mod vs unmod
par(mar = c(5, 4, 4, 4) + 0.3)              # Additional space for second y-axis
plot(unmod2_time$position, unmod2_time$time, pch = 16, col = 2, type = "l",ylim=c(0.005,0.01))   # Basic line plot in R) 
axis(side=1, at=seq(0, 420, by=10))
par(new = TRUE)  
plot(Tetra_NAI_N3_1$position, Tetra_NAI_N3_1$time, pch = 16, col = 4, type = "l",ylim=c(0.005,0.01))  

# Create first plot
par(new = TRUE)                             # Add new plot
# Create second plot without axes
plot(NAI_N3_1_reactivity$Position, NAI_N3_1_reactivity$Mod_percentage, pch = 17, col = 3,              
     axes = FALSE, xlab = "", ylab = "",type="l",ylim=c(0,0.08))
axis(side = 4, at = pretty(range(NAI_N3_1_reactivity$Mod_percentage)))      # Add second axis
mtext("NAI_N3_1 reactivity", side = 4, line = 3)             # Add second axis label
dev.off()

# time mod vs unmod
par(mar = c(5, 4, 4, 4) + 0.3)              # Additional space for second y-axis
plot(unmod2_time$position[280:320], unmod2_time$time[280:320], pch = 16, col = 2, type = "l",ylim=c(0.005,0.01))   # Basic line plot in R) 
axis(side=1, at=seq(0, 420, by=2))
par(new = TRUE)  
plot(Tetra_NAI_N3_1$position[280:320], Tetra_NAI_N3_1$time[280:320], pch = 16, col = 4, type = "l",ylim=c(0.005,0.01))  

# Create first plot
par(new = TRUE)                             # Add new plot
# Create second plot without axes
plot(NAI_N3_1_reactivity$Position[280:320], NAI_N3_1_reactivity$Mod_percentage[280:320], pch = 17, col = 3,              
     axes = FALSE, xlab = "", ylab = "",type="l",ylim=c(0,0.08))
axis(side = 4, at = pretty(range(NAI_N3_1_reactivity$Mod_percentage)))      # Add second axis
mtext("NAI_N3_1 reactivity", side = 4, line = 3)             # Add second axis label
dev.off()


# time mod vs unmod
par(mar = c(5, 4, 4, 4) + 0.3)              # Additional space for second y-axis
startn=230
endn=280
plot(unmod2_time$position[startn:endn], unmod2_time$time[startn:endn], pch = 16, col = 2, type = "l",ylim=c(0.005,0.01))   # Basic line plot in R) 
axis(side=1, at=seq(0, 420, by=2))
par(new = TRUE)  
plot(Tetra_NAI_N3_1$position[startn:endn], Tetra_NAI_N3_1$time[startn:endn], pch = 16, col = 4, type = "l",ylim=c(0.005,0.01))  

# Create first plot
par(new = TRUE)                             # Add new plot
# Create second plot without axes
plot(NAI_N3_1_reactivity$Position[startn:endn], NAI_N3_1_reactivity$Mod_percentage[startn:endn], pch = 17, col = 3,              
     axes = FALSE, xlab = "", ylab = "",type="l",ylim=c(0,0.08))
axis(side = 4, at = pretty(range(NAI_N3_1_reactivity$Mod_percentage)))      # Add second axis
mtext("NAI_N3_1 reactivity", side = 4, line = 3)             # Add second axis label
dev.off()

# time mod vs unmod
par(mar = c(5, 4, 4, 4) + 0.3)              # Additional space for second y-axis
startn=120
endn=160
plot(unmod2_time$position[startn:endn], unmod2_time$time[startn:endn], pch = 16, col = 2, type = "l",ylim=c(0.005,0.01))   # Basic line plot in R) 
axis(side=1, at=seq(startn,endn, by=2))
par(new = TRUE)  
plot(Tetra_NAI_N3_1$position[startn:endn], Tetra_NAI_N3_1$time[startn:endn], pch = 16, col = 4, type = "l",ylim=c(0.005,0.01))  

# Create first plot
par(new = TRUE)                             # Add new plot
# Create second plot without axes
plot(NAI_N3_1_reactivity$Position[startn:endn], NAI_N3_1_reactivity$Mod_percentage[startn:endn], pch = 17, col = 3,              
     axes = FALSE, xlab = "", ylab = "",type="l",ylim=c(0,0.08))
axis(side = 4, at = pretty(range(NAI_N3_1_reactivity$Mod_percentage)))      # Add second axis
mtext("NAI_N3_1 reactivity", side = 4, line = 3)             # Add second axis label
dev.off()
#unmod2_time

unmod2_time$trans=mean(unmod2_time$time)-unmod2_time$time
unmod2_time$trans
unmod2_time$trans_zscore=(unmod2_time$trans-mean(unmod2_time$trans))/sd(unmod2_time$trans)
unmod2_time$position20=unmod2_time$position-20
head(unmod2_time)
par(mar = c(5, 4, 4, 4) + 0.3)              # Additional space for second y-axis
startn=1
endn=20
plot(unmod2_time$position20[startn:endn], unmod2_time$trans_zscore[startn:endn], pch = 16, col = 2, type = "l")   # Basic line plot in R) 
#axis(side=1, at=seq(startn,endn, by=2))

# Create first plot
par(new = TRUE)                             # Add new plot
# Create second plot without axes
plot(NAI_N3_1_reactivity$Position[startn:endn], NAI_N3_1_reactivity$Mod_percentage[startn:endn], pch = 17, col = 3,              
     axes = FALSE, xlab = "", ylab = "",type="l" )
axis(side = 4, at = pretty(range(NAI_N3_1_reactivity$Mod_percentage)))      # Add second axis
mtext("NAI_N3_1 reactivity", side = 4, line = 3)             # Add second axis label
dev.off()
head(NAI_N3_1_reactivity)

#move 20bp
unmod2_time$trans=mean(unmod2_time$time)-unmod2_time$time
unmod2_time$trans
unmod2_time$position20=unmod2_time$position-22
head(unmod2_time)
head(NAI_N3_1_reactivity)
unmod2_time[22:25,]
#merge two table.
data_plot=merge(unmod2_time,NAI_N3_1_reactivity,by.x="position20",by.y="Position")
head(data_plot)
#plot 
startn=0
endn=100
plot(data_plot$position[startn:endn], data_plot$trans[startn:endn], pch = 16, col = 2, type = "l")   # Basic line plot in R) 
par(new = TRUE)                             # Add new plot
# Create second plot without axes
plot(data_plot$position[startn:endn], data_plot$Mod_percentage[startn:endn], pch = 17, col = 3,              
     axes = F, xlab = "", ylab = "",type="l")
axis(side = 4, at = pretty(range(NAI_N3_1_reactivity$Mod_percentage)))      # Add second axis
mtext("NAI_N3_1 reactivity", side = 4, line = 3)             # Add second axis label
dev.off()
```
* tetre.R

```R
library(data.table)
library(ggplot2)

# read rawdata.
##  simulate
simulate=read.table('/Users/xugang/Desktop/sequencing_center_desktop/2020/rna_structure/data_plot/tetra_rna-time.tsv',sep="\t",header=T)
## load unmod time.
unmod2_time=read.table('/Users/xugang/Desktop/sequencing_center_desktop/2020/rna_structure/data_plot/Tetra_unmod2.gene.event.sta.tsv',sep="\t",header=T)
unmod2_reactivity=read.table('/Users/xugang/Desktop/sequencing_center_desktop/2020/rna_structure/data_plot/Output_TeTrahymenaTetra_unmod2.combined.RData.csv',sep=",",header=T)
## load nai_n3 mod time.
NAI_N3_1_time=read.table('/Users/xugang/Desktop/sequencing_center_desktop/2020/rna_structure/data_plot/Tetra_NAI_N3_1.gene.event.sta.tsv',sep="\t",header=T)
NAI_N3_1_reactivity=read.table('/Users/xugang/Desktop/sequencing_center_desktop/2020/rna_structure/data_plot/Output_TeTrahymenaTetra_NAI_N3_1.combined.RData.csv',sep=",",header=T)
## load nai_n3 2 replace.
NAI_N3_2_time=read.table('/Users/xugang/Desktop/sequencing_center_desktop/2020/rna_structure/data_plot/Tetra_NAI_N3_2.gene.event.sta.tsv',sep="\t",header=T)
NAI_N3_2_reactivity=read.table('/Users/xugang/Desktop/sequencing_center_desktop/2020/rna_structure/data_plot/Output_TeTrahymenaTetra_NAI_N3_2.combined.RData.csv',sep=",",header=T)
## load treate NAI N3 25 min.
NAI_N3_25_time=read.table('/Users/xugang/Desktop/sequencing_center_desktop/2020/rna_structure/data_plot/Tetra_NAI_N3_25min.gene.event.sta.tsv',sep="\t",header=T)
NAI_N3_25_reactivity=read.table('/Users/xugang/Desktop/sequencing_center_desktop/2020/rna_structure/data_plot/Output_TeTrahymenaTetra_NAI_N3_25min.combined.RData.csv',sep=",",header=T)
## load denature data.
denature_time=read.table('/Users/xugang/Desktop/sequencing_center_desktop/2020/rna_structure/data_plot/Tetra_NAI_N3_denatured.gene.event.sta.tsv',sep="\t",header=T)
denature_reactivity=read.table('/Users/xugang/Desktop/sequencing_center_desktop/2020/rna_structure/data_plot/Output_TeTrahymenaTetra_NAI_N3_denatured.combined.RData.csv',sep=",",header=T)
## load denature normalize data
denature_time_nor=read.table('/Users/xugang/Desktop/sequencing_center_desktop/2020/rna_structure/data_plot/denature_nor.tsv',sep="\t",header=F)
head(denature_time_nor)
mean(denature_time_nor$V21,na.rm=T)
denature_time_nor_v=colMeans(denature_time_nor[,2:418],na.rm=T)
denature_time_nor_v
position_time_nor_p=seq(0,416,by = 1)
denature_time_nor_table=data.frame(position_time_nor_p,denature_time_nor_v)
head(denature_time_nor_table)


# 2. check load data.
print('simulate')
head(simulate)
print('unmod2_time')
head(unmod2_time)
print('unmod2_reactivity')
head(unmod2_reactivity)
print('NAI_N3_1_time')
head(NAI_N3_1_time)
print('NAI_N3_1_reactivity')
head(NAI_N3_1_reactivity)
print('NAI_N3_2_time')
head(NAI_N3_2_time)
print('NAI_N3_2_reactivity')
head(NAI_N3_2_reactivity)
print('NAI_N3_25_time')
head(NAI_N3_25_time)
print('NAI_N3_25_reactivity')
head(NAI_N3_25_reactivity)
print('denature_time')
head(denature_time)
print('denature_reactivity')
head(denature_reactivity)

# Zscore for reactivity
## unmod2 for reactivity
unmod2_reactivity$reactivity=(unmod2_reactivity$Mod_percentage-mean(unmod2_reactivity$Mod_percentage))/sd(unmod2_reactivity$Mod_percentage)
head(unmod2_reactivity)

# use the denature time to normolize.
unmod2_time=merge(unmod2_time,denature_time_nor_table,by.x="position",by.y="position_time_nor_p")

## time for zscore.
### unmod2 time

head(unmod2_time)
unmod2_time$time_nor_denature=unmod2_time$time-unmod2_time$denature_time_nor_v
unmod2_time$time_nor_zs=(unmod2_time$time_nor_denature-mean(unmod2_time$time_nor_denature))/sd(unmod2_time$time_nor_denature)

unmod2_time$timezs=(unmod2_time$time-mean(unmod2_time$time))/sd(unmod2_time$time)
unmod2_time$timezs_trans=(mean(unmod2_time$time)-unmod2_time$time)/sd(unmod2_time$time)
head(unmod2_time)




## NAI_N3_1 for reactivity 
NAI_N3_1_reactivity$reactivity=(NAI_N3_1_reactivity$Mod_percentage-mean(NAI_N3_1_reactivity$Mod_percentage))/sd(NAI_N3_1_reactivity$Mod_percentage)
head(NAI_N3_1_reactivity)

# use the denature time to normolize.
# use the denature time to normolize.
NAI_N3_1_time=merge(NAI_N3_1_time,denature_time_nor_table,by.x="position",by.y="position_time_nor_p")

## time for zscore.
### NAI_N3_1_time time
NAI_N3_1_time$time_nor_denature=NAI_N3_1_time$time-NAI_N3_1_time$denature_time_nor_v
NAI_N3_1_time$time_nor_zs=(NAI_N3_1_time$time_nor_denature-mean(NAI_N3_1_time$time_nor_denature))/sd(NAI_N3_1_time$time_nor_denature)


NAI_N3_1_time$timezs=(NAI_N3_1_time$time-mean(NAI_N3_1_time$time))/sd(NAI_N3_1_time$time)
NAI_N3_1_time$timezs_trans=(mean(NAI_N3_1_time$time)-NAI_N3_1_time$time)/sd(NAI_N3_1_time$time)
head(NAI_N3_1_time)

## NAI_N3_2 reactivity
NAI_N3_2_reactivity$reactivity=(NAI_N3_2_reactivity$Mod_percentage-mean(NAI_N3_2_reactivity$Mod_percentage))/sd(NAI_N3_2_reactivity$Mod_percentage)
head(NAI_N3_2_reactivity)

## time for zscore.
### NAI_N3_2_time
NAI_N3_2_time=merge(NAI_N3_2_time,denature_time_nor_table,by.x="position",by.y="position_time_nor_p")

## time for zscore.
### NAI_N3_1_time time
NAI_N3_2_time$time_nor_denature=NAI_N3_2_time$time-NAI_N3_2_time$denature_time_nor_v
NAI_N3_2_time$time_nor_zs=(NAI_N3_2_time$time_nor_denature-mean(NAI_N3_2_time$time_nor_denature))/sd(NAI_N3_2_time$time_nor_denature)


NAI_N3_2_time$timezs=(NAI_N3_2_time$time-mean(NAI_N3_2_time$time))/sd(NAI_N3_2_time$time)
NAI_N3_2_time$timezs_trans=(mean(NAI_N3_2_time$time)-NAI_N3_2_time$time)/sd(NAI_N3_2_time$time)
head(NAI_N3_2_time)

## NAI_N3_25 for reactivity
NAI_N3_25_reactivity$reactivity=(NAI_N3_25_reactivity$Mod_percentage-mean(NAI_N3_25_reactivity$Mod_percentage))/sd(NAI_N3_25_reactivity$Mod_percentage)
head(NAI_N3_25_reactivity)
## time for zscore.
### NAI_N3_25_time
NAI_N3_25_time=merge(NAI_N3_25_time,denature_time_nor_table,by.x="position",by.y="position_time_nor_p")

## time for zscore.
### NAI_N3_25_time time
NAI_N3_25_time$time_nor_denature=NAI_N3_25_time$time-NAI_N3_25_time$denature_time_nor_v
NAI_N3_25_time$time_nor_zs=(NAI_N3_25_time$time_nor_denature-mean(NAI_N3_25_time$time_nor_denature))/sd(NAI_N3_25_time$time_nor_denature)

NAI_N3_25_time$timezs=(NAI_N3_25_time$time-mean(NAI_N3_25_time$time))/sd(NAI_N3_25_time$time)
NAI_N3_25_time$timezs_trans=(mean(NAI_N3_25_time$time)-NAI_N3_25_time$time)/sd(NAI_N3_25_time$time)
head(NAI_N3_25_time)

## denature for reactivity
denature_reactivity$reactivity=(denature_reactivity$Mod_percentage-mean(denature_reactivity$Mod_percentage))/sd(denature_reactivity$Mod_percentage)
head(denature_reactivity)
## time for zscore.
### denature time
denature_time$timezs=(denature_time$time-mean(denature_time$time))/sd(denature_time$time)
denature_time$timezs_trans=(mean(denature_time$time)-denature_time$time)/sd(denature_time$time)
head(denature_time)

# plot 1-100 reactivity
head(denature_reactivity)
plot_data=denature_reactivity
plot_data=unmod2_reactivity
plot_data=NAI_N3_1_reactivity
plot_data1_100=plot_data[plot_data$Position<100,]
head(plot_data1_100)
## get data
ggplot(plot_data1_100,aes(x=Position,y=reactivity))+ geom_line()+ scale_x_continuous(breaks = seq(0, 425, by = 10))


# use denature data to remove noise.
dim(denature_time)
dim(unmod2_time)
head(unmod2_time)
## use denature time zs to unmod2 time zs.
unmod2_time$timezs2denaturezs=unmod2_time$timezs/denature_time$timezs

head(unmod2_time)
ggplot(unmod2_time,aes(x=position,y=time_nor_zs))+ geom_line()+ scale_x_continuous(breaks = seq(0, 425, by = 10))

## use denature time zs to NAI_N3_1_time zs.
NAI_N3_1_time$timezs2denaturezs=NAI_N3_1_time$timezs/denature_time$timezs
head(NAI_N3_1_time)
ggplot(NAI_N3_1_time,aes(x=position,y=timezs2denaturezs))+ geom_line()+ scale_x_continuous(breaks = seq(0, 425, by = 10))


ggplot(NAI_N3_1_time,aes(x=position,y=time_nor_zs))+ geom_line()+ scale_x_continuous(breaks = seq(0, 425, by = 10))
head(NAI_N3_1_reactivity)
ggplot(NAI_N3_1_reactivity,aes(x=Position,y=reactivity))+ geom_line()+ scale_x_continuous(breaks = seq(0, 425, by = 10))


## use denature time zs to NAI_N3_1_time 2 time zs.
NAI_N3_2_time$timezs2denaturezs=NAI_N3_2_time$timezs/denature_time$timezs
head(NAI_N3_2_time)
ggplot(NAI_N3_2_time,aes(x=position,y=timezs2denaturezs))+ geom_line()+ scale_x_continuous(breaks = seq(0, 425, by = 10))

ggplot(NAI_N3_2_time,aes(x=position,y=time_nor_zs))+ geom_line()+ scale_x_continuous(breaks = seq(0, 425, by = 10))

## use denature time zs to NAI_N3_1_time 25 time zs.
NAI_N3_25_time$timezs2denaturezs=NAI_N3_25_time$timezs/denature_time$timezs
head(NAI_N3_25_time)
ggplot(NAI_N3_25_time,aes(x=position,y=timezs2denaturezs))+ geom_line()+ scale_x_continuous(breaks = seq(0, 425, by = 10))

ggplot(NAI_N3_25_time,aes(x=position,y=time_nor_zs))+ geom_line()+ scale_x_continuous(breaks = seq(0, 425, by = 10))

png("wide.png", width = 1e5, height = 500)

plot((sin(1:10000/100)+rnorm(10000)/5),type='l')

## use denature time zs to denature_time 25 time zs.
denature_time$timezs2denaturezs=denature_time$timezs/denature_time$timezs
head(denature_time)
ggplot(denature_time,aes(x=position,y=timezs2denaturezs))+ geom_line()+ scale_x_continuous(breaks = seq(0, 425, by = 10))


# use time substract denature time.
## unmod time.
unmod2_time$timeddenature=unmod2_time$time-denature_time$time
ggplot(unmod2_time,aes(x=position,y=timeddenature))+ geom_line()+ scale_x_continuous(breaks = seq(0, 425, by = 10))

## NAI_N3_1_time
NAI_N3_1_time$timeddenature=NAI_N3_1_time$time-denature_time$time
head(NAI_N3_1_time)
ggplot(NAI_N3_1_time,aes(x=position,y=timeddenature))+ geom_line()+ scale_x_continuous(breaks = seq(0, 425, by = 10))

## NAI_N3_2_time
NAI_N3_2_time$timeddenature=NAI_N3_2_time$time-denature_time$time
head(NAI_N3_2_time)
ggplot(NAI_N3_2_time,aes(x=position,y=timeddenature))+ geom_line()+ scale_x_continuous(breaks = seq(0, 425, by = 10))

## NAINAI_N3_25_time
NAI_N3_25_time$timeddenature=NAI_N3_25_time$time-denature_time$time
head(NAI_N3_25_time)
ggplot(NAI_N3_25_time,aes(x=position,y=timeddenature))+ geom_line()+ scale_x_continuous(breaks = seq(0, 425, by = 10))

## 
denature_time$timeddenature=denature_time$time-denature_time$time
head(denature_time)
ggplot(denature_time,aes(x=position,y=timeddenature))+ geom_line()+ scale_x_continuous(breaks = seq(0, 425, by = 10))


#plot time by atcg.
head(denature_time)
head(NAI_N3_1_reactivity)
denature_time$position2=denature_time$position
denature_time2=merge(NAI_N3_1_reactivity,denature_time,by.x="Position",by.y="position2")
head(denature_time2)
write.csv(denature_time2,'/Users/xugang/Desktop/sequencing_center_desktop/2020/rna_structure/data_plot/denature_time2.csv')

ggplot(denature_time2, aes(x=NT, y=timezs)) + 
  geom_boxplot()+theme(axis.text=element_text(size=20),
                       axis.title=element_text(size=22,face="bold"))
denature_time2_merge=read.csv('/Users/xugang/Desktop/sequencing_center_desktop/2020/rna_structure/data_plot/denature_time2_2N.csv')
head(denature_time2_merge)
denature_time2_merge=denature_time2_merge[-1,]
head(denature_time2_merge)
ggplot(denature_time2_merge, aes(x=NT, y=timezs)) + 
  geom_boxplot()+theme(axis.text=element_text(size=20),
                       axis.title=element_text(size=22,face="bold"))
# plot for visualization.
## 1. merge data
### 1.1 generate new position. 
unmod2_time$position2=unmod2_time$position-20
head(unmod2_time)

### 1.2 merge data by postion.
data_plot=merge(NAI_N3_1_reactivity,unmod2_time,by.x="Position",by.y="position2")
head(data_plot)
cor(data_plot$reactivity,data_plot$time_nor_zs)

cor(data_plot$reactivity,((data_plot$time_nor_zs+3)*2))
plot(data_plot$reactivity,data_plot$time_nor_zs)
data_plot_filter=data_plot[data_plot$time_nor_zs < -0.8,]
data_plot_filter
plot(data_plot_filter$reactivity,data_plot_filter$time_nor_zs)

## 2. plot data
### 2.1 time zscore.
#### 2.1.1  whole plot
plot(data_plot$Position, data_plot$time_nor_zs, pch = 16, col = 2, type = "l",ylim=c(-3,3) )   
# Basic line plot in R) 
axis(1, at=seq(0,420,10))
par(new = TRUE)                             # Add new plot
# Create second plot without axes
plot(data_plot$Position, data_plot$reactivity , pch = 17, col = 3,              
     axes = F, xlab = "", ylab = "",type="l",ylim=c(-0.5,8))
axis(side = 4, at = pretty(range(data_plot$reactivity )))      # Add second axis
mtext("timezx", side = 4, line = 3)             # Add second axis label
dev.off()
#### 2.1.2  partial plot
startn=300
endn=400
plot(data_plot$Position[startn:endn], data_plot$timezs2denaturezs[startn:endn], pch = 16, col = 2, type = "l",ylim=c(-10,10))   # Basic line plot in R) 
par(new = TRUE)                             # Add new plot
# Create second plot without axes
plot(data_plot$Position[startn:endn], data_plot$reactivity[startn:endn] , pch = 17, col = 3,              
     axes = F, xlab = "", ylab = "",type="l",ylim=c(-0.5,4))
axis(side = 4, at = pretty(range(data_plot$reactivity[startn:endn])))      # Add second axis
mtext("timezx", side = 4, line = 3)             # Add second axis label
dev.off()





#
# calculate correlation from move step bp.
iv=c()
cordv=c()
for (i in -40:40) {
  unmod2_time$position2=unmod2_time$position - i
  head(unmod2_time)
  data_plot=merge(NAI_N3_1_reactivity,unmod2_time,by.x="Position",by.y="position2")
  head(data_plot)
  cord=cor(data_plot$reactivity,data_plot$timezs2denaturezs)
  print(i)
  print(cord)
  iv=append(iv,i)
  cordv=append(cordv,cord)
}
iv
cordv
distable=data.frame(iv,cordv)
ggplot(distable,aes(x=iv,y=cordv))+ geom_line() + scale_x_continuous(breaks = seq(-40, 40, by = 2))

# calculate correlation from move step bp.
iv=c()
cordv=c()
for (i in -40:40) {
  NAI_N3_2_time$position2=NAI_N3_2_time$position - i
  head(NAI_N3_2_time)
  data_plot=merge(NAI_N3_2_reactivity,NAI_N3_2_time,by.x="Position",by.y="position2")
  head(data_plot)
  cord=cor(data_plot$reactivity,data_plot$timezs_trans)
  print(i)
  print(cord)
  iv=append(iv,i)
  cordv=append(cordv,cord)
}
iv
cordv
distable=data.frame(iv,cordv)
ggplot(distable,aes(x=iv,y=cordv))+ geom_line() + scale_x_continuous(breaks = seq(-40, 40, by = 2))

# calculate correlation from move step bp.
iv=c()
cordv=c()
for (i in -40:40) {
  NAI_N3_25_time$position2=NAI_N3_25_time$position - i
  head(NAI_N3_25_time)
  data_plot=merge(NAI_N3_25_reactivity,NAI_N3_25_time,by.x="Position",by.y="position2")
  head(data_plot)
  cord=cor(data_plot$reactivity,data_plot$timezs_trans)
  print(i)
  print(cord)
  iv=append(iv,i)
  cordv=append(cordv,cord)
}
iv
cordv
distable=data.frame(iv,cordv)
ggplot(distable,aes(x=iv,y=cordv))+ geom_line() + scale_x_continuous(breaks = seq(-40, 40, by = 2))


# plot individual.
head(unmod2_reactivity)
ggplot(unmod2_reactivity,aes(x=Position,y=reactivity))+ geom_line()+ scale_x_continuous(breaks = seq(0, 425, by = 10))
head(unmod2_time)
ggplot(unmod2_time,aes(x=position,y=timezs))+ geom_line()+ scale_x_continuous(breaks = seq(0, 425, by = 10))
head(NAI_N3_1_reactivity)
ggplot(NAI_N3_1_reactivity,aes(x=Position,y=reactivity))+ geom_line()+ scale_x_continuous(breaks = seq(0, 425, by = 10))
head(NAI_N3_1_time)
ggplot(NAI_N3_1_time,aes(x=position,y=timezs))+ geom_line()+ scale_x_continuous(breaks = seq(0, 425, by = 10))
head(NAI_N3_2_reactivity)
ggplot(NAI_N3_2_reactivity,aes(x=Position,y=reactivity))+ geom_line()+ scale_x_continuous(breaks = seq(0, 425, by = 10))
head(NAI_N3_2_time)
ggplot(NAI_N3_2_time,aes(x=position,y=timezs))+ geom_line()+ scale_x_continuous(breaks = seq(0, 425, by = 10))
head(NAI_N3_25_reactivity)
ggplot(NAI_N3_25_reactivity,aes(x=Position,y=reactivity))+ geom_line()+ scale_x_continuous(breaks = seq(0, 425, by = 10))
head(NAI_N3_25_time)
ggplot(NAI_N3_25_time,aes(x=position,y=timezs))+ geom_line()+ scale_x_continuous(breaks = seq(0, 425, by = 10))
head(denature_reactivity)
ggplot(denature_reactivity,aes(x=Position,y=reactivity))+ geom_line()+ scale_x_continuous(breaks = seq(0, 425, by = 10))
head(denature_time)
ggplot(denature_time,aes(x=position,y=timezs))+ geom_line()+ scale_x_continuous(breaks = seq(0, 425, by = 10))

#

# plot for visualization.
## 1. merge data
### 1.1 generate new position. 
NAI_N3_1_time$position2=NAI_N3_1_time$position +25
head(NAI_N3_1_time)

### 1.2 merge data by postion.
data_plot=merge(NAI_N3_1_reactivity,NAI_N3_1_time,by.x="Position",by.y="position2")
head(data_plot)
cor(data_plot$reactivity,data_plot$timezs_trans)
## 2. plot data
### 2.1 time zscore.
#### 2.1.1  whole plot
plot(data_plot$Position, data_plot$timezs, pch = 16, col = 2, type = "l")   # Basic line plot in R) 
par(new = TRUE)                             # Add new plot
# Create second plot without axes
plot(data_plot$Position, data_plot$reactivity , pch = 17, col = 3,              
     axes = F, xlab = "", ylab = "",type="l")
axis(side = 4, at = pretty(range(data_plot$reactivity )))      # Add second axis
mtext("timezx", side = 4, line = 3)             # Add second axis label
dev.off()
#### 2.1.2  partial plot
startn=0
endn=100
plot(data_plot$Position[startn:endn], data_plot$timezs[startn:endn], pch = 16, col = 2, type = "l")   # Basic line plot in R) 
par(new = TRUE)                             # Add new plot
# Create second plot without axes
plot(data_plot$Position[startn:endn], data_plot$reactivity[startn:endn] , pch = 17, col = 3,              
     axes = F, xlab = "", ylab = "",type="l")
axis(side = 4, at = pretty(range(data_plot$reactivity[startn:endn])))      # Add second axis
mtext("timezx", side = 4, line = 3)             # Add second axis label
dev.off()

#### 2.2 reverse time zscore.

plot(data_plot$Position, data_plot$timezs_trans, pch = 16, col = 2, type = "l")   # Basic line plot in R) 
par(new = TRUE)                             # Add new plot
# Create second plot without axes
plot(data_plot$Position, data_plot$reactivity , pch = 17, col = 3,              
     axes = F, xlab = "", ylab = "",type="l")
axis(side = 4, at = pretty(range(data_plot$reactivity )))      # Add second axis
mtext("timezx", side = 4, line = 3)             # Add second axis label
dev.off()
#### 2.2.2  partial plot
startn=300
endn=400
plot(data_plot$Position[startn:endn], data_plot$timezs_trans[startn:endn], pch = 16, col = 2, type = "l")   # Basic line plot in R) 
par(new = TRUE)                             # Add new plot
# Create second plot without axes
plot(data_plot$Position[startn:endn], data_plot$reactivity[startn:endn] , pch = 17, col = 3,              
     axes = F, xlab = "", ylab = "",type="l")
axis(side = 4, at = pretty(range(data_plot$reactivity[startn:endn])))      # Add second axis
mtext("timezx", side = 4, line = 3)             # Add second axis label
dev.off()





```

* a3.statistic.R

```r
library('ggplot2')
dir='/Users/xugang/Desktop/sequencing_center_desktop/2020/rna_structure/a2-human/readcount/'
dir2='/Users/xugang/Desktop/sequencing_center_desktop/2020/rna_structure/a2-human/structure_analysis/'
######################################################################
#all transcripts
######################################################################
dtime_mfe=read.table(paste(dir2,"o1.faj5273.time.mfe.all.12.tsv",sep=''),header = T)
dtime_mfe=read.table(paste(dir2,"o1.faj5033.time.mfe.all.12.tsv",sep=''),header = T)
dtime_mfe=read.table(paste(dir2,"o1.faj05033.time.mfe.all.95th.12.tsv",sep=''),header = T)
dtime_mfe=read.table(paste(dir2,"o1.faj5272.time.mfe.all.12.tsv",sep=''),header = T)
dtime_mfe=read.table(paste(dir2,"o1.faj5272_2.time.mfe.all.12.tsv",sep=''),header = T)

# if the
colnamesv=c("name","pos","time","current","count","mfe","bin","entropy","gc")
colnames(dtime_mfe)=colnamesv
head(dtime_mfe)
###############################
species='human'
id='5033'
modify='unmodify'
bin=12
###############################
mean_count=mean(dtime_mfe$count)
mean_entropy=mean(dtime_mfe$entropy,na.rm=T)
mean_mfe=mean(dtime_mfe$mfe,na.rm=T)
mean_count
mean_entropy
mean_mfe
##########################
namev='ENST00000252486'
namev='ENST00000019317'
select_mfe=dtime_mfe[dtime_mfe$name==namev,]
head(select_mfe)
#dtime_mfe2=select_mfe[select_mfe$mfe<mean_mfe & select_mfe$count > mean_count & select_mfe$entropy > mean_entropy,]
#da4=da3[da3$mfe < mean(da3$mfe)*2 & da3$entropy > mean(da3$entropy)*1.11,]


###############################


##########################

#
dtime_mfe2=dtime_mfe[dtime_mfe$name==namev,]
dtime_mfe2=dtime_mfe[dtime_mfe$name==namev,]

da1=data.frame(dtime_mfe2$pos,dtime_mfe2$time*10000,dtime_mfe2$count)
colnames(da1)=c('pos','time10000','count')
da2=data.frame(dtime_mfe2$pos+bin+12-1,dtime_mfe2$mfe,dtime_mfe2$entropy)
colnames(da2)=c('pos','mfe','entropy')
da3=merge(da1,da2,by='pos',all=F)
head(da3)
####################
da4=da3[da3$mfe < mean(da3$mfe)*2 & da3$entropy > mean(da3$entropy)*1.11,]
da4=da3[da3$mfe < -1.031 & da3$entropy > 2.154 &da3$count>mean_count,]
#da4=da3[da3$mfe < mean(da3$mfe) & da3$entropy > mean(da3$entropy),]
#da4=da3[da3$entropy > mean(da3$entropy)*1.11,]

corv=cor(da4$time10000,da4$mfe*100)

ggplot(da4,aes(mfe,time10000 ,color=entropy))+geom_point(size=1)+ 
  ggtitle(paste("TR: ",namev, species,id, modify, "bin:", bin,"\nCor:", corv, sep=' ' )) +
  scale_color_continuous(low="red", high="black")+theme(plot.title = element_text( size = 24),
                                                        axis.text.x = element_text( size = 20),
                                                        axis.text.y = element_text( size = 20),  
                                                        axis.title.x = element_text( size = 20),
                                                        axis.title.y = element_text( size = 20))+
  geom_smooth(method='lm')

###################################################################
#loop 1



#namev='ENST00000252486'
#namev='ENST00000253047'


name_fac=factor(dtime_mfe$name)
name_arr=levels(name_fac)
cortable=data.frame(name=c(),cor=c())
for (i in name_arr){
  print(i)
  namev=i
  dtime_mfe2=dtime_mfe[dtime_mfe$name==namev,]
  da1=data.frame(dtime_mfe2$pos,dtime_mfe2$time*10000,dtime_mfe2$count)
  colnames(da1)=c('pos','time10000','count')
  da2=data.frame(dtime_mfe2$pos+bin+12-1,dtime_mfe2$mfe,dtime_mfe2$entropy)
  colnames(da2)=c('pos','mfe','entropy')
  da3=merge(da1,da2,by='pos',all=F)
  da4=da3[da3$mfe < -1.031 & da3$entropy > 2.154 &da3$count>mean_count,]
  corv=cor(da4$time10000,da4$mfe*100)
  line1=data.frame(name=namev,cor=corv)
  #line1 = c(namev,corv)
  #print(line1)
  cortable=rbind(cortable,line1)
  #cortable
}

cortable

density(cortable$cor,na.rm = T)
boxplot(cortable$cor)

ggplot(cortable,aes(x=cor))+geom_density(color="darkblue", fill="lightblue")+
  ggtitle(paste( species,id, modify, "bin:", bin,"\n Density of Cor:", sep=' ' )) +
  theme(plot.title = element_text( size = 24),
  axis.text.x = element_text( size = 20),
  axis.text.y = element_text( size = 20),  
  axis.title.x = element_text( size = 20),
  axis.title.y = element_text( size = 20))

write.csv(cortable,paste(dir2,"o1.faj5033.allcor.95th.csv",sep=''))
write.csv(cortable,paste(dir2,"o1.faj5033.allcor.csv",sep=''))
write.csv(cortable,paste(dir2,"o1.faj5273.allcor.csv",sep=''))
write.csv(cortable,paste(dir2,"o1.faj5272.allcor.csv",sep=''))
write.csv(cortable,paste(dir2,"o1.faj5272_2.allcor.csv",sep=''))

cortable=read.csv(paste(dir2,"o1.faj5033.allcor.csv",sep=''))
cortable=read.csv(paste(dir2,"o1.faj5273.allcor.csv",sep=''))
cortable=read.csv(paste(dir2,"o1.faj5272.allcor.csv",sep=''))
cortable=read.csv(paste(dir2,"o1.faj5272_2.allcor.csv",sep=''))


head(cortable,n=12)

id='5033'
modify='unmodify'
ggplot(cortable,aes(x=cor))+geom_density(color="darkblue", fill="lightblue")+
  ggtitle(paste( species,id, modify, "bin:", bin,"\n Density of Cor:", sep=' ' )) +
  theme(plot.title = element_text( size = 24),
        axis.text.x = element_text( size = 20),
        axis.text.y = element_text( size = 20),  
        axis.title.x = element_text( size = 20),
        axis.title.y = element_text( size = 20))

na.omit(cortable[cortable$cor < -0.3,])
na.omit(cortable[cortable$cor > 0.3,])

namev='ENST00000293217'
namev='ENST00000023939'
namev='ENST00000215793'
namev='ENST00000216727'
namev='ENST00000465692'
namev='ENST00000457894'
namev='ENST00000337288'
namev='ENST00000329433'
namev='ENST00000330188'
namev='ENST00000340852'
namev='ENST00000278856'
namev='ENST00000279389'
namev='ENST00000278826'
namev='ENST00000279068'
namev='ENST00000278572'
namev='ENST00000278224'
namev='ENST00000296873'

dtime_mfe2=dtime_mfe[dtime_mfe$name==namev,]

da1=data.frame(dtime_mfe2$pos,dtime_mfe2$time*10000,dtime_mfe2$count)
colnames(da1)=c('pos','time10000','count')
da2=data.frame(dtime_mfe2$pos+bin+12-1,dtime_mfe2$mfe,dtime_mfe2$entropy)
colnames(da2)=c('pos','mfe','entropy')
da3=merge(da1,da2,by='pos',all=F)
head(da3)
####################
#da4=da3[da3$mfe < mean(da3$mfe)*2 & da3$entropy > mean(da3$entropy)*1.11,]
da4=da3[da3$mfe < -1.031 & da3$entropy > 2.154 &da3$count>mean_count,]

corv=cor(da4$time10000,da4$mfe*100)

ggplot(da4,aes(mfe,time10000 ,color=entropy))+geom_point(size=1)+ 
  ggtitle(paste("TR: ",namev, species,"\n",id, modify, "bin:", bin,"\nCor:", corv, sep=' ' )) +
  scale_color_continuous(low="red", high="black")+theme(plot.title = element_text( size = 20),
                                                        axis.text.x = element_text( size = 20),
                                                        axis.text.y = element_text( size = 20),  
                                                        axis.title.x = element_text( size = 20),
                                                        axis.title.y = element_text( size = 20))+
  geom_smooth(method='lm')
```
* a1.read1.R
```r
#library(data.table)
library(ggplot2)
suppressMessages(library(dplyr))
library("data.table") 

# read rawdata.
dir='/Users/xugang/Desktop/sequencing_center_desktop/2020/rna_structure/a1-emzye/'
## load reactivity score.
unmod2_reactivity   =read.table(paste(dir,"/a5-reactivity/Output_TeTrahymenaTetra_unmod2.combined.RData.csv",sep=''),sep=",",header=T)
NAI_N3_1_reactivity =read.table(paste(dir,'/a5-reactivity/Output_TeTrahymenaTetra_NAI_N3_1.combined.RData.csv',sep=''),sep=",",header=T)
NAI_N3_2_reactivity =read.table(paste(dir,'/a5-reactivity/Output_TeTrahymenaTetra_NAI_N3_2.combined.RData.csv',sep=''),sep=",",header=T)
NAI_N3_25_reactivity=read.table(paste(dir,'/a5-reactivity/Output_TeTrahymenaTetra_NAI_N3_25min.combined.RData.csv',sep=''),sep=",",header=T)
denature_reactivity =read.table(paste(dir,'/a5-reactivity/Output_TeTrahymenaTetra_NAI_N3_denatured.combined.RData.csv',sep=''),sep=",",header=T)
head(unmod2_reactivity)
head(NAI_N3_1_reactivity)
head(NAI_N3_2_reactivity)
head(NAI_N3_25_reactivity)
head(denature_reactivity)

# read svm machine learning.
svmd=read.table(paste(dir,'/a9-ml/ml.csv',sep=''),sep=",",header = T)
head(svmd)

## plot for test.
ggplot(svmd,aes(x=pos,y=timezs))+ geom_line()+ scale_x_continuous(breaks = seq(0, 425, by = 10)) +  coord_cartesian(ylim=c(-0.1, 2)) 



## time.
unmod2_time   =read.table(paste(dir,"/a6-time/Tetra_unmod2.time.tsv",sep=''),sep="\t",header=T)
NAI_N3_1_time =read.table(paste(dir,"/a6-time/Tetra_NAI_N3_1.time.tsv",sep=''),sep="\t",header=T)
NAI_N3_2_time =read.table(paste(dir,"/a6-time/Tetra_NAI_N3_2.time.tsv",sep=''),sep="\t",header=T)
NAI_N3_25_time=read.table(paste(dir,"/a6-time/Tetra_NAI_N3_25min.time.tsv",sep=''),sep="\t",header=T)
denature_time =read.table(paste(dir,"/a6-time/Tetra_NAI_N3_denatured.time.tsv",sep=''),sep="\t",header=T)
head(unmod2_time)
head(NAI_N3_1_time)
head(NAI_N3_2_time)
head(NAI_N3_25_time)
head(denature_time)
## load read each position 
unmod2_mod_time   =read.table(paste(dir,"/a7-mod-unmod-time/Tetra_unmod2.mod.tsv",sep=''),sep="\t",header=F)
unmod2_unmod_time =read.table(paste(dir,"/a7-mod-unmod-time/Tetra_unmod2.unmod.tsv",sep=''),sep="\t",header=F)
NAI_N3_1_mod_time   =read.table(paste(dir,"/a7-mod-unmod-time/Tetra_NAI_N3_1.mod.tsv",sep=''),sep="\t",header=F)
NAI_N3_1_unmod_time =read.table(paste(dir,"/a7-mod-unmod-time/Tetra_NAI_N3_1.unmod.tsv",sep=''),sep="\t",header=F)
NAI_N3_2_mod_time   =read.table(paste(dir,"/a7-mod-unmod-time/Tetra_NAI_N3_2.mod.tsv",sep=''),sep="\t",header=F)
NAI_N3_2_unmod_time =read.table(paste(dir,"/a7-mod-unmod-time/Tetra_NAI_N3_2.unmod.tsv",sep=''),sep="\t",header=F)
NAI_N3_25_mod_time   =read.table(paste(dir,"/a7-mod-unmod-time/Tetra_NAI_N3_25min.mod.tsv",sep=''),sep="\t",header=F)
NAI_N3_25_unmod_time =read.table(paste(dir,"/a7-mod-unmod-time/Tetra_NAI_N3_25min.unmod.tsv",sep=''),sep="\t",header=F)
denature_mod_time   =read.table(paste(dir,"/a7-mod-unmod-time/Tetra_NAI_N3_denatured.mod.tsv",sep=''),sep="\t",header=F)
denature_unmod_time =read.table(paste(dir,"/a7-mod-unmod-time/Tetra_NAI_N3_denatured.unmod.tsv",sep=''),sep="\t",header=F)
head(denature_unmod_time)
head(denature_mod_time)

## load position and time.
unmod2_mod_pt   =read.table(paste(dir,"/a8-position-time/Tetra_unmod2.mod.pos.tsv",sep=''),sep="\t",header=F)
unmod2_unmod_pt =read.table(paste(dir,"/a8-position-time/Tetra_unmod2.unmod.pos.tsv",sep=''),sep="\t",header=F)
NAI_N3_1_mod_pt   =read.table(paste(dir,"/a8-position-time/Tetra_NAI_N3_1.mod.pos.tsv",sep=''),sep="\t",header=F)
NAI_N3_1_unmod_pt =read.table(paste(dir,"/a8-position-time/Tetra_NAI_N3_1.unmod.pos.tsv",sep=''),sep="\t",header=F)
NAI_N3_2_mod_pt   =read.table(paste(dir,"/a8-position-time/Tetra_NAI_N3_2.mod.pos.tsv",sep=''),sep="\t",header=F)
NAI_N3_2_unmod_pt =read.table(paste(dir,"/a8-position-time/Tetra_NAI_N3_2.unmod.pos.tsv",sep=''),sep="\t",header=F)
NAI_N3_25_mod_pt   =read.table(paste(dir,"/a8-position-time/Tetra_NAI_N3_25min.mod.pos.tsv",sep=''),sep="\t",header=F)
NAI_N3_25_unmod_pt =read.table(paste(dir,"/a8-position-time/Tetra_NAI_N3_25min.unmod.pos.tsv",sep=''),sep="\t",header=F)
denature_mod_pt   =read.table(paste(dir,"/a8-position-time/Tetra_NAI_N3_denatured.mod.pos.tsv",sep=''),sep="\t",header=F)
denature_unmod_pt =read.table(paste(dir,"/a8-position-time/Tetra_NAI_N3_denatured.unmod.pos.tsv",sep=''),sep="\t",header=F)

colnames(unmod2_mod_pt)=c('position','time','kmer')
colnames(unmod2_unmod_pt)=c('position','time','kmer')
colnames(NAI_N3_1_mod_pt)=c('position','time','kmer')
colnames(NAI_N3_1_unmod_pt)=c('position','time','kmer')
colnames(NAI_N3_2_mod_pt)=c('position','time','kmer')
colnames(NAI_N3_2_unmod_pt)=c('position','time','kmer')
colnames(NAI_N3_25_mod_pt)=c('position','time','kmer')
colnames(NAI_N3_25_unmod_pt)=c('position','time','kmer')
colnames(denature_mod_pt)=c('position','time','kmer')
colnames(denature_unmod_pt)=c('position','time','kmer')


head(unmod2_mod_pt)
dim(unmod2_mod_pt)
dim(unmod2_unmod_pt)
##
#datapro=unmod2_mod_pt
#datapro=unmod2_unmod_pt
#datapro=NAI_N3_1_mod_pt
#datapro=NAI_N3_1_unmod_pt
#datapro=NAI_N3_2_mod_pt
#datapro=NAI_N3_2_unmod_pt
#datapro=NAI_N3_25_mod_pt
datapro=NAI_N3_25_unmod_pt
#datapro=denature_mod_pt
#datapro=denature_unmod_pt
  
dataA=datapro[datapro$kmer %like% "^A", ]
dataT=datapro[datapro$kmer %like% "^T", ]
dataC=datapro[datapro$kmer %like% "^C", ]
dataG=datapro[datapro$kmer %like% "^G", ]

staA=dim(dataA)
staT=dim(dataT)
staC=dim(dataC)
staG=dim(dataG)
ggplot(dataA, aes(x=kmer, y=time)) + labs(title=paste("A: ", staA[1],sep=""))+
  geom_boxplot(outlier.shape = NA)+theme(axis.text=element_text(size=9,angle=90),
                       axis.title=element_text(size=12,face="bold"))+  coord_cartesian(ylim=c(0, 100)) 
ggplot(dataT, aes(x=kmer, y=time)) +  labs(title=paste("T: ", staT[1],sep=""))+
  geom_boxplot(outlier.shape = NA)+theme(axis.text=element_text(size=9,angle=90),
                       axis.title=element_text(size=12,face="bold"))+  coord_cartesian(ylim=c(0, 100))
ggplot(dataC, aes(x=kmer, y=time)) +  labs(title=paste("C: ", staC[1],sep=""))+
  geom_boxplot(outlier.shape = NA)+theme(axis.text=element_text(size=9,angle=90),
                       axis.title=element_text(size=12,face="bold"))+  coord_cartesian(ylim=c(0, 100))
ggplot(dataG, aes(x=kmer, y=time)) +  labs(title=paste("G: ", staG[1],sep=""))+
  geom_boxplot(outlier.shape = NA)+theme(axis.text=element_text(size=9,angle=90),
                       axis.title=element_text(size=12,face="bold"))+  coord_cartesian(ylim=c(0, 100))
## plot time for kmer.
ggplot(unmod2_mod_pt, aes(x=kmer, y=time)) + 
  geom_boxplot()+theme(axis.text=element_text(size=20),
                       axis.title=element_text(size=22,face="bold"))


# get reactivity score.

unmod2_reactivity=unmod2_reactivity %>% mutate(reactivity=(  (Mod_percentage-mean(Mod_percentage))/sd(Mod_percentage)  ) )
NAI_N3_1_reactivity =NAI_N3_1_reactivity%>% mutate(reactivity=(  (Mod_percentage-mean(Mod_percentage))/sd(Mod_percentage)  ) )
NAI_N3_2_reactivity =NAI_N3_2_reactivity%>% mutate(reactivity=(  (Mod_percentage-mean(Mod_percentage))/sd(Mod_percentage)  ) )
NAI_N3_25_reactivity=NAI_N3_25_reactivity%>% mutate(reactivity=(  (Mod_percentage-mean(Mod_percentage))/sd(Mod_percentage)  ) )
denature_reactivity =denature_reactivity%>% mutate(reactivity=(  (Mod_percentage-mean(Mod_percentage))/sd(Mod_percentage)  ) )

head(unmod2_reactivity)
# plot for test.
ggplot(unmod2_reactivity,aes(x=Position,y=reactivity))+ geom_line()+ scale_x_continuous(breaks = seq(0, 425, by = 10))
ggplot(NAI_N3_1_reactivity,aes(x=Position,y=reactivity))+ geom_line()+ scale_x_continuous(breaks = seq(0, 425, by = 10))

head(NAI_N3_1_reactivity)
write.table(NAI_N3_1_reactivity,paste(dir,"/a5-reactivity/NAI_N3_1_reactivity.tsv",sep=''),sep="\t")

# get time zs
unmod2_time   =unmod2_time   %>% mutate(timezs=(  (time-mean(time))/sd(time)  ) )
NAI_N3_1_time =NAI_N3_1_time %>% mutate(timezs=(  (time-mean(time))/sd(time)  ) )
NAI_N3_2_time =NAI_N3_2_time %>% mutate(timezs=(  (time-mean(time))/sd(time)  ) )
NAI_N3_25_time=NAI_N3_25_time%>% mutate(timezs=(  (time-mean(time))/sd(time)  ) )
denature_time =denature_time %>% mutate(timezs=(  (time-mean(time))/sd(time)  ) )

head(unmod2_time)

#plot for time
ggplot(unmod2_time,aes(x=position,y=timezs))+ geom_line()+ scale_x_continuous(breaks = seq(0, 425, by = 10))
ggplot(NAI_N3_1_time,aes(x=position,y=timezs))+ geom_line()+ scale_x_continuous(breaks = seq(0, 425, by = 10))
ggplot(NAI_N3_2_time,aes(x=position,y=timezs))+ geom_line()+ scale_x_continuous(breaks = seq(0, 425, by = 10))
ggplot(NAI_N3_25_time,aes(x=position,y=timezs))+ geom_line()+ scale_x_continuous(breaks = seq(0, 425, by = 10))
ggplot(denature_time,aes(x=position,y=timezs))+ geom_line()+ scale_x_continuous(breaks = seq(0, 425, by = 10))

# calculate background time.
denature_unmod_time_v=colMeans(denature_unmod_time[,2:418],na.rm=T)
denature_unmod_time_v
denature_unmod_time_p=seq(0,416,by = 1)
denature_unmod_time_table=data.frame(denature_unmod_time_p,denature_unmod_time_v)

denature_unmod_time_table=denature_unmod_time_table %>% mutate(time_nor_zs=(  (denature_unmod_time_v-mean(denature_unmod_time_v))/sd(denature_unmod_time_v)  ) )
head(denature_unmod_time_table)
## plot for test.
ggplot(denature_unmod_time_table,aes(x=denature_unmod_time_p,y=time_nor_zs))+ geom_line()+ scale_x_continuous(breaks = seq(0, 425, by = 10))

# normolize the data by position.
unmod2_time=merge(unmod2_time,denature_unmod_time_table,by.x="position",by.y="denature_unmod_time_p")
head(unmod2_time)
## raw time - background time and zs.
unmod2_time1=unmod2_time
unmod2_time1=unmod2_time1%>% mutate(time_rm_nor=(time-denature_unmod_time_v))%>%
  mutate(time_rm_nor_zs=( (time_rm_nor-mean(time_rm_nor))/sd(time_rm_nor)   ))
head(unmod2_time1)
# plot for test
ggplot(unmod2_time1,aes(x=position,y=time_rm_nor_zs))+ geom_line()+ scale_x_continuous(breaks = seq(0, 425, by = 10))
```
* a2-processe.R
```r
head(unmod2_time1)



# normolize the data by position.
unmod2_time=merge(unmod2_time,denature_unmod_time_table,by.x="position",by.y="denature_unmod_time_p")
NAI_N3_1_time=merge(NAI_N3_1_time,denature_unmod_time_table,by.x="position",by.y="denature_unmod_time_p")
NAI_N3_2_time=merge(NAI_N3_2_time,denature_unmod_time_table,by.x="position",by.y="denature_unmod_time_p")
NAI_N3_25_time=merge(NAI_N3_25_time,denature_unmod_time_table,by.x="position",by.y="denature_unmod_time_p")
denature_time=merge(denature_time,denature_unmod_time_table,by.x="position",by.y="denature_unmod_time_p")

head(unmod2_time)
## raw time - background time and zs.
unmod2_time1=unmod2_time
NAI_N3_1_time1=NAI_N3_1_time
NAI_N3_2_time1=NAI_N3_2_time
NAI_N3_25_time1=NAI_N3_25_time
denature_time1=denature_time

unmod2_time1=unmod2_time1%>% mutate(time_rm_nor=(time-denature_unmod_time_v))%>%
  mutate(time_rm_nor_zs=( (time_rm_nor-mean(time_rm_nor))/sd(time_rm_nor)   ))

NAI_N3_1_time1=NAI_N3_1_time1%>% mutate(time_rm_nor=(time-denature_unmod_time_v))%>%
  mutate(time_rm_nor_zs=( (time_rm_nor-mean(time_rm_nor))/sd(time_rm_nor)   ))

NAI_N3_2_time1=NAI_N3_2_time1%>% mutate(time_rm_nor=(time-denature_unmod_time_v))%>%
  mutate(time_rm_nor_zs=( (time_rm_nor-mean(time_rm_nor))/sd(time_rm_nor)   ))

NAI_N3_25_time1=NAI_N3_25_time1%>% mutate(time_rm_nor=(time-denature_unmod_time_v))%>%
  mutate(time_rm_nor_zs=( (time_rm_nor-mean(time_rm_nor))/sd(time_rm_nor)   ))

denature_time1=denature_time1%>% mutate(time_rm_nor=(time-denature_unmod_time_v))%>%
  mutate(time_rm_nor_zs=( (time_rm_nor-mean(time_rm_nor))/sd(time_rm_nor)   ))
# output
write.table(unmod2_time1,paste(dir,"/a9-rm-bk-time/unmod2_rm_bk_time.tsv",sep=''),sep="\t")
write.table(NAI_N3_1_time1,paste(dir,"/a9-rm-bk-time/NAI_N3_1_time.tsv",sep=''),sep="\t")
write.table(NAI_N3_2_time1,paste(dir,"/a9-rm-bk-time/NAI_N3_2_time.tsv",sep=''),sep="\t")
write.table(NAI_N3_25_time1,paste(dir,"/a9-rm-bk-time/NAI_N3_25_time.tsv",sep=''),sep="\t")
write.table(denature_time1,paste(dir,"/a9-rm-bk-time/denature_time.tsv",sep=''),sep="\t")


head(unmod2_time1)
head(NAI_N3_1_time1)

# plot for test
ggplot(unmod2_time1,aes(x=position,y=time_rm_nor_zs))+ geom_line()+ scale_x_continuous(breaks = seq(0, 425, by = 10))
ggplot(NAI_N3_1_time1,aes(x=position,y=time_rm_nor_zs))+ geom_line()+ scale_x_continuous(breaks = seq(0, 425, by = 10))
ggplot(NAI_N3_2_time1,aes(x=position,y=time_rm_nor_zs))+ geom_line()+ scale_x_continuous(breaks = seq(0, 425, by = 10))
ggplot(NAI_N3_25_time1,aes(x=position,y=time_rm_nor_zs))+ geom_line()+ scale_x_continuous(breaks = seq(0, 425, by = 10))
ggplot(denature_time1,aes(x=position,y=time_rm_nor_zs))+ geom_line()+ scale_x_continuous(breaks = seq(0, 425, by = 10))

unmod2_time1$color='gray'
unmod2_time1$color[unmod2_time1$time_rm_nor_zs > 1]='red'
unmod2_time1$color[unmod2_time1$time_rm_nor_zs > 0.3 & unmod2_time1$time_rm_nor_zs <= 1]='yellow'
unmod2_time1$color[unmod2_time1$time_rm_nor_zs > -1 & unmod2_time1$time_rm_nor_zs <= -0.3]='green'
unmod2_time1$color[unmod2_time1$time_rm_nor_zs < -1]='blue'
head(unmod2_time1)
write.table(unmod2_time1,paste(dir,"/a9-rm-bk-time/unmod2_rm_bk_time_color.tsv",sep=''),sep="\t")
min(unmod2_time1$time_rm_nor_zs)

#ROC calculate.
#错位15-25
```
* a1-read1-sum.R
```r
#library(data.table)
library(ggplot2)
suppressMessages(library(dplyr))
#library("data.table") 

# read rawdata.
dir='/Users/xugang/Desktop/sequencing_center_desktop/2020/rna_structure/a1-emzye/'
## load reactivity score.
unmod2_reactivity   =read.table(paste(dir,"/a5-reactivity/Output_TeTrahymenaTetra_unmod2.combined.RData.csv",sep=''),sep=",",header=T)
NAI_N3_1_reactivity =read.table(paste(dir,'/a5-reactivity/Output_TeTrahymenaTetra_NAI_N3_1.combined.RData.csv',sep=''),sep=",",header=T)
NAI_N3_2_reactivity =read.table(paste(dir,'/a5-reactivity/Output_TeTrahymenaTetra_NAI_N3_2.combined.RData.csv',sep=''),sep=",",header=T)
NAI_N3_25_reactivity=read.table(paste(dir,'/a5-reactivity/Output_TeTrahymenaTetra_NAI_N3_25min.combined.RData.csv',sep=''),sep=",",header=T)
denature_reactivity =read.table(paste(dir,'/a5-reactivity/Output_TeTrahymenaTetra_NAI_N3_denatured.combined.RData.csv',sep=''),sep=",",header=T)
head(unmod2_reactivity)
head(NAI_N3_1_reactivity)
head(NAI_N3_2_reactivity)
head(NAI_N3_25_reactivity)
head(denature_reactivity)

# read svm machine learning.
svmd=read.table(paste(dir,'/a9-ml/ml.csv',sep=''),sep=",",header = T)
head(svmd)

## plot for test.
ggplot(svmd,aes(x=pos,y=timezs))+ geom_line()+ scale_x_continuous(breaks = seq(0, 425, by = 10)) +  coord_cartesian(ylim=c(-0.1, 2)) 



## time.
unmod2_time   =read.table(paste(dir,"/a6-time-sum/Tetra_unmod2.sum.time.tsv",sep=''),sep="\t",header=T)
NAI_N3_1_time =read.table(paste(dir,"/a6-time-sum/Tetra_NAI_N3_1.sum.time.tsv",sep=''),sep="\t",header=T)
NAI_N3_2_time =read.table(paste(dir,"/a6-time-sum/Tetra_NAI_N3_2.sum.time.tsv",sep=''),sep="\t",header=T)
NAI_N3_25_time=read.table(paste(dir,"/a6-time-sum/Tetra_NAI_N3_25min.sum.time.tsv",sep=''),sep="\t",header=T)
denature_time =read.table(paste(dir,"/a6-time-sum/Tetra_NAI_N3_denatured.sum.time.tsv",sep=''),sep="\t",header=T)
head(unmod2_time)
head(NAI_N3_1_time)
head(NAI_N3_2_time)
head(NAI_N3_25_time)
head(denature_time)
## load read each position 
unmod2_mod_time   =read.table(paste(dir,"/a7-mod-unmod-time-sum/Tetra_unmod2.mod.tsv",sep=''),sep="\t",header=F)
unmod2_unmod_time =read.table(paste(dir,"/a7-mod-unmod-time-sum/Tetra_unmod2.unmod.tsv",sep=''),sep="\t",header=F)
NAI_N3_1_mod_time   =read.table(paste(dir,"/a7-mod-unmod-time-sum/Tetra_NAI_N3_1.mod.tsv",sep=''),sep="\t",header=F)
NAI_N3_1_unmod_time =read.table(paste(dir,"/a7-mod-unmod-time-sum/Tetra_NAI_N3_1.unmod.tsv",sep=''),sep="\t",header=F)
NAI_N3_2_mod_time   =read.table(paste(dir,"/a7-mod-unmod-time-sum/Tetra_NAI_N3_2.mod.tsv",sep=''),sep="\t",header=F)
NAI_N3_2_unmod_time =read.table(paste(dir,"/a7-mod-unmod-time-sum/Tetra_NAI_N3_2.unmod.tsv",sep=''),sep="\t",header=F)
NAI_N3_25_mod_time   =read.table(paste(dir,"/a7-mod-unmod-time-sum/Tetra_NAI_N3_25min.mod.tsv",sep=''),sep="\t",header=F)
NAI_N3_25_unmod_time =read.table(paste(dir,"/a7-mod-unmod-time-sum/Tetra_NAI_N3_25min.unmod.tsv",sep=''),sep="\t",header=F)
denature_mod_time   =read.table(paste(dir,"/a7-mod-unmod-time-sum/Tetra_NAI_N3_denatured.mod.tsv",sep=''),sep="\t",header=F)
denature_unmod_time =read.table(paste(dir,"/a7-mod-unmod-time-sum/Tetra_NAI_N3_denatured.unmod.tsv",sep=''),sep="\t",header=F)
head(denature_unmod_time)
head(denature_mod_time)

## load position and time.
unmod2_mod_pt   =read.table(paste(dir,"/a8-position-time-sum/Tetra_unmod2.mod.pos.tsv",sep=''),sep="\t",header=F)
unmod2_unmod_pt =read.table(paste(dir,"/a8-position-time-sum/Tetra_unmod2.unmod.pos.tsv",sep=''),sep="\t",header=F)
NAI_N3_1_mod_pt   =read.table(paste(dir,"/a8-position-time-sum/Tetra_NAI_N3_1.mod.pos.tsv",sep=''),sep="\t",header=F)
NAI_N3_1_unmod_pt =read.table(paste(dir,"/a8-position-time-sum/Tetra_NAI_N3_1.unmod.pos.tsv",sep=''),sep="\t",header=F)
NAI_N3_2_mod_pt   =read.table(paste(dir,"/a8-position-time-sum/Tetra_NAI_N3_2.mod.pos.tsv",sep=''),sep="\t",header=F)
NAI_N3_2_unmod_pt =read.table(paste(dir,"/a8-position-time-sum/Tetra_NAI_N3_2.unmod.pos.tsv",sep=''),sep="\t",header=F)
NAI_N3_25_mod_pt   =read.table(paste(dir,"/a8-position-time-sum/Tetra_NAI_N3_25min.mod.pos.tsv",sep=''),sep="\t",header=F)
NAI_N3_25_unmod_pt =read.table(paste(dir,"/a8-position-time-sum/Tetra_NAI_N3_25min.unmod.pos.tsv",sep=''),sep="\t",header=F)
denature_mod_pt   =read.table(paste(dir,"/a8-position-time-sum/Tetra_NAI_N3_denatured.mod.pos.tsv",sep=''),sep="\t",header=F)
denature_unmod_pt =read.table(paste(dir,"/a8-position-time-sum/Tetra_NAI_N3_denatured.unmod.pos.tsv",sep=''),sep="\t",header=F)

colnames(unmod2_mod_pt)=c('position','time','kmer')
colnames(unmod2_unmod_pt)=c('position','time','kmer')
colnames(NAI_N3_1_mod_pt)=c('position','time','kmer')
colnames(NAI_N3_1_unmod_pt)=c('position','time','kmer')
colnames(NAI_N3_2_mod_pt)=c('position','time','kmer')
colnames(NAI_N3_2_unmod_pt)=c('position','time','kmer')
colnames(NAI_N3_25_mod_pt)=c('position','time','kmer')
colnames(NAI_N3_25_unmod_pt)=c('position','time','kmer')
colnames(denature_mod_pt)=c('position','time','kmer')
colnames(denature_unmod_pt)=c('position','time','kmer')


head(unmod2_mod_pt)
dim(unmod2_mod_pt)
dim(unmod2_unmod_pt)
##
#datapro=unmod2_mod_pt
#datapro=unmod2_unmod_pt
#datapro=NAI_N3_1_mod_pt
#datapro=NAI_N3_1_unmod_pt
#datapro=NAI_N3_2_mod_pt
#datapro=NAI_N3_2_unmod_pt
#datapro=NAI_N3_25_mod_pt
datapro=NAI_N3_25_unmod_pt
#datapro=denature_mod_pt
#datapro=denature_unmod_pt

dataA=datapro[datapro$kmer %like% "^A", ]
dataT=datapro[datapro$kmer %like% "^T", ]
dataC=datapro[datapro$kmer %like% "^C", ]
dataG=datapro[datapro$kmer %like% "^G", ]

staA=dim(dataA)
staT=dim(dataT)
staC=dim(dataC)
staG=dim(dataG)
ggplot(dataA, aes(x=kmer, y=time)) + labs(title=paste("A: ", staA[1],sep=""))+
  geom_boxplot(outlier.shape = NA)+theme(axis.text=element_text(size=9,angle=90),
                                         axis.title=element_text(size=12,face="bold"))+  coord_cartesian(ylim=c(0, 100)) 
ggplot(dataT, aes(x=kmer, y=time)) +  labs(title=paste("T: ", staT[1],sep=""))+
  geom_boxplot(outlier.shape = NA)+theme(axis.text=element_text(size=9,angle=90),
                                         axis.title=element_text(size=12,face="bold"))+  coord_cartesian(ylim=c(0, 100))
ggplot(dataC, aes(x=kmer, y=time)) +  labs(title=paste("C: ", staC[1],sep=""))+
  geom_boxplot(outlier.shape = NA)+theme(axis.text=element_text(size=9,angle=90),
                                         axis.title=element_text(size=12,face="bold"))+  coord_cartesian(ylim=c(0, 100))
ggplot(dataG, aes(x=kmer, y=time)) +  labs(title=paste("G: ", staG[1],sep=""))+
  geom_boxplot(outlier.shape = NA)+theme(axis.text=element_text(size=9,angle=90),
                                         axis.title=element_text(size=12,face="bold"))+  coord_cartesian(ylim=c(0, 100))
## plot time for kmer.
ggplot(unmod2_mod_pt, aes(x=kmer, y=time)) + 
  geom_boxplot()+theme(axis.text=element_text(size=20),
                       axis.title=element_text(size=22,face="bold"))


# get reactivity score.

unmod2_reactivity=unmod2_reactivity %>% mutate(reactivity=(  (Mod_percentage-mean(Mod_percentage))/sd(Mod_percentage)  ) )
NAI_N3_1_reactivity =NAI_N3_1_reactivity%>% mutate(reactivity=(  (Mod_percentage-mean(Mod_percentage))/sd(Mod_percentage)  ) )
NAI_N3_2_reactivity =NAI_N3_2_reactivity%>% mutate(reactivity=(  (Mod_percentage-mean(Mod_percentage))/sd(Mod_percentage)  ) )
NAI_N3_25_reactivity=NAI_N3_25_reactivity%>% mutate(reactivity=(  (Mod_percentage-mean(Mod_percentage))/sd(Mod_percentage)  ) )
denature_reactivity =denature_reactivity%>% mutate(reactivity=(  (Mod_percentage-mean(Mod_percentage))/sd(Mod_percentage)  ) )

head(unmod2_reactivity)
# plot for test.
ggplot(unmod2_reactivity,aes(x=Position,y=reactivity))+ geom_line()+ scale_x_continuous(breaks = seq(0, 425, by = 10))
ggplot(NAI_N3_1_reactivity,aes(x=Position,y=reactivity))+ geom_line()+ scale_x_continuous(breaks = seq(0, 425, by = 20))
ggplot(NAI_N3_1_reactivity,aes(x=Position,y=Mod_percentage))+ geom_line()+ scale_x_continuous(breaks = seq(0, 425, by = 20))

head(NAI_N3_1_reactivity)
write.table(NAI_N3_1_reactivity,paste(dir,"/a5-reactivity/NAI_N3_1_reactivity.tsv",sep=''),sep="\t")

# get time zs
unmod2_time   =unmod2_time   %>% mutate(timezs=(  (time-mean(time))/sd(time)  ) )
NAI_N3_1_time =NAI_N3_1_time %>% mutate(timezs=(  (time-mean(time))/sd(time)  ) )
NAI_N3_2_time =NAI_N3_2_time %>% mutate(timezs=(  (time-mean(time))/sd(time)  ) )
NAI_N3_25_time=NAI_N3_25_time%>% mutate(timezs=(  (time-mean(time))/sd(time)  ) )
denature_time =denature_time %>% mutate(timezs=(  (time-mean(time))/sd(time)  ) )

head(unmod2_time)

#plot for time
ggplot(unmod2_time,aes(x=position,y=timezs))+ geom_line()+ scale_x_continuous(breaks = seq(0, 425, by = 10))
ggplot(NAI_N3_1_time,aes(x=position,y=timezs))+ geom_line()+ scale_x_continuous(breaks = seq(0, 425, by = 10))
ggplot(NAI_N3_2_time,aes(x=position,y=timezs))+ geom_line()+ scale_x_continuous(breaks = seq(0, 425, by = 10))
ggplot(NAI_N3_25_time,aes(x=position,y=timezs))+ geom_line()+ scale_x_continuous(breaks = seq(0, 425, by = 10))
ggplot(denature_time,aes(x=position,y=timezs))+ geom_line()+ scale_x_continuous(breaks = seq(0, 425, by = 10))

# calculate background time.
denature_unmod_time_v=colMeans(denature_unmod_time[,2:418],na.rm=T)
denature_unmod_time_v
denature_unmod_time_p=seq(0,416,by = 1)
denature_unmod_time_table=data.frame(denature_unmod_time_p,denature_unmod_time_v)

denature_unmod_time_table=denature_unmod_time_table %>% mutate(time_nor_zs=(  (denature_unmod_time_v-mean(denature_unmod_time_v))/sd(denature_unmod_time_v)  ) )
head(denature_unmod_time_table)
## plot for test.
ggplot(denature_unmod_time_table,aes(x=denature_unmod_time_p,y=time_nor_zs))+ geom_line()+ scale_x_continuous(breaks = seq(0, 425, by = 10))

# normolize the data by position.
unmod2_time=merge(unmod2_time,denature_unmod_time_table,by.x="position",by.y="denature_unmod_time_p")
head(unmod2_time)
## raw time - background time and zs.
unmod2_time1=unmod2_time
unmod2_time1=unmod2_time1%>% mutate(time_rm_nor=(time-denature_unmod_time_v))%>%
  mutate(time_rm_nor_zs=( (time_rm_nor-mean(time_rm_nor))/sd(time_rm_nor)   ))
head(unmod2_time1)
# plot for test
ggplot(unmod2_time1,aes(x=position,y=time_rm_nor_zs))+ geom_line()+ scale_x_continuous(breaks = seq(0, 425, by = 10))

```
* a3-auc.R
```r
#install.packages('cvAUC')
library(cvAUC)
dir='/Users/xugang/Desktop/sequencing_center_desktop/2020/rna_structure/a1-emzye/'
data(ROCR.simple)
ROCR.simple
auc <- AUC(ROCR.simple$predictions, ROCR.simple$labels)
# [1] 0.8341875
auc

head(ROCR.simple)

head(NAI_N3_1_reactivity)
head(unmod2_reactivity)
structure=read.table(paste(dir,'/a10-auc/tetra-structure.txt',sep=''),sep="\t",header = T)
structure=read.table(paste(dir,'/a10-auc/crystal_PDB_00079_part1.txt',sep=''),sep="\t",header = T)
structure=read.table(paste(dir,'/a10-auc/crystal_PDB_00079_part2.txt',sep=''),sep="\t",header = T)
structure=read.table(paste(dir,'/a10-auc/crystal_PDB_00805_part1.txt',sep=''),sep="\t",header = T)
structure=read.table(paste(dir,'/a10-auc/crystal_PDB_00805_part2.txt',sep=''),sep="\t",header = T)
structure=read.table(paste(dir,'/a10-auc/crystal_PDB_00805_part3.txt',sep=''),sep="\t",header = T)
structure=read.table(paste(dir,'/a10-auc/crystal_PDB_00805_part4.txt',sep=''),sep="\t",header = T)

structure=read.table(paste(dir,'/a10-auc/part-structure.txt',sep=''),sep="\t",header = T)


head(structure)
tail(structure)
head(NAI_N3_2_reactivity)
head(NAI_N3_25_reactivity)
head(denature_reactivity)
auc_data2=merge(NAI_N3_1_reactivity,structure,by.x='Position',by.y='position')
auc_data2=merge(NAI_N3_2_reactivity,structure,by.x='Position',by.y='position')
auc_data2=merge(NAI_N3_25_reactivity,structure,by.x='Position',by.y='position')
auc_data2=merge(denature_reactivity,structure,by.x='Position',by.y='position')
auc_data2=merge(unmod2_reactivity,structure,by.x='Position',by.y='position')
head(auc_data2)
#
auc_data=auc_data2[auc_data2$reactivity > 0,]
auc_data=auc_data2[auc_data2$Position > 103 & auc_data2$Position < 242 & auc_data2$reactivity > 0.1,]
head(auc_data)
auc_data


plot(auc_data$Position,auc_data$reactivity,type='l')
par(new = TRUE)  
plot(auc_data$Position,auc_data$pair*(-1), col = 3,axes = F, xlab = "", ylab = "",type="p",cex=0.2)
axis(side = 4, at = pretty(range(auc_data$pair*(-1) )))   
#auc_data[35:421,]
out=cvAUC(auc_data$reactivity*(-1), auc_data$pair)
out

#Plot fold AUCs
plot(out$perf, main=paste("ROC AUC: ", out$cvAUC,sep='') )
#plot(out$perf, col="grey82", lty=4, main="ROC")
#Plot CV AUC
plot(out$perf, col="red", avg="vertical", add=TRUE)


#
auc_test=data.frame(value=c(5.0437,-0.4610,0.8916,-0.3709716,-0.4391517,
                            -0.3771764,-0.4533397,2.1096188,0.5926826,
                            -0.4773314,-0.3443886,-0.4610671,1.0120529,
                            0.6060499,-0.4459078,-0.4259067,-0.4430279,
                            -0.2418133,7),
                    pre=c(1,0,1,0,0,0,0,0,0,0,0,0,1,1,0,0,0,0,1))
auc_test
auc_test_out=cvAUC(auc_test$value,auc_test$pre)
auc_test_out
plot(auc_test_out$perf, main=paste("ROC AUC: ", auc_test_out$cvAUC,sep='') )
#

# 0.594692
head(unmod2_time1)

shiftt = data.frame(pos=c(),auc=c())
for (i in 0:30 ){
  processed=unmod2_time1
  print(i)
  shift=i
  processed$position_nor=processed$position-shift
  head(processed)
  auc_datat=merge(processed,structure,by.x='position_nor',by.y='position')
  head(auc_datat)
  line=dim(auc_datat)[1]
  auc_data2=auc_datat$position_nor>35
  auc_datat[30:line,]
  print (line)
  aucv=AUC(auc_datat[35:line,]$time_rm_nor_zs, auc_datat[35:line,]$pair)
  print(aucv)
  shiftt=rbind(shiftt,c(shift,aucv))
}
colnames(shiftt)=c('pos','auc')
shiftt
plot(shiftt$pos,shiftt$auc)
max(shiftt$auc)

# test for ??????

shiftt = data.frame(pos=c(),auc=c())
shiftt2 = data.frame(pos2=c(),auc2=c())
head(processdata)
for (i in 0:30 ){
  #
  #i=14
  processed=processdata
  print(i)
  shift=i
  processdata$position_nor=processdata$position-shift
head(processdata)
head(structure)
  auc_datat=merge(processed,structure,by.x='position_nor',by.y='position')
  head(auc_datat)
  line=dim(auc_datat)[1]
  auc_data2=auc_datat$position_nor>35
  auc_datat[30:line,]
  print (line)
  aucv=AUC(auc_datat[35:line,]$lin, auc_datat[35:line,]$pair)
  aucv2=AUC(auc_datat[35:line,]$time, auc_datat[35:line,]$pair)
  print(aucv)
  shiftt=rbind(shiftt,c(shift,aucv))
  shiftt2=rbind(shiftt2,c(shift,aucv2))
}
colnames(shiftt)=c('pos','auc')
colnames(shiftt2)=c('pos','auc')
shiftt
plot(shiftt$pos,shiftt$auc,ylim=c(0.4,0.7))
max(shiftt$auc)
plot(shiftt$pos,shiftt2$auc)
max(shiftt2$auc)
auc_datat
write.table(auc_datat,paste(dir,'/a10-auc/line-smooth.tsv',sep=''),sep="\t")
# 2 max auc 0.575
# 3 max auc 0.5818
# 4 max auc 0.577
# 5 max auc 0.577

# test for svm

shiftt = data.frame(pos=c(),auc=c())
for (i in 0:30 ){
  processed=svmd
  print(i)
  shift=i
  processed$position_nor=processed$pos-shift
  head(processed)
  auc_datat=merge(processed,structure,by.x='position_nor',by.y='position')
  head(auc_datat)
  line=dim(auc_datat)[1]
  auc_data2=auc_datat$position_nor>35
  auc_datat[30:line,]
  print (line)
  aucv=AUC(auc_datat[35:line,]$timezs, auc_datat[35:line,]$pair)
  print(aucv)
  shiftt=rbind(shiftt,c(shift,aucv))
}
colnames(shiftt)=c('pos','auc')
shiftt
plot(shiftt$pos,shiftt$auc)
max(shiftt$auc)
min(shiftt$auc)


barplot(c(0.5865757,0.594692),names.arg=c("time", "reactivity"),ylim=c(0,0.7))
text(c(0.7,1.9),c(0.6065757,0.614692),label=c(0.5865757,0.594692))


#ks

shiftt = data.frame(pos=c(),auc=c())

for (i in 0:30 ){
  #i=14
  processed=ksdata
  print(i)
  shift=i
  processed$position_nor=processed$position-shift
  head(processed)
  head(structure)
  auc_datat=merge(processed,structure,by.x='position_nor',by.y='position')
  head(auc_datat)
  line=dim(auc_datat)[1]
  auc_data2=auc_datat$position_nor>35
  auc_datat[30:line,]
  print (line)
  aucv=AUC(auc_datat[35:line,]$df, auc_datat[35:line,]$pair)

  print(aucv)
  shiftt=rbind(shiftt,c(shift,aucv))

}
colnames(shiftt)=c('pos','auc')
shiftt
plot(shiftt$pos,shiftt$auc,ylim=c(0.4,0.7))
max(shiftt$auc)

#write.table(auc_datat,paste(dir,'/a10-auc/line-smooth.tsv',sep=''),sep="\t")

shiftt = data.frame(pos=c(),auc=c())

for (i in 0:30 ){
  #i=14
  processed=ksdata2
  print(i)
  shift=i
  processed$position_nor=processed$position-shift
  head(processed)
  head(structure)
  auc_datat=merge(processed,structure,by.x='position_nor',by.y='position')
  head(auc_datat)
  line=dim(auc_datat)[1]
  auc_data2=auc_datat$position_nor>35
  auc_datat[30:line,]
  print (line)
  aucv=AUC(auc_datat[35:line,]$df, auc_datat[35:line,]$pair)
  
  print(aucv)
  shiftt=rbind(shiftt,c(shift,aucv))
  
}
colnames(shiftt)=c('pos','auc')
shiftt
plot(shiftt$pos,shiftt$auc,ylim=c(0.4,0.7))
max(shiftt$auc)

#write.table(auc_datat,paste(dir,'/a10-auc/line-smooth.tsv',sep=''),sep="\t")



#ks smooth

shiftt = data.frame(pos=c(),auc=c())

for (i in 0:30 ){
  #i=14
  processed=ksdata
  print(i)
  shift=i
  processed$position_nor=processed$position-shift
  head(processed)
  head(structure)
  auc_datat=merge(processed,structure,by.x='position_nor',by.y='position')
  head(auc_datat)
  line=dim(auc_datat)[1]
  auc_data2=auc_datat$position_nor>35
  auc_datat[30:line,]
  print (line)
  aucv=AUC(auc_datat[35:line,]$lin, auc_datat[35:line,]$pair)
  
  print(aucv)
  shiftt=rbind(shiftt,c(shift,aucv))
  
}
colnames(shiftt)=c('pos','auc')
shiftt
plot(shiftt$pos,shiftt$auc,ylim=c(0.4,0.7))
max(shiftt$auc)

#write.table(auc_datat,paste(dir,'/a10-auc/line-smooth.tsv',sep=''),sep="\t")

shiftt = data.frame(pos=c(),auc=c())

for (i in 0:30 ){
  #i=14
  processed=ksdata2
  print(i)
  shift=i
  processed$position_nor=processed$position-shift
  head(processed)
  head(structure)
  auc_datat=merge(processed,structure,by.x='position_nor',by.y='position')
  head(auc_datat)
  line=dim(auc_datat)[1]
  auc_data2=auc_datat$position_nor>35
  auc_datat[30:line,]
  print (line)
  aucv=AUC(auc_datat[35:line,]$lin, auc_datat[35:line,]$pair)
  
  print(aucv)
  shiftt=rbind(shiftt,c(shift,aucv))
  
}
colnames(shiftt)=c('pos','auc')
shiftt
plot(shiftt$pos,shiftt$auc,ylim=c(0.4,0.7))
max(shiftt$auc)

# ks sig


shiftt = data.frame(pos=c(),auc=c())

for (i in 0:30 ){
  #i=14
  processed=ksdata
  print(i)
  shift=i
  processed$position_nor=processed$position-shift
  head(processed)
  head(structure)
  auc_datat=merge(processed,structure,by.x='position_nor',by.y='position')
  head(auc_datat)
  line=dim(auc_datat)[1]
  auc_data2=auc_datat$position_nor>35
  auc_datat[30:line,]
  print (line)
  aucv=AUC(auc_datat[35:line,]$sig, auc_datat[35:line,]$pair)
  
  print(aucv)
  shiftt=rbind(shiftt,c(shift,aucv))
  
}
colnames(shiftt)=c('pos','auc')
shiftt
plot(shiftt$pos,shiftt$auc,ylim=c(0.4,0.7))
max(shiftt$auc)

#write.table(auc_datat,paste(dir,'/a10-auc/line-smooth.tsv',sep=''),sep="\t")

shiftt = data.frame(pos=c(),auc=c())

for (i in 0:30 ){
  #i=14
  processed=ksdata2
  print(i)
  shift=i
  processed$position_nor=processed$position-shift
  head(processed)
  head(structure)
  auc_datat=merge(processed,structure,by.x='position_nor',by.y='position')
  head(auc_datat)
  line=dim(auc_datat)[1]
  auc_data2=auc_datat$position_nor>35
  auc_datat[30:line,]
  print (line)
  aucv=AUC(auc_datat[35:line,]$sig, auc_datat[35:line,]$pair)
  
  print(aucv)
  shiftt=rbind(shiftt,c(shift,aucv))
  
}
colnames(shiftt)=c('pos','auc')
shiftt
plot(shiftt$pos,shiftt$auc,ylim=c(0.4,0.7))
max(shiftt$auc)

# ks dfpv sm


shiftt = data.frame(pos=c(),auc=c())

for (i in 0:30 ){
  #i=23
  processed=ksdata
  print(i)
  shift=i
  processed$position_nor=processed$position-shift
  head(processed)
  head(structure)
  auc_datat=merge(processed,structure,by.x='position_nor',by.y='position')
  head(auc_datat)
  line=dim(auc_datat)[1]
  auc_data2=auc_datat$position_nor>35
  auc_datat[30:line,]
  print (line)
  aucv=AUC(auc_datat[35:line,]$dfpv_sm, auc_datat[35:line,]$pair)
  
  print(aucv)
  shiftt=rbind(shiftt,c(shift,aucv))
  
}
colnames(shiftt)=c('pos','auc')
shiftt
plot(shiftt$pos,shiftt$auc,ylim=c(0.4,0.7))
max(shiftt$auc)

#write.table(auc_datat,paste(dir,'/a10-auc/line-smooth.tsv',sep=''),sep="\t")
write.table(auc_datat,paste(dir,'/a10-auc/depv-smooth.tsv',sep=''),sep="\t")

shiftt = data.frame(pos=c(),auc=c())

for (i in 0:30 ){
  #i=14
  processed=ksdata2
  print(i)
  shift=i
  processed$position_nor=processed$position-shift
  head(processed)
  head(structure)
  auc_datat=merge(processed,structure,by.x='position_nor',by.y='position')
  head(auc_datat)
  line=dim(auc_datat)[1]
  auc_data2=auc_datat$position_nor>35
  auc_datat[30:line,]
  print (line)
  aucv=AUC(auc_datat[35:line,]$dfpv_sm, auc_datat[35:line,]$pair)
  
  print(aucv)
  shiftt=rbind(shiftt,c(shift,aucv))
  
}
colnames(shiftt)=c('pos','auc')
shiftt
plot(shiftt$pos,shiftt$auc,ylim=c(0.4,0.7))
max(shiftt$auc)
#
# Example of how to get CV AUC and plot the curve.
data(ROCR.hiv)
attach(ROCR.hiv)
out <- cvAUC(hiv.svm$predictions, hiv.svm$labels)
#Plot fold AUCs
plot(out$perf, col="grey82", lty=3, main="10-fold CV AUC")
#Plot CV AUC
plot(out$perf, col="red", avg="vertical", add=TRUE)
# See the ci.cvAUC documentation for an example
# of how to use the `folds` argument.
```
* a4-smooth.R
```R
library(ggplot2)
# detach("package:data.table", unload=TRUE)
detach("package:dplyr", unload=TRUE)

#prepare rwa data.
xdata=c(1,2,3,4,5,6,7,8,9,10)
tdata=data.frame(1:10,xdata)
head(tdata)

m1<-filter(xdata,filter=c(rep(1/4,4) ))
m1
plot(tdata,xlab="time",ylab="gdp")
lines(m1,col="red",cex=1.5)

# ??????????????????
plot(tdata,type="l",xlab="time",ylab="GDP")
m2<-filter(m1,filter=c(rep(1/4,4)),sides=1)
lines(2*m1-m2,col="red",cex=2)

#
#colnames()
#??? ????????????
newdata=xdata
plot(newdata,type="l",cex.axis=1.5,cex.lab=1.5,xlab="time",ylab="confidence")
a<-HoltWinters(newdata,beta=F,gamma=F)
b<-HoltWinters(newdata,alpha=0.5,beta=F,gamma=F) #????????????a
b

pdata<-predict(a,6,prediction.interval = T)
plot(a,pdata,type="o",xlab="time",ylab="confidence")

# 2???Holt_Winters????????????

tdata<-ts(xdata,start=1,freq=5)

gdp.hw<-HoltWinters(tdata,seasonal="multi")
plot(gdp.hw$fitted,type="o",main="?????????")
plot(gdp.hw,type="o")
pdata<-predict(gdp.hw,n.ahead=4*5)
pdata
ts.plot(tdata,pdata,type="o",lty=1:2,col=c("red","black"))

# read data
# normolize the data by position.
#write.table(unmod2_time1,paste(dir,"/a9-rm-bk-time/unmod2_rm_bk_time.tsv",sep=''),sep="\t")
#write.table(NAI_N3_1_time1,paste(dir,"/a9-rm-bk-time/NAI_N3_1_time.tsv",sep=''),sep="\t")
#write.table(NAI_N3_2_time1,paste(dir,"/a9-rm-bk-time/NAI_N3_2_time.tsv",sep=''),sep="\t")
#write.table(NAI_N3_25_time1,paste(dir,"/a9-rm-bk-time/NAI_N3_25_time.tsv",sep=''),sep="\t")
#write.table(denature_time1,paste(dir,"/a9-rm-bk-time/denature_time.tsv",sep=''),sep="\t")
dir='/Users/xugang/Desktop/sequencing_center_desktop/2020/rna_structure/a1-emzye/'

unmod2_time=read.table(paste(dir,"/a9-rm-bk-time/unmod2_rm_bk_time.tsv",sep=''),sep="\t")
NAI_N3_1_time=read.table(paste(dir,"/a9-rm-bk-time/NAI_N3_1_time.tsv",sep=''),sep="\t" )
NAI_N3_2_time=read.table(paste(dir,"/a9-rm-bk-time/NAI_N3_2_time.tsv",sep=''),sep="\t" )
NAI_N3_25_time=read.table(paste(dir,"/a9-rm-bk-time/NAI_N3_25_time.tsv",sep=''),sep="\t" )
denature_time=read.table(paste(dir,"/a9-rm-bk-time/denature_time.tsv",sep=''),sep="\t" )
# extract variables.
unmod2_time_blur=unmod2_time[,c("position","time_rm_nor_zs")]
NAI_N3_1_time_blur=NAI_N3_1_time[,c("position","time_rm_nor_zs")]
NAI_N3_2_time_blur=NAI_N3_2_time[,c("position","time_rm_nor_zs")]
NAI_N3_25_time_blur=NAI_N3_25_time[,c("position","time_rm_nor_zs")]
denature_time_blur=denature_time[,c("position","time_rm_nor_zs")]
#
colnames(unmod2_time_blur)=c("position",'time')
colnames(NAI_N3_1_time_blur)=c("position",'time')
colnames(NAI_N3_2_time_blur)=c("position",'time')
colnames(NAI_N3_25_time_blur)=c("position",'time')
colnames(denature_time_blur)=c("position",'time')
#
head(unmod2_time_blur)
#max=4
splitn=4
#processdata=unmod2_time_blur
#processdata=NAI_N3_1_time_blur
processdata=NAI_N3_2_time_blur
processdata=NAI_N3_25_time_blur
#head(NAI_N3_1_time_blur)
processdata$lin<-filter(processdata$time,filter=c(rep(1/splitn,splitn)))
head(processdata)


plot(processdata$position, processdata$time, pch = 16, col = 2, type = "l" )   
# Basic line plot in R) 
axis(1, at=seq(0,420,10))
par(new = TRUE)                             # Add new plot
# Create second plot without axes
plot(processdata$position, processdata$lin , pch = 17, col = 3,              
     axes = F, xlab = "", ylab = "",type="l")
axis(side = 4, at = pretty(range(processdata$lin )))      # Add second axis
mtext("timezx", side = 4, line = 3)             # Add second axis label
dev.off()
```
* a5-ks.R
```r
#library(data.table)
library(ggplot2)
#suppressMessages(library(dplyr))
#detach("package:dplyr", unload=TRUE)
#library("data.table") 
# 
#install.packages('philentropy')
#library(philentropy)
#install.packages('FNN')
#library(FNN)
#install.packages('dgof')
library(dgof)
x <- rnorm(50)
y <- runif(30)*100
x=seq(1,20,by=1)
y=seq(21,40,by=1)
p1=hist(x)
p2=hist(y)
plot( p1, col=rgb(0,0,1,1/4), xlim=c(0,40),ylim=c(0,10))
plot( p2, col=rgb(1,0,0,1/4), xlim=c(0,40),ylim=c(0,10), add=T)
# Do x and y come from the same distribution?
x
y
ks.test(x,y)
ks.test(x,y,alternative = c( "greater"))

ks.test(x,y,alternative = c( "less"))
ks.test(y,x)
ks.test(y,x,alternative = c( "greater"))
ks.test(y,x,alternative = c( "less"))

# read rawdata.
dir='/Users/xugang/Desktop/sequencing_center_desktop/2020/rna_structure/a1-emzye/'




accuracy_data=read.table(paste(dir,"/list.txt.tsv",sep=''),sep="\t",header=F)
head(accuracy_data)
accuracy_data=accuracy_data[order(accuracy_data[,1]), ]

head(accuracy_data)
nal=dim(accuracy_data)[2] * -1
accuracy_data=accuracy_data[,nal]
head(accuracy_data)
accuracy_datat=t(accuracy_data)
nal=dim(accuracy_datat)[2]
accuracy_data=t(accuracy_datat[order(accuracy_datat[,nal]),])
dim(accuracy_data)
accuracy_data=accuracy_data[c(19,1:18),c(31,1:30)]
#
linev=18
accuracy_data[linev,1]
df=data.frame(position=accuracy_data[1,c(2:31)],value=accuracy_data[linev,c(2:31)])
#plot
ggplot(data=df, aes(x=position, y=value, group=1)) +
  ggtitle(paste("cutoff: ", accuracy_data[linev,1]))+
  geom_line(color="red")+
  geom_point()+theme(axis.text=element_text(size=22,face="bold"),
                     axis.title=element_text(size=24,face="bold"),
                     legend.title=element_text(size=24,face="bold"), 
                     legend.text=element_text(size=22,face="bold"))+
  scale_x_continuous( breaks = c(0, 2, 4, 6, 8, 10, 12, 14, 16, 18, 20, 22, 24, 26, 28, 30))+
  scale_y_continuous(limits = c(0.6,0.9), breaks = c(0.6,0.65,0.7,0.75,0.8,0.85,0.9))




## load read each position 
unmod2_mod_time   =read.table(paste(dir,"/a7-mod-unmod-time-sum/Tetra_unmod2.mod.tsv",sep=''),sep="\t",header=F)
unmod2_unmod_time =read.table(paste(dir,"/a7-mod-unmod-time-sum/Tetra_unmod2.unmod.tsv",sep=''),sep="\t",header=F)
NAI_N3_1_mod_time   =read.table(paste(dir,"/a7-mod-unmod-time-sum/Tetra_NAI_N3_1.mod.tsv",sep=''),sep="\t",header=F)
NAI_N3_1_unmod_time =read.table(paste(dir,"/a7-mod-unmod-time-sum/Tetra_NAI_N3_1.unmod.tsv",sep=''),sep="\t",header=F)
NAI_N3_2_mod_time   =read.table(paste(dir,"/a7-mod-unmod-time-sum/Tetra_NAI_N3_2.mod.tsv",sep=''),sep="\t",header=F)
NAI_N3_2_unmod_time =read.table(paste(dir,"/a7-mod-unmod-time-sum/Tetra_NAI_N3_2.unmod.tsv",sep=''),sep="\t",header=F)
NAI_N3_25_mod_time   =read.table(paste(dir,"/a7-mod-unmod-time-sum/Tetra_NAI_N3_25min.mod.tsv",sep=''),sep="\t",header=F)
NAI_N3_25_unmod_time =read.table(paste(dir,"/a7-mod-unmod-time-sum/Tetra_NAI_N3_25min.unmod.tsv",sep=''),sep="\t",header=F)
denature_mod_time   =read.table(paste(dir,"/a7-mod-unmod-time-sum/Tetra_NAI_N3_denatured.mod.tsv",sep=''),sep="\t",header=F)
denature_unmod_time =read.table(paste(dir,"/a7-mod-unmod-time-sum/Tetra_NAI_N3_denatured.unmod.tsv",sep=''),sep="\t",header=F)

# rename colomn 
colnames(denature_unmod_time)=c('read_name',seq(0,439,by=1))
colnames(denature_mod_time)=c('read_name',seq(0,439,by=1))
colnames(NAI_N3_25_unmod_time)=c('read_name',seq(0,439,by=1))
colnames(NAI_N3_25_mod_time)=c('read_name',seq(0,439,by=1))
colnames(NAI_N3_2_unmod_time)=c('read_name',seq(0,439,by=1))
colnames(NAI_N3_2_mod_time)=c('read_name',seq(0,439,by=1))
colnames(NAI_N3_1_unmod_time)=c('read_name',seq(0,439,by=1))
colnames(NAI_N3_1_mod_time)=c('read_name',seq(0,439,by=1))
colnames(unmod2_unmod_time)=c('read_name',seq(0,439,by=1))
colnames(unmod2_mod_time)=c('read_name',seq(0,439,by=1))
##
head(denature_unmod_time)
head(denature_mod_time)


#normolize
## min-Max normalization:(X - min(X))/(max(X)-min(X))
## Z-score standardization: (X-u)/sd
### function for min-max normalization
min_max_nor = function(x){
  (x-min(x))/(max(x)-min(x))
}

### function for Z-score standardization
zscore_nor=function(x){
  (x-mean(x))/sd(x)
}

# extract data.
rnadata=denature_unmod_time
maintext='denature_unmod_time'
rnadata=unmod2_unmod_time
maintext='unmod2_unmod_time'
rnadata=NAI_N3_1_unmod_time
maintext='NAI_N3_1_unmod_time'

#set position
b=200
predd=rnadata[,b][!is.na(rnadata[,b])]
predd
predd_nor=min_max_nor(predd)
predd_zs=zscore_nor(predd)
log2(2)
predd_nor
predd_zs
totalrna=length(predd)
hist(predd,breaks=1000,main=paste(maintext,' rawdata Position ',b,' number:',totalrna,sep=''),freq=T,xlim=c(0,2000),xlab='')

hist(predd_nor,breaks=1000,main=paste(maintext,' min-max normalization Position ',b,' number:',totalrna,sep=''),freq=T,xlab='')
hist(predd_zs,breaks=1000,main=paste(maintext,' Z-score standardization Position ',b,' number:',totalrna,sep=''),freq=T,xlab='')
hist(log2(predd),breaks=1000,main=paste(maintext,' log2 standardization Position ',b,' number:',totalrna,sep=''),freq=T,xlab='')

h <- hist(predd,breaks=1000)
hist(predd,breaks=1000,main=paste(maintext,' Position ',b,' number:',totalrna,sep=''),freq=T,xlim=c(0,100),xlab='')
text(h$mids,h$counts,labels=h$counts,cex=0.4, adj=c(0.5, -0.5))
#

# raw data ks test.
  
shiftt = data.frame(pos=c(),DF=c(),pv=c())
shiftt2= data.frame(pos=c(),DF=c(),pv=c())
for (b in 2:418){
  
  print(b)
  denature=denature_unmod_time[,b][!is.na(denature_unmod_time[,b])]
  denature
  unmod2=unmod2_unmod_time[,b][!is.na(unmod2_unmod_time[,b])]
  unmod2
  nai_n3=NAI_N3_1_unmod_time[,b][!is.na(NAI_N3_1_unmod_time[,b])]
  nai_n3
  #test. unmod vs denature. 
  s1=ks.test(denature,unmod2,alternative = c( "greater"))
  #nai_n3 vs denature.
  s2=ks.test(denature,nai_n3,alternative = c( "greater"))
  # get info unmod vs denature
  div=s1['statistic'][[1]][[1]]
  dip=s1['p.value'][[1]]
  shiftt=rbind(shiftt,c(b-2,div,dip))
#
  div2=s2['statistic'][[1]][[1]]
  dip2=s2['p.value'][[1]]
  shiftt2=rbind(shiftt2,c(b-2,div2,dip2))
  }

colnames(shiftt)=c("position","df","pvalue")
colnames(shiftt2)=c("position","df","pvalue")
head(shiftt)
head(shiftt2)
## plot df pvalue.
hist(shiftt$df,breaks = 100)
hist(shiftt$pvalue,breaks = 100)
hist(shiftt2$df,breaks = 100)
hist(shiftt2$pvalue,breaks = 100)
plot(-log10(pvalue+10^-270)~df, data = shiftt, pch = 16, cex = .4, xlab = "DF",ylab = "-log10 P value")
plot(-log10(pvalue+10^-270)~df, data = shiftt2, pch = 16, cex = .4, xlab = "DF",ylab = "-log10 P value")
## set sig
shiftt$sig=0
shiftt$sig[shiftt$df>0.15 & shiftt$pvalue<0.01]=1
head(shiftt)
shiftt2$sig=0
shiftt2$sig[shiftt2$df>0.15 & shiftt2$pvalue<0.01]=1
head(shiftt2)
## set df and -log10 pvalue
shiftt$dfpv=shiftt$df*(-log10(shiftt$pvalue+10^-270))
shiftt2$dfpv=shiftt2$df*(-log10(shiftt2$pvalue+10^-270))

## smooth dfpv
shiftt$dfpv_sm=filter(shiftt$dfpv,filter=c(rep(1/4,4)))
shiftt2$dfpv_sm=filter(shiftt2$dfpv,filter=c(rep(1/4,4)))

#smooth
shiftt$lin<-filter(shiftt$df,filter=c(rep(1/4,4)))
shiftt2$lin<-filter(shiftt2$df,filter=c(rep(1/4,4)))

#trans data to ksdata.
ksdata=shiftt
ksdata2=shiftt2
# plot the result. unmod vs denature. 
plot(shiftt$position, shiftt$df, pch = 16, col = 2, type = "l" )   
axis(1, at=seq(0,420,10))
par(new = TRUE)                             # Add new plot
# Create second plot without axes
plot(shiftt$position, -log10(shiftt$pvalue+10^-280) , pch = 17, col = 3,              
     axes = F, xlab = "", ylab = "",type="l")
axis(side = 4, at = pretty(range(-log10(shiftt$pvalue+10^-280) )))      # Add second axis
#axis(side = 4, at = pretty(range(-1,200)))      # Add second axis
mtext("timezx", side = 4, line = 3)             # Add second axis label
dev.off()

# plot nai_n3 vs denature.
plot(shiftt2$position, shiftt2$df, pch = 16, col = 2, type = "l" )   
axis(1, at=seq(0,420,10))
par(new = TRUE)                             # Add new plot
# Create second plot without axes
plot(shiftt2$position, -log10(shiftt2$pvalue+10^-280) , pch = 17, col = 3,              
     axes = F, xlab = "", ylab = "",type="l")
axis(side = 4, at = pretty(range(-log10(shiftt2$pvalue+10^-280) )))      # Add second axis
#axis(side = 4, at = pretty(range(-1,200)))      # Add second axis
mtext("timezx", side = 4, line = 3)             # Add second axis label
dev.off()


# plot the result. unmod vs denature, smooth
plot(shiftt$position, shiftt$df, pch = 16, col = 2, type = "l" )   
axis(1, at=seq(0,420,10))
par(new = TRUE)                             # Add new plot
# Create second plot without axes
plot(shiftt$position, shiftt$lin, pch = 17, col = 4,              
     axes = F, xlab = "", ylab = "",type="l")
axis(side = 4, at = pretty(range(shiftt$lin)))      # Add second axis
#axis(side = 4, at = pretty(range(-1,200)))      # Add second axis
mtext("timezx", side = 4, line = 3)             # Add second axis label
dev.off()

# plot nai_n3 vs denature, smooth.
plot(shiftt2$position, shiftt2$df, pch = 16, col = 2, type = "l" )   
axis(1, at=seq(0,420,10))
par(new = TRUE)                             # Add new plot
# Create second plot without axes
plot(shiftt2$position, shiftt2$lin , pch = 17, col = 4,              
     axes = F, xlab = "", ylab = "",type="l")
axis(side = 4, at = pretty(range(shiftt2$lin )))      # Add second axis
#axis(side = 4, at = pretty(range(-1,200)))      # Add second axis
mtext("timezx", side = 4, line = 3)             # Add second axis label
dev.off()
  
#min-max
shiftt = data.frame(pos=c(),DF=c(),pv=c())
shiftt2= data.frame(pos=c(),DF=c(),pv=c())
for (b in 2:418){
  
  print(b)
  denature=min_max_nor(denature_unmod_time[,b][!is.na(denature_unmod_time[,b])])
  denature
  unmod2=min_max_nor(unmod2_unmod_time[,b][!is.na(unmod2_unmod_time[,b])])
  unmod2
  nai_n3=min_max_nor(NAI_N3_1_unmod_time[,b][!is.na(NAI_N3_1_unmod_time[,b])])
  nai_n3
  #test. unmod vs denature. 
  s1=ks.test(denature,unmod2,alternative = c( "greater"))
  #nai_n3 vs denature.
  s2=ks.test(denature,nai_n3,alternative = c( "greater"))
  # get info unmod vs denature
  div=s1['statistic'][[1]][[1]]
  dip=s1['p.value'][[1]]
  shiftt=rbind(shiftt,c(b-2,div,dip))
  #
  div2=s2['statistic'][[1]][[1]]
  dip2=s2['p.value'][[1]]
  shiftt2=rbind(shiftt2,c(b-2,div2,dip2))
}

 colnames(shiftt)=c("position","df","pvalue")
colnames(shiftt2)=c("position","df","pvalue")
head(shiftt)
head(shiftt2)
plot(-log10(pvalue+10^-270)~df, data = shiftt, pch = 16, cex = .4, xlab = "DF",ylab = "-log10 P value")
plot(-log10(pvalue+10^-270)~df, data = shiftt2, pch = 16, cex = .4, xlab = "DF",ylab = "-log10 P value")
## set sig
shiftt$sig=0
shiftt$sig[shiftt$df>0.15 & shiftt$pvalue<0.01]=1
head(shiftt)
shiftt2$sig=0
shiftt2$sig[shiftt2$df>0.15 & shiftt2$pvalue<0.01]=1
head(shiftt2)
## set df and -log10 pvalue
shiftt$dfpv=shiftt$df*(-log10(shiftt$pvalue+10^-270))
shiftt2$dfpv=shiftt2$df*(-log10(shiftt2$pvalue+10^-270))

## smooth dfpv
shiftt$dfpv_sm=filter(shiftt$dfpv,filter=c(rep(1/4,4)))
shiftt2$dfpv_sm=filter(shiftt2$dfpv,filter=c(rep(1/4,4)))

#smooth
shiftt$lin<-filter(shiftt$df,filter=c(rep(1/4,4)))
shiftt2$lin<-filter(shiftt2$df,filter=c(rep(1/4,4)))

#trans data to ksdata.
ksdata=shiftt
ksdata2=shiftt2

# plot the result. unmod vs denature. 
plot(shiftt$position, shiftt$df, pch = 16, col = 2, type = "l" )   
axis(1, at=seq(0,420,10))
par(new = TRUE)                             # Add new plot
# Create second plot without axes
plot(shiftt$position, -log10(shiftt$pvalue+10^-280) , pch = 17, col = 3,              
     axes = F, xlab = "", ylab = "",type="l")
axis(side = 4, at = pretty(range(-log10(shiftt$pvalue+10^-280) )))      # Add second axis
#axis(side = 4, at = pretty(range(-1,200)))      # Add second axis
mtext("timezx", side = 4, line = 3)             # Add second axis label
dev.off()

# plot nai_n3 vs denature.
plot(shiftt2$position, shiftt2$df, pch = 16, col = 2, type = "l" )   
axis(1, at=seq(0,420,10))
par(new = TRUE)                             # Add new plot
# Create second plot without axes
plot(shiftt2$position, -log10(shiftt2$pvalue+10^-280) , pch = 17, col = 3,              
     axes = F, xlab = "", ylab = "",type="l")
axis(side = 4, at = pretty(range(-log10(shiftt2$pvalue+10^-280) )))      # Add second axis
#axis(side = 4, at = pretty(range(-1,200)))      # Add second axis
mtext("timezx", side = 4, line = 3)             # Add second axis label
dev.off()


# plot the result. unmod vs denature, smooth
plot(shiftt$position, shiftt$df, pch = 16, col = 2, type = "l" )   
axis(1, at=seq(0,420,10))
par(new = TRUE)                             # Add new plot
# Create second plot without axes
plot(shiftt$position, shiftt$lin, pch = 17, col = 4,              
     axes = F, xlab = "", ylab = "",type="l")
axis(side = 4, at = pretty(range(shiftt$lin)))      # Add second axis
#axis(side = 4, at = pretty(range(-1,200)))      # Add second axis
mtext("timezx", side = 4, line = 3)             # Add second axis label
dev.off()

# plot nai_n3 vs denature, smooth.
plot(shiftt2$position, shiftt2$df, pch = 16, col = 2, type = "l" )   
axis(1, at=seq(0,420,10))
par(new = TRUE)                             # Add new plot
# Create second plot without axes
plot(shiftt2$position, shiftt2$lin , pch = 17, col = 4,              
     axes = F, xlab = "", ylab = "",type="l")
axis(side = 4, at = pretty(range(shiftt2$lin )))      # Add second axis
#axis(side = 4, at = pretty(range(-1,200)))      # Add second axis
mtext("timezx", side = 4, line = 3)             # Add second axis label
dev.off()

a=1


#zscore
shiftt = data.frame(pos=c(),DF=c(),pv=c())
shiftt2= data.frame(pos=c(),DF=c(),pv=c())
for (b in 2:418){
  # b=162
  print(b)
  denature=zscore_nor(denature_unmod_time[,b][!is.na(denature_unmod_time[,b])])
  #denature
  unmod2=zscore_nor(unmod2_unmod_time[,b][!is.na(unmod2_unmod_time[,b])])
  #unmod2
  nai_n3=zscore_nor(NAI_N3_1_unmod_time[,b][!is.na(NAI_N3_1_unmod_time[,b])])
  #nai_n3
   #test. unmod vs denature. 
  s1=ks.test(denature,unmod2,alternative = c( "greater"))
  #nai_n3 vs denature.
  s2=ks.test(denature,nai_n3,alternative = c( "greater"))

  # get info unmod vs denature
  div=s1['statistic'][[1]][[1]]
  dip=s1['p.value'][[1]]
  shiftt=rbind(shiftt,c(b-2,div,dip))
  #
  div2=s2['statistic'][[1]][[1]]
  dip2=s2['p.value'][[1]]
  shiftt2=rbind(shiftt2,c(b-2,div2,dip2))
}

colnames(shiftt)=c("position","df","pvalue")
colnames(shiftt2)=c("position","df","pvalue")

#plot b=4 
## position 202 move 12 214 but b contain readname so column is 215
## position 210  b is 223
## position 190 202 b is 203
## position 134 146 b is 147
b=223
# b=162
print(b)
denature=zscore_nor(denature_unmod_time[,b][!is.na(denature_unmod_time[,b])])
#denature
unmod2=zscore_nor(unmod2_unmod_time[,b][!is.na(unmod2_unmod_time[,b])])
#unmod2
nai_n3=zscore_nor(NAI_N3_1_unmod_time[,b][!is.na(NAI_N3_1_unmod_time[,b])])
#nai_n3
#test. unmod vs denature. 
s1=ks.test(denature,unmod2,alternative = c( "greater"))
#nai_n3 vs denature.
s2=ks.test(denature,nai_n3,alternative = c( "greater"))
s1
s2
#head(denature_unmod_time)
##get dataframe
length(denature)

denaturet=data.frame(time=denature,label=rep('denature',length(denature)))
unmod2t=data.frame(time=unmod2,label=rep('unmod',length(unmod2)))
nai_n3t=data.frame(time=nai_n3,label=rep('NAI_N3',length(nai_n3)))

dtattpot=rbind(denaturet,unmod2t,nai_n3t)
minb=min(dtattpot$time)
## ggplot2 botxplot
ggplot(dtattpot, aes(x=label, y=time, color=label)) +
  geom_boxplot()+theme(axis.text=element_text(size=22,face="bold"),
                       axis.title=element_text(size=24,face="bold"),
                       legend.title=element_text(size=24,face="bold"), 
                       legend.text=element_text(size=22,face="bold"))+
  coord_cartesian(ylim = c(minb, 8))
##
ggplot(dtattpot, aes(x=label, y=time, color=label)) +
  geom_violin(trim=FALSE)+theme(axis.text=element_text(size=22,face="bold"),
                                axis.title=element_text(size=24,face="bold"),
                                legend.title=element_text(size=24,face="bold"), 
                                legend.text=element_text(size=22,face="bold"))+
  coord_cartesian(ylim = c(minb, 8))

##

ggplot(dtattpot, aes(x=time, color=label)) +
  geom_density() +theme(axis.text=element_text(size=22,face="bold"),
                        axis.title=element_text(size=24,face="bold"),
                        legend.title=element_text(size=24,face="bold"), 
                        legend.text=element_text(size=22,face="bold"))+coord_cartesian(xlim = c(minb, 8))

# get info unmod vs denature
div=s1['statistic'][[1]][[1]]
dip=s1['p.value'][[1]]
#plot end

head(shiftt)
head(shiftt2)
plot(-log10(pvalue+10^-270)~df, data = shiftt, pch = 16, cex = .4, xlab = "DF",ylab = "-log10 P value")
plot(-log10(pvalue+10^-270)~df, data = shiftt2, pch = 16, cex = .4, xlab = "DF",ylab = "-log10 P value")
## set sig
shiftt$sig=0
shiftt$sig[shiftt$df>0.15 & shiftt$pvalue<0.01]=1
head(shiftt)
shiftt2$sig=0
shiftt2$sig[shiftt2$df>0.15 & shiftt2$pvalue<0.01]=1
head(shiftt2)
## set df and -log10 pvalue
shiftt$dfpv=shiftt$df*(-log10(shiftt$pvalue+10^-270))
shiftt2$dfpv=shiftt2$df*(-log10(shiftt2$pvalue+10^-270))

## smooth dfpv
shiftt$dfpv_sm=filter(shiftt$dfpv,filter=c(rep(1/4,4)))
shiftt2$dfpv_sm=filter(shiftt2$dfpv,filter=c(rep(1/4,4)))

#smooth
shiftt$lin<-filter(shiftt$df,filter=c(rep(1/4,4)))
shiftt2$lin<-filter(shiftt2$df,filter=c(rep(1/4,4)))

#trans data to ksdata.
ksdata=shiftt
ksdata2=shiftt2

#save file 
write.table(ksdata2,paste(dir,'/a10-auc/nai_n3_ks.tsv',sep=''),sep="\t")
write.table(ksdata,paste(dir,'/a10-auc/unmod_ks.tsv',sep=''),sep="\t")
# plot the result. unmod vs denature. 
plot(shiftt$position, shiftt$df, pch = 16, col = 2, type = "l" )   
 axis(1, at=seq(0,420,10))
par(new = TRUE)                             # Add new plot
# Create second plot without axes
plot(shiftt$position, -log10(shiftt$pvalue+10^-280) , pch = 17, col = 3,              
     axes = F, xlab = "", ylab = "",type="l")
axis(side = 4, at = pretty(range(-log10(shiftt$pvalue+10^-280) )))      # Add second axis
#axis(side = 4, at = pretty(range(-1,200)))      # Add second axis
mtext("timezx", side = 4, line = 3)             # Add second axis label
dev.off()

# plot nai_n3 vs denature.
plot(shiftt2$position, shiftt2$df, pch = 16, col = 2, type = "l" )   
axis(1, at=seq(0,420,10))
par(new = TRUE)                             # Add new plot
# Create second plot without axes
plot(shiftt2$position, -log10(shiftt2$pvalue+10^-280) , pch = 17, col = 3,              
     axes = F, xlab = "", ylab = "",type="l")
axis(side = 4, at = pretty(range(-log10(shiftt2$pvalue+10^-280) )))      # Add second axis
#axis(side = 4, at = pretty(range(-1,200)))      # Add second axis
mtext("timezx", side = 4, line = 3)             # Add second axis label
dev.off()


# plot the result. unmod vs denature, smooth
plot(shiftt$position, shiftt$df, pch = 16, col = 2, type = "l" )   
axis(1, at=seq(0,420,10))
par(new = TRUE)                             # Add new plot
# Create second plot without axes
plot(shiftt$position, shiftt$lin, pch = 17, col = 4,              
     axes = F, xlab = "", ylab = "",type="l")
axis(side = 4, at = pretty(range(shiftt$lin)))      # Add second axis
#axis(side = 4, at = pretty(range(-1,200)))      # Add second axis
mtext("timezx", side = 4, line = 3)             # Add second axis label
dev.off()

# plot nai_n3 vs denature, smooth.
plot(shiftt2$position, shiftt2$df, pch = 16, col = 2, type = "l" )   
axis(1, at=seq(0,420,10))
par(new = TRUE)                             # Add new plot
# Create second plot without axes
plot(shiftt2$position, shiftt2$lin , pch = 17, col = 4,              
     axes = F, xlab = "", ylab = "",type="l")
axis(side = 4, at = pretty(range(shiftt2$lin )))      # Add second axis
#axis(side = 4, at = pretty(range(-1,200)))      # Add second axis
mtext("timezx", side = 4, line = 3)             # Add second axis label
dev.off()

# log2
shiftt = data.frame(pos=c(),DF=c(),pv=c())
shiftt2= data.frame(pos=c(),DF=c(),pv=c())
for (b in 2:418){
  
  print(b)
  denature=denature_unmod_time[,b][!is.na(denature_unmod_time[,b])]
  denature
  unmod2=unmod2_unmod_time[,b][!is.na(unmod2_unmod_time[,b])]
  unmod2
  nai_n3=NAI_N3_1_unmod_time[,b][!is.na(NAI_N3_1_unmod_time[,b])]
  nai_n3
  #test. unmod vs denature. 
  s1=ks.test(denature,unmod2,alternative = c( "greater"))
  #nai_n3 vs denature.
  s2=ks.test(denature,nai_n3,alternative = c( "greater"))
  # get info unmod vs denature
  div=s1['statistic'][[1]][[1]]
  dip=s1['p.value'][[1]]
  shiftt=rbind(shiftt,c(b-2,div,dip))
  #
  div2=s2['statistic'][[1]][[1]]
  dip2=s2['p.value'][[1]]
  shiftt2=rbind(shiftt2,c(b-2,div2,dip2))
}

colnames(shiftt)=c("position","df","pvalue")
colnames(shiftt2)=c("position","df","pvalue")
head(shiftt)
head(shiftt2)
plot(-log10(pvalue+10^-270)~df, data = shiftt, pch = 16, cex = .4, xlab = "DF",ylab = "-log10 P value")
plot(-log10(pvalue+10^-270)~df, data = shiftt2, pch = 16, cex = .4, xlab = "DF",ylab = "-log10 P value")
## set sig
shiftt$sig=0
shiftt$sig[shiftt$df>0.15 & shiftt$pvalue<0.01]=1
head(shiftt)
shiftt2$sig=0
shiftt2$sig[shiftt2$df>0.15 & shiftt2$pvalue<0.01]=1
head(shiftt2)
## set df and -log10 pvalue
shiftt$dfpv=shiftt$df*(-log10(shiftt$pvalue+10^-270))
shiftt2$dfpv=shiftt2$df*(-log10(shiftt2$pvalue+10^-270))

## smooth dfpv
shiftt$dfpv_sm=filter(shiftt$dfpv,filter=c(rep(1/4,4)))
shiftt2$dfpv_sm=filter(shiftt2$dfpv,filter=c(rep(1/4,4)))


#smooth
shiftt$lin<-filter(shiftt$df,filter=c(rep(1/4,4)))
shiftt2$lin<-filter(shiftt2$df,filter=c(rep(1/4,4)))

#trans data to ksdata.
ksdata=shiftt
ksdata2=shiftt2
#

#
plot(shiftt$position, shiftt$lin, pch = 16, col = 2, type = "l" )   
axis(1, at=seq(0,420,10))
plot(shiftt2$position, shiftt2$lin, pch = 16, col = 2, type = "l" )   
axis(1, at=seq(0,420,10))

# plot the result. unmod vs denature. 
plot(shiftt$position, shiftt$df, pch = 16, col = 2, type = "l" )   
axis(1, at=seq(0,420,10))
par(new = TRUE)                             # Add new plot
# Create second plot without axes
plot(shiftt$position, -log10(shiftt$pvalue+10^-280) , pch = 17, col = 3,              
     axes = F, xlab = "", ylab = "",type="l")
axis(side = 4, at = pretty(range(-log10(shiftt$pvalue+10^-280) )))      # Add second axis
#axis(side = 4, at = pretty(range(-1,200)))      # Add second axis
mtext("timezx", side = 4, line = 3)             # Add second axis label
dev.off()

# plot nai_n3 vs denature.
plot(shiftt2$position, shiftt2$df, pch = 16, col = 2, type = "l" )   
axis(1, at=seq(0,420,10))
par(new = TRUE)                             # Add new plot
# Create second plot without axes
plot(shiftt2$position, -log10(shiftt2$pvalue+10^-280) , pch = 17, col = 3,              
     axes = F, xlab = "", ylab = "",type="l")
axis(side = 4, at = pretty(range(-log10(shiftt2$pvalue+10^-280) )))      # Add second axis
#axis(side = 4, at = pretty(range(-1,200)))      # Add second axis
mtext("timezx", side = 4, line = 3)             # Add second axis label
dev.off()

# plot the result. unmod vs denature, smooth
plot(shiftt$position, shiftt$df, pch = 16, col = 2, type = "l" )   
axis(1, at=seq(0,420,10))
par(new = TRUE)                             # Add new plot
# Create second plot without axes
plot(shiftt$position, shiftt$lin, pch = 17, col = 4,              
     axes = F, xlab = "", ylab = "",type="l")
axis(side = 4, at = pretty(range(shiftt$lin)))      # Add second axis
#axis(side = 4, at = pretty(range(-1,200)))      # Add second axis
mtext("timezx", side = 4, line = 3)             # Add second axis label
dev.off()

# plot nai_n3 vs denature, smooth.
plot(shiftt2$position, shiftt2$df, pch = 16, col = 2, type = "l" )   
axis(1, at=seq(0,420,10))
par(new = TRUE)                             # Add new plot
# Create second plot without axes
plot(shiftt2$position, shiftt2$lin , pch = 17, col = 4,              
     axes = F, xlab = "", ylab = "",type="l")
axis(side = 4, at = pretty(range(shiftt2$lin )))      # Add second axis
#axis(side = 4, at = pretty(range(-1,200)))      # Add second axis
mtext("timezx", side = 4, line = 3)             # Add second axis label
dev.off()



#mean plot the mean and sd distribution.
pos_m=seq(2,438,by=1)
rnadata=denature_unmod_time
maintext='denature_unmod_time'
rnadata=unmod2_unmod_time
maintext='unmod2_unmod_time'
rnadata=NAI_N3_1_unmod_time
maintext='NAI_N3_1_unmod_time'

# get the raw data signal mean and sd.
mtxout=data.frame(c(0),c(0),c(0))
colnames(mtxout)=c('pos','mean','sd')
for(b in pos_m)
{
print(b)
predd=rnadata[,b][!is.na(rnadata[,b])]

a1=mean(predd)
a2=sd(predd)
newrow=c(b-2,a1,a2)
mtxout=rbind(mtxout,newrow)
}

mtxout=mtxout[-1,]
head(mtxout)
plot(mtxout$mean,mtxout$sd,xlim=c(0,1000),ylim=c(0,1000),pch = 19,cex=0.2,main=maintext)
abline(b=1,a=0,col="red")
#text(mtxout[100,2],mtxout[100,3],'100',col='red')
points(mtxout[100,2],mtxout[100,3],col='red',pch = 19,cex=0.2)
points(mtxout[120,2],mtxout[120,3],col='red',pch = 19,cex=0.2)
points(mtxout[200,2],mtxout[200,3],col='red',pch = 19,cex=0.2)
```






