# Determination of RNA structure with nanopore long reads dwell time

## abstract
RNA molecules can fold into complex secondary structure, that are critical for many biological processes. Currently methods have been developed transcriptome-wide probing of RNA secondary structure using nuclease and chemical modifiers. We found that using nanopore sequencing with direct long-read RNA sequencing RNA secondary structures have more dwell time than denature RNAs. The current that flows through the channel of the nanopores during sequencing----the current dwell time determined paired or not. We present RNA structure analysis pipeline using nanopore sequencing long read dwell time to detect secondary structures.


## Background 
RNAs can ploy significant roles in genetic material carrying heritable information and also as a ribozyme catalyzing many biological processes 1 . RNAs can fold into complex secondary and tertiary structures to carefully regulate and refine gene expression. Assign correct structures to transcript is important. Enzymatic and chemical probes have been developed with high-throughput sequencing to generate (product) large-scale RNA structure information across transcriptomes in vitro and vivo [2-10]. 
With recent advances, Oxford Nanopore platforms have enabled amplification-free, long-read complementary (cDNA) DNA in high-throughput sequencing. The DNA and RNA molecules are directly threaded through a biological pore generating current.
In the present study, we coupled dwell time with direct RNA sequencing on nanopore sequencing to identify RNA structural.

##



