# 鲁伯埙 Results:

> 

## 实验设计

|ID |Name |
|-|-|
|7-7 |野生型|  
|7-111| 突变体|


## Raw data process pipeline 
### **Quality**

7-7-R

![7-7-r](../image/190828-lbx/7-7-r-qc.png)   

7-7-T

![7-7-t](../image/190828-lbx/7-7-t-qc.png)   

7-111-R

![7-1111-r](../image/190828-lbx/7-111-R-qc.png)   

7-111-T

![7-111-t](../image/190828-lbx/7-111-T-qc.png)  

### **rRNA**
| Iterm    | 7-111-R            | 7-7-R              | 7-111-T            | 7-7-T              |
|----------|--------------------|--------------------|--------------------|--------------------|
| Total    |  33421015          |  17057680          |  36125969          |  37219183          |
| rRNA     |  9679013 (28.96%)  |  3451860 (20.24%)  |  2389272 (6.61%)   |  376013 (1.01%)    |
| non rRNA |  23742002 (71.04%) |  13605820 (79.76%) |  33736697 (93.39%) |  36843170 (98.99%) |

### Read length distribution
7-7-R

![7-7-r](../image/190828-lbx/7-7-R_nontam.png)   

7-7-T

![7-7-t](../image/190828-lbx/7-7-T_nontam.png)   

7-111-R

![7-1111-r](../image/190828-lbx/7-111-R_nontam.png)   

7-111-T

![7-111-t](../image/190828-lbx/7-111-T_nontam.png) 

### read to RNA DNA and Intron (Using Tophat and readsNumCal_intron_v3)
| Iterm                                | 7-111-R   | 7-7-R    | 7-111-T  | 7-7-T    |
|--------------------------------------|-----------|----------|----------|----------|
| unique mapped reads of RNA           |  10548663 |  4743168 |  5635056 |  5271249 |
| unique mapped reads of DNA           |  5064442  |  456997  |  3928551 |  6091270 |
| unique mapped reads of Intron        |  207337   |  73397   |  450475  |  527807  |
| unique mapped ambiguous reads of RNA |  191480   |  287350  |  364688  |  418920  |

## Differentail translations (Xtail) Results: 

**File list**
* [lbx_results.txt](result_190626_lbx/lbx_results.txt)
* [lbx.merge.counter](result_190626_lbx/lbx.merge.counter)
* [lbxfc_results.txt](result_190626_lbx/lbxfc_results.txt)
* [lbxFC.pdf](result_190626_lbx/lbxFC.pdf)
* [lbxrs_results.txt](result_190626_lbx/lbxrs_results.txt)
* [lbxRs.pdf](result_190626_lbx/lbxRs.pdf)
* [lbxvolcano.pdf](result_190626_lbx/lbxvolcano.pdf)

|File name|Note|
|--------- |----|
|[lbx_results.txt](result_190626_lbx/lbx_results.txt)    |The results of ﬁst pipline are named with suﬃx ” v1”, which are generated by comparing mRNA and RPF log2 fold changes: The element log2FC_TE_v1 represents the log2 fold change of TE; The pvalue_v1 represent statistical signiﬁcance. The sencond pipline are named with suﬃx ” v2”, which are derived by comparing log2 ratios between two conditions: log2FC_TE_v2, and pvalue_v2 are log2 ratio of TE, and pvalues. Finally, the more conserved results (with larger-Pvalue) was select as the ﬁnal assessment of diﬀerential translation, which are named with suﬃx ” ﬁnal”. The pvalue.adjust is the estimated false discovery rate corresponding to the pvalue_final.|
|[lbx.merge.counter](result_190626_lbx/lbx.merge.counter)|Each condition reads count.|
|[lbxFC.pdf](result_190626_lbx/lbxFC.pdf)                |Figure 1: Scatter plot of log2 fold changes|
|[lbxfc_results.txt](result_190626_lbx/lbxfc_results.txt)|Figure 1: result records.|
|[lbxRs.pdf](result_190626_lbx/lbxRs.pdf)                |Figure 2: Scatter plot of log2 RPF-to-mRNA ratios|
|[lbxrs_results.txt](result_190626_lbx/lbxrs_results.txt)|Figure 2: result records.|
|[lbxvolcano.pdf](result_190626_lbx/lbxvolcano.pdf)      |It can also be useful to evaluate the fold changes cutoﬀ and p values thresholds by looking at the volcano plot.|

-----------
* blue: for genes whoes mRNA_log2FC larger than log2FC.cutoff (transcriptional level).
* red: for genes whoes RPF_log2FC larger than log2FC.cutoff (translational level).
* green: for genes changing homodirectionally at both level.
* yellow: for genes changing antidirectionally at two levels.

![lbxFC](../image/190828-lbx/lbxFC.png) 
* Figure 1: Scatter plot of log2 fold changes

Those genes in which the diﬀerence of mRNA_log2FC and RPF_log2FC did not exceed more than log2FC.cutoff are excluded. The points will be color-coded with the pvalue_final obtained with xtail (more signiﬁcant p values having darker color). By default the log2FC.cutoff is 1.

-------------------

* blue: for genes whoes log2R larger in ﬁrst condition than second condition.

* red: for genes whoes log2R larger in second condition than the ﬁrst condition.

* green: for genes whoes log2R changing homodirectionally in two condition.

* yellow: for genes whoes log2R changing antidirectionally in two conditon.

![lbxRS](../image/190828-lbx/lbxRS.png) 

* Figure 2: Scatter plot of log2 RPF-to-mRNA ratios

Those genes in which the diﬀerence of log2R in two conditions did not exceed more than log2R.cutoff are excluded. The points will be color-coded with the pvalue_final obtained with xtail (more signiﬁcant p values having darker color). By default the log2R.cutoff is 1.

## Ribo code
**File list**  

|File name|Note|
|---------|----|
|[RiboCode_ORFs_result.txt](result_190626_lbx/RiboCode_ORFs_result.txt)           |contains the information of predicted ORFs in each transcript.  |
|[RiboCode_ORFs_result_collapsed.txt](result_190626_lbx/RiboCode_ORFs_result_collapsed.txt)  | combines the ORFs with the same stop codon in different transcript isoforms: the one harboring the most upstream in-frame ATG is chosen. |

**Some column names of the result file:**

|                    Column                      |                       mean                                | 
|---------------------------------               |-------------------------------------------------------| 
| - ORF_ID                                 |  The identifier of ORFs that predicated.                                                                                    | 
| - ORF_type                               |  The type of ORF is annotated according the relative location to associated CDS. The following ORF categories are reported: | 
||  "annotated" (overlapping annotated CDS, have the same stop with annotated CDS)             |   
||  "uORF" (in upstream of annotated CDS, not overlapping annotated CDS)                       |   
||  "dORF" (in downstream of annotated CDS, not overlapping annotated CDS)                     |  
||  "Overlap_uORF" (in upstream of annotated CDS, overlapping annotated CDS)                   |  
||  "Overlap_dORF" (in downstream of annotated CDS, overlapping annotated CDS"                 | 
||  "Internal" (in internal of annotated CDS, but in a different frame relative annotated CDS) | 
||  "novel" (in non-coding genes or non-coding transcripts of coding genes).                   |  
| - ORF_tstart, ORF_tstop                  |  the beginning and end of ORF in RNA transcript (1-based coordinate)                                                        | 
| - ORF_gstart, ORF_gstop                  |  the beginning and end of ORF in genome (1-based coordinate)                                                                | 
| - pval_frame0_vs_frame1                  |  significance levels of P-site densities of frame0 greater than of frame1                                                   | 
| - pval_frame0_vs_frame2                  |  significance levels of P-site densities of frame0 greater than of frame2                                                   | 
| - pval_combined                          |  integrated P-value                                                                                                         | 


### 7-111-R
| #read_length | proportion(per total mapped reads) | predicted_psite | f0_sum | f1_sum | f2_sum | f0_percent | pvalue1                | pvalue2               | pvalue_combined        |
|--------------|------------------------------------|-----------------|--------|--------|--------|------------|------------------------|-----------------------|------------------------|
| # 30         | 23.45%                             | 12              | 54979  | 5055   | 8040   | 80.76%     | 0.00014652626009624466 | 0.0004229273113302267 | 4.3151740824280333e-07 |
| # 29         | 5.06%                              | 12              | 11074  | 3784   | 1779   | 66.56%     | 0.0002995597005840259  | 0.0003563133512190236 | 7.173383508666136e-07  |

![7-111-R](../image/190828-lbx/7-111-29nt-psite.png)   
![7-111-R](../image/190828-lbx/7-111-30nt-psite.png)     

### 7-7-R
|              |                                    |                 |        |        |        |            |                        |                        |                       | 
|--------------|------------------------------------|-----------------|--------|--------|--------|------------|------------------------|------------------------|-----------------------| 
| #read_length | proportion(per total mapped reads) | predicted_psite | f0_sum | f1_sum | f2_sum | f0_percent | pvalue1                | pvalue2                | pvalue_combined       | 
| # 28         | 22.95%                             | 12              | 51889  | 6952   | 1806   | 85.56%     | 0.00014652626009624466 | 0.00014652626009624466 | 1.516508550767744e-07 | 
| # 29         | 43.83%                             | 12              | 78951  | 3609   | 8621   | 86.59%     | 0.00014652626009624466 | 0.00014652626009624466 | 1.516508550767744e-07 | 

![7-7-R](../image/190828-lbx/7-7-R-28nt-psite.png)   
![7-7-R](../image/190828-lbx/7-7-R-29nt-psite.png)     

### 
![7-7-R](../image/190828-lbx/orfs_category.png)    

预期是111的stalling增加，您看看是否有这个现象

**Htt**

![](./result_190626_lbx/a1-htt.png)    


## RiboMiner

### Quality Control (QC)

**3-nt periodicity plot generated by RiboMiner**

* 7-7 
![](./result_190626_lbx/periodicity.28.77.png)
![](./result_190626_lbx/periodicity.29.77.png)
![](./result_190626_lbx/periodicity.30.77.png)

* 111
![](./result_190626_lbx/periodicity.30.111.png)
![](./result_190626_lbx/periodicity.31.111.png)
  
**Length distribution of read length.**

Exon means reads mapped to transcriptome (exon).intergenic region means reads mapped to intergenic region of genes. And intron means reads mapped to intron

* 7-7 

Intergenic
![](./result_190626_lbx/con_77.DNA.png)

Intron

![](./result_190626_lbx/con_77.intron.png)

Exon

![](./result_190626_lbx/con_77.RNA.png)

* 111

Intergenic

![](./result_190626_lbx/con_111.DNA.png)

Intron

![](./result_190626_lbx/con_111.intron.png)

Exon

![](./result_190626_lbx/con_111.RNA.png)

**Reads mapped to different reading frames.**

* 7-7

![](./result_190626_lbx/read_in_diff_frames_77.png)

* 111

![](./result_190626_lbx/read_in_diff_frames_111.png)

### Metagene Analysis (MA)



**A general reads distribution along transcripts.**
![](./result_190626_lbx/meta_whole_region.png)


**B. Distribution of polarity scores.** 
![](./result_190626_lbx/polarity.png)

polarity 分布如果都在0附近说明对这个转录本来说reads分布相对于转录本中心来说分布比较均匀，如果左偏，或者右偏说明有很多的转录本上的reads分布不均匀，在转录本5'端多了或者少了


**C. Ribosome density profiles along CDS regions.** 
![](./result_190626_lbx/meta_cds_5.png)
![](./result_190626_lbx/meta_cds_3.png)

left: read density on the position away from start codon. right: read density on the position away from stop codon. 



**D. Ribosome density profiles along UTR regions.** 

**这种图y值=（每个位置上的RPM）/(n codon之后的density均值）**

![](./result_190626_lbx/meta_utr_5.png)
![](./result_190626_lbx/meta_utr_3.png)

left: read density on the position away from start codon. right: read density on the position away from stop codon.

**7_111翻译起始有点慢**
**start codon多很多，考虑翻译起始受抑制吧**

**Figure 3: Ribosome density among different amino acids and codons.**

A. Relative ribosome density on different amino acids for 7-7 sample. 
![](./result_190626_lbx/aa.77.png)

B. Relative ribosome density on different codons for 7-7 sample. 
![](./result_190626_lbx/aa.77-class.png)

C. Relative ribosome density on different amino acids for 7-7 sample. 
![](./result_190626_lbx/aa.7-111.png)


D. Relative ribosome density on different codons for 7-7 sample. 
![](./result_190626_lbx/aa.7-111.class.png)

E. The change ratio of ribosome density between 7-7 sample and 7-111 sample among different amino acids. 
![](./result_190626_lbx/aa.ratio.png)

Notes: all these analysis are based on total transcripts.

#### 7-111 vs 7-7


|Name|Transcript|Translation|Number|
|-|-|-|-|
|transcript_down_translation_up:| transcript 7-111  high 7-7 transcirpt low,  |translation 7-111 low 7-7 high|320|
|transcript_only_down: |transcript 7-111  high 7-7 transcirpt low, | translation not change|1027|
|transcript_only_up: |transcript 7-111  low 7-7 transcirpt high, | translation not change|316|
|transcript_translation_down: |transcript 7-111  high 7-7 transcirpt low,|  translation 7-111 high 7-7 low|284|
|transcript_translation_up: |transcript 7-111  low 7-7 transcirpt high,|  translation 7-111 low 7-7 high|85|
|transcript_up_translation_down:| transcript 7-111  low 7-7 transcirpt high,  |translation 7-111 high 7-7 low|124|
|translation_only_down:| transcript not change, |  translation 7-111 high 7-7 low|64|
|translation_only_up: |transcript not change, |translation 7-111 low 7-7 high|68|

[transcript_down_translation_up](./result_190626_lbx/d_feature_analysis/b7-PlotRiboDensityAtEachKindAAOrCodon_select.transcript_down_translation_up.list.gene_density_on_each_kind_of_AA.pdf)

[transcript_only_down](./result_190626_lbx/d_feature_analysis/b7-PlotRiboDensityAtEachKindAAOrCodon_select.transcript_only_down.list.gene_density_on_each_kind_of_AA.pdf)

[transcript_only_up](./result_190626_lbx/d_feature_analysis/b7-PlotRiboDensityAtEachKindAAOrCodon_select.transcript_only_up.list.gene_density_on_each_kind_of_AA.pdf)

[transcript_translation_down](./result_190626_lbx/d_feature_analysis/b7-PlotRiboDensityAtEachKindAAOrCodon_select.transcript_translation_down.list.gene_density_on_each_kind_of_AA.pdf)

[transcript_translation_up](./result_190626_lbx/d_feature_analysis/b7-PlotRiboDensityAtEachKindAAOrCodon_select.transcript_translation_up.list.gene_density_on_each_kind_of_AA.pdf)

[transcript_up_translation_down](./result_190626_lbx/d_feature_analysis/b7-PlotRiboDensityAtEachKindAAOrCodon_select.transcript_up_translation_down.list.gene_density_on_each_kind_of_AA.pdf)

[translation_only_down](./result_190626_lbx/d_feature_analysis/b7-PlotRiboDensityAtEachKindAAOrCodon_select.translation_only_down.list.gene_density_on_each_kind_of_AA.pdf)

[translation_only_up](./result_190626_lbx/d_feature_analysis/b7-PlotRiboDensityAtEachKindAAOrCodon_select.translation_only_up.list.gene_density_on_each_kind_of_AA.pdf)

**Ribosome density among triplete-AA motifs.** 

A. Relative ribosome density on poly-proline acid. 

[transcript_down_translation_up](./result_190626_lbx/d_feature_analysis/b9-PPP_plot_select.transcript_down_translation_up.list.gene_mean_density_on_triAAMotifs.pdf)

[transcript_only_down](./result_190626_lbx/d_feature_analysis/b9-PPP_plot_select.transcript_only_down.list.gene_mean_density_on_triAAMotifs.pdf)

[transcript_only_up](./result_190626_lbx/d_feature_analysis/b9-PPP_plot_select.transcript_only_up.list.gene_mean_density_on_triAAMotifs.pdf)

[transcript_translation_down](./result_190626_lbx/d_feature_analysis/b9-PPP_plot_select.transcript_translation_down.list.gene_mean_density_on_triAAMotifs.pdf)

[transcript_translation_up](./result_190626_lbx/d_feature_analysis/b9-PPP_plot_select.transcript_translation_up.list.gene_mean_density_on_triAAMotifs.pdf)

[transcript_up_translation_down](./result_190626_lbx/d_feature_analysis/b9-PPP_plot_select.transcript_up_translation_down.list.gene_mean_density_on_triAAMotifs.pdf)

[translation_only_down](./result_190626_lbx/d_feature_analysis/b9-PPP_plot_select.translation_only_down.list.gene_mean_density_on_triAAMotifs.pdf)

[translation_only_up](./result_190626_lbx/d_feature_analysis/b9-PPP_plot_select.translation_only_up.list.gene_mean_density_on_triAAMotifs.pdf)


B. Relative ribosome density.

[transcript_down_translation_up](./result_190626_lbx/d_feature_analysis/c1-triple_motif_plot_select.transcript_down_translation_up.list.gene_mean_density_on_triAAMotifs.pdf)

[transcript_only_down](./result_190626_lbx/d_feature_analysis/c1-triple_motif_plot_select.transcript_only_down.list.gene_mean_density_on_triAAMotifs.pdf)

[transcript_only_up](./result_190626_lbx/d_feature_analysis/c1-triple_motif_plot_select.transcript_only_up.list.gene_mean_density_on_triAAMotifs.pdf)

[transcript_translation_down](./result_190626_lbx/d_feature_analysis/c1-triple_motif_plot_select.transcript_translation_down.list.gene_mean_density_on_triAAMotifs.pdf)

[transcript_up_translation_down](./result_190626_lbx/d_feature_analysis/c1-triple_motif_plot_select.transcript_up_translation_down.list.gene_mean_density_on_triAAMotifs.pdf)

[translation_only_down](./result_190626_lbx/d_feature_analysis/c1-triple_motif_plot_select.translation_only_down.list.gene_mean_density_on_triAAMotifs.pdf)

## Distribution of local cAI along transcripts for different gene sets.

tAI是表示tRNA gene copy numbers对翻译影响的，值越大说明翻译相对越快，CAI是评估codon usage对翻译速率影响的，值越大，说明转录本中optimal codon越读，翻译越快

[c7-tAIPlot_homo_average_start_codon.pdf](./result_190626_lbx/d_feature_analysis/c7-tAIPlot_homo_average_start_codon.pdf)

[c7-tAIPlot_homo_average_stop_codon.pdf](./result_190626_lbx/d_feature_analysis/c7-tAIPlot_homo_average_stop_codon.pdf)

[c7-tAIPlot_opposite_average_start_codon.pdf](./result_190626_lbx/d_feature_analysis/c7-tAIPlot_opposite_average_start_codon.pdf)

[c7-tAIPlot_opposite_average_stop_codon.pdf](./result_190626_lbx/d_feature_analysis/c7-tAIPlot_opposite_average_stop_codon.pdf)

[c7-tAIPlot_transcrpt_average_start_codon.pdf](./result_190626_lbx/d_feature_analysis/c7-tAIPlot_transcrpt_average_start_codon.pdf)

[c7-tAIPlot_transcrpt_average_stop_codon.pdf](./result_190626_lbx/d_feature_analysis/c7-tAIPlot_transcrpt_average_stop_codon.pdf)

[c7-tAIPlot_translation_average_start_codon.pdf](./result_190626_lbx/d_feature_analysis/c7-tAIPlot_translation_average_start_codon.pdf)

[c7-tAIPlot_translation_average_stop_codon.pdf](./result_190626_lbx/d_feature_analysis/c7-tAIPlot_translation_average_stop_codon.pdf)

## Distribution of local tAI along transcripts for different gene sets. 

[c9-cAIPlot_homo_average_start_codon.pdf](./result_190626_lbx/d_feature_analysis/c9-cAIPlot_homo_average_start_codon.pdf)

[c9-cAIPlot_homo_average_stop_codon.pdf](./result_190626_lbx/d_feature_analysis/c9-cAIPlot_homo_average_stop_codon.pdf)

[c9-cAIPlot_opposite_average_start_codon.pdf](./result_190626_lbx/d_feature_analysis/c9-cAIPlot_opposite_average_start_codon.pdf)

[c9-cAIPlot_opposite_average_stop_codon.pdf](./result_190626_lbx/d_feature_analysis/c9-cAIPlot_opposite_average_stop_codon.pdf)

[c9-cAIPlot_transcrpt_average_start_codon.pdf](./result_190626_lbx/d_feature_analysis/c9-cAIPlot_transcrpt_average_start_codon.pdf)

[c9-cAIPlot_transcrpt_average_stop_codon.pdf](./result_190626_lbx/d_feature_analysis/c9-cAIPlot_transcrpt_average_stop_codon.pdf)

[c9-cAIPlot_translation_average_start_codon.pdf](./result_190626_lbx/d_feature_analysis/c9-cAIPlot_translation_average_start_codon.pdf)

[c9-cAIPlot_translation_average_stop_codon.pdf](./result_190626_lbx/d_feature_analysis/c9-cAIPlot_translation_average_stop_codon.pdf)

## Average hydrophobicity of each position along transcripts for different gene sets. 

[d3_hydropathy_homo_average_start_codon.pdf](./result_190626_lbx/d_feature_analysis/d3_hydropathy_homo_average_start_codon.pdf)

[d3_hydropathy_homo_average_stop_codon.pdf](./result_190626_lbx/d_feature_analysis/d3_hydropathy_homo_average_stop_codon.pdf)

[d3_hydropathy_opposite_average_start_codon.pdf](./result_190626_lbx/d_feature_analysis/d3_hydropathy_opposite_average_start_codon.pdf)

[d3_hydropathy_opposite_average_stop_codon.pdf](./result_190626_lbx/d_feature_analysis/d3_hydropathy_opposite_average_stop_codon.pdf)

[d3_hydropathy_transcript_average_start_codon.pdf](./result_190626_lbx/d_feature_analysis/d3_hydropathy_transcript_average_start_codon.pdf)

[d3_hydropathy_transcript_average_stop_codon.pdf](./result_190626_lbx/d_feature_analysis/d3_hydropathy_transcript_average_stop_codon.pdf)

[d3_hydropathy_translation_average_start_codon.pdf](./result_190626_lbx/d_feature_analysis/d3_hydropathy_translation_average_start_codon.pdf)

[d3_hydropathy_translation_average_stop_codon.pdf](./result_190626_lbx/d_feature_analysis/d3_hydropathy_translation_average_stop_codon.pdf)

##  Charge amino acids of each position along transcripts for different gene sets. 

[d4_charge_homo_average_start_codon.pdf](./result_190626_lbx/d_feature_analysis/d4_charge_homo_average_start_codon.pdf)

[d4_charge_homo_average_stop_codon.pdf](./result_190626_lbx/d_feature_analysis/d4_charge_homo_average_stop_codon.pdf)

[d4_charge_opposite_average_start_codon.pdf](./result_190626_lbx/d_feature_analysis/d4_charge_opposite_average_start_codon.pdf)

[d4_charge_opposite_average_stop_codon.pdf](./result_190626_lbx/d_feature_analysis/d4_charge_opposite_average_stop_codon.pdf)

[d4_charge_transcript_average_start_codon.pdf](./result_190626_lbx/d_feature_analysis/d4_charge_transcript_average_start_codon.pdf)

[d4_charge_transcript_average_stop_codon.pdf](./result_190626_lbx/d_feature_analysis/d4_charge_transcript_average_stop_codon.pdf)

[d4_charge_translation_average_start_codon.pdf](./result_190626_lbx/d_feature_analysis/d4_charge_translation_average_start_codon.pdf)

[d4_charge_translation_average_stop_codon.pdf](./result_190626_lbx/d_feature_analysis/d4_charge_translation_average_stop_codon.pdf)

总结下问题: 1. tAI, cAI在start峰的意义  
在start codon上的可以忽略不看，因为start codon上的值是相对固定的都是AUG上计算出来的值，没什么意义  
2. global tAI，cAI纵坐标不一样，是什么意义  
tAI是表示tRNA gene copy numbers对翻译影响的，值越大说明翻译相对越快，CAI是评估codon usage对翻译速率影响的，值越大，说明转录本中optimal codon越读，翻译越快 
3. polarity score分布峰宽增加有什么意义?   
polarity 分布如果都在0附近说明对这个转录本来说reads分布相对于转录本中心来说分布比较均匀，如果左偏，或者右偏说明有很多的转录本上的reads分布不均匀，在转录本5'端多了或者少了  
4. 是否可以初步分析下QQQ是否有stalling现象。非常感谢！  

